// react-native-scanbot-barcode-scanner-sdk/lib/module/impl/BarcodeScannerImpl.js
import { NativeEventEmitter, Platform } from 'react-native';
import { BarcodeDocumentParserResult, BarcodeScannerResult, BarcodeScannerUiResult, MultipleScanningMode, SingleScanningMode, createSBError, handleImageInput, mapRTUUIResult, withSBErrorHandling } from '../types';
import { ScanbotBarcodeSDKModule } from './scanbotBarcodeSDKModule';
const isIOS = Platform.OS === 'ios';

/**
 * @internal
 * @hidden
 */
export const ScanbotBarcodeImpl = {
  async startScanner(configuration) {
    const barcodeItemMapperEventName = 'barcodeItemMapperEvent';
    let barcodeItemMapperCallback = null;
    let barcodeItemMapperEventEmitter;
    if (configuration.useCase instanceof SingleScanningMode || configuration.useCase instanceof MultipleScanningMode) {
      barcodeItemMapperCallback = configuration.useCase.barcodeInfoMapping.barcodeItemMapper;
      if (barcodeItemMapperCallback) {
        barcodeItemMapperEventEmitter = new NativeEventEmitter(ScanbotBarcodeSDKModule);
        barcodeItemMapperEventEmitter.removeAllListeners(barcodeItemMapperEventName);
        barcodeItemMapperEventEmitter.addListener(barcodeItemMapperEventName, barcodeItem => {
          const barcodeItemUuid = `${barcodeItem.text}${barcodeItem.upcEanExtension}_${barcodeItem.format}`;
          barcodeItemMapperCallback(barcodeItem, barcodeMappedData => ScanbotBarcodeSDKModule.onBarcodeItemMapper(barcodeItemUuid, barcodeMappedData), () => ScanbotBarcodeSDKModule.onBarcodeItemMapper(barcodeItemUuid, null));
        });

        // On iOS, the communication with the native part is throwing an error (startBarcodeScannerV2) because of the barcodeItemMapper
        // callback in the configuration class (not parsable), that's why we need to remove it before executing startBarcodeScannerV2 method
        configuration.useCase.barcodeInfoMapping.barcodeItemMapper = null;
      }
    }
    try {
      const barcodeResult = await ScanbotBarcodeSDKModule.startBarcodeScanner(isIOS ? JSON.stringify(configuration) : configuration, !!barcodeItemMapperCallback);
      return mapRTUUIResult(barcodeResult, BarcodeScannerUiResult);
    } catch (error) {
      throw createSBError(error);
    } finally {
      barcodeItemMapperEventEmitter?.removeAllListeners(barcodeItemMapperEventName);
      barcodeItemMapperEventEmitter = undefined;
    }
  },
  scanFromImage(params) {
    return withSBErrorHandling(async () => new BarcodeScannerResult(await ScanbotBarcodeSDKModule.scanBarcodesFromImage({
      configuration: params.configuration,
      image: handleImageInput(params.image)
    })));
  },
  scanFromPdf(params) {
    return withSBErrorHandling(async () => new BarcodeScannerResult(await ScanbotBarcodeSDKModule.scanBarcodesFromPdf(params)));
  },
  parseDocument(params) {
    return withSBErrorHandling(async () => new BarcodeDocumentParserResult(await ScanbotBarcodeSDKModule.parseBarcodeDocument(params)));
  }
};
