// react-native-scanbot-barcode-scanner-sdk/lib/module/types/core/image/ImageRefTypes.js
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from core/schemas/ImageRefTypes.yaml

import { ToJsonConfiguration } from '../common/JsonSerializationTypes';
import { RefCountedObjectProfile } from '../utils/ObjectPoolTypes';

/**
Color conversion to apply during image loading.

- `GRAY`:
   Load image as grayscale.
- `COLOR`:
   Load image as 3-channel BGR.
- `ANY_COLOR`:
   Load image as 3-channel BGR if the source color space is color, otherwise as single-channel grayscale. Alpha channel is discarded.
- `UNCHANGED`:
   Load image as-is, preserving the alpha channel if it exists.
*/

/** @hidden */
export const ColorConversionValues = ['GRAY', 'COLOR', 'ANY_COLOR', 'UNCHANGED'];

/**
Canvas color to use when converting images with alpha channel to images without alpha channel.

- `WHITE`:
   White canvas color.
- `BLACK`:
   Black canvas color.
*/

/** @hidden */
export const CanvasColorValues = ['WHITE', 'BLACK'];

/**
Image Info.
*/
export class ImageInfo {
  /**
    Image height in pixels.
    */

  /**
    Image width in pixels.
    */

  /**
    Byte size of the non-hibernating image.
    */

  /** @param source {@displayType `DeepPartial<ImageInfo>`} */
  constructor(source = {}) {
    if (source.height !== undefined) {
      this.height = source.height;
    } else {
      throw new Error('height must be present in constructor argument');
    }
    if (source.width !== undefined) {
      this.width = source.width;
    } else {
      throw new Error('width must be present in constructor argument');
    }
    if (source.maxByteSize !== undefined) {
      this.maxByteSize = source.maxByteSize;
    } else {
      throw new Error('maxByteSize must be present in constructor argument');
    }
  }
  serialize(config = new ToJsonConfiguration()) {
    return {
      height: this.height,
      width: this.width,
      maxByteSize: this.maxByteSize
    };
  }
}

/**
Image Ref Path Load Mode.

- `EAGER`:
   Image is immediately loaded into memory.
- `LAZY`:
   Image is loaded into memory the first time it's requested. Specified path must exist at the time of loading.
- `LAZY_WITH_COPY`:
   Image is loaded into memory the first time it's requested. Specified path is copied into internal directory and so isn't required to exist at the time of loading.
*/

/** @hidden */
export const PathLoadModeValues = ['EAGER', 'LAZY', 'LAZY_WITH_COPY'];

/**
Image Ref Buffer Load Mode.

- `EAGER`:
   Image is immediately decoded.
- `LAZY`:
   Image is decoded the first time it's requested.
*/

/** @hidden */
export const BufferLoadModeValues = ['EAGER', 'LAZY'];

/**
Image Encoding Format.

- `JPEG`:
   JPEG encoding.
- `PNG`:
   PNG encoding.
*/

/** @hidden */
export const EncodingFormatValues = ['JPEG', 'PNG'];

/**
Encryption/Decryption mode to be used when an image is loaded/saved.

- `REQUIRED`:
   Encryption/Decryption is used when loading/saving an image. A crypting provider must first be set using `ImageRef.setStorageCryptingProvider`. If no provider is set, the loading/saving operation will fail.
- `DISABLED`:
   Encryption is not used when loading or saving images. Trying to load an encrypted image will fail.
- `AUTO`:
   When saving, images are encrypted only if a crypting provider is set using `ImageRef.setStorageCryptingProvider`.
   When loading, images will be decrypted if they're encrypted and a crypting provider is set, and loaded as-is otherwise.
*/

/** @hidden */
export const EncryptionModeValues = ['REQUIRED', 'DISABLED', 'AUTO'];

/**
Basic options for loading image.
*/
export class BasicImageLoadOptions {
  /**
    If the rect is not empty, the image will be cropped to this rect before processing.
    */
  cropRect = {
    x: 0,
    y: 0,
    width: 0,
    height: 0
  };
  /**
    CanvasColor color to use when converting images with alpha channel to images without alpha channel.
    
    Default is WHITE
    */
  canvasColor = 'WHITE';

  /** @param source {@displayType `DeepPartial<BasicImageLoadOptions>`} */
  constructor(source = {}) {
    if (source.cropRect !== undefined) {
      this.cropRect = {
        x: source.cropRect.x,
        y: source.cropRect.y,
        width: source.cropRect.width,
        height: source.cropRect.height
      };
    }
    if (source.canvasColor !== undefined) {
      this.canvasColor = source.canvasColor;
    }
  }
}

/**
Options for loading images that come from sensor.
*/
export class RawImageLoadOptions {
  /**
    If the rect is not empty, the image will be cropped to this rect before processing.
    */
  cropRect = {
    x: 0,
    y: 0,
    width: 0,
    height: 0
  };
  /**
    Location of the image origin in the image coordinate system. 
    During loading images are flipped such that after the flip their new origin is in their top-left corner. 
    For example, an image coming from the front camera might have its origin set to the top-right, which will flip it along its vertical axis to create a mirror effect.
    
    The origin is interpreted net of the image orientation. The origin rotates together with the image. The origin value is interpreted and the resulting flip executed AFTER the image matrix has been reoriented to have a neutral orientation.
    
    Default is TOP_LEFT
    */
  origin = 'TOP_LEFT';
  /**
    CanvasColor color to use when converting images with alpha channel to images without alpha channel.
    
    Default is WHITE
    */
  canvasColor = 'WHITE';
  /**
    Rotation that should be applied to the image to recover correct orientation. Is applied before cropping.
    
    Default is NONE
    */
  orientation = 'NONE';

  /** @param source {@displayType `DeepPartial<RawImageLoadOptions>`} */
  constructor(source = {}) {
    if (source.cropRect !== undefined) {
      this.cropRect = {
        x: source.cropRect.x,
        y: source.cropRect.y,
        width: source.cropRect.width,
        height: source.cropRect.height
      };
    }
    if (source.origin !== undefined) {
      this.origin = source.origin;
    }
    if (source.canvasColor !== undefined) {
      this.canvasColor = source.canvasColor;
    }
    if (source.orientation !== undefined) {
      this.orientation = source.orientation;
    }
  }
}

/**
Options for loading image from path.
*/
export class PathImageLoadOptions {
  /**
    If the rect is not empty, the image will be cropped to this rect before processing.
    */
  cropRect = {
    x: 0,
    y: 0,
    width: 0,
    height: 0
  };
  /**
    Color conversion to apply during image loading.
    
    Default is ANY_COLOR
    */
  colorConversion = 'ANY_COLOR';
  /**
    CanvasColor color to use when converting images with alpha channel to images without alpha channel.
    
    Default is WHITE
    */
  canvasColor = 'WHITE';
  /**
    Load mode.
    
    Default is EAGER
    */
  loadMode = 'EAGER';
  /**
    Encryption mode.
    
    Default is AUTO
    */
  encryptionMode = 'AUTO';

  /** @param source {@displayType `DeepPartial<PathImageLoadOptions>`} */
  constructor(source = {}) {
    if (source.cropRect !== undefined) {
      this.cropRect = {
        x: source.cropRect.x,
        y: source.cropRect.y,
        width: source.cropRect.width,
        height: source.cropRect.height
      };
    }
    if (source.colorConversion !== undefined) {
      this.colorConversion = source.colorConversion;
    }
    if (source.canvasColor !== undefined) {
      this.canvasColor = source.canvasColor;
    }
    if (source.loadMode !== undefined) {
      this.loadMode = source.loadMode;
    }
    if (source.encryptionMode !== undefined) {
      this.encryptionMode = source.encryptionMode;
    }
  }
}

/**
Options for loading image from buffer.
*/
export class BufferImageLoadOptions {
  /**
    If the rect is not empty, the image will be cropped to this rect before processing.
    */
  cropRect = {
    x: 0,
    y: 0,
    width: 0,
    height: 0
  };
  /**
    Color conversion to apply during image loading.
    
    Default is ANY_COLOR
    */
  colorConversion = 'ANY_COLOR';
  /**
    CanvasColor color to use when converting images with alpha channel to images without alpha channel.
    
    Default is WHITE
    */
  canvasColor = 'WHITE';
  /**
    Load mode.
    
    Default is EAGER
    */
  loadMode = 'EAGER';

  /** @param source {@displayType `DeepPartial<BufferImageLoadOptions>`} */
  constructor(source = {}) {
    if (source.cropRect !== undefined) {
      this.cropRect = {
        x: source.cropRect.x,
        y: source.cropRect.y,
        width: source.cropRect.width,
        height: source.cropRect.height
      };
    }
    if (source.colorConversion !== undefined) {
      this.colorConversion = source.colorConversion;
    }
    if (source.canvasColor !== undefined) {
      this.canvasColor = source.canvasColor;
    }
    if (source.loadMode !== undefined) {
      this.loadMode = source.loadMode;
    }
  }
}

/**
Options for saving image to a path.
*/
export class SaveImageOptions {
  /**
    Quality parameter is for jpeg only and is in range 0 to 100. If -1, then settings from hibernation are used.
    In case when the Image Ref was created with a lazy load mode and originally has the same format as the requested to save,
    then setting quality to -1 leads to simply copying from source to destination,
    which is time efficient and prevents quality loss caused by jpeg encoding/decoding.
    
    Default is -1
    */
  quality = -1;
  /**
    Encryption mode.
    
    Default is AUTO
    */
  encryptionMode = 'AUTO';

  /** @param source {@displayType `DeepPartial<SaveImageOptions>`} */
  constructor(source = {}) {
    if (source.quality !== undefined) {
      this.quality = source.quality;
    }
    if (source.encryptionMode !== undefined) {
      this.encryptionMode = source.encryptionMode;
    }
  }
}

/**
Options for encoding image.
*/
export class EncodeImageOptions {
  /**
    Quality parameter is for jpeg only and is in range 0 to 100. If -1, then settings from hibernation are used.
    In case when the Image Ref was created with a lazy load mode and originally has the same format as the requested to save,
    then setting quality to -1 leads to simply copying from source to destination,
    which is time efficient and prevents quality loss caused by jpeg encoding/decoding.
    
    Default is -1
    */
  quality = -1;
  /**
    Image format.
    
    Default is JPEG
    */
  format = 'JPEG';

  /** @param source {@displayType `DeepPartial<EncodeImageOptions>`} */
  constructor(source = {}) {
    if (source.quality !== undefined) {
      this.quality = source.quality;
    }
    if (source.format !== undefined) {
      this.format = source.format;
    }
  }
}

/**
The type of source which originated the underlying image.

- `API`:
   ImageRef was created by a call to an SDK feature, such as a scanner, image processor, etc.
- `PLATFORM_IMAGE`:
   ImageRef was created from a platform image, e.g. Bitmap on Android, UIImage on iOS.
- `CAMERA`:
   ImageRef was created from data from the camera.
- `FILE`:
   ImageRef was created from a file.
- `BUFFER`:
   ImageRef was created from an encoded buffer.
- `OTHER`:
   ImageRef was created from another type, e.g. from custom loader.
*/

/** @hidden */
export const ImageSourceTypeValues = ['API', 'PLATFORM_IMAGE', 'CAMERA', 'FILE', 'BUFFER', 'OTHER'];

/**
Description of source from which the ImageRef was created.
*/
export class ImageSource {
  /**
    Source type from which the ImageRef was created.
    */

  /**
    File from which the ImageRef was created. Non-null only if the image ref was created from file.
    */

  /** @param source {@displayType `DeepPartial<ImageSource>`} */
  constructor(source = {}) {
    if (source.type !== undefined) {
      this.type = source.type;
    } else {
      throw new Error('type must be present in constructor argument');
    }
    if (source.filePath !== undefined) {
      this.filePath = source.filePath != null ? source.filePath : null;
    } else {
      throw new Error('filePath must be present in constructor argument');
    }
  }
  serialize(config = new ToJsonConfiguration()) {
    return {
      type: this.type,
      filePath: this.filePath != null ? this.filePath : null
    };
  }
}

/**
ImageRef profile part specific to image information.
*/
export class ImageProfile {
  /**
    Memory consumption of a memory-backed bitmap. Zero, if the image is hibernating.
    
    Default is 0
    */
  bitmapMemoryConsumption = 0;
  /**
    Memory consumption of the hibernation buffer. Zero, if the image is not hibernating, or is hibernated to a file.
    
    Default is 0
    */
  hibernationMemoryConsumption = 0;

  /** @param source {@displayType `DeepPartial<ImageProfile>`} */
  constructor(source = {}) {
    if (source.bitmapMemoryConsumption !== undefined) {
      this.bitmapMemoryConsumption = source.bitmapMemoryConsumption;
    }
    if (source.hibernationMemoryConsumption !== undefined) {
      this.hibernationMemoryConsumption = source.hibernationMemoryConsumption;
    }
  }
  serialize(config = new ToJsonConfiguration()) {
    return {
      bitmapMemoryConsumption: this.bitmapMemoryConsumption,
      hibernationMemoryConsumption: this.hibernationMemoryConsumption
    };
  }
}

/**
ImageRef profile which provides detailed information about stored object.
*/
export class ImageRefProfile {
  /**
    Information about the strong and serialized references to the image.
    */

  /**
    Image-specific information, e.g. memory consumption.
    */

  /**
    Description of source from which the ImageRef was created. Non-null only if detailed profiling has been previously enabled by a call to enableDetailedProfiling.
    */

  /** @param source {@displayType `DeepPartial<ImageRefProfile>`} */
  constructor(source = {}) {
    if (source.refInfo !== undefined) {
      this.refInfo = new RefCountedObjectProfile(source.refInfo);
    } else {
      throw new Error('refInfo must be present in constructor argument');
    }
    if (source.imageInfo !== undefined) {
      this.imageInfo = new ImageProfile(source.imageInfo);
    } else {
      throw new Error('imageInfo must be present in constructor argument');
    }
    if (source.imageSource !== undefined) {
      this.imageSource = source.imageSource != null ? new ImageSource(source.imageSource) : null;
    } else {
      throw new Error('imageSource must be present in constructor argument');
    }
  }
  serialize(config = new ToJsonConfiguration()) {
    return {
      refInfo: this.refInfo.serialize(config),
      imageInfo: this.imageInfo.serialize(config),
      imageSource: this.imageSource != null ? this.imageSource.serialize(config) : null
    };
  }
}

/**
Snapshot of all alive ImageRefs.
*/
export class ImageRefPoolSnapshot {
  /**
    Detailed profiles of all alive ImageRefs.
    */

  /**
    Estimation of total memory consumption of ImageRefs.
    
    Default is 0
    */
  totalMemoryConsumption = 0;

  /** @param source {@displayType `DeepPartial<ImageRefPoolSnapshot>`} */
  constructor(source = {}) {
    if (source.imageRefProfiles !== undefined) {
      this.imageRefProfiles = source.imageRefProfiles.map(it => {
        return new ImageRefProfile(it);
      });
    } else {
      throw new Error('imageRefProfiles must be present in constructor argument');
    }
    if (source.totalMemoryConsumption !== undefined) {
      this.totalMemoryConsumption = source.totalMemoryConsumption;
    }
  }
  serialize(config = new ToJsonConfiguration()) {
    return {
      imageRefProfiles: this.imageRefProfiles.map(it => {
        return it.serialize(config);
      }),
      totalMemoryConsumption: this.totalMemoryConsumption
    };
  }
}

/**
difference between two snapshots.
*/
export class ImageRefPoolSnapshotsDiff {
  /**
    Difference between total memory consumption in two snapshots.
    
    Default is 0
    */
  totalMemoryConsumptionDiff = 0;
  /**
    Unique ids of ImageRefs that are not present in the second snapshot and present in the first snapshot.
    */

  /**
    Profiles of ImageRefs that are present in the second snapshot, but not in the first snapshot.
    */

  /**
    Profiles of ImageRefs that are present in both snapshots but are different. Profile is considered modified if number of references or memory consumption is different.
    */

  /** @param source {@displayType `DeepPartial<ImageRefPoolSnapshotsDiff>`} */
  constructor(source = {}) {
    if (source.totalMemoryConsumptionDiff !== undefined) {
      this.totalMemoryConsumptionDiff = source.totalMemoryConsumptionDiff;
    }
    if (source.removed !== undefined) {
      this.removed = source.removed.map(it => {
        return it;
      });
    } else {
      throw new Error('removed must be present in constructor argument');
    }
    if (source.added !== undefined) {
      this.added = source.added.map(it => {
        return it;
      });
    } else {
      throw new Error('added must be present in constructor argument');
    }
    if (source.modified !== undefined) {
      this.modified = source.modified.map(it => {
        return it;
      });
    } else {
      throw new Error('modified must be present in constructor argument');
    }
  }
  serialize(config = new ToJsonConfiguration()) {
    return {
      totalMemoryConsumptionDiff: this.totalMemoryConsumptionDiff,
      removed: this.removed.map(it => {
        return it;
      }),
      added: this.added.map(it => {
        return it;
      }),
      modified: this.modified.map(it => {
        return it;
      })
    };
  }
}
