// react-native-scanbot-sdk/src/ui_v2/check/CheckScannerUserGuidance.ts
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/check/CheckScannerUserGuidance.yaml

import { BackgroundStyle, StyledText } from '../../ui_v2/common/Common';
import { DeepPartial, PartiallyConstructible } from '../../utils/utils';

/**
Configure the captions of the user guidance hints for different scanning states.
*/
export class CheckScannerGuidanceStates extends PartiallyConstructible {
  /**
    The user guidance text displayed when no check is found. Initial state.
    
    Default is "?checkUserGuidanceNoCheckFound"
    */
  public noCheckFound: string = '?checkUserGuidanceNoCheckFound';
  /**
    The user guidance text displayed when a check presence was detected and scanning is in progress.
    
    Default is "?checkUserGuidanceScanningProgress"
    */
  public scanningProgress: string = '?checkUserGuidanceScanningProgress';
  /**
    The user guidance text displayed when it is too dark to capture an adequate image.
    
    Default is "?checkUserGuidanceTooDark"
    */
  public tooDark: string = '?checkUserGuidanceTooDark';
  /**
    The user guidance text displayed when the detected check is too far away and appears too small.
    
    Default is "?checkUserGuidanceTooSmall"
    */
  public tooSmall: string = '?checkUserGuidanceTooSmall';
  /**
    The user guidance text displayed when the detected check is not in a good perspective (device tilted).
    
    Default is "?checkUserGuidanceBadPerspective"
    */
  public badPerspective: string = '?checkUserGuidanceBadPerspective';

  /** @param source {@displayType `DeepPartial<CheckScannerGuidanceStates>`} */
  public constructor(source: DeepPartial<CheckScannerGuidanceStates> = {}) {
    super();
    if (source.noCheckFound !== undefined) {
      this.noCheckFound = source.noCheckFound;
    }
    if (source.scanningProgress !== undefined) {
      this.scanningProgress = source.scanningProgress;
    }
    if (source.tooDark !== undefined) {
      this.tooDark = source.tooDark;
    }
    if (source.tooSmall !== undefined) {
      this.tooSmall = source.tooSmall;
    }
    if (source.badPerspective !== undefined) {
      this.badPerspective = source.badPerspective;
    }
  }
}

/**
Configuration of the hints guiding users through the check scanning process.
*/
export class CheckScannerGuidanceConfiguration extends PartiallyConstructible {
  /**
    Determines whether the user guidance hints should be visible (enabled).
    
    Default is true
    */
  public visibility: boolean = true;
  /**
    Configure the text style for the user guidance hints.
    */
  public title: StyledText = new StyledText({
    text: '?checkUserGuidanceNoCheckFound',
    color: '?sbColorOnPrimary',
  });
  /**
    Configure the background style for the user guidance hints.
    */
  public background: BackgroundStyle = new BackgroundStyle({
    strokeColor: '#00000000',
    fillColor: '?sbColorSurfaceLow',
  });
  /**
    Configure the captions of the user guidance hints for different states.
    */
  public statesTitles: CheckScannerGuidanceStates = new CheckScannerGuidanceStates({});

  /** @param source {@displayType `DeepPartial<CheckScannerGuidanceConfiguration>`} */
  public constructor(source: DeepPartial<CheckScannerGuidanceConfiguration> = {}) {
    super();
    if (source.visibility !== undefined) {
      this.visibility = source.visibility;
    }
    if (source.title !== undefined) {
      this.title = new StyledText(source.title);
    }
    if (source.background !== undefined) {
      this.background = new BackgroundStyle(source.background);
    }
    if (source.statesTitles !== undefined) {
      this.statesTitles = new CheckScannerGuidanceStates(source.statesTitles);
    }
  }
}
