// react-native-scanbot-sdk/lib/commonjs/ui_v2/vin/VINScannerScreenConfiguration.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VinScannerScreenConfiguration = void 0;
var _ActionBarConfiguration = require("../../ui_v2/common/ActionBarConfiguration");
var _CameraConfiguration = require("../../ui_v2/common/CameraConfiguration");
var _CameraPermission = require("../../ui_v2/common/CameraPermission");
var _Common = require("../../ui_v2/common/Common");
var _ScanCompletionOverlay = require("../../ui_v2/common/ScanCompletionOverlay");
var _ScanbotAlertDialog = require("../../ui_v2/common/ScanbotAlertDialog");
var _TopBarConfiguration = require("../../ui_v2/common/TopBarConfiguration");
var _UserGuidanceConfiguration = require("../../ui_v2/common/UserGuidanceConfiguration");
var _ViewFinderConfiguration = require("../../ui_v2/common/ViewFinderConfiguration");
var _VINScannerIntroScreenConfiguration = require("../../ui_v2/vin/VINScannerIntroScreenConfiguration");
var _VINScannerScreenTextLocalization = require("../../ui_v2/vin/VINScannerScreenTextLocalization");
var _Geometry = require("../../utils/geometry/Geometry");
var _utils = require("../../utils/utils");
var _VinScannerTypes = require("../../vin/VinScannerTypes");
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/vin/VINScannerScreenConfiguration.yaml

/**
Configuration of the screen for detecting VIN data.
*/
class VinScannerScreenConfiguration extends _utils.PartiallyConstructible {
  /**
    Version number of the configuration object.
    
    Default is "1.0"
    */
  version = '1.0';
  /**
    The configuration object should be applied for this screen.
    
    Default is "VINScanner"
    */
  screen = 'VINScanner';
  /**
    Configuration of the all strings for VIN scanner screen.
    */
  localization = new _VINScannerScreenTextLocalization.VinScannerScreenTextLocalization({});
  /**
    Define the screen's base color values from which other colors are derived.
    */
  palette = new _Common.Palette({});
  /**
    The background color of the VIN scanner screen.
    
    Default is "?sbColorSurfaceLow"
    */
  backgroundColor = '?sbColorSurfaceLow';
  /**
    Configuration of the camera behavior.
    */
  cameraConfiguration = new _CameraConfiguration.CameraConfiguration({
    orientationLockMode: 'PORTRAIT'
  });
  /**
    Configuration of the dialog for requesting camera permissions.
    */
  cameraPermission = new _CameraPermission.CameraPermissionScreen({
    statusBarMode: 'DARK',
    background: '?sbColorSurface',
    iconBackground: '?sbColorOutline',
    icon: new _Common.IconStyle({
      visible: true,
      color: '?sbColorOnSurface'
    }),
    enableCameraButton: new _Common.ButtonConfiguration({
      visible: true,
      text: '?cameraPermissionEnableCameraButton',
      accessibilityDescription: '?accessibilityDescriptionCameraPermissionEnableCameraButton',
      background: new _Common.BackgroundStyle({
        strokeColor: '?sbColorPrimary',
        fillColor: '?sbColorPrimary',
        strokeWidth: 0.0
      }),
      foreground: new _Common.ForegroundStyle({
        iconVisible: false,
        color: '?sbColorOnPrimary',
        useShadow: false
      })
    }),
    closeButton: new _Common.ButtonConfiguration({
      visible: true,
      text: '?cameraPermissionCloseButton',
      accessibilityDescription: '?accessibilityDescriptionCameraPermissionCloseButton',
      background: new _Common.BackgroundStyle({
        strokeColor: '#00000000',
        fillColor: '#00000000',
        strokeWidth: 0.0
      }),
      foreground: new _Common.ForegroundStyle({
        iconVisible: false,
        color: '?sbColorPrimary',
        useShadow: false
      })
    }),
    enableCameraTitle: new _Common.StyledText({
      text: '?cameraPermissionEnableCameraTitle',
      color: '?sbColorOnSurface'
    }),
    enableCameraExplanation: new _Common.StyledText({
      text: '?cameraPermissionEnableCameraExplanation',
      color: '?sbColorOnSurfaceVariant'
    })
  });
  /**
    Configuration of the top user guidance.
    */
  topUserGuidance = new _UserGuidanceConfiguration.UserGuidanceConfiguration({
    title: new _Common.StyledText({
      text: '?topUserGuidance'
    })
  });
  /**
    Configuration of the user guidance below finder view.
    */
  finderViewUserGuidance = new _UserGuidanceConfiguration.UserGuidanceConfiguration({
    title: new _Common.StyledText({
      text: '?finderViewUserGuidance'
    })
  });
  /**
    Configuration of the top bar on the VIN scanner screen.
    */
  topBar = new _TopBarConfiguration.TopBarConfiguration({
    title: new _Common.StyledText({
      visible: true,
      text: '?topBarTitle'
    }),
    cancelButton: new _Common.ButtonConfiguration({
      text: '?topBarCancelButton',
      accessibilityDescription: '?accessibilityDescriptionCancelButton',
      background: new _Common.BackgroundStyle({
        strokeColor: '#00000000',
        fillColor: '#00000000',
        strokeWidth: 0.0
      }),
      foreground: new _Common.ForegroundStyle({
        color: '?sbColorOnPrimary'
      })
    })
  });
  /**
    Configuration of the button in the top bar that opens the introduction screen.
    */
  topBarOpenIntroScreenButton = new _Common.IconButton({
    color: '?sbColorOnPrimary',
    accessibilityDescription: '?accessibilityDescriptionOpenIntroScreenButton'
  });
  /**
    Configuration of the introduction screen for the VIN scanner screen.
    */
  introScreen = new _VINScannerIntroScreenConfiguration.VinScannerIntroScreenConfiguration({});
  /**
    Configuration of the action bar.
    */
  actionBar = new _ActionBarConfiguration.ActionBarConfiguration({
    flipCameraButton: new _Common.RoundButton({
      visible: false,
      accessibilityDescription: '?accessibilityDescriptionFlipCameraButton',
      backgroundColor: '?sbColorSurfaceHigh',
      foregroundColor: '?sbColorOnPrimary',
      activeBackgroundColor: '?sbColorWarning',
      activeForegroundColor: '#1C1B1F'
    })
  });
  /**
    Configuration of the view finder.
    */
  viewFinder = new _ViewFinderConfiguration.PermanentViewFinderConfiguration({
    style: new _ViewFinderConfiguration.FinderCorneredStyle({
      strokeColor: '?sbColorSurface',
      strokeWidth: 2.0
    }),
    aspectRatio: new _Geometry.AspectRatio({
      width: 5.0,
      height: 1.0
    }),
    preferredHeight: 56.0
  });
  /**
    Configuration of the scanner.
    */
  scannerConfiguration = new _VinScannerTypes.VinScannerConfiguration({});
  /**
    Whether word boxes should be displayed or not.
    
    Default is true
    */
  shouldShowWordBoxes = true;
  /**
    The color of word boxes filling. Works best with an alpha < 0.5.
    
    Default is "#33CCE54C"
    */
  wordBoxesFillColor = '#33CCE54C';
  /**
    The color of word boxes border lines.
    
    Default is "#33CCE599"
    */
  wordBoxesLineColor = '#33CCE599';
  /**
    If the confirmation alert dialog is enabled.
    
    Default is false
    */
  confirmationAlertDialogEnabled = false;
  /**
    Configuration of the confirmation alert dialog.
    */
  confirmationAlertDialog = new _ScanbotAlertDialog.ScanbotAlertDialog({
    title: new _Common.StyledText({
      text: '?vinConfirmationAlertTitle',
      color: '?sbColorOnSurface'
    }),
    subtitle: new _Common.StyledText({
      color: '?sbColorOnSurfaceVariant'
    }),
    sheetColor: '?sbColorSurface',
    modalOverlayColor: '?sbColorModalOverlay',
    dividerColor: '?sbColorOutline',
    okButton: new _Common.ButtonConfiguration({
      visible: true,
      text: '?vinConfirmationAlertSubmitButton',
      accessibilityDescription: '?accessibilityDescriptionConfirmationSubmitButton',
      background: new _Common.BackgroundStyle({
        strokeColor: '?sbColorPrimary',
        fillColor: '?sbColorPrimary',
        strokeWidth: 1.0
      }),
      foreground: new _Common.ForegroundStyle({
        iconVisible: true,
        color: '?sbColorOnPrimary',
        useShadow: false
      })
    }),
    cancelButton: new _Common.ButtonConfiguration({
      visible: true,
      text: '?vinConfirmationAlertCancelButton',
      accessibilityDescription: '?accessibilityDescriptionConfirmationCancelButton',
      background: new _Common.BackgroundStyle({
        strokeColor: '#00000000',
        fillColor: '#00000000',
        strokeWidth: 1.0
      }),
      foreground: new _Common.ForegroundStyle({
        iconVisible: false,
        color: '?sbColorPrimary',
        useShadow: false
      })
    })
  });
  /**
    Configuration of the result overlay.
    */
  successOverlay = new _ScanCompletionOverlay.ScanCompletionOverlay({});
  /**
    Configuration of the scan confirmation sound.
    */
  sound = new _Common.Sound({});
  /**
    Configuration of the vibration feedback.
    */
  vibration = new _Common.Vibration({});
  /**
    If enabled, the screen will not turn off while the barcode scanner is active. (Android only).
    
    Default is true
    */
  keepScreenOn = true;

  /** @param source {@displayType `DeepPartial<VinScannerScreenConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.version !== undefined) {
      this.version = source.version;
    }
    if (source.screen !== undefined) {
      this.screen = source.screen;
    }
    if (source.localization !== undefined) {
      this.localization = new _VINScannerScreenTextLocalization.VinScannerScreenTextLocalization(source.localization);
    }
    if (source.palette !== undefined) {
      this.palette = new _Common.Palette(source.palette);
    }
    if (source.backgroundColor !== undefined) {
      this.backgroundColor = source.backgroundColor;
    }
    if (source.cameraConfiguration !== undefined) {
      this.cameraConfiguration = new _CameraConfiguration.CameraConfiguration(source.cameraConfiguration);
    }
    if (source.cameraPermission !== undefined) {
      this.cameraPermission = new _CameraPermission.CameraPermissionScreen(source.cameraPermission);
    }
    if (source.topUserGuidance !== undefined) {
      this.topUserGuidance = new _UserGuidanceConfiguration.UserGuidanceConfiguration(source.topUserGuidance);
    }
    if (source.finderViewUserGuidance !== undefined) {
      this.finderViewUserGuidance = new _UserGuidanceConfiguration.UserGuidanceConfiguration(source.finderViewUserGuidance);
    }
    if (source.topBar !== undefined) {
      this.topBar = new _TopBarConfiguration.TopBarConfiguration(source.topBar);
    }
    if (source.topBarOpenIntroScreenButton !== undefined) {
      this.topBarOpenIntroScreenButton = new _Common.IconButton(source.topBarOpenIntroScreenButton);
    }
    if (source.introScreen !== undefined) {
      this.introScreen = new _VINScannerIntroScreenConfiguration.VinScannerIntroScreenConfiguration(source.introScreen);
    }
    if (source.actionBar !== undefined) {
      this.actionBar = new _ActionBarConfiguration.ActionBarConfiguration(source.actionBar);
    }
    if (source.viewFinder !== undefined) {
      this.viewFinder = new _ViewFinderConfiguration.PermanentViewFinderConfiguration(source.viewFinder);
    }
    if (source.scannerConfiguration !== undefined) {
      this.scannerConfiguration = new _VinScannerTypes.VinScannerConfiguration(source.scannerConfiguration);
    }
    if (source.shouldShowWordBoxes !== undefined) {
      this.shouldShowWordBoxes = source.shouldShowWordBoxes;
    }
    if (source.wordBoxesFillColor !== undefined) {
      this.wordBoxesFillColor = source.wordBoxesFillColor;
    }
    if (source.wordBoxesLineColor !== undefined) {
      this.wordBoxesLineColor = source.wordBoxesLineColor;
    }
    if (source.confirmationAlertDialogEnabled !== undefined) {
      this.confirmationAlertDialogEnabled = source.confirmationAlertDialogEnabled;
    }
    if (source.confirmationAlertDialog !== undefined) {
      this.confirmationAlertDialog = new _ScanbotAlertDialog.ScanbotAlertDialog(source.confirmationAlertDialog);
    }
    if (source.successOverlay !== undefined) {
      this.successOverlay = new _ScanCompletionOverlay.ScanCompletionOverlay(source.successOverlay);
    }
    if (source.sound !== undefined) {
      this.sound = new _Common.Sound(source.sound);
    }
    if (source.vibration !== undefined) {
      this.vibration = new _Common.Vibration(source.vibration);
    }
    if (source.keepScreenOn !== undefined) {
      this.keepScreenOn = source.keepScreenOn;
    }
  }
}
exports.VinScannerScreenConfiguration = VinScannerScreenConfiguration;
