// react-native-scanbot-sdk/src/documents/CreditCardDocumentModel.ts
/// Auto-generated. Modifications will be overwritten.
/// Generated from core/document-schemas/CreditCardDocumentModel.yaml

import { Field, GenericDocument } from './GenericDocument';

/// All possible root document types.
export type CreditCardDocumentModelRootType = typeof CreditCardDocumentType;

export const CreditCardDocumentType = 'CreditCard';

/** Credit Card Document */
export class CreditCard {
  private _document: GenericDocument;

  get document(): GenericDocument {
    return this._document;
  }

  constructor(document: GenericDocument) {
    if (document.type.name !== this.requiredDocumentType()) {
      throw new Error(
        `Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`
      );
    }

    this._document = document;
  }

  requiredDocumentType(): string {
    return CreditCardDocumentType;
  }

  /** The card number of the recognized credit card */
  get cardNumber(): Field {
    return this.document.fields.find((f: Field) => f.type.name === 'CardNumber') as Field;
  }

  /** The cardholder name of the recognized credit card */
  get cardholderName(): Field | undefined {
    return this.document.fields.find((f: Field) => f.type.name === 'CardholderName');
  }

  /** The expiry date of the recognized credit card */
  get expiryDate(): Field | undefined {
    return this.document.fields.find((f: Field) => f.type.name === 'ExpiryDate');
  }
}
