// react-native-scanbot-barcode-scanner-sdk/lib/module/types/base/errors.js
export class SBError extends Error {
  constructor(message) {
    super(message);
  }
}
export class InvalidLicenseError extends SBError {
  type = 'InvalidLicense';
  constructor(message) {
    super(message);
  }
}
export class NullPointerError extends SBError {
  type = 'NullPointer';
  constructor(message) {
    super(message);
  }
}
export class InvalidArgumentError extends SBError {
  type = 'InvalidArgument';
  constructor(message) {
    super(message);
  }
}
export class InvalidImageRefError extends SBError {
  type = 'InvalidImageRef';
  constructor(message) {
    super(message);
  }
}
export class ComponentUnavailableError extends SBError {
  type = 'ComponentUnavailable';
  constructor(message) {
    super(message);
  }
}
export class IllegalStateError extends SBError {
  type = 'IllegalState';
  constructor(message) {
    super(message);
  }
}
export class IOError extends SBError {
  type = 'IOError';
  constructor(message) {
    super(message);
  }
}
export class InvalidDataError extends SBError {
  type = 'InvalidData';
  constructor(message) {
    super(message);
  }
}
export class OutOfMemoryError extends SBError {
  type = 'OutOfMemory';
  constructor(message) {
    super(message);
  }
}
export class TimeoutError extends SBError {
  type = 'Timeout';
  constructor(message) {
    super(message);
  }
}
export class UnknownError extends SBError {
  type = 'Unknown';
  constructor(message) {
    super(message);
  }
}
export class ProcessError extends SBError {
  type = 'ProcessError';
  constructor(message, code) {
    super(message);
    this.code = code;
  }
}

/**
 * @internal
 * @hidden
 */
export function createSBError(err) {
  const code = Number(err.code);
  const message = err.message || err.errorMessage || 'An unknown error occurred';
  switch (code) {
    case 1:
      return new UnknownError(message);
    case 2:
      return new InvalidLicenseError(message);
    case 3:
      return new NullPointerError(message);
    case 4:
      return new InvalidArgumentError(message);
    case 5:
      return new InvalidImageRefError(message);
    case 6:
      return new ComponentUnavailableError(message);
    case 7:
      return new IllegalStateError(message);
    case 8:
      return new IOError(message);
    case 9:
      return new InvalidDataError(message);
    case 11:
      return new OutOfMemoryError(message);
    case 12:
      return new TimeoutError(message);
    default:
      {
        if (code >= 100) {
          return new ProcessError(message, code);
        }
        return new UnknownError(message);
      }
  }
}
