// react-native-scanbot-sdk/lib/commonjs/ui_v2/creditcard/CreditCardScannerUIResult.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CreditCardScannerUiResult = void 0;
var _GenericDocument = require("../../documents/GenericDocument");
var _JsonSerializationTypes = require("../../utils/json/JsonSerializationTypes");
var _utils = require("../../utils/utils");
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/creditcard/CreditCardScannerUIResult.yaml

/**
Configuration of the credit card scanner RTU UI result.
*/
class CreditCardScannerUiResult extends _utils.PartiallyConstructible {
  /**
    The status of the credit card recognition step.
    */

  /**
    Generic document containing credit card data. Not present, if status is FAIL.
    */

  /** @param source {@displayType `DeepPartial<CreditCardScannerUiResult>`} */
  constructor(source = {}) {
    super();
    if (source.recognitionStatus !== undefined) {
      this.recognitionStatus = source.recognitionStatus;
    } else {
      throw new Error('recognitionStatus must be present in constructor argument');
    }
    if (source.creditCard !== undefined) {
      this.creditCard = source.creditCard != null ? new _GenericDocument.GenericDocument(source.creditCard) : null;
    } else {
      throw new Error('creditCard must be present in constructor argument');
    }
  }
  async serialize(config = new _JsonSerializationTypes.ToJsonConfiguration()) {
    return {
      recognitionStatus: this.recognitionStatus,
      creditCard: this.creditCard != null ? await this.creditCard.serialize(config) : null
    };
  }
  _released = false;
  release() {
    if (this._released) {
      return;
    }
    {
      if (this.creditCard != null) {
        this.creditCard.release();
      }
    }
    this._released = true;
  }
  async encodeImages() {
    if (this.creditCard != null) {
      await this.creditCard.encodeImages();
    }
  }
}
exports.CreditCardScannerUiResult = CreditCardScannerUiResult;
