// react-native-scanbot-sdk/lib/commonjs/ui_v2/creditcard/CreditCardScannerIntroScreenConfiguration.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CreditCardScannerIntroScreenConfiguration = exports.CreditCardScannerIntroImage = exports.CreditCardNoImage = exports.CreditCardIntroTwoSidesImage = exports.CreditCardIntroOneSideImage = exports.CreditCardIntroCustomImage = void 0;
var _Common = require("../../ui_v2/common/Common");
var _utils = require("../../utils/utils");
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/creditcard/CreditCardScannerIntroScreenConfiguration.yaml
/**
The image for the introduction screen of a credit card scanner screen.
*/
/** @internal */
let CreditCardScannerIntroImage = exports.CreditCardScannerIntroImage = void 0;
(function (_CreditCardScannerIntroImage) {
  function From(source) {
    const _type = source._type;
    switch (_type) {
      case 'CreditCardIntroOneSideImage':
        return new CreditCardIntroOneSideImage(source);
      case 'CreditCardIntroTwoSidesImage':
        return new CreditCardIntroTwoSidesImage(source);
      case 'CreditCardNoImage':
        return new CreditCardNoImage(source);
      case 'CreditCardIntroCustomImage':
        return new CreditCardIntroCustomImage(source);
      default:
        throw `Unknown child class name: ${_type}`;
    }
  }
  _CreditCardScannerIntroImage.From = From;
})(CreditCardScannerIntroImage || (exports.CreditCardScannerIntroImage = CreditCardScannerIntroImage = {}));
/**
The image for the introduction screen with one side of a credit card.
*/
class CreditCardIntroOneSideImage extends _utils.PartiallyConstructible {
  _type = 'CreditCardIntroOneSideImage';

  /** @param source {@displayType `DeepPartial<CreditCardIntroOneSideImage>`} */
  constructor(source = {}) {
    super();
  }
}

/**
The image for the introduction screen with two sides of a credit card.
*/
exports.CreditCardIntroOneSideImage = CreditCardIntroOneSideImage;
class CreditCardIntroTwoSidesImage extends _utils.PartiallyConstructible {
  _type = 'CreditCardIntroTwoSidesImage';

  /** @param source {@displayType `DeepPartial<CreditCardIntroTwoSidesImage>`} */
  constructor(source = {}) {
    super();
  }
}

/**
No image for the introduction screen.
*/
exports.CreditCardIntroTwoSidesImage = CreditCardIntroTwoSidesImage;
class CreditCardNoImage extends _utils.PartiallyConstructible {
  _type = 'CreditCardNoImage';

  /** @param source {@displayType `DeepPartial<CreditCardNoImage>`} */
  constructor(source = {}) {
    super();
  }
}

/**
A custom image for the introduction screen.
*/
exports.CreditCardNoImage = CreditCardNoImage;
class CreditCardIntroCustomImage extends _utils.PartiallyConstructible {
  _type = 'CreditCardIntroCustomImage';
  /**
    The web or file URI to the image.
    */

  /** @param source {@displayType `DeepPartial<CreditCardIntroCustomImage>`} */
  constructor(source = {}) {
    super();
    if (source.uri !== undefined) {
      this.uri = source.uri;
    } else {
      throw new Error('uri must be present in constructor argument');
    }
  }
}

/**
Configuration of the introduction screen for the credit card scanner.
*/
exports.CreditCardIntroCustomImage = CreditCardIntroCustomImage;
class CreditCardScannerIntroScreenConfiguration extends _utils.PartiallyConstructible {
  /**
    The image for the introduction screen.
    */
  image = new CreditCardIntroOneSideImage({});
  /**
    The background color of the introduction screen.
    
    Default is "?sbColorSurface"
    */
  backgroundColor = '?sbColorSurface';
  /**
    The divider color of the introduction screen.
    
    Default is "?sbColorOutline"
    */
  dividerColor = '?sbColorOutline';
  /**
    The handle color of the introduction screen.
    
    Default is "?sbColorOutline"
    */
  handlerColor = '?sbColorOutline';
  /**
    Determines whether the introduction screen should automatically be shown or not when the scanning session starts.
    
    Default is false
    */
  showAutomatically = false;
  /**
    The title of the introduction screen, located in the top bar.
    */
  title = new _Common.StyledText({
    text: '?introScreenTitle',
    color: '?sbColorOnSurface'
  });
  /**
    The text explanation of the introduction screen.
    */
  explanation = new _Common.StyledText({
    text: '?introScreenText',
    color: '?sbColorOnSurface'
  });
  /**
    Configuration of the 'Done' / 'Start scanning' button.
    */
  doneButton = new _Common.ButtonConfiguration({
    text: '?introScreenDoneButton',
    accessibilityDescription: '?accessibilityDescriptionIntroScreenDoneButton',
    background: new _Common.BackgroundStyle({
      strokeColor: '#00000000',
      fillColor: '?sbColorPrimary',
      strokeWidth: 0.0
    }),
    foreground: new _Common.ForegroundStyle({
      color: '?sbColorOnPrimary'
    })
  });

  /** @param source {@displayType `DeepPartial<CreditCardScannerIntroScreenConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.image !== undefined) {
      this.image = CreditCardScannerIntroImage.From(source.image);
    }
    if (source.backgroundColor !== undefined) {
      this.backgroundColor = source.backgroundColor;
    }
    if (source.dividerColor !== undefined) {
      this.dividerColor = source.dividerColor;
    }
    if (source.handlerColor !== undefined) {
      this.handlerColor = source.handlerColor;
    }
    if (source.showAutomatically !== undefined) {
      this.showAutomatically = source.showAutomatically;
    }
    if (source.title !== undefined) {
      this.title = new _Common.StyledText(source.title);
    }
    if (source.explanation !== undefined) {
      this.explanation = new _Common.StyledText(source.explanation);
    }
    if (source.doneButton !== undefined) {
      this.doneButton = new _Common.ButtonConfiguration(source.doneButton);
    }
  }
}
exports.CreditCardScannerIntroScreenConfiguration = CreditCardScannerIntroScreenConfiguration;
