// react-native-scanbot-sdk/src/components/document-scanner-view/ScanbotDocumentScannerView.tsx
import React, { forwardRef, useCallback, useImperativeHandle, useRef } from 'react';
import { NativeSyntheticEvent, Platform, StyleSheet, View } from 'react-native';
import { DocumentDetectionStatus } from '../../document_scanner';
import NativeDocumentScannerView, {
  Commands,
  DocumentScannerEvent,
} from '../spec/ScanbotDocumentScannerViewNativeComponent';
import {
  ScanbotDocumentScannerViewHandle,
  ScanbotDocumentScannerViewProperties,
} from './ScanbotDocumentScannerViewProperties';

export const ScanbotDocumentScannerView = forwardRef<
  ScanbotDocumentScannerViewHandle,
  ScanbotDocumentScannerViewProperties
>(({ onDocumentScannerResult, onDetectionResult, ...props }, forwardedRef) => {
  const viewRef = useRef(null);

  useImperativeHandle(forwardedRef, () => {
    return {
      freezeCamera() {
        if (viewRef.current) {
          Commands.freezeCamera(viewRef.current);
        }
      },

      unfreezeCamera() {
        if (viewRef.current) {
          Commands.unfreezeCamera(viewRef.current);
        }
      },

      snapDocument(acquireFocus?: boolean) {
        if (viewRef.current) {
          Commands.snapDocument(viewRef.current, acquireFocus === true);
        }
      },
    };
  }, []);

  const _onDocumentScannerResult = useCallback(
    (event: NativeSyntheticEvent<DocumentScannerEvent>) => {
      const result = event.nativeEvent.result;
      if (result) {
        onDocumentScannerResult(result);
      }
    },
    [onDocumentScannerResult]
  );

  const _onDetectionResult = useCallback(
    (event: NativeSyntheticEvent<DocumentScannerEvent>) => {
      if (event.nativeEvent.result) {
        onDetectionResult &&
          onDetectionResult(
            (Platform.OS === 'ios'
              ? JSON.parse(event.nativeEvent.result)
              : event.nativeEvent.result) as DocumentDetectionStatus
          );
      }
    },
    [onDetectionResult]
  );

  return (
    <View style={[styles.container, props.style]}>
      <NativeDocumentScannerView
        ref={viewRef}
        style={styles.container}
        onDocumentScannerResult={_onDocumentScannerResult}
        onDetectionResult={_onDetectionResult}
        acceptedAngleScore={props.acceptedAngleScore}
        acceptedBrightnessThreshold={props.acceptedBrightnessThreshold}
        acceptedSizeScore={props.acceptedSizeScore}
        autoSnappingEnabled={props.autoSnappingEnabled}
        autoSnappingSensitivity={props.autoSnappingSensitivity}
        autoSnappingDelay={props.autoSnappingDelay}
        cameraModule={props.cameraModule}
        cameraPreviewMode={props.cameraPreviewMode}
        flashEnabled={props.flashEnabled}
        ignoreOrientationMismatch={props.ignoreOrientationMismatch}
        photoQualityPrioritization={props.photoQualityPrioritization}
        touchToFocusEnabled={props.touchToFocusEnabled}
        hardwareButtonsEnabled={props.hardwareButtonsEnabled}
        requiredAspectRatios={props.requiredAspectRatios}
        finderEnabled={props.finderEnabled}
        finderLineColor={props.finderLineColor}
        finderLineWidth={props.finderLineWidth}
        finderOverlayColor={props.finderOverlayColor}
        finderCornerRadius={props.finderCornerRadius}
        finderMinimumPadding={props.finderMinimumPadding}
        finderAspectRatio={props.finderAspectRatio}
        polygonEnabled={props.polygonEnabled}
        polygonBackgroundColor={props.polygonBackgroundColor}
        polygonBackgroundColorOK={props.polygonBackgroundColorOK}
        polygonColor={props.polygonColor}
        polygonColorOK={props.polygonColorOK}
        polygonLineWidth={props.polygonLineWidth}
        polygonCornerRadius={props.polygonCornerRadius}
        polygonAutoSnapProgressColor={props.polygonAutoSnapProgressColor}
        polygonAutoSnapProgressLineWidth={props.polygonAutoSnapProgressLineWidth}
        polygonAutoSnapProgressEnabled={props.polygonAutoSnapProgressEnabled}
      />
    </View>
  );
});

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});
