// react-native-scanbot-barcode-scanner-sdk/lib/commonjs/types/core/generic_document/BarcodeDocumentModel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VCard = exports.SwissQR = exports.SEPA = exports.MedicalCertificate = exports.IDCardPDF417 = exports.HIBC = exports.GS1 = exports.DEMedicalPlan = exports.BritishColumbiaDriverLicense = exports.BoardingPass = exports.AAMVA = void 0;
/// Auto-generated. Modifications will be overwritten.
/// Generated from core/document-schemas/BarcodeDocumentModel.yaml

/** Boarding Pass */
class BoardingPass {
  get document() {
    return this._document;
  }
  static DOCUMENT_TYPE = 'BoardingPass';
  static ELECTRONIC_TICKET_INDICATOR_FIELD_NAME = 'ElectronicTicketIndicator';
  static ELECTRONIC_TICKET_INDICATOR_FIELD_NAME_NORMALIZED = 'BoardingPass.ElectronicTicketIndicator';
  static FORMAT_CODE_FIELD_NAME = 'FormatCode';
  static FORMAT_CODE_FIELD_NAME_NORMALIZED = 'BoardingPass.FormatCode';
  static NUMBER_OF_LEGS_FIELD_NAME = 'NumberOfLegs';
  static NUMBER_OF_LEGS_FIELD_NAME_NORMALIZED = 'BoardingPass.NumberOfLegs';
  static PASSENGER_NAME_FIELD_NAME = 'PassengerName';
  static PASSENGER_NAME_FIELD_NAME_NORMALIZED = 'BoardingPass.PassengerName';
  static SECURITY_DATA_FIELD_NAME = 'SecurityData';
  static SECURITY_DATA_FIELD_NAME_NORMALIZED = 'BoardingPass.SecurityData';
  static SECURITY_DATA_TYPE_FIELD_NAME = 'SecurityDataType';
  static SECURITY_DATA_TYPE_FIELD_NAME_NORMALIZED = 'BoardingPass.SecurityDataType';
  constructor(document) {
    if (document.type.name !== this.requiredDocumentType()) {
      throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
    }
    this._document = document;
  }
  requiredDocumentType() {
    return BoardingPass.DOCUMENT_TYPE;
  }

  /** Electronic Ticket */
  get electronicTicketIndicator() {
    return this.document.fields.find(f => f.type.name === 'ElectronicTicketIndicator');
  }

  /** Format Code. */
  get formatCode() {
    return this.document.fields.find(f => f.type.name === 'FormatCode');
  }

  /** Number of encoded legs. */
  get numberOfLegs() {
    return this.document.fields.find(f => f.type.name === 'NumberOfLegs');
  }

  /** Passenger Name */
  get passengerName() {
    return this.document.fields.find(f => f.type.name === 'PassengerName');
  }

  /** Security Data */
  get securityData() {
    return this.document.fields.find(f => f.type.name === 'SecurityData');
  }

  /** Security Data Type */
  get securityDataType() {
    return this.document.fields.find(f => f.type.name === 'SecurityDataType');
  }

  /** An array of all children of type "Leg". */
  get legs() {
    return this.document.children.filter(c => c.type.name === 'Leg').map(c => new BoardingPass.Leg(c));
  }
}
exports.BoardingPass = BoardingPass;
(function (_BoardingPass) {
  class Leg {
    get document() {
      return this._document;
    }
    static DOCUMENT_TYPE = 'Leg';
    static AIRLINE_DESIGNATOR_OF_BOARDING_PASS_ISSUER_FIELD_NAME = 'AirlineDesignatorOfBoardingPassIssuer';
    static AIRLINE_DESIGNATOR_OF_BOARDING_PASS_ISSUER_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.AirlineDesignatorOfBoardingPassIssuer';
    static AIRLINE_NUMERIC_CODE_FIELD_NAME = 'AirlineNumericCode';
    static AIRLINE_NUMERIC_CODE_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.AirlineNumericCode';
    static BAGGAGE_TAG_LICENSE_PLATE_NUMBERS_FIELD_NAME = 'BaggageTagLicensePlateNumbers';
    static BAGGAGE_TAG_LICENSE_PLATE_NUMBERS_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.BaggageTagLicensePlateNumbers';
    static CHECK_IN_SEQUENCE_NUMBER_FIELD_NAME = 'CheckInSequenceNumber';
    static CHECK_IN_SEQUENCE_NUMBER_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.CheckInSequenceNumber';
    static COMPARTMENT_CODE_FIELD_NAME = 'CompartmentCode';
    static COMPARTMENT_CODE_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.CompartmentCode';
    static DATE_OF_BOARDING_PASS_ISSUANCE_JULIAN_FIELD_NAME = 'DateOfBoardingPassIssuanceJulian';
    static DATE_OF_BOARDING_PASS_ISSUANCE_JULIAN_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.DateOfBoardingPassIssuanceJulian';
    static DATE_OF_FLIGHT_JULIAN_FIELD_NAME = 'DateOfFlightJulian';
    static DATE_OF_FLIGHT_JULIAN_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.DateOfFlightJulian';
    static DEPARTURE_AIRPORT_CODE_FIELD_NAME = 'DepartureAirportCode';
    static DEPARTURE_AIRPORT_CODE_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.DepartureAirportCode';
    static DESTINATION_AIRPORT_CODE_FIELD_NAME = 'DestinationAirportCode';
    static DESTINATION_AIRPORT_CODE_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.DestinationAirportCode';
    static DOCUMENT_FORM_SERIAL_NUMBER_FIELD_NAME = 'DocumentFormSerialNumber';
    static DOCUMENT_FORM_SERIAL_NUMBER_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.DocumentFormSerialNumber';
    static DOCUMENT_TYPE_FIELD_NAME = 'DocumentType';
    static DOCUMENT_TYPE_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.DocumentType';
    static FAST_TRACK_FIELD_NAME = 'FastTrack';
    static FAST_TRACK_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.FastTrack';
    static FIRST_NON_CONSECUTIVE_BAGGAGE_TAG_LICENSE_PLATE_NUMBER_FIELD_NAME = 'FirstNonConsecutiveBaggageTagLicensePlateNumber';
    static FIRST_NON_CONSECUTIVE_BAGGAGE_TAG_LICENSE_PLATE_NUMBER_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.FirstNonConsecutiveBaggageTagLicensePlateNumber';
    static FLIGHT_NUMBER_FIELD_NAME = 'FlightNumber';
    static FLIGHT_NUMBER_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.FlightNumber';
    static FOR_INDIVIDUAL_AIRLINE_USE_FIELD_NAME = 'ForIndividualAirlineUse';
    static FOR_INDIVIDUAL_AIRLINE_USE_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.ForIndividualAirlineUse';
    static FREE_BAGGAGE_ALLOWANCE_FIELD_NAME = 'FreeBaggageAllowance';
    static FREE_BAGGAGE_ALLOWANCE_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.FreeBaggageAllowance';
    static FREQUENT_FLYER_AIRLINE_DESIGNATOR_FIELD_NAME = 'FrequentFlyerAirlineDesignator';
    static FREQUENT_FLYER_AIRLINE_DESIGNATOR_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.FrequentFlyerAirlineDesignator';
    static FREQUENT_FLYER_NUMBER_FIELD_NAME = 'FrequentFlyerNumber';
    static FREQUENT_FLYER_NUMBER_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.FrequentFlyerNumber';
    static IDAD_INDICATOR_FIELD_NAME = 'IDADIndicator';
    static IDAD_INDICATOR_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.IDADIndicator';
    static INTERNATIONAL_DOCUMENTATION_VERIFICATION_FIELD_NAME = 'InternationalDocumentationVerification';
    static INTERNATIONAL_DOCUMENTATION_VERIFICATION_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.InternationalDocumentationVerification';
    static MARKETING_CARRIER_DESIGNATOR_FIELD_NAME = 'MarketingCarrierDesignator';
    static MARKETING_CARRIER_DESIGNATOR_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.MarketingCarrierDesignator';
    static OPERATING_CARRIER_DESIGNATOR_FIELD_NAME = 'OperatingCarrierDesignator';
    static OPERATING_CARRIER_DESIGNATOR_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.OperatingCarrierDesignator';
    static OPERATING_CARRIER_PNR_CODE_FIELD_NAME = 'OperatingCarrierPNRCode';
    static OPERATING_CARRIER_PNR_CODE_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.OperatingCarrierPNRCode';
    static PASSENGER_DESCRIPTION_FIELD_NAME = 'PassengerDescription';
    static PASSENGER_DESCRIPTION_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.PassengerDescription';
    static PASSENGER_STATUS_FIELD_NAME = 'PassengerStatus';
    static PASSENGER_STATUS_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.PassengerStatus';
    static SEAT_NUMBER_FIELD_NAME = 'SeatNumber';
    static SEAT_NUMBER_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.SeatNumber';
    static SECOND_NON_CONSECUTIVE_BAGGAGE_TAG_LICENSE_PLATE_NUMBER_FIELD_NAME = 'SecondNonConsecutiveBaggageTagLicensePlateNumber';
    static SECOND_NON_CONSECUTIVE_BAGGAGE_TAG_LICENSE_PLATE_NUMBER_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.SecondNonConsecutiveBaggageTagLicensePlateNumber';
    static SELECTEE_INDICATOR_FIELD_NAME = 'SelecteeIndicator';
    static SELECTEE_INDICATOR_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.SelecteeIndicator';
    static SOURCE_OF_BOARDING_PASS_ISSUANCE_FIELD_NAME = 'SourceOfBoardingPassIssuance';
    static SOURCE_OF_BOARDING_PASS_ISSUANCE_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.SourceOfBoardingPassIssuance';
    static SOURCE_OF_CHECK_IN_FIELD_NAME = 'SourceOfCheckIn';
    static SOURCE_OF_CHECK_IN_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.SourceOfCheckIn';
    static VERSION_NUMBER_FIELD_NAME = 'VersionNumber';
    static VERSION_NUMBER_FIELD_NAME_NORMALIZED = 'BoardingPass.Leg.VersionNumber';
    constructor(document) {
      if (document.type.name !== this.requiredDocumentType()) {
        throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
      }
      this._document = document;
    }
    requiredDocumentType() {
      return BoardingPass.Leg.DOCUMENT_TYPE;
    }

    /** Airline Designator Of Boarding Pass Issuer */
    get airlineDesignatorOfBoardingPassIssuer() {
      return this.document.fields.find(f => f.type.name === 'AirlineDesignatorOfBoardingPassIssuer');
    }

    /** Airline Numeric Code */
    get airlineNumericCode() {
      return this.document.fields.find(f => f.type.name === 'AirlineNumericCode');
    }

    /** Baggage Tag License Plate Numbers */
    get baggageTagLicensePlateNumbers() {
      return this.document.fields.find(f => f.type.name === 'BaggageTagLicensePlateNumbers');
    }

    /** Check In Sequence Number */
    get checkInSequenceNumber() {
      return this.document.fields.find(f => f.type.name === 'CheckInSequenceNumber');
    }

    /** Compartment Code */
    get compartmentCode() {
      return this.document.fields.find(f => f.type.name === 'CompartmentCode');
    }

    /** Date Of Boarding Pass Issuance Julian */
    get dateOfBoardingPassIssuanceJulian() {
      return this.document.fields.find(f => f.type.name === 'DateOfBoardingPassIssuanceJulian');
    }

    /** Date Of Flight Julian */
    get dateOfFlightJulian() {
      return this.document.fields.find(f => f.type.name === 'DateOfFlightJulian');
    }

    /** Departure Airport Code */
    get departureAirportCode() {
      return this.document.fields.find(f => f.type.name === 'DepartureAirportCode');
    }

    /** Destination Airport Code */
    get destinationAirportCode() {
      return this.document.fields.find(f => f.type.name === 'DestinationAirportCode');
    }

    /** Document Form Serial Number */
    get documentFormSerialNumber() {
      return this.document.fields.find(f => f.type.name === 'DocumentFormSerialNumber');
    }

    /** Document Type */
    get documentType() {
      return this.document.fields.find(f => f.type.name === 'DocumentType');
    }

    /** Fast Track */
    get fastTrack() {
      return this.document.fields.find(f => f.type.name === 'FastTrack');
    }

    /** First Non Consecutive Baggage Tag License Plate Number */
    get firstNonConsecutiveBaggageTagLicensePlateNumber() {
      return this.document.fields.find(f => f.type.name === 'FirstNonConsecutiveBaggageTagLicensePlateNumber');
    }

    /** Flight Number */
    get flightNumber() {
      return this.document.fields.find(f => f.type.name === 'FlightNumber');
    }

    /** For Individual Airline Use */
    get forIndividualAirlineUse() {
      return this.document.fields.find(f => f.type.name === 'ForIndividualAirlineUse');
    }

    /** Free Baggage Allowance */
    get freeBaggageAllowance() {
      return this.document.fields.find(f => f.type.name === 'FreeBaggageAllowance');
    }

    /** Frequent Flyer Airline Designator */
    get frequentFlyerAirlineDesignator() {
      return this.document.fields.find(f => f.type.name === 'FrequentFlyerAirlineDesignator');
    }

    /** Frequent Flyer Number */
    get frequentFlyerNumber() {
      return this.document.fields.find(f => f.type.name === 'FrequentFlyerNumber');
    }

    /** IDAD Indicator */
    get idadIndicator() {
      return this.document.fields.find(f => f.type.name === 'IDADIndicator');
    }

    /** International Documentation Verification */
    get internationalDocumentationVerification() {
      return this.document.fields.find(f => f.type.name === 'InternationalDocumentationVerification');
    }

    /** Marketing Carrier Designator */
    get marketingCarrierDesignator() {
      return this.document.fields.find(f => f.type.name === 'MarketingCarrierDesignator');
    }

    /** Operating Carrier Designator */
    get operatingCarrierDesignator() {
      return this.document.fields.find(f => f.type.name === 'OperatingCarrierDesignator');
    }

    /** Operating Carrier PNR Code */
    get operatingCarrierPNRCode() {
      return this.document.fields.find(f => f.type.name === 'OperatingCarrierPNRCode');
    }

    /** Passenger Description */
    get passengerDescription() {
      return this.document.fields.find(f => f.type.name === 'PassengerDescription');
    }

    /** Passenger Status */
    get passengerStatus() {
      return this.document.fields.find(f => f.type.name === 'PassengerStatus');
    }

    /** Seat Number */
    get seatNumber() {
      return this.document.fields.find(f => f.type.name === 'SeatNumber');
    }

    /** Second Non Consecutive Baggage Tag License Plate Number */
    get secondNonConsecutiveBaggageTagLicensePlateNumber() {
      return this.document.fields.find(f => f.type.name === 'SecondNonConsecutiveBaggageTagLicensePlateNumber');
    }

    /** Selectee Indicator */
    get selecteeIndicator() {
      return this.document.fields.find(f => f.type.name === 'SelecteeIndicator');
    }

    /** Source Of Boarding Pass Issuance */
    get sourceOfBoardingPassIssuance() {
      return this.document.fields.find(f => f.type.name === 'SourceOfBoardingPassIssuance');
    }

    /** Source Of Check In */
    get sourceOfCheckIn() {
      return this.document.fields.find(f => f.type.name === 'SourceOfCheckIn');
    }

    /** VersionNumber */
    get versionNumber() {
      return this.document.fields.find(f => f.type.name === 'VersionNumber');
    }
  }
  _BoardingPass.Leg = Leg;
})(BoardingPass || (exports.BoardingPass = BoardingPass = {}));
/** SwissQR */
class SwissQR {
  get document() {
    return this._document;
  }
  static DOCUMENT_TYPE = 'SwissQR';
  static ADDITIONAL_BILLING_INFORMATION_FIELD_NAME = 'AdditionalBillingInformation';
  static ADDITIONAL_BILLING_INFORMATION_FIELD_NAME_NORMALIZED = 'SwissQR.AdditionalBillingInformation';
  static ADDITIONAL_INFO_TRAILER_FIELD_NAME = 'AdditionalInfoTrailer';
  static ADDITIONAL_INFO_TRAILER_FIELD_NAME_NORMALIZED = 'SwissQR.AdditionalInfoTrailer';
  static ADDITIONAL_INFO_UNSTRUCTURED_FIELD_NAME = 'AdditionalInfoUnstructured';
  static ADDITIONAL_INFO_UNSTRUCTURED_FIELD_NAME_NORMALIZED = 'SwissQR.AdditionalInfoUnstructured';
  static ALTERNATIVE_PROCEDURE_PARAMETER_FIELD_NAME = 'AlternativeProcedureParameter';
  static ALTERNATIVE_PROCEDURE_PARAMETER_FIELD_NAME_NORMALIZED = 'SwissQR.AlternativeProcedureParameter';
  static AMOUNT_FIELD_NAME = 'Amount';
  static AMOUNT_FIELD_NAME_NORMALIZED = 'SwissQR.Amount';
  static CURRENCY_FIELD_NAME = 'Currency';
  static CURRENCY_FIELD_NAME_NORMALIZED = 'SwissQR.Currency';
  static DEBTOR_ADDRESS_TYPE_FIELD_NAME = 'DebtorAddressType';
  static DEBTOR_ADDRESS_TYPE_FIELD_NAME_NORMALIZED = 'SwissQR.DebtorAddressType';
  static DEBTOR_BUILDING_OR_ADDRESS_LINE_2_FIELD_NAME = 'DebtorBuildingOrAddressLine2';
  static DEBTOR_BUILDING_OR_ADDRESS_LINE_2_FIELD_NAME_NORMALIZED = 'SwissQR.DebtorBuildingOrAddressLine2';
  static DEBTOR_COUNTRY_FIELD_NAME = 'DebtorCountry';
  static DEBTOR_COUNTRY_FIELD_NAME_NORMALIZED = 'SwissQR.DebtorCountry';
  static DEBTOR_NAME_FIELD_NAME = 'DebtorName';
  static DEBTOR_NAME_FIELD_NAME_NORMALIZED = 'SwissQR.DebtorName';
  static DEBTOR_PLACE_FIELD_NAME = 'DebtorPlace';
  static DEBTOR_PLACE_FIELD_NAME_NORMALIZED = 'SwissQR.DebtorPlace';
  static DEBTOR_POSTAL_CODE_FIELD_NAME = 'DebtorPostalCode';
  static DEBTOR_POSTAL_CODE_FIELD_NAME_NORMALIZED = 'SwissQR.DebtorPostalCode';
  static DEBTOR_STREET_OR_ADDRESS_LINE_1_FIELD_NAME = 'DebtorStreetOrAddressLine1';
  static DEBTOR_STREET_OR_ADDRESS_LINE_1_FIELD_NAME_NORMALIZED = 'SwissQR.DebtorStreetOrAddressLine1';
  static DUE_DATE_FIELD_NAME = 'DueDate';
  static DUE_DATE_FIELD_NAME_NORMALIZED = 'SwissQR.DueDate';
  static ENCODING_FIELD_NAME = 'Encoding';
  static ENCODING_FIELD_NAME_NORMALIZED = 'SwissQR.Encoding';
  static FINAL_PAYEE_ADDRESS_TYPE_FIELD_NAME = 'FinalPayeeAddressType';
  static FINAL_PAYEE_ADDRESS_TYPE_FIELD_NAME_NORMALIZED = 'SwissQR.FinalPayeeAddressType';
  static FINAL_PAYEE_BUILDING_OR_ADDRESS_LINE_2_FIELD_NAME = 'FinalPayeeBuildingOrAddressLine2';
  static FINAL_PAYEE_BUILDING_OR_ADDRESS_LINE_2_FIELD_NAME_NORMALIZED = 'SwissQR.FinalPayeeBuildingOrAddressLine2';
  static FINAL_PAYEE_COUNTRY_FIELD_NAME = 'FinalPayeeCountry';
  static FINAL_PAYEE_COUNTRY_FIELD_NAME_NORMALIZED = 'SwissQR.FinalPayeeCountry';
  static FINAL_PAYEE_NAME_FIELD_NAME = 'FinalPayeeName';
  static FINAL_PAYEE_NAME_FIELD_NAME_NORMALIZED = 'SwissQR.FinalPayeeName';
  static FINAL_PAYEE_PLACE_FIELD_NAME = 'FinalPayeePlace';
  static FINAL_PAYEE_PLACE_FIELD_NAME_NORMALIZED = 'SwissQR.FinalPayeePlace';
  static FINAL_PAYEE_POSTAL_CODE_FIELD_NAME = 'FinalPayeePostalCode';
  static FINAL_PAYEE_POSTAL_CODE_FIELD_NAME_NORMALIZED = 'SwissQR.FinalPayeePostalCode';
  static FINAL_PAYEE_STREET_OR_ADDRESS_LINE_1_FIELD_NAME = 'FinalPayeeStreetOrAddressLine1';
  static FINAL_PAYEE_STREET_OR_ADDRESS_LINE_1_FIELD_NAME_NORMALIZED = 'SwissQR.FinalPayeeStreetOrAddressLine1';
  static IBAN_FIELD_NAME = 'IBAN';
  static IBAN_FIELD_NAME_NORMALIZED = 'SwissQR.IBAN';
  static MAJOR_VERSION_FIELD_NAME = 'MajorVersion';
  static MAJOR_VERSION_FIELD_NAME_NORMALIZED = 'SwissQR.MajorVersion';
  static PAYEE_ADDRESS_TYPE_FIELD_NAME = 'PayeeAddressType';
  static PAYEE_ADDRESS_TYPE_FIELD_NAME_NORMALIZED = 'SwissQR.PayeeAddressType';
  static PAYEE_BUILDING_OR_ADDRESS_LINE_2_FIELD_NAME = 'PayeeBuildingOrAddressLine2';
  static PAYEE_BUILDING_OR_ADDRESS_LINE_2_FIELD_NAME_NORMALIZED = 'SwissQR.PayeeBuildingOrAddressLine2';
  static PAYEE_COUNTRY_FIELD_NAME = 'PayeeCountry';
  static PAYEE_COUNTRY_FIELD_NAME_NORMALIZED = 'SwissQR.PayeeCountry';
  static PAYEE_NAME_FIELD_NAME = 'PayeeName';
  static PAYEE_NAME_FIELD_NAME_NORMALIZED = 'SwissQR.PayeeName';
  static PAYEE_PLACE_FIELD_NAME = 'PayeePlace';
  static PAYEE_PLACE_FIELD_NAME_NORMALIZED = 'SwissQR.PayeePlace';
  static PAYEE_POSTAL_CODE_FIELD_NAME = 'PayeePostalCode';
  static PAYEE_POSTAL_CODE_FIELD_NAME_NORMALIZED = 'SwissQR.PayeePostalCode';
  static PAYEE_STREET_OR_ADDRESS_LINE_1_FIELD_NAME = 'PayeeStreetOrAddressLine1';
  static PAYEE_STREET_OR_ADDRESS_LINE_1_FIELD_NAME_NORMALIZED = 'SwissQR.PayeeStreetOrAddressLine1';
  static PAYMENT_REFERENCE_FIELD_NAME = 'PaymentReference';
  static PAYMENT_REFERENCE_FIELD_NAME_NORMALIZED = 'SwissQR.PaymentReference';
  static PAYMENT_REFERENCE_TYPE_FIELD_NAME = 'PaymentReferenceType';
  static PAYMENT_REFERENCE_TYPE_FIELD_NAME_NORMALIZED = 'SwissQR.PaymentReferenceType';
  constructor(document) {
    if (document.type.name !== this.requiredDocumentType()) {
      throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
    }
    this._document = document;
  }
  requiredDocumentType() {
    return SwissQR.DOCUMENT_TYPE;
  }

  /** Additional Billing Information */
  get additionalBillingInformation() {
    return this.document.fields.find(f => f.type.name === 'AdditionalBillingInformation');
  }

  /** Additional Info Trailer */
  get additionalInfoTrailer() {
    return this.document.fields.find(f => f.type.name === 'AdditionalInfoTrailer');
  }

  /** Additional Info Unstructured */
  get additionalInfoUnstructured() {
    return this.document.fields.find(f => f.type.name === 'AdditionalInfoUnstructured');
  }

  /** Alternative Procedure Parameter */
  get alternativeProcedureParameter() {
    return this.document.fields.find(f => f.type.name === 'AlternativeProcedureParameter');
  }

  /** Amount */
  get amount() {
    return this.document.fields.find(f => f.type.name === 'Amount');
  }

  /** Currency */
  get currency() {
    return this.document.fields.find(f => f.type.name === 'Currency');
  }

  /** Debtor Address Type */
  get debtorAddressType() {
    return this.document.fields.find(f => f.type.name === 'DebtorAddressType');
  }

  /** Debtor Building Or Address Line 2 */
  get debtorBuildingOrAddressLine2() {
    return this.document.fields.find(f => f.type.name === 'DebtorBuildingOrAddressLine2');
  }

  /** Debtor Country */
  get debtorCountry() {
    return this.document.fields.find(f => f.type.name === 'DebtorCountry');
  }

  /** Debtor Name */
  get debtorName() {
    return this.document.fields.find(f => f.type.name === 'DebtorName');
  }

  /** Debtor Place */
  get debtorPlace() {
    return this.document.fields.find(f => f.type.name === 'DebtorPlace');
  }

  /** Debtor Postal Code */
  get debtorPostalCode() {
    return this.document.fields.find(f => f.type.name === 'DebtorPostalCode');
  }

  /** Debtor Street Or Address Line 1 */
  get debtorStreetOrAddressLine1() {
    return this.document.fields.find(f => f.type.name === 'DebtorStreetOrAddressLine1');
  }

  /** Due Date */
  get dueDate() {
    return this.document.fields.find(f => f.type.name === 'DueDate');
  }

  /** Encoding */
  get encoding() {
    return this.document.fields.find(f => f.type.name === 'Encoding');
  }

  /** Final Payee Address Type */
  get finalPayeeAddressType() {
    return this.document.fields.find(f => f.type.name === 'FinalPayeeAddressType');
  }

  /** Final Payee Building Or Address Line 2 */
  get finalPayeeBuildingOrAddressLine2() {
    return this.document.fields.find(f => f.type.name === 'FinalPayeeBuildingOrAddressLine2');
  }

  /** Final Payee Country */
  get finalPayeeCountry() {
    return this.document.fields.find(f => f.type.name === 'FinalPayeeCountry');
  }

  /** Final Payee Name */
  get finalPayeeName() {
    return this.document.fields.find(f => f.type.name === 'FinalPayeeName');
  }

  /** Final Payee Place */
  get finalPayeePlace() {
    return this.document.fields.find(f => f.type.name === 'FinalPayeePlace');
  }

  /** Final Payee Postal Code */
  get finalPayeePostalCode() {
    return this.document.fields.find(f => f.type.name === 'FinalPayeePostalCode');
  }

  /** Final Payee Street Or Address Line 1 */
  get finalPayeeStreetOrAddressLine1() {
    return this.document.fields.find(f => f.type.name === 'FinalPayeeStreetOrAddressLine1');
  }

  /** IBAN */
  get iban() {
    return this.document.fields.find(f => f.type.name === 'IBAN');
  }

  /** Major Version */
  get majorVersion() {
    return this.document.fields.find(f => f.type.name === 'MajorVersion');
  }

  /** Payee Address Type */
  get payeeAddressType() {
    return this.document.fields.find(f => f.type.name === 'PayeeAddressType');
  }

  /** Payee Building Or Address Line 2 */
  get payeeBuildingOrAddressLine2() {
    return this.document.fields.find(f => f.type.name === 'PayeeBuildingOrAddressLine2');
  }

  /** Payee Country */
  get payeeCountry() {
    return this.document.fields.find(f => f.type.name === 'PayeeCountry');
  }

  /** Payee Name */
  get payeeName() {
    return this.document.fields.find(f => f.type.name === 'PayeeName');
  }

  /** Payee Place */
  get payeePlace() {
    return this.document.fields.find(f => f.type.name === 'PayeePlace');
  }

  /** Payee Postal Code */
  get payeePostalCode() {
    return this.document.fields.find(f => f.type.name === 'PayeePostalCode');
  }

  /** Payee Street Or Address Line 1 */
  get payeeStreetOrAddressLine1() {
    return this.document.fields.find(f => f.type.name === 'PayeeStreetOrAddressLine1');
  }

  /** Payment Reference */
  get paymentReference() {
    return this.document.fields.find(f => f.type.name === 'PaymentReference');
  }

  /** Payment Reference Type */
  get paymentReferenceType() {
    return this.document.fields.find(f => f.type.name === 'PaymentReferenceType');
  }
}

/** Medical Plan */
exports.SwissQR = SwissQR;
class DEMedicalPlan {
  get document() {
    return this._document;
  }
  static DOCUMENT_TYPE = 'DEMedicalPlan';
  static CURRENT_PAGE_FIELD_NAME = 'CurrentPage';
  static CURRENT_PAGE_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.CurrentPage';
  static DOCUMENT_VERSION_NUMBER_FIELD_NAME = 'DocumentVersionNumber';
  static DOCUMENT_VERSION_NUMBER_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.DocumentVersionNumber';
  static GUID_FIELD_NAME = 'GUID';
  static GUID_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.GUID';
  static LANGUAGE_COUNTRY_CODE_FIELD_NAME = 'LanguageCountryCode';
  static LANGUAGE_COUNTRY_CODE_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.LanguageCountryCode';
  static PATCH_VERSION_NUMBER_FIELD_NAME = 'PatchVersionNumber';
  static PATCH_VERSION_NUMBER_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.PatchVersionNumber';
  static TOTAL_NUMBER_OF_PAGES_FIELD_NAME = 'TotalNumberOfPages';
  static TOTAL_NUMBER_OF_PAGES_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.TotalNumberOfPages';
  constructor(document) {
    if (document.type.name !== this.requiredDocumentType()) {
      throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
    }
    this._document = document;
  }
  requiredDocumentType() {
    return DEMedicalPlan.DOCUMENT_TYPE;
  }

  /** Current Page */
  get currentPage() {
    return this.document.fields.find(f => f.type.name === 'CurrentPage');
  }

  /** Document Version Number */
  get documentVersionNumber() {
    return this.document.fields.find(f => f.type.name === 'DocumentVersionNumber');
  }

  /** GUID */
  get guid() {
    return this.document.fields.find(f => f.type.name === 'GUID');
  }

  /** Language Country Code */
  get languageCountryCode() {
    return this.document.fields.find(f => f.type.name === 'LanguageCountryCode');
  }

  /** Patch Version Number */
  get patchVersionNumber() {
    return this.document.fields.find(f => f.type.name === 'PatchVersionNumber');
  }

  /** Total Number Of Pages */
  get totalNumberOfPages() {
    return this.document.fields.find(f => f.type.name === 'TotalNumberOfPages');
  }

  /** The child document of type "Patient". */
  get patient() {
    const child = this.document.children.find(c => c.type.name === 'Patient');
    return new DEMedicalPlan.Patient(child);
  }

  /** The child document of type "Doctor". */
  get doctor() {
    const child = this.document.children.find(c => c.type.name === 'Doctor');
    return new DEMedicalPlan.Doctor(child);
  }

  /** An array of all children of type "Subheading". */
  get subheadings() {
    return this.document.children.filter(c => c.type.name === 'Subheading').map(c => new DEMedicalPlan.Subheading(c));
  }
}
exports.DEMedicalPlan = DEMedicalPlan;
(function (_DEMedicalPlan) {
  class Patient {
    get document() {
      return this._document;
    }
    static DOCUMENT_TYPE = 'Patient';
    static ALLERGIES_AND_INTOLERANCES_FIELD_NAME = 'AllergiesAndIntolerances';
    static ALLERGIES_AND_INTOLERANCES_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Patient.AllergiesAndIntolerances';
    static BIRTH_DATE_FIELD_NAME = 'BirthDate';
    static BIRTH_DATE_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Patient.BirthDate';
    static BREAST_FEEDING_FIELD_NAME = 'BreastFeeding';
    static BREAST_FEEDING_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Patient.BreastFeeding';
    static CREATININE_VALUE_FIELD_NAME = 'CreatinineValue';
    static CREATININE_VALUE_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Patient.CreatinineValue';
    static FIRST_NAME_FIELD_NAME = 'FirstName';
    static FIRST_NAME_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Patient.FirstName';
    static GENDER_FIELD_NAME = 'Gender';
    static GENDER_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Patient.Gender';
    static HEIGHT_FIELD_NAME = 'Height';
    static HEIGHT_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Patient.Height';
    static LAST_NAME_FIELD_NAME = 'LastName';
    static LAST_NAME_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Patient.LastName';
    static PATIENT_FREE_TEXT_FIELD_NAME = 'PatientFreeText';
    static PATIENT_FREE_TEXT_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Patient.PatientFreeText';
    static PATIENT_ID_FIELD_NAME = 'PatientID';
    static PATIENT_ID_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Patient.PatientID';
    static PRE_NAME_FIELD_NAME = 'PreName';
    static PRE_NAME_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Patient.PreName';
    static PREGNANT_FIELD_NAME = 'Pregnant';
    static PREGNANT_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Patient.Pregnant';
    static SUFFIX_FIELD_NAME = 'Suffix';
    static SUFFIX_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Patient.Suffix';
    static TITLE_FIELD_NAME = 'Title';
    static TITLE_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Patient.Title';
    static WEIGHT_FIELD_NAME = 'Weight';
    static WEIGHT_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Patient.Weight';
    constructor(document) {
      if (document.type.name !== this.requiredDocumentType()) {
        throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
      }
      this._document = document;
    }
    requiredDocumentType() {
      return DEMedicalPlan.Patient.DOCUMENT_TYPE;
    }

    /** Allergies And Intolerances */
    get allergiesAndIntolerances() {
      return this.document.fields.find(f => f.type.name === 'AllergiesAndIntolerances');
    }

    /** Birth Date */
    get birthDate() {
      return this.document.fields.find(f => f.type.name === 'BirthDate');
    }

    /** Breast Feeding */
    get breastFeeding() {
      return this.document.fields.find(f => f.type.name === 'BreastFeeding');
    }

    /** Creatinine Value */
    get creatinineValue() {
      return this.document.fields.find(f => f.type.name === 'CreatinineValue');
    }

    /** First Name */
    get firstName() {
      return this.document.fields.find(f => f.type.name === 'FirstName');
    }

    /** Gender */
    get gender() {
      return this.document.fields.find(f => f.type.name === 'Gender');
    }

    /** Height */
    get height() {
      return this.document.fields.find(f => f.type.name === 'Height');
    }

    /** Last Name */
    get lastName() {
      return this.document.fields.find(f => f.type.name === 'LastName');
    }

    /** Patient Free Text */
    get patientFreeText() {
      return this.document.fields.find(f => f.type.name === 'PatientFreeText');
    }

    /** Patient ID */
    get patientID() {
      return this.document.fields.find(f => f.type.name === 'PatientID');
    }

    /** Pre Name */
    get preName() {
      return this.document.fields.find(f => f.type.name === 'PreName');
    }

    /** Pregnant */
    get pregnant() {
      return this.document.fields.find(f => f.type.name === 'Pregnant');
    }

    /** Name Suffix */
    get suffix() {
      return this.document.fields.find(f => f.type.name === 'Suffix');
    }

    /** Title */
    get title() {
      return this.document.fields.find(f => f.type.name === 'Title');
    }

    /** Weight */
    get weight() {
      return this.document.fields.find(f => f.type.name === 'Weight');
    }
  }
  _DEMedicalPlan.Patient = Patient;
  class Doctor {
    get document() {
      return this._document;
    }
    static DOCUMENT_TYPE = 'Doctor';
    static DOCTOR_NUMBER_FIELD_NAME = 'DoctorNumber';
    static DOCTOR_NUMBER_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Doctor.DoctorNumber';
    static EMAIL_FIELD_NAME = 'Email';
    static EMAIL_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Doctor.Email';
    static HOSPITAL_ID_FIELD_NAME = 'HospitalID';
    static HOSPITAL_ID_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Doctor.HospitalID';
    static ISSUER_NAME_FIELD_NAME = 'IssuerName';
    static ISSUER_NAME_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Doctor.IssuerName';
    static ISSUING_DATE_AND_TIME_FIELD_NAME = 'IssuingDateAndTime';
    static ISSUING_DATE_AND_TIME_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Doctor.IssuingDateAndTime';
    static PHARMACY_ID_FIELD_NAME = 'PharmacyID';
    static PHARMACY_ID_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Doctor.PharmacyID';
    static PLACE_FIELD_NAME = 'Place';
    static PLACE_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Doctor.Place';
    static POSTAL_CODE_FIELD_NAME = 'PostalCode';
    static POSTAL_CODE_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Doctor.PostalCode';
    static STREET_FIELD_NAME = 'Street';
    static STREET_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Doctor.Street';
    static TELEPHONE_NUMBER_FIELD_NAME = 'TelephoneNumber';
    static TELEPHONE_NUMBER_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Doctor.TelephoneNumber';
    constructor(document) {
      if (document.type.name !== this.requiredDocumentType()) {
        throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
      }
      this._document = document;
    }
    requiredDocumentType() {
      return DEMedicalPlan.Doctor.DOCUMENT_TYPE;
    }

    /** Doctor Number */
    get doctorNumber() {
      return this.document.fields.find(f => f.type.name === 'DoctorNumber');
    }

    /** Email */
    get email() {
      return this.document.fields.find(f => f.type.name === 'Email');
    }

    /** Hospital ID */
    get hospitalID() {
      return this.document.fields.find(f => f.type.name === 'HospitalID');
    }

    /** Issuer Name */
    get issuerName() {
      return this.document.fields.find(f => f.type.name === 'IssuerName');
    }

    /** Issuing Date And Time */
    get issuingDateAndTime() {
      return this.document.fields.find(f => f.type.name === 'IssuingDateAndTime');
    }

    /** Pharmacy ID */
    get pharmacyID() {
      return this.document.fields.find(f => f.type.name === 'PharmacyID');
    }

    /** Place */
    get place() {
      return this.document.fields.find(f => f.type.name === 'Place');
    }

    /** Postal Code */
    get postalCode() {
      return this.document.fields.find(f => f.type.name === 'PostalCode');
    }

    /** Street */
    get street() {
      return this.document.fields.find(f => f.type.name === 'Street');
    }

    /** Telephone Number */
    get telephoneNumber() {
      return this.document.fields.find(f => f.type.name === 'TelephoneNumber');
    }
  }
  _DEMedicalPlan.Doctor = Doctor;
  class Subheading {
    get document() {
      return this._document;
    }
    static DOCUMENT_TYPE = 'Subheading';
    static GENERAL_NOTE_FIELD_NAME = 'GeneralNote';
    static GENERAL_NOTE_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.GeneralNote';
    static KEY_WORDS_FIELD_NAME = 'KeyWords';
    static KEY_WORDS_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.KeyWords';
    static SUBHEADING_FREE_TEXT_FIELD_NAME = 'SubheadingFreeText';
    static SUBHEADING_FREE_TEXT_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.SubheadingFreeText';
    constructor(document) {
      if (document.type.name !== this.requiredDocumentType()) {
        throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
      }
      this._document = document;
    }
    requiredDocumentType() {
      return DEMedicalPlan.Subheading.DOCUMENT_TYPE;
    }

    /** General Note */
    get generalNotes() {
      return this.document.fields.filter(f => f.type.name === 'GeneralNote');
    }

    /** Key Words */
    get keyWords() {
      return this.document.fields.find(f => f.type.name === 'KeyWords');
    }

    /** Subheading Free Text */
    get subheadingFreeText() {
      return this.document.fields.find(f => f.type.name === 'SubheadingFreeText');
    }

    /** An array of all children of type "Medicine". */
    get medicines() {
      return this.document.children.filter(c => c.type.name === 'Medicine').map(c => new DEMedicalPlan.Subheading.Medicine(c));
    }

    /** An array of all children of type "Prescription". */
    get prescriptions() {
      return this.document.children.filter(c => c.type.name === 'Prescription').map(c => new DEMedicalPlan.Subheading.Prescription(c));
    }
  }
  _DEMedicalPlan.Subheading = Subheading;
  (function (_Subheading) {
    class Medicine {
      get document() {
        return this._document;
      }
      static DOCUMENT_TYPE = 'Medicine';
      static DOSAGE_FORM_FIELD_NAME = 'DosageForm';
      static DOSAGE_FORM_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.Medicine.DosageForm';
      static DOSAGE_FORM_FREE_TEXT_FIELD_NAME = 'DosageFormFreeText';
      static DOSAGE_FORM_FREE_TEXT_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.Medicine.DosageFormFreeText';
      static DOSAGE_FREE_TEXT_FIELD_NAME = 'DosageFreeText';
      static DOSAGE_FREE_TEXT_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.Medicine.DosageFreeText';
      static DOSING_UNIT_FIELD_NAME = 'DosingUnit';
      static DOSING_UNIT_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.Medicine.DosingUnit';
      static DOSING_UNIT_FREE_TEXT_FIELD_NAME = 'DosingUnitFreeText';
      static DOSING_UNIT_FREE_TEXT_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.Medicine.DosingUnitFreeText';
      static DRUG_NAME_FIELD_NAME = 'DrugName';
      static DRUG_NAME_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.Medicine.DrugName';
      static EVENING_FIELD_NAME = 'Evening';
      static EVENING_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.Medicine.Evening';
      static GENERAL_NOTES_FIELD_NAME = 'GeneralNotes';
      static GENERAL_NOTES_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.Medicine.GeneralNotes';
      static MIDDAY_FIELD_NAME = 'Midday';
      static MIDDAY_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.Medicine.Midday';
      static MORNING_FIELD_NAME = 'Morning';
      static MORNING_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.Medicine.Morning';
      static NIGHT_FIELD_NAME = 'Night';
      static NIGHT_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.Medicine.Night';
      static PHARMACEUTICAL_NUMBER_FIELD_NAME = 'PharmaceuticalNumber';
      static PHARMACEUTICAL_NUMBER_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.Medicine.PharmaceuticalNumber';
      static REASON_FOR_TREATMENT_FIELD_NAME = 'ReasonForTreatment';
      static REASON_FOR_TREATMENT_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.Medicine.ReasonForTreatment';
      static RELEVANT_INFO_FIELD_NAME = 'RelevantInfo';
      static RELEVANT_INFO_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.Medicine.RelevantInfo';
      constructor(document) {
        if (document.type.name !== this.requiredDocumentType()) {
          throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
        }
        this._document = document;
      }
      requiredDocumentType() {
        return DEMedicalPlan.Subheading.Medicine.DOCUMENT_TYPE;
      }

      /** Dosage Form */
      get dosageForm() {
        return this.document.fields.find(f => f.type.name === 'DosageForm');
      }

      /** Dosage Form Free Text */
      get dosageFormFreeText() {
        return this.document.fields.find(f => f.type.name === 'DosageFormFreeText');
      }

      /** Dosage Free Text */
      get dosageFreeText() {
        return this.document.fields.find(f => f.type.name === 'DosageFreeText');
      }

      /** Dosing Unit */
      get dosingUnit() {
        return this.document.fields.find(f => f.type.name === 'DosingUnit');
      }

      /** Dosing Unit Free Text */
      get dosingUnitFreeText() {
        return this.document.fields.find(f => f.type.name === 'DosingUnitFreeText');
      }

      /** Drug Name */
      get drugName() {
        return this.document.fields.find(f => f.type.name === 'DrugName');
      }

      /** Evening Intake Dose */
      get evening() {
        return this.document.fields.find(f => f.type.name === 'Evening');
      }

      /** General Notes */
      get generalNotes() {
        return this.document.fields.find(f => f.type.name === 'GeneralNotes');
      }

      /** Midday Intake Dose */
      get midday() {
        return this.document.fields.find(f => f.type.name === 'Midday');
      }

      /** Morning Intake Dose */
      get morning() {
        return this.document.fields.find(f => f.type.name === 'Morning');
      }

      /** Night Intake Dose */
      get night() {
        return this.document.fields.find(f => f.type.name === 'Night');
      }

      /** Pharmaceutical Number */
      get pharmaceuticalNumber() {
        return this.document.fields.find(f => f.type.name === 'PharmaceuticalNumber');
      }

      /** Reason For Treatment */
      get reasonForTreatment() {
        return this.document.fields.find(f => f.type.name === 'ReasonForTreatment');
      }

      /** Relevant Info */
      get relevantInfo() {
        return this.document.fields.find(f => f.type.name === 'RelevantInfo');
      }

      /** An array of all children of type "Substance". */
      get substances() {
        return this.document.children.filter(c => c.type.name === 'Substance').map(c => new DEMedicalPlan.Subheading.Medicine.Substance(c));
      }
    }
    _Subheading.Medicine = Medicine;
    (function (_Medicine) {
      class Substance {
        get document() {
          return this._document;
        }
        static DOCUMENT_TYPE = 'Substance';
        static ACTIVE_SUBSTANCE_FIELD_NAME = 'ActiveSubstance';
        static ACTIVE_SUBSTANCE_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.Medicine.Substance.ActiveSubstance';
        static POTENCY_FIELD_NAME = 'Potency';
        static POTENCY_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.Medicine.Substance.Potency';
        constructor(document) {
          if (document.type.name !== this.requiredDocumentType()) {
            throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
          }
          this._document = document;
        }
        requiredDocumentType() {
          return DEMedicalPlan.Subheading.Medicine.Substance.DOCUMENT_TYPE;
        }

        /** Active Substance */
        get activeSubstance() {
          return this.document.fields.find(f => f.type.name === 'ActiveSubstance');
        }

        /** Potency */
        get potency() {
          return this.document.fields.find(f => f.type.name === 'Potency');
        }
      }
      _Medicine.Substance = Substance;
    })(Medicine || (Medicine = _Subheading.Medicine || (_Subheading.Medicine = {})));
    class Prescription {
      get document() {
        return this._document;
      }
      static DOCUMENT_TYPE = 'Prescription';
      static GENERAL_INFORMATION_FIELD_NAME = 'GeneralInformation';
      static GENERAL_INFORMATION_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.Prescription.GeneralInformation';
      static PRESCRIPTION_FREE_TEXT_FIELD_NAME = 'PrescriptionFreeText';
      static PRESCRIPTION_FREE_TEXT_FIELD_NAME_NORMALIZED = 'DEMedicalPlan.Subheading.Prescription.PrescriptionFreeText';
      constructor(document) {
        if (document.type.name !== this.requiredDocumentType()) {
          throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
        }
        this._document = document;
      }
      requiredDocumentType() {
        return DEMedicalPlan.Subheading.Prescription.DOCUMENT_TYPE;
      }

      /** General Information */
      get generalInformation() {
        return this.document.fields.find(f => f.type.name === 'GeneralInformation');
      }

      /** Prescription Free Text */
      get prescriptionFreeText() {
        return this.document.fields.find(f => f.type.name === 'PrescriptionFreeText');
      }
    }
    _Subheading.Prescription = Prescription;
  })(Subheading || (Subheading = _DEMedicalPlan.Subheading || (_DEMedicalPlan.Subheading = {})));
})(DEMedicalPlan || (exports.DEMedicalPlan = DEMedicalPlan = {}));
/** ID Card */
class IDCardPDF417 {
  get document() {
    return this._document;
  }
  static DOCUMENT_TYPE = 'IDCardPDF417';
  static BIRTH_DATE_FIELD_NAME = 'BirthDate';
  static BIRTH_DATE_FIELD_NAME_NORMALIZED = 'IDCardPDF417.BirthDate';
  static DATE_EXPIRED_FIELD_NAME = 'DateExpired';
  static DATE_EXPIRED_FIELD_NAME_NORMALIZED = 'IDCardPDF417.DateExpired';
  static DATE_ISSUED_FIELD_NAME = 'DateIssued';
  static DATE_ISSUED_FIELD_NAME_NORMALIZED = 'IDCardPDF417.DateIssued';
  static DOCUMENT_CODE_FIELD_NAME = 'DocumentCode';
  static DOCUMENT_CODE_FIELD_NAME_NORMALIZED = 'IDCardPDF417.DocumentCode';
  static FIRST_NAME_FIELD_NAME = 'FirstName';
  static FIRST_NAME_FIELD_NAME_NORMALIZED = 'IDCardPDF417.FirstName';
  static LAST_NAME_FIELD_NAME = 'LastName';
  static LAST_NAME_FIELD_NAME_NORMALIZED = 'IDCardPDF417.LastName';
  static OPTIONAL_FIELD_NAME = 'Optional';
  static OPTIONAL_FIELD_NAME_NORMALIZED = 'IDCardPDF417.Optional';
  constructor(document) {
    if (document.type.name !== this.requiredDocumentType()) {
      throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
    }
    this._document = document;
  }
  requiredDocumentType() {
    return IDCardPDF417.DOCUMENT_TYPE;
  }

  /** Birth Date. The format is ISO8601 with delimiters */
  get birthDate() {
    return this.document.fields.find(f => f.type.name === 'BirthDate');
  }

  /** Date Expired. The format is ISO8601 with delimiters */
  get dateExpired() {
    return this.document.fields.find(f => f.type.name === 'DateExpired');
  }

  /** Date Issued. The format is ISO8601 with delimiters */
  get dateIssued() {
    return this.document.fields.find(f => f.type.name === 'DateIssued');
  }

  /** Document Code */
  get documentCode() {
    return this.document.fields.find(f => f.type.name === 'DocumentCode');
  }

  /** First Name */
  get firstName() {
    return this.document.fields.find(f => f.type.name === 'FirstName');
  }

  /** Last Name */
  get lastName() {
    return this.document.fields.find(f => f.type.name === 'LastName');
  }

  /** Optional */
  get optional() {
    return this.document.fields.find(f => f.type.name === 'Optional');
  }
}

/** GS1 */
exports.IDCardPDF417 = IDCardPDF417;
class GS1 {
  get document() {
    return this._document;
  }
  static DOCUMENT_TYPE = 'GS1';
  constructor(document) {
    if (document.type.name !== this.requiredDocumentType()) {
      throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
    }
    this._document = document;
  }
  requiredDocumentType() {
    return GS1.DOCUMENT_TYPE;
  }

  /** An array of all children of type "Element". */
  get elements() {
    return this.document.children.filter(c => c.type.name === 'Element').map(c => new GS1.Element(c));
  }
}
exports.GS1 = GS1;
(function (_GS) {
  class Element {
    get document() {
      return this._document;
    }
    static DOCUMENT_TYPE = 'Element';
    static APPLICATION_IDENTIFIER_FIELD_NAME = 'ApplicationIdentifier';
    static APPLICATION_IDENTIFIER_FIELD_NAME_NORMALIZED = 'GS1.Element.ApplicationIdentifier';
    static DATA_TITLE_FIELD_NAME = 'DataTitle';
    static DATA_TITLE_FIELD_NAME_NORMALIZED = 'GS1.Element.DataTitle';
    static ELEMENT_DESCRIPTION_FIELD_NAME = 'ElementDescription';
    static ELEMENT_DESCRIPTION_FIELD_NAME_NORMALIZED = 'GS1.Element.ElementDescription';
    static RAW_VALUE_FIELD_NAME = 'RawValue';
    static RAW_VALUE_FIELD_NAME_NORMALIZED = 'GS1.Element.RawValue';
    constructor(document) {
      if (document.type.name !== this.requiredDocumentType()) {
        throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
      }
      this._document = document;
    }
    requiredDocumentType() {
      return GS1.Element.DOCUMENT_TYPE;
    }

    /** Application Identifier */
    get applicationIdentifier() {
      return this.document.fields.find(f => f.type.name === 'ApplicationIdentifier');
    }

    /** Data Title */
    get dataTitle() {
      return this.document.fields.find(f => f.type.name === 'DataTitle');
    }

    /** Description */
    get elementDescription() {
      return this.document.fields.find(f => f.type.name === 'ElementDescription');
    }

    /** Raw Value */
    get rawValue() {
      return this.document.fields.find(f => f.type.name === 'RawValue');
    }

    /** An array of all children of type "ValidationError". */
    get validationErrors() {
      return this.document.children.filter(c => c.type.name === 'ValidationError').map(c => new GS1.Element.ValidationError(c));
    }
  }
  _GS.Element = Element;
  (function (_Element) {
    class ValidationError {
      get document() {
        return this._document;
      }
      static DOCUMENT_TYPE = 'ValidationError';
      static CODE_FIELD_NAME = 'Code';
      static CODE_FIELD_NAME_NORMALIZED = 'GS1.Element.ValidationError.Code';
      static REASON_FIELD_NAME = 'Reason';
      static REASON_FIELD_NAME_NORMALIZED = 'GS1.Element.ValidationError.Reason';
      constructor(document) {
        if (document.type.name !== this.requiredDocumentType()) {
          throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
        }
        this._document = document;
      }
      requiredDocumentType() {
        return GS1.Element.ValidationError.DOCUMENT_TYPE;
      }

      /** Error Code */
      get code() {
        return this.document.fields.find(f => f.type.name === 'Code');
      }

      /** Reason */
      get reason() {
        return this.document.fields.find(f => f.type.name === 'Reason');
      }
    }
    _Element.ValidationError = ValidationError;
  })(Element || (Element = _GS.Element || (_GS.Element = {})));
})(GS1 || (exports.GS1 = GS1 = {}));
/** SEPA */
class SEPA {
  get document() {
    return this._document;
  }
  static DOCUMENT_TYPE = 'SEPA';
  static AMOUNT_FIELD_NAME = 'Amount';
  static AMOUNT_FIELD_NAME_NORMALIZED = 'SEPA.Amount';
  static CHARACTER_SET_FIELD_NAME = 'CharacterSet';
  static CHARACTER_SET_FIELD_NAME_NORMALIZED = 'SEPA.CharacterSet';
  static IDENTIFICATION_FIELD_NAME = 'Identification';
  static IDENTIFICATION_FIELD_NAME_NORMALIZED = 'SEPA.Identification';
  static INFORMATION_FIELD_NAME = 'Information';
  static INFORMATION_FIELD_NAME_NORMALIZED = 'SEPA.Information';
  static PURPOSE_FIELD_NAME = 'Purpose';
  static PURPOSE_FIELD_NAME_NORMALIZED = 'SEPA.Purpose';
  static RECEIVER_BIC_FIELD_NAME = 'ReceiverBIC';
  static RECEIVER_BIC_FIELD_NAME_NORMALIZED = 'SEPA.ReceiverBIC';
  static RECEIVER_IBAN_FIELD_NAME = 'ReceiverIBAN';
  static RECEIVER_IBAN_FIELD_NAME_NORMALIZED = 'SEPA.ReceiverIBAN';
  static RECEIVER_NAME_FIELD_NAME = 'ReceiverName';
  static RECEIVER_NAME_FIELD_NAME_NORMALIZED = 'SEPA.ReceiverName';
  static REMITTANCE_FIELD_NAME = 'Remittance';
  static REMITTANCE_FIELD_NAME_NORMALIZED = 'SEPA.Remittance';
  static SERVICE_TAG_FIELD_NAME = 'ServiceTag';
  static SERVICE_TAG_FIELD_NAME_NORMALIZED = 'SEPA.ServiceTag';
  static VERSION_FIELD_NAME = 'Version';
  static VERSION_FIELD_NAME_NORMALIZED = 'SEPA.Version';
  constructor(document) {
    if (document.type.name !== this.requiredDocumentType()) {
      throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
    }
    this._document = document;
  }
  requiredDocumentType() {
    return SEPA.DOCUMENT_TYPE;
  }

  /** Amount */
  get amount() {
    return this.document.fields.find(f => f.type.name === 'Amount');
  }

  /** Character Set */
  get characterSet() {
    return this.document.fields.find(f => f.type.name === 'CharacterSet');
  }

  /** Identification */
  get identification() {
    return this.document.fields.find(f => f.type.name === 'Identification');
  }

  /** Information */
  get information() {
    return this.document.fields.find(f => f.type.name === 'Information');
  }

  /** Amount */
  get purpose() {
    return this.document.fields.find(f => f.type.name === 'Purpose');
  }

  /** Receiver BIC */
  get receiverBIC() {
    return this.document.fields.find(f => f.type.name === 'ReceiverBIC');
  }

  /** Receiver IBAN */
  get receiverIBAN() {
    return this.document.fields.find(f => f.type.name === 'ReceiverIBAN');
  }

  /** Receiver Name */
  get receiverName() {
    return this.document.fields.find(f => f.type.name === 'ReceiverName');
  }

  /** Remittance */
  get remittance() {
    return this.document.fields.find(f => f.type.name === 'Remittance');
  }

  /** Service Tag */
  get serviceTag() {
    return this.document.fields.find(f => f.type.name === 'ServiceTag');
  }

  /** Version */
  get version() {
    return this.document.fields.find(f => f.type.name === 'Version');
  }
}

/** Medical Certificate */
exports.SEPA = SEPA;
class MedicalCertificate {
  get document() {
    return this._document;
  }
  static DOCUMENT_TYPE = 'MedicalCertificate';
  static ACCIDENT_FIELD_NAME = 'Accident';
  static ACCIDENT_FIELD_NAME_NORMALIZED = 'MedicalCertificate.Accident';
  static ASSIGNED_TO_ACCIDENT_INSURANCE_DOCTOR_FIELD_NAME = 'AssignedToAccidentInsuranceDoctor';
  static ASSIGNED_TO_ACCIDENT_INSURANCE_DOCTOR_FIELD_NAME_NORMALIZED = 'MedicalCertificate.AssignedToAccidentInsuranceDoctor';
  static BIRTH_DATE_FIELD_NAME = 'BirthDate';
  static BIRTH_DATE_FIELD_NAME_NORMALIZED = 'MedicalCertificate.BirthDate';
  static CHILD_NEEDS_CARE_FROM_FIELD_NAME = 'ChildNeedsCareFrom';
  static CHILD_NEEDS_CARE_FROM_FIELD_NAME_NORMALIZED = 'MedicalCertificate.ChildNeedsCareFrom';
  static CHILD_NEEDS_CARE_UNTIL_FIELD_NAME = 'ChildNeedsCareUntil';
  static CHILD_NEEDS_CARE_UNTIL_FIELD_NAME_NORMALIZED = 'MedicalCertificate.ChildNeedsCareUntil';
  static DIAGNOSE_FIELD_NAME = 'Diagnose';
  static DIAGNOSE_FIELD_NAME_NORMALIZED = 'MedicalCertificate.Diagnose';
  static DIAGNOSED_ON_FIELD_NAME = 'DiagnosedOn';
  static DIAGNOSED_ON_FIELD_NAME_NORMALIZED = 'MedicalCertificate.DiagnosedOn';
  static DOCTOR_NUMBER_FIELD_NAME = 'DoctorNumber';
  static DOCTOR_NUMBER_FIELD_NAME_NORMALIZED = 'MedicalCertificate.DoctorNumber';
  static DOCUMENT_DATE_FIELD_NAME = 'DocumentDate';
  static DOCUMENT_DATE_FIELD_NAME_NORMALIZED = 'MedicalCertificate.DocumentDate';
  static FIRST_NAME_FIELD_NAME = 'FirstName';
  static FIRST_NAME_FIELD_NAME_NORMALIZED = 'MedicalCertificate.FirstName';
  static HEALTH_INSURANCE_NUMBER_FIELD_NAME = 'HealthInsuranceNumber';
  static HEALTH_INSURANCE_NUMBER_FIELD_NAME_NORMALIZED = 'MedicalCertificate.HealthInsuranceNumber';
  static INCAPABLE_OF_WORK_SINCE_FIELD_NAME = 'IncapableOfWorkSince';
  static INCAPABLE_OF_WORK_SINCE_FIELD_NAME_NORMALIZED = 'MedicalCertificate.IncapableOfWorkSince';
  static INCAPABLE_OF_WORK_UNTIL_FIELD_NAME = 'IncapableOfWorkUntil';
  static INCAPABLE_OF_WORK_UNTIL_FIELD_NAME_NORMALIZED = 'MedicalCertificate.IncapableOfWorkUntil';
  static INITIAL_CERTIFICATE_FIELD_NAME = 'InitialCertificate';
  static INITIAL_CERTIFICATE_FIELD_NAME_NORMALIZED = 'MedicalCertificate.InitialCertificate';
  static INSURED_PERSON_NUMBER_FIELD_NAME = 'InsuredPersonNumber';
  static INSURED_PERSON_NUMBER_FIELD_NAME_NORMALIZED = 'MedicalCertificate.InsuredPersonNumber';
  static LAST_NAME_FIELD_NAME = 'LastName';
  static LAST_NAME_FIELD_NAME_NORMALIZED = 'MedicalCertificate.LastName';
  static PLACE_OF_OPERATION_NUMBER_FIELD_NAME = 'PlaceOfOperationNumber';
  static PLACE_OF_OPERATION_NUMBER_FIELD_NAME_NORMALIZED = 'MedicalCertificate.PlaceOfOperationNumber';
  static RENEWED_CERTIFICATE_FIELD_NAME = 'RenewedCertificate';
  static RENEWED_CERTIFICATE_FIELD_NAME_NORMALIZED = 'MedicalCertificate.RenewedCertificate';
  static REQUIRES_CARE_FIELD_NAME = 'RequiresCare';
  static REQUIRES_CARE_FIELD_NAME_NORMALIZED = 'MedicalCertificate.RequiresCare';
  static STATUS_FIELD_NAME = 'Status';
  static STATUS_FIELD_NAME_NORMALIZED = 'MedicalCertificate.Status';
  static WORK_ACCIDENT_FIELD_NAME = 'WorkAccident';
  static WORK_ACCIDENT_FIELD_NAME_NORMALIZED = 'MedicalCertificate.WorkAccident';
  constructor(document) {
    if (document.type.name !== this.requiredDocumentType()) {
      throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
    }
    this._document = document;
  }
  requiredDocumentType() {
    return MedicalCertificate.DOCUMENT_TYPE;
  }

  /** Accident */
  get accident() {
    return this.document.fields.find(f => f.type.name === 'Accident');
  }

  /** Assigned To Accident Insurance Doctor */
  get assignedToAccidentInsuranceDoctor() {
    return this.document.fields.find(f => f.type.name === 'AssignedToAccidentInsuranceDoctor');
  }

  /** Birth Date. The format is ISO8601 with delimiters */
  get birthDate() {
    return this.document.fields.find(f => f.type.name === 'BirthDate');
  }

  /** Child Needs Care From */
  get childNeedsCareFrom() {
    return this.document.fields.find(f => f.type.name === 'ChildNeedsCareFrom');
  }

  /** Child Needs Care Until */
  get childNeedsCareUntil() {
    return this.document.fields.find(f => f.type.name === 'ChildNeedsCareUntil');
  }

  /** Diagnose */
  get diagnose() {
    return this.document.fields.find(f => f.type.name === 'Diagnose');
  }

  /** Diagnosed On. The format is ISO8601 with delimiters */
  get diagnosedOn() {
    return this.document.fields.find(f => f.type.name === 'DiagnosedOn');
  }

  /** Doctor Number */
  get doctorNumber() {
    return this.document.fields.find(f => f.type.name === 'DoctorNumber');
  }

  /** Document Date. The format is ISO8601 with delimiters */
  get documentDate() {
    return this.document.fields.find(f => f.type.name === 'DocumentDate');
  }

  /** First Name */
  get firstName() {
    return this.document.fields.find(f => f.type.name === 'FirstName');
  }

  /** Health Insurance Number */
  get healthInsuranceNumber() {
    return this.document.fields.find(f => f.type.name === 'HealthInsuranceNumber');
  }

  /** Incapable Of Work Since. The format is ISO8601 with delimiters */
  get incapableOfWorkSince() {
    return this.document.fields.find(f => f.type.name === 'IncapableOfWorkSince');
  }

  /** Incapable Of Work Until. The format is ISO8601 with delimiters */
  get incapableOfWorkUntil() {
    return this.document.fields.find(f => f.type.name === 'IncapableOfWorkUntil');
  }

  /** Initial Certificate */
  get initialCertificate() {
    return this.document.fields.find(f => f.type.name === 'InitialCertificate');
  }

  /** Insured Person Number */
  get insuredPersonNumber() {
    return this.document.fields.find(f => f.type.name === 'InsuredPersonNumber');
  }

  /** Last Name */
  get lastName() {
    return this.document.fields.find(f => f.type.name === 'LastName');
  }

  /** Place Of Operation Number */
  get placeOfOperationNumber() {
    return this.document.fields.find(f => f.type.name === 'PlaceOfOperationNumber');
  }

  /** Renewed Certificate */
  get renewedCertificate() {
    return this.document.fields.find(f => f.type.name === 'RenewedCertificate');
  }

  /** Requires Care */
  get requiresCare() {
    return this.document.fields.find(f => f.type.name === 'RequiresCare');
  }

  /** Status */
  get status() {
    return this.document.fields.find(f => f.type.name === 'Status');
  }

  /** Work Accident */
  get workAccident() {
    return this.document.fields.find(f => f.type.name === 'WorkAccident');
  }
}

/** VCard */
exports.MedicalCertificate = MedicalCertificate;
class VCard {
  get document() {
    return this._document;
  }
  static DOCUMENT_TYPE = 'VCard';
  constructor(document) {
    if (document.type.name !== this.requiredDocumentType()) {
      throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
    }
    this._document = document;
  }
  requiredDocumentType() {
    return VCard.DOCUMENT_TYPE;
  }

  /** The child document of type "Version". */
  get version() {
    const child = this.document.children.find(c => c.type.name === 'Version');
    return new VCard.Version(child);
  }

  /** An array of all children of type "Source". */
  get sources() {
    return this.document.children.filter(c => c.type.name === 'Source').map(c => new VCard.Source(c));
  }

  /** The child document of type "Kind". */
  get kind() {
    const child = this.document.children.find(c => c.type.name === 'Kind');
    return child ? new VCard.Kind(child) : undefined;
  }

  /** An array of all children of type "XML". */
  get xmls() {
    return this.document.children.filter(c => c.type.name === 'XML').map(c => new VCard.XML(c));
  }

  /** The child document of type "Name". */
  get name() {
    const child = this.document.children.find(c => c.type.name === 'Name');
    return child ? new VCard.Name(child) : undefined;
  }

  /** The child document of type "FormattedName". */
  get formattedName() {
    const child = this.document.children.find(c => c.type.name === 'FormattedName');
    return child ? new VCard.FormattedName(child) : undefined;
  }

  /** An array of all children of type "Nickname". */
  get nicknames() {
    return this.document.children.filter(c => c.type.name === 'Nickname').map(c => new VCard.Nickname(c));
  }

  /** The child document of type "Birthday". */
  get birthday() {
    const child = this.document.children.find(c => c.type.name === 'Birthday');
    return child ? new VCard.Birthday(child) : undefined;
  }

  /** The child document of type "Anniversary". */
  get anniversary() {
    const child = this.document.children.find(c => c.type.name === 'Anniversary');
    return child ? new VCard.Anniversary(child) : undefined;
  }

  /** The child document of type "Gender". */
  get gender() {
    const child = this.document.children.find(c => c.type.name === 'Gender');
    return child ? new VCard.Gender(child) : undefined;
  }

  /** An array of all children of type "DeliveryAddress". */
  get deliveryAddresses() {
    return this.document.children.filter(c => c.type.name === 'DeliveryAddress').map(c => new VCard.DeliveryAddress(c));
  }

  /** An array of all children of type "Label". */
  get labels() {
    return this.document.children.filter(c => c.type.name === 'Label').map(c => new VCard.Label(c));
  }

  /** An array of all children of type "Photo". */
  get photos() {
    return this.document.children.filter(c => c.type.name === 'Photo').map(c => new VCard.Photo(c));
  }

  /** An array of all children of type "TelephoneNumber". */
  get telephoneNumbers() {
    return this.document.children.filter(c => c.type.name === 'TelephoneNumber').map(c => new VCard.TelephoneNumber(c));
  }

  /** An array of all children of type "Email". */
  get emails() {
    return this.document.children.filter(c => c.type.name === 'Email').map(c => new VCard.Email(c));
  }

  /** An array of all children of type "IMPP". */
  get impps() {
    return this.document.children.filter(c => c.type.name === 'IMPP').map(c => new VCard.IMPP(c));
  }

  /** An array of all children of type "Language". */
  get languages() {
    return this.document.children.filter(c => c.type.name === 'Language').map(c => new VCard.Language(c));
  }

  /** An array of all children of type "TimeZone". */
  get timeZones() {
    return this.document.children.filter(c => c.type.name === 'TimeZone').map(c => new VCard.TimeZone(c));
  }

  /** An array of all children of type "GeoLocation". */
  get geoLocations() {
    return this.document.children.filter(c => c.type.name === 'GeoLocation').map(c => new VCard.GeoLocation(c));
  }

  /** An array of all children of type "Title". */
  get titles() {
    return this.document.children.filter(c => c.type.name === 'Title').map(c => new VCard.Title(c));
  }

  /** An array of all children of type "Role". */
  get roles() {
    return this.document.children.filter(c => c.type.name === 'Role').map(c => new VCard.Role(c));
  }

  /** An array of all children of type "Logo". */
  get logos() {
    return this.document.children.filter(c => c.type.name === 'Logo').map(c => new VCard.Logo(c));
  }

  /** An array of all children of type "Agent". */
  get agents() {
    return this.document.children.filter(c => c.type.name === 'Agent').map(c => new VCard.Agent(c));
  }

  /** An array of all children of type "Organisation". */
  get organisations() {
    return this.document.children.filter(c => c.type.name === 'Organisation').map(c => new VCard.Organisation(c));
  }

  /** An array of all children of type "Member". */
  get members() {
    return this.document.children.filter(c => c.type.name === 'Member').map(c => new VCard.Member(c));
  }

  /** An array of all children of type "Related". */
  get relateds() {
    return this.document.children.filter(c => c.type.name === 'Related').map(c => new VCard.Related(c));
  }

  /** An array of all children of type "Category". */
  get categories() {
    return this.document.children.filter(c => c.type.name === 'Category').map(c => new VCard.Category(c));
  }

  /** An array of all children of type "Note". */
  get notes() {
    return this.document.children.filter(c => c.type.name === 'Note').map(c => new VCard.Note(c));
  }

  /** The child document of type "ProductId". */
  get productId() {
    const child = this.document.children.find(c => c.type.name === 'ProductId');
    return child ? new VCard.ProductId(child) : undefined;
  }

  /** The child document of type "Revision". */
  get revision() {
    const child = this.document.children.find(c => c.type.name === 'Revision');
    return child ? new VCard.Revision(child) : undefined;
  }

  /** An array of all children of type "Sound". */
  get sounds() {
    return this.document.children.filter(c => c.type.name === 'Sound').map(c => new VCard.Sound(c));
  }

  /** The child document of type "UID". */
  get uid() {
    const child = this.document.children.find(c => c.type.name === 'UID');
    return child ? new VCard.UID(child) : undefined;
  }

  /** An array of all children of type "ClientPIDMap". */
  get clientPIDMaps() {
    return this.document.children.filter(c => c.type.name === 'ClientPIDMap').map(c => new VCard.ClientPIDMap(c));
  }

  /** An array of all children of type "URL". */
  get urls() {
    return this.document.children.filter(c => c.type.name === 'URL').map(c => new VCard.URL(c));
  }

  /** An array of all children of type "PublicKey". */
  get publicKeys() {
    return this.document.children.filter(c => c.type.name === 'PublicKey').map(c => new VCard.PublicKey(c));
  }

  /** An array of all children of type "BusyTimeURL". */
  get busyTimeURLs() {
    return this.document.children.filter(c => c.type.name === 'BusyTimeURL').map(c => new VCard.BusyTimeURL(c));
  }

  /** An array of all children of type "CalendarURIForRequests". */
  get calendarURIForRequestss() {
    return this.document.children.filter(c => c.type.name === 'CalendarURIForRequests').map(c => new VCard.CalendarURIForRequests(c));
  }

  /** An array of all children of type "CalendarURI". */
  get calendarURIs() {
    return this.document.children.filter(c => c.type.name === 'CalendarURI').map(c => new VCard.CalendarURI(c));
  }

  /** The child document of type "SortString". */
  get sortString() {
    const child = this.document.children.find(c => c.type.name === 'SortString');
    return child ? new VCard.SortString(child) : undefined;
  }

  /** The child document of type "Classification". */
  get classification() {
    const child = this.document.children.find(c => c.type.name === 'Classification');
    return child ? new VCard.Classification(child) : undefined;
  }

  /** An array of all children of type "Custom". */
  get customs() {
    return this.document.children.filter(c => c.type.name === 'Custom').map(c => new VCard.Custom(c));
  }
}
exports.VCard = VCard;
(function (_VCard) {
  class Entry {
    get document() {
      return this._document;
    }
    static RAW_VALUE_FIELD_NAME = 'RawValue';
    static RAW_VALUE_FIELD_NAME_NORMALIZED = 'VCard.Entry.RawValue';
    static TYPE_MODIFIER_FIELD_NAME = 'TypeModifier';
    static TYPE_MODIFIER_FIELD_NAME_NORMALIZED = 'VCard.Entry.TypeModifier';
    static VALUE_FIELD_NAME = 'Value';
    static VALUE_FIELD_NAME_NORMALIZED = 'VCard.Entry.Value';
    constructor(document) {
      if (document.type.name !== this.requiredDocumentType()) {
        throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
      }
      this._document = document;
    }
    /** Raw Value */
    get rawValue() {
      return this.document.fields.find(f => f.type.name === 'RawValue');
    }

    /** Type Modifier */
    get typeModifiers() {
      return this.document.fields.filter(f => f.type.name === 'TypeModifier');
    }

    /** Value */
    get values() {
      return this.document.fields.filter(f => f.type.name === 'Value');
    }
  }
  _VCard.Entry = Entry;
  class Version extends Entry {
    static DOCUMENT_TYPE = 'Version';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Version.DOCUMENT_TYPE;
    }
  }
  _VCard.Version = Version;
  class Source extends Entry {
    static DOCUMENT_TYPE = 'Source';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Source.DOCUMENT_TYPE;
    }
  }
  _VCard.Source = Source;
  class Kind extends Entry {
    static DOCUMENT_TYPE = 'Kind';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Kind.DOCUMENT_TYPE;
    }
  }
  _VCard.Kind = Kind;
  class XML extends Entry {
    static DOCUMENT_TYPE = 'XML';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.XML.DOCUMENT_TYPE;
    }
  }
  _VCard.XML = XML;
  class Name extends Entry {
    static DOCUMENT_TYPE = 'Name';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Name.DOCUMENT_TYPE;
    }
  }
  _VCard.Name = Name;
  class FormattedName extends Entry {
    static DOCUMENT_TYPE = 'FormattedName';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.FormattedName.DOCUMENT_TYPE;
    }
  }
  _VCard.FormattedName = FormattedName;
  class Nickname extends Entry {
    static DOCUMENT_TYPE = 'Nickname';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Nickname.DOCUMENT_TYPE;
    }
  }
  _VCard.Nickname = Nickname;
  class Birthday extends Entry {
    static DOCUMENT_TYPE = 'Birthday';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Birthday.DOCUMENT_TYPE;
    }
  }
  _VCard.Birthday = Birthday;
  class Anniversary extends Entry {
    static DOCUMENT_TYPE = 'Anniversary';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Anniversary.DOCUMENT_TYPE;
    }
  }
  _VCard.Anniversary = Anniversary;
  class Gender extends Entry {
    static DOCUMENT_TYPE = 'Gender';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Gender.DOCUMENT_TYPE;
    }
  }
  _VCard.Gender = Gender;
  class DeliveryAddress extends Entry {
    static DOCUMENT_TYPE = 'DeliveryAddress';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.DeliveryAddress.DOCUMENT_TYPE;
    }
  }
  _VCard.DeliveryAddress = DeliveryAddress;
  class Label extends Entry {
    static DOCUMENT_TYPE = 'Label';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Label.DOCUMENT_TYPE;
    }
  }
  _VCard.Label = Label;
  class Photo extends Entry {
    static DOCUMENT_TYPE = 'Photo';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Photo.DOCUMENT_TYPE;
    }
  }
  _VCard.Photo = Photo;
  class TelephoneNumber extends Entry {
    static DOCUMENT_TYPE = 'TelephoneNumber';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.TelephoneNumber.DOCUMENT_TYPE;
    }
  }
  _VCard.TelephoneNumber = TelephoneNumber;
  class Email extends Entry {
    static DOCUMENT_TYPE = 'Email';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Email.DOCUMENT_TYPE;
    }
  }
  _VCard.Email = Email;
  class IMPP extends Entry {
    static DOCUMENT_TYPE = 'IMPP';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.IMPP.DOCUMENT_TYPE;
    }
  }
  _VCard.IMPP = IMPP;
  class Language extends Entry {
    static DOCUMENT_TYPE = 'Language';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Language.DOCUMENT_TYPE;
    }
  }
  _VCard.Language = Language;
  class TimeZone extends Entry {
    static DOCUMENT_TYPE = 'TimeZone';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.TimeZone.DOCUMENT_TYPE;
    }
  }
  _VCard.TimeZone = TimeZone;
  class GeoLocation extends Entry {
    static DOCUMENT_TYPE = 'GeoLocation';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.GeoLocation.DOCUMENT_TYPE;
    }
  }
  _VCard.GeoLocation = GeoLocation;
  class Title extends Entry {
    static DOCUMENT_TYPE = 'Title';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Title.DOCUMENT_TYPE;
    }
  }
  _VCard.Title = Title;
  class Role extends Entry {
    static DOCUMENT_TYPE = 'Role';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Role.DOCUMENT_TYPE;
    }
  }
  _VCard.Role = Role;
  class Logo extends Entry {
    static DOCUMENT_TYPE = 'Logo';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Logo.DOCUMENT_TYPE;
    }
  }
  _VCard.Logo = Logo;
  class Agent extends Entry {
    static DOCUMENT_TYPE = 'Agent';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Agent.DOCUMENT_TYPE;
    }
  }
  _VCard.Agent = Agent;
  class Organisation extends Entry {
    static DOCUMENT_TYPE = 'Organisation';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Organisation.DOCUMENT_TYPE;
    }
  }
  _VCard.Organisation = Organisation;
  class Member extends Entry {
    static DOCUMENT_TYPE = 'Member';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Member.DOCUMENT_TYPE;
    }
  }
  _VCard.Member = Member;
  class Related extends Entry {
    static DOCUMENT_TYPE = 'Related';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Related.DOCUMENT_TYPE;
    }
  }
  _VCard.Related = Related;
  class Category extends Entry {
    static DOCUMENT_TYPE = 'Category';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Category.DOCUMENT_TYPE;
    }
  }
  _VCard.Category = Category;
  class Note extends Entry {
    static DOCUMENT_TYPE = 'Note';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Note.DOCUMENT_TYPE;
    }
  }
  _VCard.Note = Note;
  class ProductId extends Entry {
    static DOCUMENT_TYPE = 'ProductId';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.ProductId.DOCUMENT_TYPE;
    }
  }
  _VCard.ProductId = ProductId;
  class Revision extends Entry {
    static DOCUMENT_TYPE = 'Revision';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Revision.DOCUMENT_TYPE;
    }
  }
  _VCard.Revision = Revision;
  class Sound extends Entry {
    static DOCUMENT_TYPE = 'Sound';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Sound.DOCUMENT_TYPE;
    }
  }
  _VCard.Sound = Sound;
  class UID extends Entry {
    static DOCUMENT_TYPE = 'UID';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.UID.DOCUMENT_TYPE;
    }
  }
  _VCard.UID = UID;
  class ClientPIDMap extends Entry {
    static DOCUMENT_TYPE = 'ClientPIDMap';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.ClientPIDMap.DOCUMENT_TYPE;
    }
  }
  _VCard.ClientPIDMap = ClientPIDMap;
  class URL extends Entry {
    static DOCUMENT_TYPE = 'URL';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.URL.DOCUMENT_TYPE;
    }
  }
  _VCard.URL = URL;
  class PublicKey extends Entry {
    static DOCUMENT_TYPE = 'PublicKey';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.PublicKey.DOCUMENT_TYPE;
    }
  }
  _VCard.PublicKey = PublicKey;
  class BusyTimeURL extends Entry {
    static DOCUMENT_TYPE = 'BusyTimeURL';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.BusyTimeURL.DOCUMENT_TYPE;
    }
  }
  _VCard.BusyTimeURL = BusyTimeURL;
  class CalendarURIForRequests extends Entry {
    static DOCUMENT_TYPE = 'CalendarURIForRequests';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.CalendarURIForRequests.DOCUMENT_TYPE;
    }
  }
  _VCard.CalendarURIForRequests = CalendarURIForRequests;
  class CalendarURI extends Entry {
    static DOCUMENT_TYPE = 'CalendarURI';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.CalendarURI.DOCUMENT_TYPE;
    }
  }
  _VCard.CalendarURI = CalendarURI;
  class SortString extends Entry {
    static DOCUMENT_TYPE = 'SortString';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.SortString.DOCUMENT_TYPE;
    }
  }
  _VCard.SortString = SortString;
  class Classification extends Entry {
    static DOCUMENT_TYPE = 'Classification';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Classification.DOCUMENT_TYPE;
    }
  }
  _VCard.Classification = Classification;
  class Custom extends Entry {
    static DOCUMENT_TYPE = 'Custom';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return VCard.Custom.DOCUMENT_TYPE;
    }
  }
  _VCard.Custom = Custom;
})(VCard || (exports.VCard = VCard = {}));
/** British Columbia (Canada) driver license */
class BritishColumbiaDriverLicense {
  get document() {
    return this._document;
  }
  static DOCUMENT_TYPE = 'BritishColumbiaDriverLicense';
  static ADDRESS_FIELD_NAME = 'Address';
  static ADDRESS_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.Address';
  static BIRTH_DATE_FIELD_NAME = 'BirthDate';
  static BIRTH_DATE_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.BirthDate';
  static CARD_EXPIRY_FIELD_NAME = 'CardExpiry';
  static CARD_EXPIRY_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.CardExpiry';
  static CARD_NUMBER_FIELD_NAME = 'CardNumber';
  static CARD_NUMBER_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.CardNumber';
  static CITY_FIELD_NAME = 'City';
  static CITY_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.City';
  static ERROR_CONTROL_CODE_FIELD_NAME = 'ErrorControlCode';
  static ERROR_CONTROL_CODE_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.ErrorControlCode';
  static EYE_COLOR_FIELD_NAME = 'EyeColor';
  static EYE_COLOR_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.EyeColor';
  static GIVEN_NAMES_FIELD_NAME = 'GivenNames';
  static GIVEN_NAMES_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.GivenNames';
  static HAIR_COLOR_FIELD_NAME = 'HairColor';
  static HAIR_COLOR_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.HairColor';
  static HEIGHT_FIELD_NAME = 'Height';
  static HEIGHT_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.Height';
  static LAST_NAME_FIELD_NAME = 'LastName';
  static LAST_NAME_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.LastName';
  static PERSONAL_HEALTH_NUMBER_FIELD_NAME = 'PersonalHealthNumber';
  static PERSONAL_HEALTH_NUMBER_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.PersonalHealthNumber';
  static POSTAL_CODE_FIELD_NAME = 'PostalCode';
  static POSTAL_CODE_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.PostalCode';
  static PROVINCE_FIELD_NAME = 'Province';
  static PROVINCE_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.Province';
  static PROVINCE_CODE_FIELD_NAME = 'ProvinceCode';
  static PROVINCE_CODE_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.ProvinceCode';
  static SECURITY_FUNCTION_FIELD_NAME = 'SecurityFunction';
  static SECURITY_FUNCTION_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.SecurityFunction';
  static SECURITY_VERSION_FIELD_NAME = 'SecurityVersion';
  static SECURITY_VERSION_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.SecurityVersion';
  static SEX_FIELD_NAME = 'Sex';
  static SEX_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.Sex';
  static VERSION_FIELD_NAME = 'Version';
  static VERSION_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.Version';
  static WEIGHT_FIELD_NAME = 'Weight';
  static WEIGHT_FIELD_NAME_NORMALIZED = 'BritishColumbiaDriverLicense.Weight';
  constructor(document) {
    if (document.type.name !== this.requiredDocumentType()) {
      throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
    }
    this._document = document;
  }
  requiredDocumentType() {
    return BritishColumbiaDriverLicense.DOCUMENT_TYPE;
  }

  /** Address */
  get address() {
    return this.document.fields.find(f => f.type.name === 'Address');
  }

  /** Birth date in CCYYMMDD format */
  get birthDate() {
    return this.document.fields.find(f => f.type.name === 'BirthDate');
  }

  /** Card expiry in YYMM format */
  get cardExpiry() {
    return this.document.fields.find(f => f.type.name === 'CardExpiry');
  }

  /** Card Number */
  get cardNumber() {
    return this.document.fields.find(f => f.type.name === 'CardNumber');
  }

  /** City */
  get city() {
    return this.document.fields.find(f => f.type.name === 'City');
  }

  /** Error control code */
  get errorControlCode() {
    return this.document.fields.find(f => f.type.name === 'ErrorControlCode');
  }

  /** Eye color */
  get eyeColor() {
    return this.document.fields.find(f => f.type.name === 'EyeColor');
  }

  /** Given names */
  get givenNames() {
    return this.document.fields.find(f => f.type.name === 'GivenNames');
  }

  /** Hair color */
  get hairColor() {
    return this.document.fields.find(f => f.type.name === 'HairColor');
  }

  /** Height */
  get height() {
    return this.document.fields.find(f => f.type.name === 'Height');
  }

  /** Last name */
  get lastName() {
    return this.document.fields.find(f => f.type.name === 'LastName');
  }

  /** Personal health number */
  get personalHealthNumber() {
    return this.document.fields.find(f => f.type.name === 'PersonalHealthNumber');
  }

  /** Postal code */
  get postalCode() {
    return this.document.fields.find(f => f.type.name === 'PostalCode');
  }

  /** Province */
  get province() {
    return this.document.fields.find(f => f.type.name === 'Province');
  }

  /** Province code */
  get provinceCode() {
    return this.document.fields.find(f => f.type.name === 'ProvinceCode');
  }

  /** Security function */
  get securityFunction() {
    return this.document.fields.find(f => f.type.name === 'SecurityFunction');
  }

  /** Security Version */
  get securityVersion() {
    return this.document.fields.find(f => f.type.name === 'SecurityVersion');
  }

  /** Sex */
  get sex() {
    return this.document.fields.find(f => f.type.name === 'Sex');
  }

  /** Version */
  get version() {
    return this.document.fields.find(f => f.type.name === 'Version');
  }

  /** Weight */
  get weight() {
    return this.document.fields.find(f => f.type.name === 'Weight');
  }
}

/** AAMVA */
exports.BritishColumbiaDriverLicense = BritishColumbiaDriverLicense;
class AAMVA {
  get document() {
    return this._document;
  }
  static DOCUMENT_TYPE = 'AAMVA';
  static ISSUER_IDENTIFICATION_NUMBER_FIELD_NAME = 'IssuerIdentificationNumber';
  static ISSUER_IDENTIFICATION_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.IssuerIdentificationNumber';
  static JURISDICTION_VERSION_NUMBER_FIELD_NAME = 'JurisdictionVersionNumber';
  static JURISDICTION_VERSION_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.JurisdictionVersionNumber';
  static VERSION_FIELD_NAME = 'Version';
  static VERSION_FIELD_NAME_NORMALIZED = 'AAMVA.Version';
  constructor(document) {
    if (document.type.name !== this.requiredDocumentType()) {
      throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
    }
    this._document = document;
  }
  requiredDocumentType() {
    return AAMVA.DOCUMENT_TYPE;
  }

  /** Issuer Identification Number */
  get issuerIdentificationNumber() {
    return this.document.fields.find(f => f.type.name === 'IssuerIdentificationNumber');
  }

  /** Jurisdiction Version Number */
  get jurisdictionVersionNumber() {
    return this.document.fields.find(f => f.type.name === 'JurisdictionVersionNumber');
  }

  /** Version */
  get version() {
    return this.document.fields.find(f => f.type.name === 'Version');
  }

  /** The child document of type "TitleData". */
  get titleData() {
    const child = this.document.children.find(c => c.type.name === 'TitleData');
    return child ? new AAMVA.TitleData(child) : undefined;
  }

  /** The child document of type "RegistrationData". */
  get registrationData() {
    const child = this.document.children.find(c => c.type.name === 'RegistrationData');
    return child ? new AAMVA.RegistrationData(child) : undefined;
  }

  /** The child document of type "MotorCarrierData". */
  get motorCarrierData() {
    const child = this.document.children.find(c => c.type.name === 'MotorCarrierData');
    return child ? new AAMVA.MotorCarrierData(child) : undefined;
  }

  /** The child document of type "RegistrantAndVehicleData". */
  get registrantAndVehicleData() {
    const child = this.document.children.find(c => c.type.name === 'RegistrantAndVehicleData');
    return child ? new AAMVA.RegistrantAndVehicleData(child) : undefined;
  }

  /** The child document of type "VehicleOwnerData". */
  get vehicleOwnerData() {
    const child = this.document.children.find(c => c.type.name === 'VehicleOwnerData');
    return child ? new AAMVA.VehicleOwnerData(child) : undefined;
  }

  /** The child document of type "VehicleData". */
  get vehicleData() {
    const child = this.document.children.find(c => c.type.name === 'VehicleData');
    return child ? new AAMVA.VehicleData(child) : undefined;
  }

  /** The child document of type "VehicleSafetyInspectionData". */
  get vehicleSafetyInspectionData() {
    const child = this.document.children.find(c => c.type.name === 'VehicleSafetyInspectionData');
    return child ? new AAMVA.VehicleSafetyInspectionData(child) : undefined;
  }

  /** The child document of type "DriverLicense". */
  get driverLicense() {
    const child = this.document.children.find(c => c.type.name === 'DriverLicense');
    return child ? new AAMVA.DriverLicense(child) : undefined;
  }

  /** The child document of type "IDCard". */
  get idCard() {
    const child = this.document.children.find(c => c.type.name === 'IDCard');
    return child ? new AAMVA.IDCard(child) : undefined;
  }

  /** The child document of type "EnhancedDriverLicense". */
  get enhancedDriverLicense() {
    const child = this.document.children.find(c => c.type.name === 'EnhancedDriverLicense');
    return child ? new AAMVA.EnhancedDriverLicense(child) : undefined;
  }

  /** The child document of type "RawDocument". */
  get rawDocument() {
    const child = this.document.children.find(c => c.type.name === 'RawDocument');
    return new AAMVA.RawDocument(child);
  }
}
exports.AAMVA = AAMVA;
(function (_AAMVA) {
  class TitleData {
    get document() {
      return this._document;
    }
    static DOCUMENT_TYPE = 'TitleData';
    static ADDRESS_CITY_FIELD_NAME = 'AddressCity';
    static ADDRESS_CITY_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.AddressCity';
    static ADDRESS_JURISDICTION_CODE_FIELD_NAME = 'AddressJurisdictionCode';
    static ADDRESS_JURISDICTION_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.AddressJurisdictionCode';
    static ADDRESS_STREET_FIELD_NAME = 'AddressStreet';
    static ADDRESS_STREET_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.AddressStreet';
    static ADDRESS_ZIP_CODE_FIELD_NAME = 'AddressZipCode';
    static ADDRESS_ZIP_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.AddressZipCode';
    static BUSINESS_NAME_FIELD_NAME = 'BusinessName';
    static BUSINESS_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.BusinessName';
    static FAMILY_NAME_FIELD_NAME = 'FamilyName';
    static FAMILY_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.FamilyName';
    static FIRST_LIEN_HOLDER_ID_FIELD_NAME = 'FirstLienHolderId';
    static FIRST_LIEN_HOLDER_ID_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.FirstLienHolderId';
    static FIRST_LIEN_HOLDER_NAME_FIELD_NAME = 'FirstLienHolderName';
    static FIRST_LIEN_HOLDER_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.FirstLienHolderName';
    static GIVEN_NAME_FIELD_NAME = 'GivenName';
    static GIVEN_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.GivenName';
    static NEW_USED_INDICATOR_FIELD_NAME = 'NewUsedIndicator';
    static NEW_USED_INDICATOR_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.NewUsedIndicator';
    static ODOMETER_DATE_FIELD_NAME = 'OdometerDate';
    static ODOMETER_DATE_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.OdometerDate';
    static ODOMETER_DISCLOSURE_FIELD_NAME = 'OdometerDisclosure';
    static ODOMETER_DISCLOSURE_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.OdometerDisclosure';
    static ODOMETER_READING_KILOMETERS_FIELD_NAME = 'OdometerReadingKilometers';
    static ODOMETER_READING_KILOMETERS_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.OdometerReadingKilometers';
    static ODOMETER_READING_MILEAGE_FIELD_NAME = 'OdometerReadingMileage';
    static ODOMETER_READING_MILEAGE_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.OdometerReadingMileage';
    static PREVIOUS_TITLE_NUMBER_FIELD_NAME = 'PreviousTitleNumber';
    static PREVIOUS_TITLE_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.PreviousTitleNumber';
    static PREVIOUS_TITLING_JURISDICTION_FIELD_NAME = 'PreviousTitlingJurisdiction';
    static PREVIOUS_TITLING_JURISDICTION_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.PreviousTitlingJurisdiction';
    static TITLE_BRAND_FIELD_NAME = 'TitleBrand';
    static TITLE_BRAND_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.TitleBrand';
    static TITLE_ISSUE_DATE_FIELD_NAME = 'TitleIssueDate';
    static TITLE_ISSUE_DATE_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.TitleIssueDate';
    static TITLE_NUMBER_FIELD_NAME = 'TitleNumber';
    static TITLE_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.TitleNumber';
    static TITLING_JURISDICTION_FIELD_NAME = 'TitlingJurisdiction';
    static TITLING_JURISDICTION_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.TitlingJurisdiction';
    static VEHICLE_BODY_STYLE_FIELD_NAME = 'VehicleBodyStyle';
    static VEHICLE_BODY_STYLE_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.VehicleBodyStyle';
    static VEHICLE_COLOR_FIELD_NAME = 'VehicleColor';
    static VEHICLE_COLOR_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.VehicleColor';
    static VEHICLE_IDENTIFICATION_NUMBER_FIELD_NAME = 'VehicleIdentificationNumber';
    static VEHICLE_IDENTIFICATION_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.VehicleIdentificationNumber';
    static VEHICLE_MAKE_FIELD_NAME = 'VehicleMake';
    static VEHICLE_MAKE_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.VehicleMake';
    static VEHICLE_MODEL_FIELD_NAME = 'VehicleModel';
    static VEHICLE_MODEL_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.VehicleModel';
    static VEHICLE_MODEL_YEAR_FIELD_NAME = 'VehicleModelYear';
    static VEHICLE_MODEL_YEAR_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.VehicleModelYear';
    static VEHICLE_PURCHASE_DATE_FIELD_NAME = 'VehiclePurchaseDate';
    static VEHICLE_PURCHASE_DATE_FIELD_NAME_NORMALIZED = 'AAMVA.TitleData.VehiclePurchaseDate';
    constructor(document) {
      if (document.type.name !== this.requiredDocumentType()) {
        throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
      }
      this._document = document;
    }
    requiredDocumentType() {
      return AAMVA.TitleData.DOCUMENT_TYPE;
    }

    /** Data element ID "NAT". City portion of the owner’s address. */
    get addressCity() {
      return this.document.fields.find(f => f.type.name === 'AddressCity');
    }

    /** Data element ID "NAU". Jurisdiction portion of the owner’s address. */
    get addressJurisdictionCode() {
      return this.document.fields.find(f => f.type.name === 'AddressJurisdictionCode');
    }

    /** Data element ID "NAR". Street portion of the owner’s address. */
    get addressStreet() {
      return this.document.fields.find(f => f.type.name === 'AddressStreet');
    }

    /** Data element ID "NAV". The ZIP code or Postal code portion of the owner’s address. */
    get addressZipCode() {
      return this.document.fields.find(f => f.type.name === 'AddressZipCode');
    }

    /** Data element ID "BBC". The name of business that owns the vehicle. */
    get businessName() {
      return this.document.fields.find(f => f.type.name === 'BusinessName');
    }

    /** Data element ID "NAA". Family name (commonly called surname or last name) of the owner of the vehicle. */
    get familyName() {
      return this.document.fields.find(f => f.type.name === 'FamilyName');
    }

    /** Data element ID "LAF". A code that uniquely identifies the first holder of a lien. */
    get firstLienHolderId() {
      return this.document.fields.find(f => f.type.name === 'FirstLienHolderId');
    }

    /** Data element ID "LAA". Name of the first lien holder of the vehicle. */
    get firstLienHolderName() {
      return this.document.fields.find(f => f.type.name === 'FirstLienHolderName');
    }

    /** Data element ID "NAE". Given name or names (includes all of what are commonly referred to as first and middle names) of the owner of the vehicle. */
    get givenName() {
      return this.document.fields.find(f => f.type.name === 'GivenName');
    }

    /** Data element ID "TAW". This code represents whether the vehicle/vessel is new or used. Note: jurisdictions’ definitions of these classifications may vary a little due to state regulations on demo vehicles, slates between dealers, application of state taxes, etc. N = New, U = Used. */
    get newUsedIndicator() {
      return this.document.fields.find(f => f.type.name === 'NewUsedIndicator');
    }

    /** Data element ID "TAH". The date the odometer reading was recorded by the jurisdiction. */
    get odometerDate() {
      return this.document.fields.find(f => f.type.name === 'OdometerDate');
    }

    /** Data element ID "TAG". This is the federal odometer mileage disclosure. The mandatory information is: (1) Actual vehicle mileage; (2) Mileage exceeds mechanical limitations; (3) Not actual mileage; (4) Mileage disclosure not required. */
    get odometerDisclosure() {
      return this.document.fields.find(f => f.type.name === 'OdometerDisclosure');
    }

    /** Data element ID "TAI". This is the odometer reading registered with the DMV either at the time of titling or registration renewal in kilometers. */
    get odometerReadingKilometers() {
      return this.document.fields.find(f => f.type.name === 'OdometerReadingKilometers');
    }

    /** Data element ID "TAF". This is the odometer reading registered with the DMV either at the time of titling or registration renewal. */
    get odometerReadingMileage() {
      return this.document.fields.find(f => f.type.name === 'OdometerReadingMileage');
    }

    /** Data element ID "TAZ". The title number assigned to the vehicle by the previous titling jurisdiction. */
    get previousTitleNumber() {
      return this.document.fields.find(f => f.type.name === 'PreviousTitleNumber');
    }

    /** Data element ID "TPJ". The code for the jurisdiction (U.S., Canadian, or Mexican) that titled the vehicle immediately prior to the current titling jurisdiction. */
    get previousTitlingJurisdiction() {
      return this.document.fields.find(f => f.type.name === 'PreviousTitlingJurisdiction');
    }

    /** Data element ID "TAY". Code providing information about the brand applied to the title. */
    get titleBrand() {
      return this.document.fields.find(f => f.type.name === 'TitleBrand');
    }

    /** Data element ID "TAV". The date the jurisdiction’s titling authority issued a title to the owner of the vehicle. The format is CCYYMMDD. */
    get titleIssueDate() {
      return this.document.fields.find(f => f.type.name === 'TitleIssueDate');
    }

    /** Data element ID "TAA". The date the jurisdiction’s titling authority issued a title to the owner of the vehicle. The format is CCYYMMDD. */
    get titleNumber() {
      return this.document.fields.find(f => f.type.name === 'TitleNumber');
    }

    /** Data element ID "TAC". A unique set of alphanumeric characters assigned by the titling jurisdiction to the certificate of title of a vehicle. */
    get titlingJurisdiction() {
      return this.document.fields.find(f => f.type.name === 'TitlingJurisdiction');
    }

    /** Data element ID "VAO". The general configuration or shape of a vehicle distinguished by characteristics such as number of doors, seats, windows, roofline, and type of top. The vehicle body type is 2-character alphanumeric. */
    get vehicleBodyStyle() {
      return this.document.fields.find(f => f.type.name === 'VehicleBodyStyle');
    }

    /** Data element ID "VBD". Where the vehicle/vessel is one color, this is the appropriate code describing that color. When the vehicle is two colors, this is the code for the top-most or front-most color. */
    get vehicleColor() {
      return this.document.fields.find(f => f.type.name === 'VehicleColor');
    }

    /** Data element ID "VAD". A unique combination of alphanumeric characters that identifies a specific vehicle or component. The VIN is affixed to the vehicle in specific locations and formulated by the manufacturer. State agencies under some controlled instances may assign a VIN to a vehicle. */
    get vehicleIdentificationNumber() {
      return this.document.fields.find(f => f.type.name === 'VehicleIdentificationNumber');
    }

    /** Data element ID "VAK". The distinctive (coded) name applied to a group of vehicles by a manufacturer. */
    get vehicleMake() {
      return this.document.fields.find(f => f.type.name === 'VehicleMake');
    }

    /** Data element ID "VAM". A code denoting a family of vehicles (within a make), which has a degree of similarity in construction, such as body, chassis, etc. The field does not necessarily contain a standard code; it may contain a value provided by the originator of the field. */
    get vehicleModel() {
      return this.document.fields.find(f => f.type.name === 'VehicleModel');
    }

    /** Data element ID "VAL". The year that is assigned to a vehicle by the manufacturer. The format is CCYY. */
    get vehicleModelYear() {
      return this.document.fields.find(f => f.type.name === 'VehicleModelYear');
    }

    /** Data element ID "TAU". The date a vehicle was purchased by the current owner. The format is ISO8601 with delimiters. */
    get vehiclePurchaseDate() {
      return this.document.fields.find(f => f.type.name === 'VehiclePurchaseDate');
    }
  }
  _AAMVA.TitleData = TitleData;
  class RegistrationData {
    get document() {
      return this._document;
    }
    static DOCUMENT_TYPE = 'RegistrationData';
    static ADDRESS_CITY_FIELD_NAME = 'AddressCity';
    static ADDRESS_CITY_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.AddressCity';
    static ADDRESS_JURISDICTION_CODE_FIELD_NAME = 'AddressJurisdictionCode';
    static ADDRESS_JURISDICTION_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.AddressJurisdictionCode';
    static ADDRESS_STREET_FIELD_NAME = 'AddressStreet';
    static ADDRESS_STREET_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.AddressStreet';
    static ADDRESS_ZIP_CODE_FIELD_NAME = 'AddressZipCode';
    static ADDRESS_ZIP_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.AddressZipCode';
    static AXLES_FIELD_NAME = 'Axles';
    static AXLES_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.Axles';
    static BUSINESS_NAME_FIELD_NAME = 'BusinessName';
    static BUSINESS_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.BusinessName';
    static FUEL_FIELD_NAME = 'Fuel';
    static FUEL_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.Fuel';
    static GROSS_VEHICLE_WEIGHT_FIELD_NAME = 'GrossVehicleWeight';
    static GROSS_VEHICLE_WEIGHT_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.GrossVehicleWeight';
    static REGISTRANT_FAMILY_NAME_FIELD_NAME = 'RegistrantFamilyName';
    static REGISTRANT_FAMILY_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.RegistrantFamilyName';
    static REGISTRANT_GIVEN_NAME_FIELD_NAME = 'RegistrantGivenName';
    static REGISTRANT_GIVEN_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.RegistrantGivenName';
    static REGISTRATION_EXPIRY_DATE_FIELD_NAME = 'RegistrationExpiryDate';
    static REGISTRATION_EXPIRY_DATE_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.RegistrationExpiryDate';
    static REGISTRATION_ISSUE_DATE_FIELD_NAME = 'RegistrationIssueDate';
    static REGISTRATION_ISSUE_DATE_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.RegistrationIssueDate';
    static REGISTRATION_PLATE_NUMBER_FIELD_NAME = 'RegistrationPlateNumber';
    static REGISTRATION_PLATE_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.RegistrationPlateNumber';
    static REGISTRATION_WINDOW_STICKER_DECAL_FIELD_NAME = 'RegistrationWindowStickerDecal';
    static REGISTRATION_WINDOW_STICKER_DECAL_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.RegistrationWindowStickerDecal';
    static REGISTRATION_YEAR_FIELD_NAME = 'RegistrationYear';
    static REGISTRATION_YEAR_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.RegistrationYear';
    static VEHICLE_BODY_STYLE_FIELD_NAME = 'VehicleBodyStyle';
    static VEHICLE_BODY_STYLE_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.VehicleBodyStyle';
    static VEHICLE_COLOR_FIELD_NAME = 'VehicleColor';
    static VEHICLE_COLOR_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.VehicleColor';
    static VEHICLE_IDENTIFICATION_NUMBER_FIELD_NAME = 'VehicleIdentificationNumber';
    static VEHICLE_IDENTIFICATION_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.VehicleIdentificationNumber';
    static VEHICLE_MAKE_FIELD_NAME = 'VehicleMake';
    static VEHICLE_MAKE_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.VehicleMake';
    static VEHICLE_MODEL_FIELD_NAME = 'VehicleModel';
    static VEHICLE_MODEL_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.VehicleModel';
    static VEHICLE_MODEL_YEAR_FIELD_NAME = 'VehicleModelYear';
    static VEHICLE_MODEL_YEAR_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.VehicleModelYear';
    static VEHICLE_USE_FIELD_NAME = 'VehicleUse';
    static VEHICLE_USE_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrationData.VehicleUse';
    constructor(document) {
      if (document.type.name !== this.requiredDocumentType()) {
        throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
      }
      this._document = document;
    }
    requiredDocumentType() {
      return AAMVA.RegistrationData.DOCUMENT_TYPE;
    }

    /** Data element ID "RBK". City portion of the owner’s address. */
    get addressCity() {
      return this.document.fields.find(f => f.type.name === 'AddressCity');
    }

    /** Data element ID "RBL". Jurisdiction portion of the owner’s address. */
    get addressJurisdictionCode() {
      return this.document.fields.find(f => f.type.name === 'AddressJurisdictionCode');
    }

    /** Data element ID "RBI". Street portion of the owner’s address. */
    get addressStreet() {
      return this.document.fields.find(f => f.type.name === 'AddressStreet');
    }

    /** Data element ID "RBM". The Zip code or Postal code of the vehicle owner’s residence address. */
    get addressZipCode() {
      return this.document.fields.find(f => f.type.name === 'AddressZipCode');
    }

    /** Data element ID "VBC". The number of common axles of rotation of one or more wheels of a vehicle, whether power driven or freely rotating. */
    get axles() {
      return this.document.fields.find(f => f.type.name === 'Axles');
    }

    /** Data element ID "BBC". The business name of the first registrant of a vehicle. */
    get businessName() {
      return this.document.fields.find(f => f.type.name === 'BusinessName');
    }

    /** Data element ID "FUL". The type of fuel used by the vehicle. In most cases, the fuel type would be diesel. */
    get fuel() {
      return this.document.fields.find(f => f.type.name === 'Fuel');
    }

    /** Data element ID "VAT". The unladen weight of the vehicle (e.g., the single-unit truck, truck combination) plus the weight of the load being carried at a specific point in time. */
    get grossVehicleWeight() {
      return this.document.fields.find(f => f.type.name === 'GrossVehicleWeight');
    }

    /** Data element ID "RBD". Family name (commonly called surname or last name) of the registered owner of a vehicle. */
    get registrantFamilyName() {
      return this.document.fields.find(f => f.type.name === 'RegistrantFamilyName');
    }

    /** Data element ID "RBE". Given name or names (includes all of what are commonly referred to as first and middle names) of the registered owner of a vehicle. */
    get registrantGivenName() {
      return this.document.fields.find(f => f.type.name === 'RegistrantGivenName');
    }

    /** Data element ID "RAG". The date in which the registration expired. The format is ISO8601 with delimiters. */
    get registrationExpiryDate() {
      return this.document.fields.find(f => f.type.name === 'RegistrationExpiryDate');
    }

    /** Data element ID "RBB". The date in which the registration was issued. The format is ISO8601 with delimiters. */
    get registrationIssueDate() {
      return this.document.fields.find(f => f.type.name === 'RegistrationIssueDate');
    }

    /** Data element ID "RAM". The characters assigned to a registration plate or tag affixed to the vehicle, assigned by the jurisdiction. */
    get registrationPlateNumber() {
      return this.document.fields.find(f => f.type.name === 'RegistrationPlateNumber');
    }

    /** Data element ID "RBU". A unique number printed on the tab/decal and stored as part of the registration record. */
    get registrationWindowStickerDecal() {
      return this.document.fields.find(f => f.type.name === 'RegistrationWindowStickerDecal');
    }

    /** Data element ID "RBT". The year of registration. */
    get registrationYear() {
      return this.document.fields.find(f => f.type.name === 'RegistrationYear');
    }

    /** Data element ID "VAO". The general configuration or shape of a vehicle distinguished by characteristics such as number of doors, seats, windows, roofline, and type of top. The vehicle body type is 2-character alphanumeric. */
    get vehicleBodyStyle() {
      return this.document.fields.find(f => f.type.name === 'VehicleBodyStyle');
    }

    /** Data element ID "VBD". Where the vehicle is one color, this is the appropriate code describing that color. When the vehicle is two colors, this is the code for the top-most or front-most color. */
    get vehicleColor() {
      return this.document.fields.find(f => f.type.name === 'VehicleColor');
    }

    /** Data element ID "VAD". A unique combination of alphanumeric characters that identifies a specific vehicle or component. The VIN is affixed to the vehicle in specific locations and formulated by the manufacturer. State agencies under some controlled instances my assign a VIN to a vehicle. */
    get vehicleIdentificationNumber() {
      return this.document.fields.find(f => f.type.name === 'VehicleIdentificationNumber');
    }

    /** Data element ID "VAK". The distinctive (coded) name applied to a group of vehicles by a manufacturer. */
    get vehicleMake() {
      return this.document.fields.find(f => f.type.name === 'VehicleMake');
    }

    /** Data element ID "VAM". A code denoting a family of vehicles (within a make), which has a degree of similarity in construction, such as body, chassis, etc. The field does not necessarily contain a standard code; it may contain a value provided by the originator of the field. */
    get vehicleModel() {
      return this.document.fields.find(f => f.type.name === 'VehicleModel');
    }

    /** Data element ID "VAL". The year which is assigned to a vehicle by the manufacturer. The format is CCYY. */
    get vehicleModelYear() {
      return this.document.fields.find(f => f.type.name === 'VehicleModelYear');
    }

    /** Data element ID "VPC". Indicates the use of the vehicle. */
    get vehicleUse() {
      return this.document.fields.find(f => f.type.name === 'VehicleUse');
    }
  }
  _AAMVA.RegistrationData = RegistrationData;
  class MotorCarrierData {
    get document() {
      return this._document;
    }
    static DOCUMENT_TYPE = 'MotorCarrierData';
    static CARRIER_NAME_FIELD_NAME = 'CarrierName';
    static CARRIER_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.MotorCarrierData.CarrierName';
    static CITY_FIELD_NAME = 'City';
    static CITY_FIELD_NAME_NORMALIZED = 'AAMVA.MotorCarrierData.City';
    static JURISDICTION_FIELD_NAME = 'Jurisdiction';
    static JURISDICTION_FIELD_NAME_NORMALIZED = 'AAMVA.MotorCarrierData.Jurisdiction';
    static STREET_ADDRESS_FIELD_NAME = 'StreetAddress';
    static STREET_ADDRESS_FIELD_NAME_NORMALIZED = 'AAMVA.MotorCarrierData.StreetAddress';
    static USDOT_NUMBER_FIELD_NAME = 'USDOTNumber';
    static USDOT_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.MotorCarrierData.USDOTNumber';
    static ZIP_FIELD_NAME = 'Zip';
    static ZIP_FIELD_NAME_NORMALIZED = 'AAMVA.MotorCarrierData.Zip';
    constructor(document) {
      if (document.type.name !== this.requiredDocumentType()) {
        throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
      }
      this._document = document;
    }
    requiredDocumentType() {
      return AAMVA.MotorCarrierData.DOCUMENT_TYPE;
    }

    /** Data element ID "MAA". The name of the carrier responsible for safety. This can be an individual, partnership or corporation responsible for the transportation of persons or property. This is the name that is recognized by law. */
    get carrierName() {
      return this.document.fields.find(f => f.type.name === 'CarrierName');
    }

    /** Data element ID "MAL". This is the city for the mailing address of the individual carrier. This information is utilized by the base jurisdiction to send information to the carrier that purchased the IRP credentials. */
    get city() {
      return this.document.fields.find(f => f.type.name === 'City');
    }

    /** Data element ID "MAI". This is the jurisdiction of the residential address of the individual carrier. This information is utilized by the base jurisdiction to send information to the carrier that purchased the IRP credentials. */
    get jurisdiction() {
      return this.document.fields.find(f => f.type.name === 'Jurisdiction');
    }

    /** Data element ID "MAK". This is the mailing address of the individual carrier. This information is utilized by the base jurisdiction to send information to the carrier that purchased the IRP credentials. */
    get streetAddress() {
      return this.document.fields.find(f => f.type.name === 'StreetAddress');
    }

    /** Data element ID "MAN". A unique identifier assigned to the carrier responsible for safety issued by the U.S. Department of Transportation – Federal Motor Carrier Safety Administration. */
    get usdotNumber() {
      return this.document.fields.find(f => f.type.name === 'USDOTNumber');
    }

    /** Data element ID "MAO". The ZIP or Postal code of the resident address of the vehicle owner. */
    get zip() {
      return this.document.fields.find(f => f.type.name === 'Zip');
    }
  }
  _AAMVA.MotorCarrierData = MotorCarrierData;
  class RegistrantAndVehicleData {
    get document() {
      return this._document;
    }
    static DOCUMENT_TYPE = 'RegistrantAndVehicleData';
    static ADDRESS_FIELD_NAME = 'Address';
    static ADDRESS_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrantAndVehicleData.Address';
    static BASE_JURISDICTION_REGISTERED_WEIGHT_FIELD_NAME = 'BaseJurisdictionRegisteredWeight';
    static BASE_JURISDICTION_REGISTERED_WEIGHT_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrantAndVehicleData.BaseJurisdictionRegisteredWeight';
    static CARRIER_NAME_REGISTRANT_FIELD_NAME = 'CarrierNameRegistrant';
    static CARRIER_NAME_REGISTRANT_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrantAndVehicleData.CarrierNameRegistrant';
    static CITY_FIELD_NAME = 'City';
    static CITY_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrantAndVehicleData.City';
    static GROSS_VEHICLE_WEIGHT_FIELD_NAME = 'GrossVehicleWeight';
    static GROSS_VEHICLE_WEIGHT_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrantAndVehicleData.GrossVehicleWeight';
    static JURISDICTION_FIELD_NAME = 'Jurisdiction';
    static JURISDICTION_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrantAndVehicleData.Jurisdiction';
    static MODEL_YEAR_FIELD_NAME = 'ModelYear';
    static MODEL_YEAR_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrantAndVehicleData.ModelYear';
    static NUMBER_OF_SEATS_FIELD_NAME = 'NumberOfSeats';
    static NUMBER_OF_SEATS_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrantAndVehicleData.NumberOfSeats';
    static REGISTRATION_DECAL_NUMBER_FIELD_NAME = 'RegistrationDecalNumber';
    static REGISTRATION_DECAL_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrantAndVehicleData.RegistrationDecalNumber';
    static REGISTRATION_ENFORCEMENT_DATE_FIELD_NAME = 'RegistrationEnforcementDate';
    static REGISTRATION_ENFORCEMENT_DATE_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrantAndVehicleData.RegistrationEnforcementDate';
    static REGISTRATION_EXPIRATION_DATE_FIELD_NAME = 'RegistrationExpirationDate';
    static REGISTRATION_EXPIRATION_DATE_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrantAndVehicleData.RegistrationExpirationDate';
    static REGISTRATION_ISSUE_DATE_FIELD_NAME = 'RegistrationIssueDate';
    static REGISTRATION_ISSUE_DATE_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrantAndVehicleData.RegistrationIssueDate';
    static REGISTRATION_PLATE_NUMBER_FIELD_NAME = 'RegistrationPlateNumber';
    static REGISTRATION_PLATE_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrantAndVehicleData.RegistrationPlateNumber';
    static REGISTRATION_YEAR_FIELD_NAME = 'RegistrationYear';
    static REGISTRATION_YEAR_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrantAndVehicleData.RegistrationYear';
    static TYPE_OF_VEHICLE_FIELD_NAME = 'TypeOfVehicle';
    static TYPE_OF_VEHICLE_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrantAndVehicleData.TypeOfVehicle';
    static UNIT_NUMBER_FIELD_NAME = 'UnitNumber';
    static UNIT_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrantAndVehicleData.UnitNumber';
    static VEHICLE_IDENTIFICATION_NUMBER_FIELD_NAME = 'VehicleIdentificationNumber';
    static VEHICLE_IDENTIFICATION_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrantAndVehicleData.VehicleIdentificationNumber';
    static VEHICLE_MAKE_FIELD_NAME = 'VehicleMake';
    static VEHICLE_MAKE_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrantAndVehicleData.VehicleMake';
    static ZIP_CODE_FIELD_NAME = 'ZipCode';
    static ZIP_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.RegistrantAndVehicleData.ZipCode';
    constructor(document) {
      if (document.type.name !== this.requiredDocumentType()) {
        throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
      }
      this._document = document;
    }
    requiredDocumentType() {
      return AAMVA.RegistrantAndVehicleData.DOCUMENT_TYPE;
    }

    /** Data element ID "RBI". The first line of the registrant’s residence address. */
    get address() {
      return this.document.fields.find(f => f.type.name === 'Address');
    }

    /** Data element ID "RAU". The declared base jurisdiction registration weight. */
    get baseJurisdictionRegisteredWeight() {
      return this.document.fields.find(f => f.type.name === 'BaseJurisdictionRegisteredWeight');
    }

    /** Data element ID "RBC". The name of the first registrant of a vehicle. Registrant’s name may be a combined individual name or the name of a business */
    get carrierNameRegistrant() {
      return this.document.fields.find(f => f.type.name === 'CarrierNameRegistrant');
    }

    /** Data element ID "RBK". The registrant’s residence city name. */
    get city() {
      return this.document.fields.find(f => f.type.name === 'City');
    }

    /** Data element ID "VAT". The unladen weight of the vehicle (e.g., single-unit truck, truck combination) plus the weight of the maximum load for which vehicle registration fees have been paid within a particular jurisdiction. */
    get grossVehicleWeight() {
      return this.document.fields.find(f => f.type.name === 'GrossVehicleWeight');
    }

    /** Data element ID "RBL". The state or province of the registrant’s residence address. */
    get jurisdiction() {
      return this.document.fields.find(f => f.type.name === 'Jurisdiction');
    }

    /** Data element ID "VAL". The year which is assigned to a vehicle by the manufacturer. The format is YY. */
    get modelYear() {
      return this.document.fields.find(f => f.type.name === 'ModelYear');
    }

    /** Data element IDs "RAP", "VBC". The seat capacity of a commercial vehicle designed for transportation of more than then passengers. The number of common axles of rotation of one or more wheels of a vehicle, whether power design or freely rotating. */
    get numberOfSeats() {
      return this.document.fields.find(f => f.type.name === 'NumberOfSeats');
    }

    /** Data element ID "RAD". The number assigned to the registration decal in those jurisdictions that issue decals. */
    get registrationDecalNumber() {
      return this.document.fields.find(f => f.type.name === 'RegistrationDecalNumber');
    }

    /** Data element ID "RAF". The registration enforcement date is the date that the current registration was enforced. This may or may not be the original registration date. The format is ISO8601 with delimiters */
    get registrationEnforcementDate() {
      return this.document.fields.find(f => f.type.name === 'RegistrationEnforcementDate');
    }

    /** Data element ID "RAG". The date in which the registration expired. The format is ISO8601 with delimiters. */
    get registrationExpirationDate() {
      return this.document.fields.find(f => f.type.name === 'RegistrationExpirationDate');
    }

    /** Data element ID "IFJ". The date in which the registration was issued. The format is ISO8601 with delimiters. */
    get registrationIssueDate() {
      return this.document.fields.find(f => f.type.name === 'RegistrationIssueDate');
    }

    /** Data element ID "RAM". The characters assigned to a registration plate or tag affixed to the vehicle, assigned by the jurisdiction. */
    get registrationPlateNumber() {
      return this.document.fields.find(f => f.type.name === 'RegistrationPlateNumber');
    }

    /** Data element ID "RBT". This field is the registration year assigned by the jurisdiction. The format is CCYY. */
    get registrationYear() {
      return this.document.fields.find(f => f.type.name === 'RegistrationYear');
    }

    /** Data element ID "VBB". The type of vehicle operated for the transportation of persons or property in the furtherance of any commercial or industrial enterprise, for hire or not for hire. Not all states will use all values. */
    get typeOfVehicle() {
      return this.document.fields.find(f => f.type.name === 'TypeOfVehicle');
    }

    /** Data element ID "IEG". A number, assigned by the registrant of the commercial vehicle or trailer, to identify the vehicle or trailer in the fleet. No two units in a fleet can have the same number. A.K.A vehicle unit number or owner’s equipment number. */
    get unitNumber() {
      return this.document.fields.find(f => f.type.name === 'UnitNumber');
    }

    /** Data element ID "VAD". A unique combination of alphanumeric characters that identifies a specific vehicle or component. The VIN is affixed to the vehicle in specific locations and formulated by the manufacturer. State agencies under some controlled instances may assign a VIN to a vehicle. */
    get vehicleIdentificationNumber() {
      return this.document.fields.find(f => f.type.name === 'VehicleIdentificationNumber');
    }

    /** Data element ID "VAK". The distinctive (coded) name applied to a group of vehicles by a manufacturer. */
    get vehicleMake() {
      return this.document.fields.find(f => f.type.name === 'VehicleMake');
    }

    /** Data element ID "RBM". The ZIP or Postal code of the resident address of the registrant. */
    get zipCode() {
      return this.document.fields.find(f => f.type.name === 'ZipCode');
    }
  }
  _AAMVA.RegistrantAndVehicleData = RegistrantAndVehicleData;
  class VehicleOwnerData {
    get document() {
      return this._document;
    }
    static DOCUMENT_TYPE = 'VehicleOwnerData';
    static FIRST_OWNER_ID_NUMBER_FIELD_NAME = 'FirstOwnerIdNumber';
    static FIRST_OWNER_ID_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.FirstOwnerIdNumber';
    static FIRST_OWNER_LAST_NAME_FIELD_NAME = 'FirstOwnerLastName';
    static FIRST_OWNER_LAST_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.FirstOwnerLastName';
    static FIRST_OWNER_LEGAL_STATUS_FIELD_NAME = 'FirstOwnerLegalStatus';
    static FIRST_OWNER_LEGAL_STATUS_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.FirstOwnerLegalStatus';
    static FIRST_OWNER_MIDDLE_NAME_FIELD_NAME = 'FirstOwnerMiddleName';
    static FIRST_OWNER_MIDDLE_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.FirstOwnerMiddleName';
    static FIRST_OWNER_NAME_FIELD_NAME = 'FirstOwnerName';
    static FIRST_OWNER_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.FirstOwnerName';
    static FIRST_OWNER_TOTAL_NAME_FIELD_NAME = 'FirstOwnerTotalName';
    static FIRST_OWNER_TOTAL_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.FirstOwnerTotalName';
    static MAILING_ADDRESS_1_FIELD_NAME = 'MailingAddress1';
    static MAILING_ADDRESS_1_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.MailingAddress1';
    static MAILING_ADDRESS_2_FIELD_NAME = 'MailingAddress2';
    static MAILING_ADDRESS_2_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.MailingAddress2';
    static MAILING_CITY_FIELD_NAME = 'MailingCity';
    static MAILING_CITY_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.MailingCity';
    static MAILING_JURISDICTION_CODE_FIELD_NAME = 'MailingJurisdictionCode';
    static MAILING_JURISDICTION_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.MailingJurisdictionCode';
    static MAILING_ZIP_CODE_FIELD_NAME = 'MailingZipCode';
    static MAILING_ZIP_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.MailingZipCode';
    static RESIDENCE_ADDRESS_1_FIELD_NAME = 'ResidenceAddress1';
    static RESIDENCE_ADDRESS_1_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.ResidenceAddress1';
    static RESIDENCE_ADDRESS_2_FIELD_NAME = 'ResidenceAddress2';
    static RESIDENCE_ADDRESS_2_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.ResidenceAddress2';
    static RESIDENCE_CITY_FIELD_NAME = 'ResidenceCity';
    static RESIDENCE_CITY_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.ResidenceCity';
    static RESIDENCE_JURISDICTION_CODE_FIELD_NAME = 'ResidenceJurisdictionCode';
    static RESIDENCE_JURISDICTION_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.ResidenceJurisdictionCode';
    static RESIDENCE_ZIP_CODE_FIELD_NAME = 'ResidenceZipCode';
    static RESIDENCE_ZIP_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.ResidenceZipCode';
    static SECOND_OWNER_ID_NUMBER_FIELD_NAME = 'SecondOwnerIdNumber';
    static SECOND_OWNER_ID_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.SecondOwnerIdNumber';
    static SECOND_OWNER_LAST_NAME_FIELD_NAME = 'SecondOwnerLastName';
    static SECOND_OWNER_LAST_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.SecondOwnerLastName';
    static SECOND_OWNER_LEGAL_STATUS_FIELD_NAME = 'SecondOwnerLegalStatus';
    static SECOND_OWNER_LEGAL_STATUS_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.SecondOwnerLegalStatus';
    static SECOND_OWNER_MIDDLE_NAME_FIELD_NAME = 'SecondOwnerMiddleName';
    static SECOND_OWNER_MIDDLE_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.SecondOwnerMiddleName';
    static SECOND_OWNER_NAME_FIELD_NAME = 'SecondOwnerName';
    static SECOND_OWNER_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.SecondOwnerName';
    static SECOND_OWNER_TOTAL_NAME_FIELD_NAME = 'SecondOwnerTotalName';
    static SECOND_OWNER_TOTAL_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleOwnerData.SecondOwnerTotalName';
    constructor(document) {
      if (document.type.name !== this.requiredDocumentType()) {
        throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
      }
      this._document = document;
    }
    requiredDocumentType() {
      return AAMVA.VehicleOwnerData.DOCUMENT_TYPE;
    }

    /** Data element ID "NAX". The unique customer number/ID of the first vehicle owner. */
    get firstOwnerIdNumber() {
      return this.document.fields.find(f => f.type.name === 'FirstOwnerIdNumber');
    }

    /** Data element ID "NAB". Last Name or Surname of the Owner. Hyphenated names acceptable, spaces between names acceptable, but no other use of special symbols. */
    get firstOwnerLastName() {
      return this.document.fields.find(f => f.type.name === 'FirstOwnerLastName');
    }

    /** Data element ID "NBA". The legal status of the first vehicle owner. This is only used when a vehicle has multiple owners. A legal status may be ("AND", "OR"). */
    get firstOwnerLegalStatus() {
      return this.document.fields.find(f => f.type.name === 'FirstOwnerLegalStatus');
    }

    /** Data element ID "NAD". Middle Name(s) or Initial(s) of the Owner. Hyphenated names acceptable, spaces, between names acceptable, but no other use of special symbols. */
    get firstOwnerMiddleName() {
      return this.document.fields.find(f => f.type.name === 'FirstOwnerMiddleName');
    }

    /** Data element ID "NAC". First Name or Given Name of the Owner. Hyphenated names acceptable, but no other use of special symbols. */
    get firstOwnerName() {
      return this.document.fields.find(f => f.type.name === 'FirstOwnerName');
    }

    /** Data element ID "NAA". Name of the (or one of the) individual(s) who owns the Vehicle as defined in the ANSI D- 20 Data Element Dictionary. (Lastname@Firstname@MI@Suffix if any.) */
    get firstOwnerTotalName() {
      return this.document.fields.find(f => f.type.name === 'FirstOwnerTotalName');
    }

    /** Data element ID "NAR". Street address line 1. (Mailing) */
    get mailingAddress1() {
      return this.document.fields.find(f => f.type.name === 'MailingAddress1');
    }

    /** Data element ID "NAS". Street address line 2. (Mailing) */
    get mailingAddress2() {
      return this.document.fields.find(f => f.type.name === 'MailingAddress2');
    }

    /** Data element ID "NAT". Name of city for mailing address. */
    get mailingCity() {
      return this.document.fields.find(f => f.type.name === 'MailingCity');
    }

    /** Data element ID "NAU". Jurisdiction code for mailing address. Conforms to Canadian, Mexican and US jurisdictions as appropriate. Codes for provinces (Canada) and states (US and Mexico). */
    get mailingJurisdictionCode() {
      return this.document.fields.find(f => f.type.name === 'MailingJurisdictionCode');
    }

    /** Data element ID "NAV". The ZIP code or Postal code used for mailing. (As used by Canadian, Mexican and US jurisdictions.) */
    get mailingZipCode() {
      return this.document.fields.find(f => f.type.name === 'MailingZipCode');
    }

    /** Data element ID "NAM". Street address line 1. (Mailing) */
    get residenceAddress1() {
      return this.document.fields.find(f => f.type.name === 'ResidenceAddress1');
    }

    /** Data element ID "NAN". Street address line 2. (Mailing) */
    get residenceAddress2() {
      return this.document.fields.find(f => f.type.name === 'ResidenceAddress2');
    }

    /** Data element ID "NAO". Name of city for mailing address. */
    get residenceCity() {
      return this.document.fields.find(f => f.type.name === 'ResidenceCity');
    }

    /** Data element ID "NAP". Jurisdiction code for mailing address. Conforms to Canadian, Mexican and US jurisdictions as appropriate. Codes for provinces (Canada) and states (US and Mexico). */
    get residenceJurisdictionCode() {
      return this.document.fields.find(f => f.type.name === 'ResidenceJurisdictionCode');
    }

    /** Data element ID "NAQ". The ZIP code or Postal code used for mailing. (As used by Canadian, Mexican and US jurisdictions). */
    get residenceZipCode() {
      return this.document.fields.find(f => f.type.name === 'ResidenceZipCode');
    }

    /** Data element ID "NAY". The unique customer number/ID of the second vehicle owner. */
    get secondOwnerIdNumber() {
      return this.document.fields.find(f => f.type.name === 'SecondOwnerIdNumber');
    }

    /** Data element ID "NAF". Last Name or Surname of the Owner. Hyphenated names acceptable, spaces between names acceptable, but no other use of special symbols. */
    get secondOwnerLastName() {
      return this.document.fields.find(f => f.type.name === 'SecondOwnerLastName');
    }

    /** Data element ID "NBB". The legal status of the second vehicle owner. This is only used when a vehicle has multiple owners. A legal status may be ("AND", "OR"). */
    get secondOwnerLegalStatus() {
      return this.document.fields.find(f => f.type.name === 'SecondOwnerLegalStatus');
    }

    /** Data element ID "NAH". Middle Name(s) or Initial(s) of the Owner. Hyphenated names acceptable, spaces between names acceptable, but no other use of special symbols. */
    get secondOwnerMiddleName() {
      return this.document.fields.find(f => f.type.name === 'SecondOwnerMiddleName');
    }

    /** Data element ID "NAG". First Name or Given Name of the Owner. Hyphenated names acceptable, but no other use of special symbols. */
    get secondOwnerName() {
      return this.document.fields.find(f => f.type.name === 'SecondOwnerName');
    }

    /** Data element ID "NAE". Name of the (or one of the) individual(s) who owns the Vehicle as defined in the ANSI D- 20 Data Element Dictionary. (Lastname@Firstname@MI@Suffix if any.) */
    get secondOwnerTotalName() {
      return this.document.fields.find(f => f.type.name === 'SecondOwnerTotalName');
    }
  }
  _AAMVA.VehicleOwnerData = VehicleOwnerData;
  class VehicleData {
    get document() {
      return this._document;
    }
    static DOCUMENT_TYPE = 'VehicleData';
    static BODY_STYLE_FIELD_NAME = 'BodyStyle';
    static BODY_STYLE_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.BodyStyle';
    static DATE_JUNKED_FIELD_NAME = 'DateJunked';
    static DATE_JUNKED_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.DateJunked';
    static DATE_RECOVERED_FIELD_NAME = 'DateRecovered';
    static DATE_RECOVERED_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.DateRecovered';
    static DATE_STOLEN_FIELD_NAME = 'DateStolen';
    static DATE_STOLEN_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.DateStolen';
    static ENGINE_DISPLACEMENT_FIELD_NAME = 'EngineDisplacement';
    static ENGINE_DISPLACEMENT_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.EngineDisplacement';
    static ENGINE_SIZE_FIELD_NAME = 'EngineSize';
    static ENGINE_SIZE_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.EngineSize';
    static FUEL_TYPE_FIELD_NAME = 'FuelType';
    static FUEL_TYPE_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.FuelType';
    static HORSEPOWER_FIELD_NAME = 'Horsepower';
    static HORSEPOWER_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.Horsepower';
    static IFTA_INDICATOR_FIELD_NAME = 'IFTAIndicator';
    static IFTA_INDICATOR_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.IFTAIndicator';
    static IRP_INDICATOR_FIELD_NAME = 'IRPIndicator';
    static IRP_INDICATOR_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.IRPIndicator';
    static JUNKED_INDICATOR_FIELD_NAME = 'JunkedIndicator';
    static JUNKED_INDICATOR_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.JunkedIndicator';
    static MSRP_FIELD_NAME = 'MSRP';
    static MSRP_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.MSRP';
    static MAJOR_CODE_FIELD_NAME = 'MajorCode';
    static MAJOR_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.MajorCode';
    static MAKE_YEAR_FIELD_NAME = 'MakeYear';
    static MAKE_YEAR_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.MakeYear';
    static MANUFACTURE_GROSS_WEIGHT_FIELD_NAME = 'ManufactureGrossWeight';
    static MANUFACTURE_GROSS_WEIGHT_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.ManufactureGrossWeight';
    static MINOR_CODE_FIELD_NAME = 'MinorCode';
    static MINOR_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.MinorCode';
    static NUMBER_OF_AXLES_FIELD_NAME = 'NumberOfAxles';
    static NUMBER_OF_AXLES_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.NumberOfAxles';
    static NUMBER_OF_CYLINDERS_FIELD_NAME = 'NumberOfCylinders';
    static NUMBER_OF_CYLINDERS_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.NumberOfCylinders';
    static NUMBER_OF_DOORS_FIELD_NAME = 'NumberOfDoors';
    static NUMBER_OF_DOORS_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.NumberOfDoors';
    static STOLEN_INDICATOR_FIELD_NAME = 'StolenIndicator';
    static STOLEN_INDICATOR_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.StolenIndicator';
    static TRANSMISSION_CODE_FIELD_NAME = 'TransmissionCode';
    static TRANSMISSION_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.TransmissionCode';
    static UNLADEN_WEIGHT_FIELD_NAME = 'UnladenWeight';
    static UNLADEN_WEIGHT_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.UnladenWeight';
    static VLT_CLAC_FROM_DATE_FIELD_NAME = 'VLTClacFromDate';
    static VLT_CLAC_FROM_DATE_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.VLTClacFromDate';
    static VEHICLE_ID_NUMBER_FIELD_NAME = 'VehicleIdNumber';
    static VEHICLE_ID_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.VehicleIdNumber';
    static VEHICLE_IDENTIFICATION_NUMBER_FIELD_NAME = 'VehicleIdentificationNumber';
    static VEHICLE_IDENTIFICATION_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.VehicleIdentificationNumber';
    static VEHICLE_MAKE_FIELD_NAME = 'VehicleMake';
    static VEHICLE_MAKE_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.VehicleMake';
    static VEHICLE_MODEL_FIELD_NAME = 'VehicleModel';
    static VEHICLE_MODEL_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.VehicleModel';
    static VEHICLE_STATUS_CODE_FIELD_NAME = 'VehicleStatusCode';
    static VEHICLE_STATUS_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.VehicleStatusCode';
    static VEHICLE_TYPE_CODE_FIELD_NAME = 'VehicleTypeCode';
    static VEHICLE_TYPE_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleData.VehicleTypeCode';
    constructor(document) {
      if (document.type.name !== this.requiredDocumentType()) {
        throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
      }
      this._document = document;
    }
    requiredDocumentType() {
      return AAMVA.VehicleData.DOCUMENT_TYPE;
    }

    /** Data element ID "VAO". Vehicle manufacture body style. */
    get bodyStyle() {
      return this.document.fields.find(f => f.type.name === 'BodyStyle');
    }

    /** Data element ID "VAG". Date vehicle reported junked. The format is ISO8601 with delimiters. */
    get dateJunked() {
      return this.document.fields.find(f => f.type.name === 'DateJunked');
    }

    /** Data element ID "VAJ". Date vehicle reported recovered. The format is ISO8601 with delimiters. */
    get dateRecovered() {
      return this.document.fields.find(f => f.type.name === 'DateRecovered');
    }

    /** Data element ID "VAI". Date vehicle reported stolen. The format is ISO8601 with delimiters. */
    get dateStolen() {
      return this.document.fields.find(f => f.type.name === 'DateStolen');
    }

    /** Data element ID "VAW". Manufacturer’s rated engine displacement. */
    get engineDisplacement() {
      return this.document.fields.find(f => f.type.name === 'EngineDisplacement');
    }

    /** Data element ID "VAR". The size of a vehicle’s engine. */
    get engineSize() {
      return this.document.fields.find(f => f.type.name === 'EngineSize');
    }

    /** Data element ID "VAN". Type of fuel the vehicle utilizes. */
    get fuelType() {
      return this.document.fields.find(f => f.type.name === 'FuelType');
    }

    /** Data element ID "VAU". Manufacturer’s rated horsepower. */
    get horsepower() {
      return this.document.fields.find(f => f.type.name === 'Horsepower');
    }

    /** Data element ID "VAY". International fuel tax indicator */
    get iftaIndicator() {
      return this.document.fields.find(f => f.type.name === 'IFTAIndicator');
    }

    /** Data element ID "VAX". International registration plan indicator. */
    get irpIndicator() {
      return this.document.fields.find(f => f.type.name === 'IRPIndicator');
    }

    /** Data element ID "VAF". Vehicle has been junked. */
    get junkedIndicator() {
      return this.document.fields.find(f => f.type.name === 'JunkedIndicator');
    }

    /** Data element ID "VAE". Manufacturer’s Suggested Retail Price. No decimal places. Right Justified Zero or space fill. */
    get msrp() {
      return this.document.fields.find(f => f.type.name === 'MSRP');
    }

    /** Data element ID "VAA". State to provide definition. */
    get majorCode() {
      return this.document.fields.find(f => f.type.name === 'MajorCode');
    }

    /** Data element ID "VAL". Vehicle manufacture year. */
    get makeYear() {
      return this.document.fields.find(f => f.type.name === 'MakeYear');
    }

    /** Data element ID "VAT". Manufacturer’s gross vehicle weight rating. */
    get manufactureGrossWeight() {
      return this.document.fields.find(f => f.type.name === 'ManufactureGrossWeight');
    }

    /** Data element ID "VAB". State to provide definition. */
    get minorCode() {
      return this.document.fields.find(f => f.type.name === 'MinorCode');
    }

    /** Data element ID "VBC". Number of axles the vehicle has. */
    get numberOfAxles() {
      return this.document.fields.find(f => f.type.name === 'NumberOfAxles');
    }

    /** Data element ID "VAQ". Number of cylinders the vehicle has. */
    get numberOfCylinders() {
      return this.document.fields.find(f => f.type.name === 'NumberOfCylinders');
    }

    /** Data element ID "VAP". Number of doors on the vehicle. */
    get numberOfDoors() {
      return this.document.fields.find(f => f.type.name === 'NumberOfDoors');
    }

    /** Data element ID "VAH". Indicates stolen vehicle. */
    get stolenIndicator() {
      return this.document.fields.find(f => f.type.name === 'StolenIndicator');
    }

    /** Data element ID "VAC". Type of transmission the vehicle carries. */
    get transmissionCode() {
      return this.document.fields.find(f => f.type.name === 'TransmissionCode');
    }

    /** Data element ID "VAV". Gross weight of the vehicle unloaded. */
    get unladenWeight() {
      return this.document.fields.find(f => f.type.name === 'UnladenWeight');
    }

    /** Data element ID "VAZ". Vehicle license tax calculation from date of purchase. */
    get vltClacFromDate() {
      return this.document.fields.find(f => f.type.name === 'VLTClacFromDate');
    }

    /** Data element ID "VBA". Unique number to identify the vehicle record. */
    get vehicleIdNumber() {
      return this.document.fields.find(f => f.type.name === 'VehicleIdNumber');
    }

    /** Data element ID "VAD". A unique combination of alphanumeric characters that identifies a specific vehicle or component. The VIN is affixed to the vehicle in specific locations and formulated by the manufacturer. State agencies under some controlled instances may assign a VIN to a vehicle. */
    get vehicleIdentificationNumber() {
      return this.document.fields.find(f => f.type.name === 'VehicleIdentificationNumber');
    }

    /** Data element ID "VAK". The distinctive (coded) name applied to a group of vehicles by a manufacturer. */
    get vehicleMake() {
      return this.document.fields.find(f => f.type.name === 'VehicleMake');
    }

    /** Data element ID "VAM". Vehicle manufacture model. */
    get vehicleModel() {
      return this.document.fields.find(f => f.type.name === 'VehicleModel');
    }

    /** Data element ID "VAS". This is the status of the vehicle (e.g., active, suspend, etc.) */
    get vehicleStatusCode() {
      return this.document.fields.find(f => f.type.name === 'VehicleStatusCode');
    }

    /** Data element ID "VBB". EPA vehicle classification. */
    get vehicleTypeCode() {
      return this.document.fields.find(f => f.type.name === 'VehicleTypeCode');
    }
  }
  _AAMVA.VehicleData = VehicleData;
  class VehicleSafetyInspectionData {
    get document() {
      return this._document;
    }
    static DOCUMENT_TYPE = 'VehicleSafetyInspectionData';
    static INSPECTION_ADDRESS_FIELD_NAME = 'InspectionAddress';
    static INSPECTION_ADDRESS_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleSafetyInspectionData.InspectionAddress';
    static INSPECTION_AIR_POLLUTION_DEVICE_CONDITIONS_FIELD_NAME = 'InspectionAirPollutionDeviceConditions';
    static INSPECTION_AIR_POLLUTION_DEVICE_CONDITIONS_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleSafetyInspectionData.InspectionAirPollutionDeviceConditions';
    static INSPECTION_FACILITY_IDENTIFIER_FIELD_NAME = 'InspectionFacilityIdentifier';
    static INSPECTION_FACILITY_IDENTIFIER_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleSafetyInspectionData.InspectionFacilityIdentifier';
    static INSPECTION_FORM_OR_STICKER_NUMBER_CURRENT_FIELD_NAME = 'InspectionFormOrStickerNumberCurrent';
    static INSPECTION_FORM_OR_STICKER_NUMBER_CURRENT_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleSafetyInspectionData.InspectionFormOrStickerNumberCurrent';
    static INSPECTION_FORM_OR_STICKER_NUMBER_PREVIOUS_FIELD_NAME = 'InspectionFormOrStickerNumberPrevious';
    static INSPECTION_FORM_OR_STICKER_NUMBER_PREVIOUS_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleSafetyInspectionData.InspectionFormOrStickerNumberPrevious';
    static INSPECTION_SMOG_CERTIFICATE_INDICATOR_FIELD_NAME = 'InspectionSmogCertificateIndicator';
    static INSPECTION_SMOG_CERTIFICATE_INDICATOR_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleSafetyInspectionData.InspectionSmogCertificateIndicator';
    static INSPECTION_STATION_NUMBER_FIELD_NAME = 'InspectionStationNumber';
    static INSPECTION_STATION_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleSafetyInspectionData.InspectionStationNumber';
    static INSPECTOR_IDENTIFICATION_NUMBER_FIELD_NAME = 'InspectorIdentificationNumber';
    static INSPECTOR_IDENTIFICATION_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleSafetyInspectionData.InspectorIdentificationNumber';
    static ODOMETER_READING_AT_INSPECTION_FIELD_NAME = 'OdometerReadingAtInspection';
    static ODOMETER_READING_AT_INSPECTION_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleSafetyInspectionData.OdometerReadingAtInspection';
    static VEHICLE_BODY_TYPE_FIELD_NAME = 'VehicleBodyType';
    static VEHICLE_BODY_TYPE_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleSafetyInspectionData.VehicleBodyType';
    static VEHICLE_MAKE_FIELD_NAME = 'VehicleMake';
    static VEHICLE_MAKE_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleSafetyInspectionData.VehicleMake';
    static VEHICLE_MODEL_YEAR_FIELD_NAME = 'VehicleModelYear';
    static VEHICLE_MODEL_YEAR_FIELD_NAME_NORMALIZED = 'AAMVA.VehicleSafetyInspectionData.VehicleModelYear';
    constructor(document) {
      if (document.type.name !== this.requiredDocumentType()) {
        throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
      }
      this._document = document;
    }
    requiredDocumentType() {
      return AAMVA.VehicleSafetyInspectionData.DOCUMENT_TYPE;
    }

    /** Data element ID "IAN". The street name and number, city, state and zip code of the inspection facility. */
    get inspectionAddress() {
      return this.document.fields.find(f => f.type.name === 'InspectionAddress');
    }

    /** Data element ID "IPD". Identifies whether the pollution control devices meet the minimum inspection criteria. */
    get inspectionAirPollutionDeviceConditions() {
      return this.document.fields.find(f => f.type.name === 'InspectionAirPollutionDeviceConditions');
    }

    /** Data element ID "IFI". The unique number assigned to an inspection facility. */
    get inspectionFacilityIdentifier() {
      return this.document.fields.find(f => f.type.name === 'InspectionFacilityIdentifier');
    }

    /** Data element ID "INC". A unique number assigned to a current vehicle inspection form for identification purposes or a preprinted unique number on the motor vehicle inspection sticker currently issued to a motor vehicle which has passed inspection. */
    get inspectionFormOrStickerNumberCurrent() {
      return this.document.fields.find(f => f.type.name === 'InspectionFormOrStickerNumberCurrent');
    }

    /** Data element ID "INP". The number of the last inspection form excluding the current inspection or the certification number of the last inspection sticker, excluding the current inspection. */
    get inspectionFormOrStickerNumberPrevious() {
      return this.document.fields.find(f => f.type.name === 'InspectionFormOrStickerNumberPrevious');
    }

    /** Data element ID "ISC". An indicator that specifies whether or not the vehicle has a current smog certificate. */
    get inspectionSmogCertificateIndicator() {
      return this.document.fields.find(f => f.type.name === 'InspectionSmogCertificateIndicator');
    }

    /** Data element ID "ISN". Station number performing the inspection. */
    get inspectionStationNumber() {
      return this.document.fields.find(f => f.type.name === 'InspectionStationNumber');
    }

    /** Data element ID "IIN". A unique number assigned to each licensed vehicle inspector. */
    get inspectorIdentificationNumber() {
      return this.document.fields.find(f => f.type.name === 'InspectorIdentificationNumber');
    }

    /** Data element ID "ORI". The vehicle’s odometer reading (to the nearest mile or kilometer) at the time of inspection. */
    get odometerReadingAtInspection() {
      return this.document.fields.find(f => f.type.name === 'OdometerReadingAtInspection');
    }

    /** Data element ID "VAO". The general configuration or shape of a vehicle distinguished by characteristics such as number of doors, seats, windows, roofline, and type of top. The vehicle body type is 2- character alphanumeric. */
    get vehicleBodyType() {
      return this.document.fields.find(f => f.type.name === 'VehicleBodyType');
    }

    /** Data element ID "VAK". The distinctive (coded) name applied to a group of vehicles by a manufacturer. */
    get vehicleMake() {
      return this.document.fields.find(f => f.type.name === 'VehicleMake');
    }

    /** Data element ID "VAL". The year which is assigned to a vehicle by the manufacturer. The format is CCYY. */
    get vehicleModelYear() {
      return this.document.fields.find(f => f.type.name === 'VehicleModelYear');
    }
  }
  _AAMVA.VehicleSafetyInspectionData = VehicleSafetyInspectionData;
  class DLID {
    get document() {
      return this._document;
    }
    static ADDRESS_CITY_FIELD_NAME = 'AddressCity';
    static ADDRESS_CITY_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.AddressCity';
    static ADDRESS_JURISDICTION_CODE_FIELD_NAME = 'AddressJurisdictionCode';
    static ADDRESS_JURISDICTION_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.AddressJurisdictionCode';
    static ADDRESS_POSTAL_CODE_FIELD_NAME = 'AddressPostalCode';
    static ADDRESS_POSTAL_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.AddressPostalCode';
    static ADDRESS_STREET_1_FIELD_NAME = 'AddressStreet1';
    static ADDRESS_STREET_1_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.AddressStreet1';
    static ADDRESS_STREET_2_FIELD_NAME = 'AddressStreet2';
    static ADDRESS_STREET_2_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.AddressStreet2';
    static ADULT_INFORMATION_FIELD_NAME = 'AdultInformation';
    static ADULT_INFORMATION_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.AdultInformation';
    static ALIAS_FAMILY_NAME_FIELD_NAME = 'AliasFamilyName';
    static ALIAS_FAMILY_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.AliasFamilyName';
    static ALIAS_GIVEN_NAME_FIELD_NAME = 'AliasGivenName';
    static ALIAS_GIVEN_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.AliasGivenName';
    static ALIAS_MIDDLE_NAME_FIELD_NAME = 'AliasMiddleName';
    static ALIAS_MIDDLE_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.AliasMiddleName';
    static ALIAS_PREFIX_NAME_FIELD_NAME = 'AliasPrefixName';
    static ALIAS_PREFIX_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.AliasPrefixName';
    static ALIAS_SUFFIX_NAME_FIELD_NAME = 'AliasSuffixName';
    static ALIAS_SUFFIX_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.AliasSuffixName';
    static CARD_REVISION_DATE_FIELD_NAME = 'CardRevisionDate';
    static CARD_REVISION_DATE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.CardRevisionDate';
    static COMPLIANCE_TYPE_FIELD_NAME = 'ComplianceType';
    static COMPLIANCE_TYPE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.ComplianceType';
    static COUNTRY_IDENTIFICATION_FIELD_NAME = 'CountryIdentification';
    static COUNTRY_IDENTIFICATION_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.CountryIdentification';
    static CUSTOMER_FAMILY_NAME_FIELD_NAME = 'CustomerFamilyName';
    static CUSTOMER_FAMILY_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.CustomerFamilyName';
    static CUSTOMER_FIRST_NAME_FIELD_NAME = 'CustomerFirstName';
    static CUSTOMER_FIRST_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.CustomerFirstName';
    static CUSTOMER_GIVEN_NAMES_FIELD_NAME = 'CustomerGivenNames';
    static CUSTOMER_GIVEN_NAMES_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.CustomerGivenNames';
    static CUSTOMER_ID_NUMBER_FIELD_NAME = 'CustomerIdNumber';
    static CUSTOMER_ID_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.CustomerIdNumber';
    static CUSTOMER_MIDDLE_NAME_FIELD_NAME = 'CustomerMiddleName';
    static CUSTOMER_MIDDLE_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.CustomerMiddleName';
    static DATE_OF_BIRTH_FIELD_NAME = 'DateOfBirth';
    static DATE_OF_BIRTH_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DateOfBirth';
    static DOCUMENT_DISCRIMINATOR_FIELD_NAME = 'DocumentDiscriminator';
    static DOCUMENT_DISCRIMINATOR_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DocumentDiscriminator';
    static DOCUMENT_EXPIRATION_DATE_FIELD_NAME = 'DocumentExpirationDate';
    static DOCUMENT_EXPIRATION_DATE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DocumentExpirationDate';
    static DOCUMENT_ISSUE_DATE_FIELD_NAME = 'DocumentIssueDate';
    static DOCUMENT_ISSUE_DATE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DocumentIssueDate';
    static DRIVER_ALIAS_DATE_OF_BIRTH_FIELD_FIELD_NAME = 'DriverAliasDateOfBirthField';
    static DRIVER_ALIAS_DATE_OF_BIRTH_FIELD_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverAliasDateOfBirthField';
    static DRIVER_ALIAS_FIRST_NAME_FIELD_NAME = 'DriverAliasFirstName';
    static DRIVER_ALIAS_FIRST_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverAliasFirstName';
    static DRIVER_ALIAS_LAST_NAME_FIELD_NAME = 'DriverAliasLastName';
    static DRIVER_ALIAS_LAST_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverAliasLastName';
    static DRIVER_ALIAS_SOCIAL_SECURITY_NUMBER_FIELD_NAME = 'DriverAliasSocialSecurityNumber';
    static DRIVER_ALIAS_SOCIAL_SECURITY_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverAliasSocialSecurityNumber';
    static DRIVER_LAST_NAME_FIELD_NAME = 'DriverLastName';
    static DRIVER_LAST_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverLastName';
    static DRIVER_LICENSE_CLASSIFICATION_CODE_FIELD_NAME = 'DriverLicenseClassificationCode';
    static DRIVER_LICENSE_CLASSIFICATION_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverLicenseClassificationCode';
    static DRIVER_LICENSE_ENDORSEMENTS_CODE_FIELD_NAME = 'DriverLicenseEndorsementsCode';
    static DRIVER_LICENSE_ENDORSEMENTS_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverLicenseEndorsementsCode';
    static DRIVER_LICENSE_NAME_FIELD_NAME = 'DriverLicenseName';
    static DRIVER_LICENSE_NAME_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverLicenseName';
    static DRIVER_LICENSE_RESTRICTION_CODE_FIELD_NAME = 'DriverLicenseRestrictionCode';
    static DRIVER_LICENSE_RESTRICTION_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverLicenseRestrictionCode';
    static DRIVER_NAME_PREFIX_FIELD_NAME = 'DriverNamePrefix';
    static DRIVER_NAME_PREFIX_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverNamePrefix';
    static DRIVER_NAME_SUFFIX_FIELD_NAME = 'DriverNameSuffix';
    static DRIVER_NAME_SUFFIX_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverNameSuffix';
    static DRIVER_PERMIT_CLASSIFICATION_CODE_FIELD_NAME = 'DriverPermitClassificationCode';
    static DRIVER_PERMIT_CLASSIFICATION_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverPermitClassificationCode';
    static DRIVER_PERMIT_ENDORSEMENT_CODE_FIELD_NAME = 'DriverPermitEndorsementCode';
    static DRIVER_PERMIT_ENDORSEMENT_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverPermitEndorsementCode';
    static DRIVER_PERMIT_ISSUE_DATE_FIELD_NAME = 'DriverPermitIssueDate';
    static DRIVER_PERMIT_ISSUE_DATE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverPermitIssueDate';
    static DRIVER_PERMIT_RESTRICTION_CODE_FIELD_NAME = 'DriverPermitRestrictionCode';
    static DRIVER_PERMIT_RESTRICTION_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverPermitRestrictionCode';
    static DRIVER_RESIDENCE_CITY_FIELD_NAME = 'DriverResidenceCity';
    static DRIVER_RESIDENCE_CITY_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverResidenceCity';
    static DRIVER_RESIDENCE_JURISDICTION_CODE_FIELD_NAME = 'DriverResidenceJurisdictionCode';
    static DRIVER_RESIDENCE_JURISDICTION_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverResidenceJurisdictionCode';
    static DRIVER_RESIDENCE_POSTAL_CODE_FIELD_NAME = 'DriverResidencePostalCode';
    static DRIVER_RESIDENCE_POSTAL_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverResidencePostalCode';
    static DRIVER_RESIDENCE_STREET_ADDRESS_1_FIELD_NAME = 'DriverResidenceStreetAddress1';
    static DRIVER_RESIDENCE_STREET_ADDRESS_1_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverResidenceStreetAddress1';
    static DRIVER_RESIDENCE_STREET_ADDRESS_2_FIELD_NAME = 'DriverResidenceStreetAddress2';
    static DRIVER_RESIDENCE_STREET_ADDRESS_2_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.DriverResidenceStreetAddress2';
    static FAMILY_NAME_TRUNCATION_FIELD_NAME = 'FamilyNameTruncation';
    static FAMILY_NAME_TRUNCATION_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.FamilyNameTruncation';
    static FEDERAL_COMMERCIAL_VEHICLE_CODES_FIELD_NAME = 'FederalCommercialVehicleCodes';
    static FEDERAL_COMMERCIAL_VEHICLE_CODES_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.FederalCommercialVehicleCodes';
    static FIRST_NAME_TRUNCATION_FIELD_NAME = 'FirstNameTruncation';
    static FIRST_NAME_TRUNCATION_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.FirstNameTruncation';
    static HAZMAT_ENDORSEMENT_EXPIRATION_DATE_FIELD_NAME = 'HAZMATEndorsementExpirationDate';
    static HAZMAT_ENDORSEMENT_EXPIRATION_DATE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.HAZMATEndorsementExpirationDate';
    static HAIR_COLOR_FIELD_NAME = 'HairColor';
    static HAIR_COLOR_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.HairColor';
    static HEIGHT_FIELD_NAME = 'Height';
    static HEIGHT_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.Height';
    static INVENTORY_CONTROL_NUMBER_FIELD_NAME = 'InventoryControlNumber';
    static INVENTORY_CONTROL_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.InventoryControlNumber';
    static ISSUE_TIME_STAMP_FIELD_NAME = 'IssueTimeStamp';
    static ISSUE_TIME_STAMP_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.IssueTimeStamp';
    static JURISDICTION_SPECIFIC_ENDORSEMENT_CODE_DESCRIPTION_FIELD_NAME = 'JurisdictionSpecificEndorsementCodeDescription';
    static JURISDICTION_SPECIFIC_ENDORSEMENT_CODE_DESCRIPTION_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.JurisdictionSpecificEndorsementCodeDescription';
    static JURISDICTION_SPECIFIC_ENDORSEMENT_CODES_FIELD_NAME = 'JurisdictionSpecificEndorsementCodes';
    static JURISDICTION_SPECIFIC_ENDORSEMENT_CODES_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.JurisdictionSpecificEndorsementCodes';
    static JURISDICTION_SPECIFIC_RESTRICTION_CODE_DESCRIPTION_FIELD_NAME = 'JurisdictionSpecificRestrictionCodeDescription';
    static JURISDICTION_SPECIFIC_RESTRICTION_CODE_DESCRIPTION_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.JurisdictionSpecificRestrictionCodeDescription';
    static JURISDICTION_SPECIFIC_RESTRICTION_CODES_FIELD_NAME = 'JurisdictionSpecificRestrictionCodes';
    static JURISDICTION_SPECIFIC_RESTRICTION_CODES_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.JurisdictionSpecificRestrictionCodes';
    static JURISDICTION_SPECIFIC_VEHICLE_CLASS_FIELD_NAME = 'JurisdictionSpecificVehicleClass';
    static JURISDICTION_SPECIFIC_VEHICLE_CLASS_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.JurisdictionSpecificVehicleClass';
    static JURISDICTION_SPECIFIC_VEHICLE_CLASSIFICATION_DESCRIPTION_FIELD_NAME = 'JurisdictionSpecificVehicleClassificationDescription';
    static JURISDICTION_SPECIFIC_VEHICLE_CLASSIFICATION_DESCRIPTION_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.JurisdictionSpecificVehicleClassificationDescription';
    static LIMITED_DURATION_DOCUMENT_INDICATOR_FIELD_NAME = 'LimitedDurationDocumentIndicator';
    static LIMITED_DURATION_DOCUMENT_INDICATOR_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.LimitedDurationDocumentIndicator';
    static MIDDLE_NAME_TRUNCATION_FIELD_NAME = 'MiddleNameTruncation';
    static MIDDLE_NAME_TRUNCATION_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.MiddleNameTruncation';
    static NAME_SUFFIX_FIELD_NAME = 'NameSuffix';
    static NAME_SUFFIX_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.NameSuffix';
    static NON_RESIDENT_INDICATOR_FIELD_NAME = 'NonResidentIndicator';
    static NON_RESIDENT_INDICATOR_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.NonResidentIndicator';
    static NUMBER_OF_DUPLICATES_FIELD_NAME = 'NumberOfDuplicates';
    static NUMBER_OF_DUPLICATES_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.NumberOfDuplicates';
    static ORGAN_DONOR_FIELD_NAME = 'OrganDonor';
    static ORGAN_DONOR_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.OrganDonor';
    static ORGAN_DONOR_INDICATOR_FIELD_NAME = 'OrganDonorIndicator';
    static ORGAN_DONOR_INDICATOR_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.OrganDonorIndicator';
    static PERMIT_EXPIRATION_DATE_FIELD_NAME = 'PermitExpirationDate';
    static PERMIT_EXPIRATION_DATE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.PermitExpirationDate';
    static PERMIT_IDENTIFIER_FIELD_NAME = 'PermitIdentifier';
    static PERMIT_IDENTIFIER_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.PermitIdentifier';
    static PHYSICAL_DESCRIPTION_EYE_COLOR_FIELD_NAME = 'PhysicalDescriptionEyeColor';
    static PHYSICAL_DESCRIPTION_EYE_COLOR_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.PhysicalDescriptionEyeColor';
    static PHYSICAL_DESCRIPTION_HEIGHT_FIELD_NAME = 'PhysicalDescriptionHeight';
    static PHYSICAL_DESCRIPTION_HEIGHT_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.PhysicalDescriptionHeight';
    static PHYSICAL_DESCRIPTION_SEX_FIELD_NAME = 'PhysicalDescriptionSex';
    static PHYSICAL_DESCRIPTION_SEX_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.PhysicalDescriptionSex';
    static PHYSICAL_DESCRIPTION_WEIGHT_RANGE_FIELD_NAME = 'PhysicalDescriptionWeightRange';
    static PHYSICAL_DESCRIPTION_WEIGHT_RANGE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.PhysicalDescriptionWeightRange';
    static PLACE_OF_BIRTH_FIELD_NAME = 'PlaceOfBirth';
    static PLACE_OF_BIRTH_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.PlaceOfBirth';
    static RACE_ETHNICITY_FIELD_NAME = 'RaceEthnicity';
    static RACE_ETHNICITY_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.RaceEthnicity';
    static SOCIAL_SECURITY_NUMBER_FIELD_NAME = 'SocialSecurityNumber';
    static SOCIAL_SECURITY_NUMBER_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.SocialSecurityNumber';
    static STANDARD_ENDORSEMENT_CODE_FIELD_NAME = 'StandardEndorsementCode';
    static STANDARD_ENDORSEMENT_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.StandardEndorsementCode';
    static STANDARD_RESTRICTION_CODE_FIELD_NAME = 'StandardRestrictionCode';
    static STANDARD_RESTRICTION_CODE_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.StandardRestrictionCode';
    static STANDARD_VEHICLE_CLASSIFICATION_FIELD_NAME = 'StandardVehicleClassification';
    static STANDARD_VEHICLE_CLASSIFICATION_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.StandardVehicleClassification';
    static UNDER_18_UNTIL_FIELD_NAME = 'Under18Until';
    static UNDER_18_UNTIL_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.Under18Until';
    static UNDER_19_UNTIL_FIELD_NAME = 'Under19Until';
    static UNDER_19_UNTIL_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.Under19Until';
    static UNDER_21_UNTIL_FIELD_NAME = 'Under21Until';
    static UNDER_21_UNTIL_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.Under21Until';
    static UNIQUE_CUSTOMER_IDENTIFIER_FIELD_NAME = 'UniqueCustomerIdentifier';
    static UNIQUE_CUSTOMER_IDENTIFIER_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.UniqueCustomerIdentifier';
    static VETERAN_INDICATOR_FIELD_NAME = 'VeteranIndicator';
    static VETERAN_INDICATOR_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.VeteranIndicator';
    static WEIGHT_KILOGRAMS_FIELD_NAME = 'WeightKilograms';
    static WEIGHT_KILOGRAMS_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.WeightKilograms';
    static WEIGHT_POUNDS_FIELD_NAME = 'WeightPounds';
    static WEIGHT_POUNDS_FIELD_NAME_NORMALIZED = 'AAMVA.DLID.WeightPounds';
    constructor(document) {
      if (document.type.name !== this.requiredDocumentType()) {
        throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
      }
      this._document = document;
    }
    /** Data element ID "DAI". City portion of the cardholder address. */
    get addressCity() {
      return this.document.fields.find(f => f.type.name === 'AddressCity');
    }

    /** Data element ID "DAJ". State portion of the cardholder address. */
    get addressJurisdictionCode() {
      return this.document.fields.find(f => f.type.name === 'AddressJurisdictionCode');
    }

    /** Data element ID "DAK". Postal code portion of the cardholder address in the U.S. and Canada. If the trailing portion of the postal code in the U.S. is not known, zeros will be used to fill the trailing set of numbers up to nine (9) digits. */
    get addressPostalCode() {
      return this.document.fields.find(f => f.type.name === 'AddressPostalCode');
    }

    /** Data element ID "DAG". Street portion of the cardholder address. */
    get addressStreet1() {
      return this.document.fields.find(f => f.type.name === 'AddressStreet1');
    }

    /** Data element ID "DAH". Second line of street portion of the cardholder address. */
    get addressStreet2() {
      return this.document.fields.find(f => f.type.name === 'AddressStreet2');
    }

    /** Data element ID "DCJ". A string of letters and/or numbers that identifies when, where, and by whom a driver license/ID card was made. If audit information is not used on the card or the MRT, it must be included in the driver record. */
    get adultInformation() {
      return this.document.fields.find(f => f.type.name === 'AdultInformation');
    }

    /** Data element ID "DBN". Other family name by which cardholder is known. */
    get aliasFamilyName() {
      return this.document.fields.find(f => f.type.name === 'AliasFamilyName');
    }

    /** Data element ID "DBG". Other given name by which cardholder is known. */
    get aliasGivenName() {
      return this.document.fields.find(f => f.type.name === 'AliasGivenName');
    }

    /** Data element ID "DBQ". Other middle name by which cardholder is known. */
    get aliasMiddleName() {
      return this.document.fields.find(f => f.type.name === 'AliasMiddleName');
    }

    /** Data element ID "DBR". Other prefix by which cardholder is known. */
    get aliasPrefixName() {
      return this.document.fields.find(f => f.type.name === 'AliasPrefixName');
    }

    /** Data element ID "DBS". Other suffix by which cardholder is known. */
    get aliasSuffixName() {
      return this.document.fields.find(f => f.type.name === 'AliasSuffixName');
    }

    /** Data element ID "DDB". DHS required field that indicates date of the most recent version change or modification to the visible format of the DL/ID. The format is ISO8601 with delimiters. */
    get cardRevisionDate() {
      return this.document.fields.find(f => f.type.name === 'CardRevisionDate');
    }

    /** Data element ID "DDA". DHS required field that indicates compliance: "F" = compliant; and, "N" = non-compliant. */
    get complianceType() {
      return this.document.fields.find(f => f.type.name === 'ComplianceType');
    }

    /** Data element ID "DCG". Country in which DL/ID is issued. U.S. = USA, Canada = CAN. */
    get countryIdentification() {
      return this.document.fields.find(f => f.type.name === 'CountryIdentification');
    }

    /** Data element ID "DCS". Family name of the cardholder. (Family name is sometimes also called “last name” or “surname.”) Collect full name for record, print as many characters as possible on portrait side of DL/ID. */
    get customerFamilyName() {
      return this.document.fields.find(f => f.type.name === 'CustomerFamilyName');
    }

    /** Data element ID "DAC". First name of the cardholder. */
    get customerFirstName() {
      return this.document.fields.find(f => f.type.name === 'CustomerFirstName');
    }

    /** Data element ID "DCT". Given names of the cardholder. (Given names include all names other than the Family Name. This includes all those names sometimes also called “first” and “middle” names.) */
    get customerGivenNames() {
      return this.document.fields.find(f => f.type.name === 'CustomerGivenNames');
    }

    /** Data element ID "DAQ". The number assigned or calculated by the issuing authority. */
    get customerIdNumber() {
      return this.document.fields.find(f => f.type.name === 'CustomerIdNumber');
    }

    /** Data element ID "DAD". Middle name(s) of the cardholder. In the case of multiple middle names they shall be separated by a comma “,”. */
    get customerMiddleName() {
      return this.document.fields.find(f => f.type.name === 'CustomerMiddleName');
    }

    /** Data element ID "DBB". Date on which the cardholder was born. The format is ISO8601 with delimiters. */
    get dateOfBirth() {
      return this.document.fields.find(f => f.type.name === 'DateOfBirth');
    }

    /** Data element ID "DCF". Number must uniquely identify a particular document issued to that customer from others that may have been issued in the past. This number may serve multiple purposes of document discrimination, audit information number, and/or inventory control. */
    get documentDiscriminator() {
      return this.document.fields.find(f => f.type.name === 'DocumentDiscriminator');
    }

    /** Data element ID "DBA". Date on which the driving and identification privileges granted by the document are no longer valid. The format is ISO8601 with delimtiers. */
    get documentExpirationDate() {
      return this.document.fields.find(f => f.type.name === 'DocumentExpirationDate');
    }

    /** Data element ID "DBD". Date on which the document was issued. The format is ISO8601 with delimiters. */
    get documentIssueDate() {
      return this.document.fields.find(f => f.type.name === 'DocumentIssueDate');
    }

    /** Data element ID "DBL". Alternative dates given as date of birth. */
    get driverAliasDateOfBirthField() {
      return this.document.fields.find(f => f.type.name === 'DriverAliasDateOfBirthField');
    }

    /** Data element ID "DBP". Alternative fist name or given name of the individual holding the Driver License or ID. Hyphenated names acceptable, but no other use of special symbols. */
    get driverAliasFirstName() {
      return this.document.fields.find(f => f.type.name === 'DriverAliasFirstName');
    }

    /** Data element ID "DBO". Alternative last name or surname of the individual holding the Driver License or ID. Hyphenated names acceptable, but no other use of special symbols. */
    get driverAliasLastName() {
      return this.document.fields.find(f => f.type.name === 'DriverAliasLastName');
    }

    /** Data element ID "DBM". DriverAliasSocialSecurityNumber */
    get driverAliasSocialSecurityNumber() {
      return this.document.fields.find(f => f.type.name === 'DriverAliasSocialSecurityNumber');
    }

    /** Data element ID "DAB". Last name or surname of the individual holding the Driver License or ID. Hyphenated names acceptable, but no other use of special symbols. */
    get driverLastName() {
      return this.document.fields.find(f => f.type.name === 'DriverLastName');
    }

    /** Data element ID "DAR". A=Class A; B=Class B; C=Class C (Class A, B and C are defined by Federal Highway regulations); M=Class M motorcycle as defined by AAMVA; others are defined by DL Classification Code Standards. */
    get driverLicenseClassificationCode() {
      return this.document.fields.find(f => f.type.name === 'DriverLicenseClassificationCode');
    }

    /** Data element ID "DAT". Any endorsements on a driver license which authorize the operation of specified types of vehicles or the operation of vehicles carrying specified loads. Endorsements shall be specific to classification of a driver license. */
    get driverLicenseEndorsementsCode() {
      return this.document.fields.find(f => f.type.name === 'DriverLicenseEndorsementsCode');
    }

    /** Data element ID "DAA". Name of the individual holding the Driver License or ID as defined in ANSI D20 Data Dictionary. (Lastname@Firstname@MI@ suffix if any) (Machine, Mag Stripe uses ‘$’ and Bar Code uses ‘,’ in place of ‘@’) Firstname, Middle Initial, Lastname (Human) */
    get driverLicenseName() {
      return this.document.fields.find(f => f.type.name === 'DriverLicenseName');
    }

    /** Data element ID "DAS". A restriction applicable to a driver license. */
    get driverLicenseRestrictionCode() {
      return this.document.fields.find(f => f.type.name === 'DriverLicenseRestrictionCode');
    }

    /** Data element ID "DAF". Prefix to Driver Name. Not defined in ANSI D20. Freeform as defined by issuing jurisdiction. */
    get driverNamePrefix() {
      return this.document.fields.find(f => f.type.name === 'DriverNamePrefix');
    }

    /** Data element ID "DAE". An affix occurring at the end of a word, e.g.; Sr., Jr., II, III, IV, etc. */
    get driverNameSuffix() {
      return this.document.fields.find(f => f.type.name === 'DriverNameSuffix');
    }

    /** Data element ID "PAA". Identifies the type of permit as defined by ANSI D20. */
    get driverPermitClassificationCode() {
      return this.document.fields.find(f => f.type.name === 'DriverPermitClassificationCode');
    }

    /** Data element ID "PAF". Permit endorsements as defined by ANSI D20. */
    get driverPermitEndorsementCode() {
      return this.document.fields.find(f => f.type.name === 'DriverPermitEndorsementCode');
    }

    /** Data element ID "PAD".  Date permit was issued. The format is ISO8601 with delimiters. */
    get driverPermitIssueDate() {
      return this.document.fields.find(f => f.type.name === 'DriverPermitIssueDate');
    }

    /** Data element ID "PAE". Permit restrictions as defined by ANSI D20. */
    get driverPermitRestrictionCode() {
      return this.document.fields.find(f => f.type.name === 'DriverPermitRestrictionCode');
    }

    /** Data element ID "DAN". Name of city for mailing address. */
    get driverResidenceCity() {
      return this.document.fields.find(f => f.type.name === 'DriverResidenceCity');
    }

    /** Data element ID "DAO". Jurisdiction code for mailing address. Conforms to Canadian, Mexican and US Jurisdictions as appropriate. Codes for provinces (Canada) and states (US and Mexico). */
    get driverResidenceJurisdictionCode() {
      return this.document.fields.find(f => f.type.name === 'DriverResidenceJurisdictionCode');
    }

    /** Data element ID "DAP". Postal code of residence */
    get driverResidencePostalCode() {
      return this.document.fields.find(f => f.type.name === 'DriverResidencePostalCode');
    }

    /** Data element ID "DAL". Street address line 1 (mailing). */
    get driverResidenceStreetAddress1() {
      return this.document.fields.find(f => f.type.name === 'DriverResidenceStreetAddress1');
    }

    /** Data element ID "DAM". Street address line 2 (mailing). */
    get driverResidenceStreetAddress2() {
      return this.document.fields.find(f => f.type.name === 'DriverResidenceStreetAddress2');
    }

    /** Data element ID "DDE". A code that indicates whether a field has been truncated (T), has not been truncated (N), or – unknown whether truncated (U). */
    get familyNameTruncation() {
      return this.document.fields.find(f => f.type.name === 'FamilyNameTruncation');
    }

    /** Data element ID "DCH". Federal Commercial Vehicle Codes */
    get federalCommercialVehicleCodes() {
      return this.document.fields.find(f => f.type.name === 'FederalCommercialVehicleCodes');
    }

    /** Data element ID "DDF". A code that indicates whether a field has been truncated (T), has not been truncated (N), or – unknown whether truncated (U). */
    get firstNameTruncation() {
      return this.document.fields.find(f => f.type.name === 'FirstNameTruncation');
    }

    /** Data element ID "DDC". Date on which the hazardous material endorsement granted by the document is no longer valid. The format is ISO8601 with delimiters. */
    get hazmatEndorsementExpirationDate() {
      return this.document.fields.find(f => f.type.name === 'HAZMATEndorsementExpirationDate');
    }

    /** Data element ID "DAZ". Color of cardholder's hair. ANSI D-20 codes converted to human readable format according to the ANSI D20 Data Dictionary. */
    get hairColor() {
      return this.document.fields.find(f => f.type.name === 'HairColor');
    }

    /** Data element ID "DAV". Height in centimeters */
    get height() {
      return this.document.fields.find(f => f.type.name === 'Height');
    }

    /** Data element ID "DCK". A string of letters and/or numbers that is affixed to the raw materials (card stock, laminate, etc.) used in producing driver licenses and ID cards. (DHS recommended field). */
    get inventoryControlNumber() {
      return this.document.fields.find(f => f.type.name === 'InventoryControlNumber');
    }

    /** Data element ID "DBE". A string used by some jurisdictions to validate the document against their data base. */
    get issueTimeStamp() {
      return this.document.fields.find(f => f.type.name === 'IssueTimeStamp');
    }

    /** Data element ID "DCQ". Text that explains the jurisdiction-specific code(s) that indicates additional driving privileges granted to the cardholder beyond the vehicle class. */
    get jurisdictionSpecificEndorsementCodeDescription() {
      return this.document.fields.find(f => f.type.name === 'JurisdictionSpecificEndorsementCodeDescription');
    }

    /** Data element ID "DCD". Jurisdiction-specific codes that represent additional privileges granted to the cardholder beyond the vehicle class (such as transportation of passengers, hazardous materials, operation of motorcycles, etc.). */
    get jurisdictionSpecificEndorsementCodes() {
      return this.document.fields.find(f => f.type.name === 'JurisdictionSpecificEndorsementCodes');
    }

    /** Data element ID "DCR". Text describing the jurisdiction-specific restriction code(s) that curtail driving privileges. */
    get jurisdictionSpecificRestrictionCodeDescription() {
      return this.document.fields.find(f => f.type.name === 'JurisdictionSpecificRestrictionCodeDescription');
    }

    /** Data element ID "DCB". Jurisdiction-specific codes that represent restrictions to driving privileges (such as airbrakes, automatic transmission, daylight only, etc.). */
    get jurisdictionSpecificRestrictionCodes() {
      return this.document.fields.find(f => f.type.name === 'JurisdictionSpecificRestrictionCodes');
    }

    /** Data element ID "DCA". Jurisdiction-specific vehicle class / group code, designating the type of vehicle the cardholder has privilege to drive. */
    get jurisdictionSpecificVehicleClass() {
      return this.document.fields.find(f => f.type.name === 'JurisdictionSpecificVehicleClass');
    }

    /** Data element ID "DCP". Text that explains the jurisdiction-specific code(s) for classifications of vehicles cardholder is authorized to drive. */
    get jurisdictionSpecificVehicleClassificationDescription() {
      return this.document.fields.find(f => f.type.name === 'JurisdictionSpecificVehicleClassificationDescription');
    }

    /** Data element ID "DDD". DHS required field that indicates that the cardholder has temporary lawful status = "1". */
    get limitedDurationDocumentIndicator() {
      return this.document.fields.find(f => f.type.name === 'LimitedDurationDocumentIndicator');
    }

    /** Data element ID "DDG". A code that indicates whether a field has been truncated (T), has not been truncated (N), or – unknown whether truncated (U). */
    get middleNameTruncation() {
      return this.document.fields.find(f => f.type.name === 'MiddleNameTruncation');
    }

    /** Data element ID "DCU". Name Suffix (If jurisdiction participates in systems requiring name suffix (PDPS, CDLIS, etc.), the suffix must be collected and displayed on the DL/ID and in the MRT). Collect full name for record, print as many characters as possible on portrait side of DL/ID. JR (Junior), SR (Senior), 1ST or I (First), 2ND or II (Second),  3RD or III (Third),  4TH or IV (Fourth), 5TH or V (Fifth), 6TH or VI (Sixth), 7TH or VII (Seventh), 8TH or VIII (Eighth), 9TH or IX (Ninth). */
    get nameSuffix() {
      return this.document.fields.find(f => f.type.name === 'NameSuffix');
    }

    /** Data element ID "DBI". "Y"; Used by some jurisdictions to indicate holder of the document is a non-resident. */
    get nonResidentIndicator() {
      return this.document.fields.find(f => f.type.name === 'NonResidentIndicator');
    }

    /** Data element ID "DBF". Number of duplicate cards issued for a license or ID if any. */
    get numberOfDuplicates() {
      return this.document.fields.find(f => f.type.name === 'NumberOfDuplicates');
    }

    /** Data element ID "DBH". Organ Donor */
    get organDonor() {
      return this.document.fields.find(f => f.type.name === 'OrganDonor');
    }

    /** Data element ID "DDK". Field that indicates that the cardholder is an organ donor = "1". */
    get organDonorIndicator() {
      return this.document.fields.find(f => f.type.name === 'OrganDonorIndicator');
    }

    /** Data element ID "PAB". Date permit expires, The format is ISO8601 with delimiters. */
    get permitExpirationDate() {
      return this.document.fields.find(f => f.type.name === 'PermitExpirationDate');
    }

    /** Data element ID "PAC". Type of permit. */
    get permitIdentifier() {
      return this.document.fields.find(f => f.type.name === 'PermitIdentifier');
    }

    /** Data element ID "DAY". Color of cardholder's eyes. ANSI D-20 codes converted to human readable format according to the ANSI D20 Data Dictionary. */
    get physicalDescriptionEyeColor() {
      return this.document.fields.find(f => f.type.name === 'PhysicalDescriptionEyeColor');
    }

    /** Data element ID "DAU". Height of cardholder. Inches (in): number of inches followed by " in" or Centimeters (cm): number of centimeters followed by " cm." */
    get physicalDescriptionHeight() {
      return this.document.fields.find(f => f.type.name === 'PhysicalDescriptionHeight');
    }

    /** Data element ID "DBC". Gender of the cardholder. "Male" or "Female". */
    get physicalDescriptionSex() {
      return this.document.fields.find(f => f.type.name === 'PhysicalDescriptionSex');
    }

    /** Data element ID "DCE". Indicates the approximate weight range of the cardholder: 0 = up to 31 kg (up to 70 lbs), 1 = 32 – 45 kg (71 – 100 lbs), 2 = 46 - 59 kg (101 – 130 lbs), 3 = 60 - 70 kg (131 – 160 lbs), 4 = 71 - 86 kg (161 – 190 lbs), 5 = 87 - 100 kg (191 – 220 lbs), 6 = 101 - 113 kg (221 – 250 lbs), 7 = 114 - 127 kg (251 – 280 lbs), 8 = 128 – 145 kg (281 – 320 lbs), 9 = 146+ kg (321+ lbs). */
    get physicalDescriptionWeightRange() {
      return this.document.fields.find(f => f.type.name === 'PhysicalDescriptionWeightRange');
    }

    /** Data element ID "DCI". Country and municipality and/or state/province. */
    get placeOfBirth() {
      return this.document.fields.find(f => f.type.name === 'PlaceOfBirth');
    }

    /** Data element ID "DCL". Codes for race or ethnicity of the cardholder. ANSI D-20 codes converted to human readable format according to the ANSI D20 Data Dictionary. */
    get raceEthnicity() {
      return this.document.fields.find(f => f.type.name === 'RaceEthnicity');
    }

    /** Data element ID "DBK". The number assigned to an individual by the Social Security Administration. */
    get socialSecurityNumber() {
      return this.document.fields.find(f => f.type.name === 'SocialSecurityNumber');
    }

    /** Data element ID "DCN". Standard endorsement code(s) for cardholder. See codes in D20. This data element is a placeholder for future efforts to standardize endorsement codes. */
    get standardEndorsementCode() {
      return this.document.fields.find(f => f.type.name === 'StandardEndorsementCode');
    }

    /** Data element ID "DCO". Standard restriction code(s) for cardholder. See codes in D20. This data element is a placeholder for future efforts to standardize restriction codes. */
    get standardRestrictionCode() {
      return this.document.fields.find(f => f.type.name === 'StandardRestrictionCode');
    }

    /** Data element ID "DCM". Standard vehicle classification code(s) for cardholder. This data element is a placeholder for future efforts to standardize vehicle classifications. */
    get standardVehicleClassification() {
      return this.document.fields.find(f => f.type.name === 'StandardVehicleClassification');
    }

    /** Data element ID "DDH". Date on which the cardholder turns 18 years old. The format is ISO8601 with delimiters. */
    get under18Until() {
      return this.document.fields.find(f => f.type.name === 'Under18Until');
    }

    /** Data element ID "DDI". Date on which the cardholder turns 19 years old. The format is ISO8601 with delimiters. */
    get under19Until() {
      return this.document.fields.find(f => f.type.name === 'Under19Until');
    }

    /** Data element ID "DDJ". Date on which the cardholder turns 21 years old. The format is ISO8601 with delimiters. */
    get under21Until() {
      return this.document.fields.find(f => f.type.name === 'Under21Until');
    }

    /** Data element ID "DBJ". A number or alphanumeric string used by some jurisdictions to identify a "customer" across multiple data bases. */
    get uniqueCustomerIdentifier() {
      return this.document.fields.find(f => f.type.name === 'UniqueCustomerIdentifier');
    }

    /** Data element ID "DDL". Field that indicates that the cardholder is a veteran = "1". */
    get veteranIndicator() {
      return this.document.fields.find(f => f.type.name === 'VeteranIndicator');
    }

    /** Data element ID "DAX". Cardholder weight in kilograms Ex. 84 kg = "084" */
    get weightKilograms() {
      return this.document.fields.find(f => f.type.name === 'WeightKilograms');
    }

    /** Data element ID "DAW". Cardholder weight in pounds Ex. 185 lb = "185" */
    get weightPounds() {
      return this.document.fields.find(f => f.type.name === 'WeightPounds');
    }
  }
  _AAMVA.DLID = DLID;
  class DriverLicense extends DLID {
    static DOCUMENT_TYPE = 'DriverLicense';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return AAMVA.DriverLicense.DOCUMENT_TYPE;
    }
  }
  _AAMVA.DriverLicense = DriverLicense;
  class IDCard extends DLID {
    static DOCUMENT_TYPE = 'IDCard';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return AAMVA.IDCard.DOCUMENT_TYPE;
    }
  }
  _AAMVA.IDCard = IDCard;
  class EnhancedDriverLicense extends DLID {
    static DOCUMENT_TYPE = 'EnhancedDriverLicense';
    constructor(document) {
      super(document);
    }
    requiredDocumentType() {
      return AAMVA.EnhancedDriverLicense.DOCUMENT_TYPE;
    }
  }
  _AAMVA.EnhancedDriverLicense = EnhancedDriverLicense;
  class RawDocument {
    get document() {
      return this._document;
    }
    static DOCUMENT_TYPE = 'RawDocument';
    constructor(document) {
      if (document.type.name !== this.requiredDocumentType()) {
        throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
      }
      this._document = document;
    }
    requiredDocumentType() {
      return AAMVA.RawDocument.DOCUMENT_TYPE;
    }
  }
  _AAMVA.RawDocument = RawDocument;
})(AAMVA || (exports.AAMVA = AAMVA = {}));
/** HIBC */
class HIBC {
  get document() {
    return this._document;
  }
  static DOCUMENT_TYPE = 'HIBC';
  static DATE_OF_MANUFACTURE_FIELD_NAME = 'DateOfManufacture';
  static DATE_OF_MANUFACTURE_FIELD_NAME_NORMALIZED = 'HIBC.DateOfManufacture';
  static EXPIRY_DATE_DAY_FIELD_NAME = 'ExpiryDateDay';
  static EXPIRY_DATE_DAY_FIELD_NAME_NORMALIZED = 'HIBC.ExpiryDateDay';
  static EXPIRY_DATE_HOUR_FIELD_NAME = 'ExpiryDateHour';
  static EXPIRY_DATE_HOUR_FIELD_NAME_NORMALIZED = 'HIBC.ExpiryDateHour';
  static EXPIRY_DATE_JULIAN_DAY_FIELD_NAME = 'ExpiryDateJulianDay';
  static EXPIRY_DATE_JULIAN_DAY_FIELD_NAME_NORMALIZED = 'HIBC.ExpiryDateJulianDay';
  static EXPIRY_DATE_MONTH_FIELD_NAME = 'ExpiryDateMonth';
  static EXPIRY_DATE_MONTH_FIELD_NAME_NORMALIZED = 'HIBC.ExpiryDateMonth';
  static EXPIRY_DATE_YEAR_FIELD_NAME = 'ExpiryDateYear';
  static EXPIRY_DATE_YEAR_FIELD_NAME_NORMALIZED = 'HIBC.ExpiryDateYear';
  static HAS_PRIMARY_DATA_FIELD_NAME = 'HasPrimaryData';
  static HAS_PRIMARY_DATA_FIELD_NAME_NORMALIZED = 'HIBC.HasPrimaryData';
  static HAS_SECONDARY_DATA_FIELD_NAME = 'HasSecondaryData';
  static HAS_SECONDARY_DATA_FIELD_NAME_NORMALIZED = 'HIBC.HasSecondaryData';
  static LABELER_IDENTIFICATION_CODE_FIELD_NAME = 'LabelerIdentificationCode';
  static LABELER_IDENTIFICATION_CODE_FIELD_NAME_NORMALIZED = 'HIBC.LabelerIdentificationCode';
  static LABELERS_PRODUCT_OR_CATALOG_NUMBER_FIELD_NAME = 'LabelersProductOrCatalogNumber';
  static LABELERS_PRODUCT_OR_CATALOG_NUMBER_FIELD_NAME_NORMALIZED = 'HIBC.LabelersProductOrCatalogNumber';
  static LINK_CHARACTER_FIELD_NAME = 'LinkCharacter';
  static LINK_CHARACTER_FIELD_NAME_NORMALIZED = 'HIBC.LinkCharacter';
  static LOT_NUMBER_FIELD_NAME = 'LotNumber';
  static LOT_NUMBER_FIELD_NAME_NORMALIZED = 'HIBC.LotNumber';
  static QUANTITY_FIELD_NAME = 'Quantity';
  static QUANTITY_FIELD_NAME_NORMALIZED = 'HIBC.Quantity';
  static SERIAL_NUMBER_FIELD_NAME = 'SerialNumber';
  static SERIAL_NUMBER_FIELD_NAME_NORMALIZED = 'HIBC.SerialNumber';
  static UNIT_OF_MEASURE_ID_FIELD_NAME = 'UnitOfMeasureID';
  static UNIT_OF_MEASURE_ID_FIELD_NAME_NORMALIZED = 'HIBC.UnitOfMeasureID';
  constructor(document) {
    if (document.type.name !== this.requiredDocumentType()) {
      throw new Error(`Expected document type ${this.requiredDocumentType()}, got ${document.type.name}`);
    }
    this._document = document;
  }
  requiredDocumentType() {
    return HIBC.DOCUMENT_TYPE;
  }

  /** Date of Manufacture */
  get dateOfManufacture() {
    return this.document.fields.find(f => f.type.name === 'DateOfManufacture');
  }

  /** Expiry Date Day */
  get expiryDateDay() {
    return this.document.fields.find(f => f.type.name === 'ExpiryDateDay');
  }

  /** Expiry Date Hour */
  get expiryDateHour() {
    return this.document.fields.find(f => f.type.name === 'ExpiryDateHour');
  }

  /** Expiry Date Julian Day */
  get expiryDateJulianDay() {
    return this.document.fields.find(f => f.type.name === 'ExpiryDateJulianDay');
  }

  /** Expiry Date Month */
  get expiryDateMonth() {
    return this.document.fields.find(f => f.type.name === 'ExpiryDateMonth');
  }

  /** Expiry Date Year */
  get expiryDateYear() {
    return this.document.fields.find(f => f.type.name === 'ExpiryDateYear');
  }

  /** Information whether primary data is present. */
  get hasPrimaryData() {
    return this.document.fields.find(f => f.type.name === 'HasPrimaryData');
  }

  /** Information whether secondary data is present. */
  get hasSecondaryData() {
    return this.document.fields.find(f => f.type.name === 'HasSecondaryData');
  }

  /** Labeler Identification Code (LIC) */
  get labelerIdentificationCode() {
    return this.document.fields.find(f => f.type.name === 'LabelerIdentificationCode');
  }

  /** Labelers Product or Catalog Number (PCN). Alphanumeric data. */
  get labelersProductOrCatalogNumber() {
    return this.document.fields.find(f => f.type.name === 'LabelersProductOrCatalogNumber');
  }

  /** Link character to link standalone primary and secondary barcodes together. */
  get linkCharacter() {
    return this.document.fields.find(f => f.type.name === 'LinkCharacter');
  }

  /** Lot Number */
  get lotNumber() {
    return this.document.fields.find(f => f.type.name === 'LotNumber');
  }

  /** Quantity */
  get quantity() {
    return this.document.fields.find(f => f.type.name === 'Quantity');
  }

  /** Serial Number */
  get serialNumber() {
    return this.document.fields.find(f => f.type.name === 'SerialNumber');
  }

  /** Unit of Measure ID, Numeric value only, 0 through 9, where 0 always represents 
  a single unit. 1 to 8 are used to indicate different packaging levels above the unit of use.
  The value 9 is used for variable quantity containers when manual key entry or scan of
  a secondary will be used to collect specific quantity data.
  The labeler should ensure consistency in this field within their packaging process.
  */
  get unitOfMeasureID() {
    return this.document.fields.find(f => f.type.name === 'UnitOfMeasureID');
  }
}

/// All possible root document types.
exports.HIBC = HIBC;
