// react-native-scanbot-sdk/src/ui_v2/document/DocumentScannerTextLocalization.ts
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/document/DocumentScannerTextLocalization.yaml

import { DeepPartial, PartiallyConstructible } from '../../utils/utils';

/**
Configuration of all the strings for the standalone cropping screen.
*/
export class DocumentScannerTextLocalization extends PartiallyConstructible {
  /**
    The 'cancel' button title, located in the top bar of the camera screen and the reorder pages screen.
    
    Default is "Cancel"
    */
  public cameraTopBarCancelButtonTitle: string = 'Cancel';
  /**
    The title of the camera screen, located in the top bar.
    
    Default is "Scan Document"
    */
  public cameraTopBarTitle: string = 'Scan Document';
  /**
    The text for the static user guidance, located just below the top bar of the camera screen.
    
    Default is "Scan each page of your document."
    */
  public cameraTopGuidance: string = 'Scan each page of your document.';
  /**
    The user guidance text that is initially displayed, before a document is detected.
    
    Default is "Please hold your device over a document."
    */
  public cameraUserGuidanceStart: string = 'Please hold your device over a document.';
  /**
    The user guidance text displayed when no document is found.
    
    Default is "No document found"
    */
  public cameraUserGuidanceNoDocumentFound: string = 'No document found';
  /**
    The user guidance text displayed a detected documents aspect ratio is not within the acceptable range.
    
    Default is "Bad aspect ratio"
    */
  public cameraUserGuidanceBadAspectRatio: string = 'Bad aspect ratio';
  /**
    The user guidance text displayed when a landscape document is detected when the camera is in portrait mode (and vice versa).
    
    Default is "Orientation mismatch"
    */
  public cameraUserGuidanceOrientationMismatch: string = 'Orientation mismatch';
  /**
    The user guidance text displayed when the document pitch/skew angle is unacceptable.
    
    Default is "Bad angles"
    */
  public cameraUserGuidanceBadAngles: string = 'Bad angles';
  /**
    The user guidance text displayed when the background is too noisy.
    
    Default is "Its too noisy. Try another background for the document"
    */
  public cameraUserGuidanceTooNoisy: string =
    'Its too noisy. Try another background for the document';
  /**
    The user guidance text displayed when the document is not centered in the finder view.
    
    Default is "Document is off center"
    */
  public cameraUserGuidanceTextHintOffCenter: string = 'Document is off center';
  /**
    The user guidance text displayed when the document is too far from camera.
    
    Default is "Document is too small. Move camera closer."
    */
  public cameraUserGuidanceTooSmall: string = 'Document is too small. Move camera closer.';
  /**
    The user guidance text displayed when it is too dark to capture an adequate image.
    
    Default is "Its too dark. Add more light."
    */
  public cameraUserGuidanceTooDark: string = 'Its too dark. Add more light.';
  /**
    The user guidance text displayed when the device is in energy saving mode. iOS only.
    
    Default is "Energy save mode is on"
    */
  public cameraUserGuidanceEnergySaveMode: string = 'Energy save mode is on';
  /**
    The user guidance text displayed when the document is ready to be captured in 'auto snapping' mode.
    
    Default is "Hold still capturing..."
    */
  public cameraUserGuidanceReadyToCapture: string = 'Hold still capturing...';
  /**
    The user guidance text displayed when the document is ready to be captured in 'manual snapping' mode.
    
    Default is "Ready to capture!"
    */
  public cameraUserGuidanceReadyToCaptureManual: string = 'Ready to capture!';
  /**
    The 'import image' button title, located in the bottom bar of the camera screen.
    
    Default is "Import"
    */
  public cameraImportButtonTitle: string = 'Import';
  /**
    The 'flashlight on' button title, located in the bottom bar of the camera screen.
    
    Default is "On"
    */
  public cameraTorchOnButtonTitle: string = 'On';
  /**
    The 'flashlight off' button title, located in the bottom bar of the camera screen.
    
    Default is "Off"
    */
  public cameraTorchOffButtonTitle: string = 'Off';
  /**
    The 'auto snapping' mode button title, located in the bottom bar of the camera screen.
    
    Default is "Auto"
    */
  public cameraAutoSnapButtonTitle: string = 'Auto';
  /**
    The 'manual snapping' mode button title, located in the bottom bar of the camera screen.
    
    Default is "Manual"
    */
  public cameraManualSnapButtonTitle: string = 'Manual';
  /**
    The 'preview' button title, located in the bottom bar of the camera screen. '%d' denotes the total number of captured pages.
    
    Default is "%d Pages"
    */
  public cameraPreviewButtonTitle: string = '%d Pages';
  /**
    The 'done' button title of the introduction screen, located in the top bar.
    
    Default is "Done"
    */
  public cameraIntroDoneButton: string = 'Done';
  /**
    The title of the introduction screen, located in the top bar.
    
    Default is "How to scan a document"
    */
  public cameraIntroTitle: string = 'How to scan a document';
  /**
    The subtitle of the introduction screen.
    
    Default is "Follow the steps below to create a high-quality document scan"
    */
  public cameraIntroSubtitle: string =
    'Follow the steps below to create a high-quality document scan';
  /**
    The first scanning step text of the introduction screen.
    
    Default is "1. Place your document on a flat surface."
    */
  public cameraIntroItem1: string = '1. Place your document on a flat surface.';
  /**
    The second scanning step text of the introduction screen.
    
    Default is "2. Hold your phone above the document."
    */
  public cameraIntroItem2: string = '2. Hold your phone above the document.';
  /**
    The third scanning step text of the introduction screen.
    
    Default is "3. Follow the on-screen guidance to find the optimal position."
    */
  public cameraIntroItem3: string =
    '3. Follow the on-screen guidance to find the optimal position.';
  /**
    The fourth scanning step text of the introduction screen.
    
    Default is "4. Once you reach the optimal position, you can manually scan the document or let the app scan it automatically."
    */
  public cameraIntroItem4: string =
    '4. Once you reach the optimal position, you can manually scan the document or let the app scan it automatically.';
  /**
    The progress overlay title that is displayed on the camera screen during processing.
    
    Default is "Please wait..."
    */
  public cameraProgressOverlayTitle: string = 'Please wait...';
  /**
    The title of the cancel camera screen alert dialog.
    
    Default is "Cancel?"
    */
  public cameraCancelAlertTitle: string = 'Cancel?';
  /**
    The subtitle of the cancel camera screen alert dialog.
    
    Default is "Canceling will delete all the pages scanned so far. Are you sure you want to cancel?"
    */
  public cameraCancelAlertSubtitle: string =
    'Canceling will delete all the pages scanned so far. Are you sure you want to cancel?';
  /**
    The 'no' button title of the cancel camera screen alert dialog.
    
    Default is "No"
    */
  public cameraCancelNoButtonTitle: string = 'No';
  /**
    The 'yes' button title of the cancel camera screen alert dialog.
    
    Default is "Yes, Cancel"
    */
  public cameraCancelYesButtonTitle: string = 'Yes, Cancel';
  /**
    The title of the page limit reached alert dialog.
    
    Default is "You reached the limit!"
    */
  public cameraLimitReachedAlertTitle: string = 'You reached the limit!';
  /**
    The subtitle of the page limit reached alert dialog.
    
    Default is "You have scanned the maximum number of pages and cannot add any more. Please delete at least one page to perform a new scan."
    */
  public cameraLimitReachedAlertSubtitle: string =
    'You have scanned the maximum number of pages and cannot add any more. Please delete at least one page to perform a new scan.';
  /**
    The 'OK' button title of the page limit reached alert dialog.
    
    Default is "Ok"
    */
  public cameraLimitReachedOkButtonTitle: string = 'Ok';
  /**
    The hint text of the acknowledgement screen explaining that the quality of the scanned page is unacceptable.
    
    Default is "The quality of your scan does not seem sufficient."
    */
  public acknowledgementScreenBadDocumentHint: string =
    'The quality of your scan does not seem sufficient.';
  /**
    The 'retake' button title of the acknowledgement screen.
    
    Default is "Retake"
    */
  public acknowledgementRetakeButtonTitle: string = 'Retake';
  /**
    The 'accept' button title of the acknowledgement screen.
    
    Default is "Use Scan"
    */
  public acknowledgementAcceptButtonTitle: string = 'Use Scan';
  /**
    The title of the review screen, located in the top bar. '%d' denotes the total number of captured pages.
    
    Default is "Review (%d)"
    */
  public reviewScreenTitle: string = 'Review (%d)';
  /**
    The 'back' button title, located in the top bar of the review screen.
    
    Default is "Back"
    */
  public reviewTopBarBackButtonTitle: string = 'Back';
  /**
    The page count label text. The first occurrence of '%d' denotes the current page number, while the second occurrence of '%d' denotes the total number of captured pages.
    
    Default is "Page %d/%d"
    */
  public reviewScreenPageCount: string = 'Page %d/%d';
  /**
    The 'add' button title, located in the bottom bar of the review screen.
    
    Default is "Add"
    */
  public reviewScreenAddButtonTitle: string = 'Add';
  /**
    The 'retake' button title, located in the bottom bar of the review screen.
    
    Default is "Retake"
    */
  public reviewScreenRetakeButtonTitle: string = 'Retake';
  /**
    The 'crop' button title, located in the bottom bar of the review screen.
    
    Default is "Crop"
    */
  public reviewScreenCropButtonTitle: string = 'Crop';
  /**
    The 'rotate' button title, located in the bottom bar of the review screen.
    
    Default is "Rotate"
    */
  public reviewScreenRotateButtonTitle: string = 'Rotate';
  /**
    The 'delete' button title, located in the bottom bar of the review screen.
    
    Default is "Delete"
    */
  public reviewScreenDeleteButtonTitle: string = 'Delete';
  /**
    The 'submit' button title, located in the bottom bar of the review screen.
    
    Default is "Submit"
    */
  public reviewScreenSubmitButtonTitle: string = 'Submit';
  /**
    The 'delete all' menu button title, accessed via the 'more' button of the review screen.
    
    Default is "Delete all"
    */
  public reviewScreenDeleteAllButtonTitle: string = 'Delete all';
  /**
    The 'reorder pages' menu button title, accessed via the 'more' button of the review screen.
    
    Default is "Reorder pages"
    */
  public reviewScreenReorderPagesButtonTitle: string = 'Reorder pages';
  /**
    The 'close' button title of the zoom overlay for the review screen.
    
    Default is "Close"
    */
  public zoomOverlayCancelButtonText: string = 'Close';
  /**
    The title of the delete page alert dialog for the review screen.
    
    Default is "Delete page?"
    */
  public reviewDeletePageAlertTitle: string = 'Delete page?';
  /**
    The title of the delete all pages alert dialog for the review screen.
    
    Default is "Delete all?"
    */
  public reviewDeleteAllPagesAlertTitle: string = 'Delete all?';
  /**
    The subtitle of the delete page alert dialog for the review screen.
    
    Default is "Are you sure you want to delete the page?"
    */
  public reviewDeletePageAlertSubTitle: string = 'Are you sure you want to delete the page?';
  /**
    The subtitle of the delete all pages alert dialog for the review screen.
    
    Default is "Are you sure you want to delete all pages?"
    */
  public reviewDeleteAllPagesAlertSubtitle: string = 'Are you sure you want to delete all pages?';
  /**
    The 'confirm' button title of the delete page alert dialog for the review screen.
    
    Default is "Yes, Delete"
    */
  public reviewDeletePageAlertConfirmButtonTitle: string = 'Yes, Delete';
  /**
    The 'confirm' button title of the delete all pages alert dialog for the review screen.
    
    Default is "Yes, Delete all"
    */
  public reviewDeleteAllPagesAlertDeleteButtonTitle: string = 'Yes, Delete all';
  /**
    The 'delete and retake' button title of the delete page alert dialog for the review screen.
    
    Default is "Delete and Retake"
    */
  public reviewDeletePageAlertDeleteRetakeButtonTitle: string = 'Delete and Retake';
  /**
    The 'cancel' button title of the delete page alert dialog for the review screen.
    
    Default is "Cancel"
    */
  public reviewDeletePageAlertCancelButtonTitle: string = 'Cancel';
  /**
    The 'cancel' button title of the delete all pages alert dialog for the review screen.
    
    Default is "Cancel"
    */
  public reviewDeleteAllPagesAlertCancelButtonTitle: string = 'Cancel';
  /**
    The title of the acknowledgement screen, located in the top bar.
    
    Default is "Acknowledgement"
    */
  public acknowledgementTitle: string = 'Acknowledgement';
  /**
    The title of the reorder pages screen, located in the top bar.
    
    Default is "Reorder Pages"
    */
  public reorderPageTitle: string = 'Reorder Pages';
  /**
    The text for the static user guidance, located just below the top bar of the reorder pages screen.
    
    Default is "Drag pages to reorder"
    */
  public reorderPageGuidanceTitle: string = 'Drag pages to reorder';
  /**
    The title below each page object of the reorder pages screen. '%d' denotes the current page number.
    
    Default is "%d"
    */
  public reorderPageText: string = '%d';
  /**
    The 'confirm' button title, located in the top bar of the reorder pages screen.
    
    Default is "Done"
    */
  public reorderTopBarConfirmButtonTitle: string = 'Done';
  /**
    The 'cancel' button title, located in the top bar of the reorder pages screen.
    
    Default is "Cancel"
    */
  public reorderTopBarCancelButtonTitle: string = 'Cancel';
  /**
    The 'confirm' button title, located in the top bar of the cropping screen.
    
    Default is "Done"
    */
  public croppingTopBarConfirmButtonTitle: string = 'Done';
  /**
    The 'cancel' button title, located in the top bar of the cropping screen.
    
    Default is "Cancel"
    */
  public croppingTopBarCancelButtonTitle: string = 'Cancel';
  /**
    The 'detect document' button title, located in the bottom bar of the cropping screen.
    
    Default is "Detect"
    */
  public croppingDetectButtonTitle: string = 'Detect';
  /**
    The 'rotate page' button title, located in the bottom bar of the cropping screen.
    
    Default is "Rotate"
    */
  public croppingRotateButtonTitle: string = 'Rotate';
  /**
    The 'reset detection' button title, located in the bottom bar of the cropping screen.
    
    Default is "Reset"
    */
  public croppingResetButtonTitle: string = 'Reset';
  /**
    The title of the cropping screen, located in the top bar.
    
    Default is "Crop"
    */
  public croppingScreenTitle: string = 'Crop';
  /**
    The accessibility hint for the 'introduction' button, located in the top bar of the camera screen.
    
    Default is "Tap to open introduction screen"
    */
  public accessibilityDescriptionCameraTopBarIntroButton: string =
    'Tap to open introduction screen';
  /**
    The accessibility hint for the 'cancel' button, located in the top bar of the camera screen.
    
    Default is "Tap to close scanner screen"
    */
  public accessibilityDescriptionCameraTopBarCancelButton: string = 'Tap to close scanner screen';
  /**
    The accessibility hint for the 'import image' button, located in the bottom bar of the camera screen.
    
    Default is "Tap to import image"
    */
  public accessibilityDescriptionCameraImportButton: string = 'Tap to import image';
  /**
    The accessibility hint for the 'flashlight on' button, located in the bottom bar of the camera screen.
    
    Default is "Tap to disable torch"
    */
  public accessibilityDescriptionCameraTorchOnButton: string = 'Tap to disable torch';
  /**
    The accessibility hint for the 'flashlight off' button, located in the bottom bar of the camera screen.
    
    Default is "Tap to enable torch"
    */
  public accessibilityDescriptionCameraTorchOffButton: string = 'Tap to enable torch';
  /**
    The accessibility hint for the 'shutter' button, located in the bottom bar of the camera screen.
    
    Default is "Tap to take image"
    */
  public accessibilityDescriptionCameraShutterButton: string = 'Tap to take image';
  /**
    The accessibility hint for the 'auto snapping' mode button, located in the bottom bar of the camera screen.
    
    Default is "Tap to enable manual snapping mode"
    */
  public accessibilityDescriptionCameraAutoSnapButton: string =
    'Tap to enable manual snapping mode';
  /**
    The accessibility hint for the 'manual snapping' mode button, located in the bottom bar of the camera screen.
    
    Default is "Tap to enable auto snapping mode"
    */
  public accessibilityDescriptionCameraManualSnapButton: string =
    'Tap to enable auto snapping mode';
  /**
    The accessibility hint for the 'preview' button, located in the bottom bar of the camera screen.
    
    Default is "Tap to preview scanned pages"
    */
  public accessibilityDescriptionCameraPreviewButton: string = 'Tap to preview scanned pages';
  /**
    The accessibility hint for the 'done' button of the introduction screen.
    
    Default is "Tap to close introduction screen"
    */
  public accessibilityDescriptionCameraIntroDoneButton: string = 'Tap to close introduction screen';
  /**
    The accessibility hint for the 'retake' button of the acknowledgement screen.
    
    Default is "Tap to retake the image"
    */
  public accessibilityDescriptionAcknowledgementRetakeButton: string = 'Tap to retake the image';
  /**
    The accessibility hint for the 'accept' button of the acknowledgement screen.
    
    Default is "Tap to accept the image regardless of its quality"
    */
  public accessibilityDescriptionAcknowledgementAcceptButton: string =
    'Tap to accept the image regardless of its quality';
  /**
    The accessibility hint for the 'confirm' button, located in the top bar of the cropping screen.
    
    Default is "Tap to apply changes"
    */
  public accessibilityDescriptionCroppingTopBarConfirmButton: string = 'Tap to apply changes';
  /**
    The accessibility hint for the 'cancel' button, located in the top bar of the cropping screen.
    
    Default is "Tap to close screen without applying changes"
    */
  public accessibilityDescriptionCroppingTopBarCancelButton: string =
    'Tap to close screen without applying changes';
  /**
    The accessibility hint for the 'detect document' button, located in the bottom bar of the cropping screen.
    
    Default is "Tap to detect document on the image"
    */
  public accessibilityDescriptionCroppingDetectButton: string =
    'Tap to detect document on the image';
  /**
    The accessibility hint for the 'rotate page' button, located in the bottom bar of the cropping screen.
    
    Default is "Tap to rotate document"
    */
  public accessibilityDescriptionCroppingRotateButton: string = 'Tap to rotate document';
  /**
    The accessibility hint for the 'reset detection' button, located in the bottom bar of the cropping screen.
    
    Default is "Tap to reset document contour to whole image"
    */
  public accessibilityDescriptionCroppingResetButton: string =
    'Tap to reset document contour to whole image';
  /**
    The accessibility hint for the 'confirm' button, located in the top bar of the reorder pages screen.
    
    Default is "Tap to apply changes"
    */
  public accessibilityDescriptionReorderTopBarConfirmButton: string = 'Tap to apply changes';
  /**
    The accessibility hint for the 'cancel' button, located in the top bar of the reorder pages screen.
    
    Default is "Tap to close screen without applying changes"
    */
  public accessibilityDescriptionReorderTopBarCancelButton: string =
    'Tap to close screen without applying changes';
  /**
    The accessibility hint for the 'next page' button of the review screen.
    
    Default is "Tap to switch to the new page"
    */
  public accessibilityDescriptionReviewNextPageButton: string = 'Tap to switch to the new page';
  /**
    The accessibility hint for the 'previous page' button of the review screen.
    
    Default is "Tap to switch to the previous page"
    */
  public accessibilityDescriptionReviewPreviousPageButton: string =
    'Tap to switch to the previous page';
  /**
    The accessibility hint for the 'add' button, located in the bottom bar of the review screen.
    
    Default is "Tap to Add new page"
    */
  public accessibilityDescriptionReviewAddButton: string = 'Tap to Add new page';
  /**
    The accessibility hint for the 'retake' button, located in the bottom bar of the review screen.
    
    Default is "Tap to Retake current selected page"
    */
  public accessibilityDescriptionReviewRetakeButton: string = 'Tap to Retake current selected page';
  /**
    The accessibility hint for the 'crop' button, located in the bottom bar of the review screen.
    
    Default is "Tap to Crop selected page"
    */
  public accessibilityDescriptionReviewCropButton: string = 'Tap to Crop selected page';
  /**
    The accessibility hint for the 'rotate' button, located in the bottom bar of the review screen.
    
    Default is "Tap to Rotate selected page clockwise"
    */
  public accessibilityDescriptionReviewRotateButton: string =
    'Tap to Rotate selected page clockwise';
  /**
    The accessibility hint for the 'delete' button, located in the bottom bar of the review screen.
    
    Default is "Tap to Delete selected page"
    */
  public accessibilityDescriptionReviewDeleteButton: string = 'Tap to Delete selected page';
  /**
    The accessibility hint for the 'submit' button, located in the bottom bar of the review screen.
    
    Default is "Tap to Submit document"
    */
  public accessibilityDescriptionReviewSubmitButton: string = 'Tap to Submit document';
  /**
    The accessibility hint for the 'more' button, located in the top bar of the review screen.
    
    Default is "Tap to show additional options"
    */
  public accessibilityDescriptionReviewMoreButton: string = 'Tap to show additional options';
  /**
    The accessibility hint for the 'delete all' menu button, accessed via the 'more' button of the review screen.
    
    Default is "Tap to Delete all pages"
    */
  public accessibilityDescriptionReviewDeleteAllButton: string = 'Tap to Delete all pages';
  /**
    The accessibility hint for the 'reorder pages' menu button, accessed via the 'more' button of the review screen.
    
    Default is "Tap to Reorder pages"
    */
  public accessibilityDescriptionReviewReorderPagesButton: string = 'Tap to Reorder pages';
  /**
    The accessibility hint for the 'close' button of the zoom overlay for the review screen.
    
    Default is "Tap to Close zoom overlay"
    */
  public accessibilityDescriptionZoomOverlayCancelButton: string = 'Tap to Close zoom overlay';
  /**
    The accessibility hint for the 'zoom' button of the review screen.
    
    Default is "Tap to open zoom overlay for selected page"
    */
  public accessibilityDescriptionReviewZoomButton: string =
    'Tap to open zoom overlay for selected page';
  /**
    The accessibility hint for the 'back' button, located in the top bar of the review screen.
    
    Default is "Tap to Close review screen"
    */
  public accessibilityDescriptionReviewTopBarBackButton: string = 'Tap to Close review screen';
  /**
    The accessibility hint for the 'confirm' button of the delete page alert dialog for the review screen.
    
    Default is "Tap to Delete page"
    */
  public accessibilityDescriptionReviewDeletePageAlertConfirmButton: string = 'Tap to Delete page';
  /**
    The accessibility hint for the 'confirm' button of the delete all pages alert dialog for the review screen.
    
    Default is "Tap to Delete all pages"
    */
  public accessibilityDescriptionReviewDeleteAllPagesAlertDeleteButton: string =
    'Tap to Delete all pages';
  /**
    The accessibility hint for the 'delete and retake' button of the delete page alert dialog for the review screen.
    
    Default is "Tap to Delete page and Retake it"
    */
  public accessibilityDescriptionReviewDeletePageAlertDeleteRetakeButton: string =
    'Tap to Delete page and Retake it';
  /**
    The accessibility hint for the 'cancel' button of the delete page alert dialog for the review screen.
    
    Default is "Tap to Cancel operation"
    */
  public accessibilityDescriptionReviewDeletePageAlertCancelButton: string =
    'Tap to Cancel operation';
  /**
    The accessibility hint for the 'cancel' button of the delete all pages alert dialog for the review screen.
    
    Default is "Tap to Cancel operation"
    */
  public accessibilityDescriptionReviewDeleteAllPagesAlertCancelButton: string =
    'Tap to Cancel operation';
  /**
    The accessibility hint for the 'OK' button of the page limit reached alert dialog.
    
    Default is "Tap to close alert dialog"
    */
  public accessibilityDescriptionCameraLimitReachedOkButton: string = 'Tap to close alert dialog';
  /**
    The accessibility hint for the 'yes' button of the cancel camera screen alert dialog.
    
    Default is "Tap to cancel document"
    */
  public accessibilityDescriptionCameraCancelYesButton: string = 'Tap to cancel document';
  /**
    The accessibility hint for the 'no' button of the cancel camera screen alert dialog.
    
    Default is "Tap to close alert dialog"
    */
  public accessibilityDescriptionCameraCancelNoButton: string = 'Tap to close alert dialog';
  /**
    The title of the camera permission dialog.
    
    Default is "Camera permission denied!"
    */
  public cameraPermissionEnableCameraTitle: string = 'Camera permission denied!';
  /**
    The explanation text of the camera permission dialog.
    
    Default is "Please allow the usage of the camera to start the scanning process."
    */
  public cameraPermissionEnableCameraExplanation: string =
    'Please allow the usage of the camera to start the scanning process.';
  /**
    The 'enable' button title of the camera permission dialog.
    
    Default is "Grant permission"
    */
  public cameraPermissionEnableCameraButton: string = 'Grant permission';
  /**
    The 'close' button title of the camera permission dialog.
    
    Default is "Close"
    */
  public cameraPermissionCloseButton: string = 'Close';
  /**
    The accessibility hint for the 'enable' button of the camera permission dialog.
    
    Default is "Tap to grant camera permission"
    */
  public accessibilityDescriptionCameraPermissionEnableCameraButton: string =
    'Tap to grant camera permission';
  /**
    The accessibility hint for the 'close' button of the camera permission dialog.
    
    Default is "Close screen without granting permission"
    */
  public accessibilityDescriptionCameraPermissionCloseButton: string =
    'Close screen without granting permission';

  /** @param source {@displayType `DeepPartial<DocumentScannerTextLocalization>`} */
  public constructor(source: DeepPartial<DocumentScannerTextLocalization> = {}) {
    super();
    if (source.cameraTopBarCancelButtonTitle !== undefined) {
      this.cameraTopBarCancelButtonTitle = source.cameraTopBarCancelButtonTitle;
    }
    if (source.cameraTopBarTitle !== undefined) {
      this.cameraTopBarTitle = source.cameraTopBarTitle;
    }
    if (source.cameraTopGuidance !== undefined) {
      this.cameraTopGuidance = source.cameraTopGuidance;
    }
    if (source.cameraUserGuidanceStart !== undefined) {
      this.cameraUserGuidanceStart = source.cameraUserGuidanceStart;
    }
    if (source.cameraUserGuidanceNoDocumentFound !== undefined) {
      this.cameraUserGuidanceNoDocumentFound = source.cameraUserGuidanceNoDocumentFound;
    }
    if (source.cameraUserGuidanceBadAspectRatio !== undefined) {
      this.cameraUserGuidanceBadAspectRatio = source.cameraUserGuidanceBadAspectRatio;
    }
    if (source.cameraUserGuidanceOrientationMismatch !== undefined) {
      this.cameraUserGuidanceOrientationMismatch = source.cameraUserGuidanceOrientationMismatch;
    }
    if (source.cameraUserGuidanceBadAngles !== undefined) {
      this.cameraUserGuidanceBadAngles = source.cameraUserGuidanceBadAngles;
    }
    if (source.cameraUserGuidanceTooNoisy !== undefined) {
      this.cameraUserGuidanceTooNoisy = source.cameraUserGuidanceTooNoisy;
    }
    if (source.cameraUserGuidanceTextHintOffCenter !== undefined) {
      this.cameraUserGuidanceTextHintOffCenter = source.cameraUserGuidanceTextHintOffCenter;
    }
    if (source.cameraUserGuidanceTooSmall !== undefined) {
      this.cameraUserGuidanceTooSmall = source.cameraUserGuidanceTooSmall;
    }
    if (source.cameraUserGuidanceTooDark !== undefined) {
      this.cameraUserGuidanceTooDark = source.cameraUserGuidanceTooDark;
    }
    if (source.cameraUserGuidanceEnergySaveMode !== undefined) {
      this.cameraUserGuidanceEnergySaveMode = source.cameraUserGuidanceEnergySaveMode;
    }
    if (source.cameraUserGuidanceReadyToCapture !== undefined) {
      this.cameraUserGuidanceReadyToCapture = source.cameraUserGuidanceReadyToCapture;
    }
    if (source.cameraUserGuidanceReadyToCaptureManual !== undefined) {
      this.cameraUserGuidanceReadyToCaptureManual = source.cameraUserGuidanceReadyToCaptureManual;
    }
    if (source.cameraImportButtonTitle !== undefined) {
      this.cameraImportButtonTitle = source.cameraImportButtonTitle;
    }
    if (source.cameraTorchOnButtonTitle !== undefined) {
      this.cameraTorchOnButtonTitle = source.cameraTorchOnButtonTitle;
    }
    if (source.cameraTorchOffButtonTitle !== undefined) {
      this.cameraTorchOffButtonTitle = source.cameraTorchOffButtonTitle;
    }
    if (source.cameraAutoSnapButtonTitle !== undefined) {
      this.cameraAutoSnapButtonTitle = source.cameraAutoSnapButtonTitle;
    }
    if (source.cameraManualSnapButtonTitle !== undefined) {
      this.cameraManualSnapButtonTitle = source.cameraManualSnapButtonTitle;
    }
    if (source.cameraPreviewButtonTitle !== undefined) {
      this.cameraPreviewButtonTitle = source.cameraPreviewButtonTitle;
    }
    if (source.cameraIntroDoneButton !== undefined) {
      this.cameraIntroDoneButton = source.cameraIntroDoneButton;
    }
    if (source.cameraIntroTitle !== undefined) {
      this.cameraIntroTitle = source.cameraIntroTitle;
    }
    if (source.cameraIntroSubtitle !== undefined) {
      this.cameraIntroSubtitle = source.cameraIntroSubtitle;
    }
    if (source.cameraIntroItem1 !== undefined) {
      this.cameraIntroItem1 = source.cameraIntroItem1;
    }
    if (source.cameraIntroItem2 !== undefined) {
      this.cameraIntroItem2 = source.cameraIntroItem2;
    }
    if (source.cameraIntroItem3 !== undefined) {
      this.cameraIntroItem3 = source.cameraIntroItem3;
    }
    if (source.cameraIntroItem4 !== undefined) {
      this.cameraIntroItem4 = source.cameraIntroItem4;
    }
    if (source.cameraProgressOverlayTitle !== undefined) {
      this.cameraProgressOverlayTitle = source.cameraProgressOverlayTitle;
    }
    if (source.cameraCancelAlertTitle !== undefined) {
      this.cameraCancelAlertTitle = source.cameraCancelAlertTitle;
    }
    if (source.cameraCancelAlertSubtitle !== undefined) {
      this.cameraCancelAlertSubtitle = source.cameraCancelAlertSubtitle;
    }
    if (source.cameraCancelNoButtonTitle !== undefined) {
      this.cameraCancelNoButtonTitle = source.cameraCancelNoButtonTitle;
    }
    if (source.cameraCancelYesButtonTitle !== undefined) {
      this.cameraCancelYesButtonTitle = source.cameraCancelYesButtonTitle;
    }
    if (source.cameraLimitReachedAlertTitle !== undefined) {
      this.cameraLimitReachedAlertTitle = source.cameraLimitReachedAlertTitle;
    }
    if (source.cameraLimitReachedAlertSubtitle !== undefined) {
      this.cameraLimitReachedAlertSubtitle = source.cameraLimitReachedAlertSubtitle;
    }
    if (source.cameraLimitReachedOkButtonTitle !== undefined) {
      this.cameraLimitReachedOkButtonTitle = source.cameraLimitReachedOkButtonTitle;
    }
    if (source.acknowledgementScreenBadDocumentHint !== undefined) {
      this.acknowledgementScreenBadDocumentHint = source.acknowledgementScreenBadDocumentHint;
    }
    if (source.acknowledgementRetakeButtonTitle !== undefined) {
      this.acknowledgementRetakeButtonTitle = source.acknowledgementRetakeButtonTitle;
    }
    if (source.acknowledgementAcceptButtonTitle !== undefined) {
      this.acknowledgementAcceptButtonTitle = source.acknowledgementAcceptButtonTitle;
    }
    if (source.reviewScreenTitle !== undefined) {
      this.reviewScreenTitle = source.reviewScreenTitle;
    }
    if (source.reviewTopBarBackButtonTitle !== undefined) {
      this.reviewTopBarBackButtonTitle = source.reviewTopBarBackButtonTitle;
    }
    if (source.reviewScreenPageCount !== undefined) {
      this.reviewScreenPageCount = source.reviewScreenPageCount;
    }
    if (source.reviewScreenAddButtonTitle !== undefined) {
      this.reviewScreenAddButtonTitle = source.reviewScreenAddButtonTitle;
    }
    if (source.reviewScreenRetakeButtonTitle !== undefined) {
      this.reviewScreenRetakeButtonTitle = source.reviewScreenRetakeButtonTitle;
    }
    if (source.reviewScreenCropButtonTitle !== undefined) {
      this.reviewScreenCropButtonTitle = source.reviewScreenCropButtonTitle;
    }
    if (source.reviewScreenRotateButtonTitle !== undefined) {
      this.reviewScreenRotateButtonTitle = source.reviewScreenRotateButtonTitle;
    }
    if (source.reviewScreenDeleteButtonTitle !== undefined) {
      this.reviewScreenDeleteButtonTitle = source.reviewScreenDeleteButtonTitle;
    }
    if (source.reviewScreenSubmitButtonTitle !== undefined) {
      this.reviewScreenSubmitButtonTitle = source.reviewScreenSubmitButtonTitle;
    }
    if (source.reviewScreenDeleteAllButtonTitle !== undefined) {
      this.reviewScreenDeleteAllButtonTitle = source.reviewScreenDeleteAllButtonTitle;
    }
    if (source.reviewScreenReorderPagesButtonTitle !== undefined) {
      this.reviewScreenReorderPagesButtonTitle = source.reviewScreenReorderPagesButtonTitle;
    }
    if (source.zoomOverlayCancelButtonText !== undefined) {
      this.zoomOverlayCancelButtonText = source.zoomOverlayCancelButtonText;
    }
    if (source.reviewDeletePageAlertTitle !== undefined) {
      this.reviewDeletePageAlertTitle = source.reviewDeletePageAlertTitle;
    }
    if (source.reviewDeleteAllPagesAlertTitle !== undefined) {
      this.reviewDeleteAllPagesAlertTitle = source.reviewDeleteAllPagesAlertTitle;
    }
    if (source.reviewDeletePageAlertSubTitle !== undefined) {
      this.reviewDeletePageAlertSubTitle = source.reviewDeletePageAlertSubTitle;
    }
    if (source.reviewDeleteAllPagesAlertSubtitle !== undefined) {
      this.reviewDeleteAllPagesAlertSubtitle = source.reviewDeleteAllPagesAlertSubtitle;
    }
    if (source.reviewDeletePageAlertConfirmButtonTitle !== undefined) {
      this.reviewDeletePageAlertConfirmButtonTitle = source.reviewDeletePageAlertConfirmButtonTitle;
    }
    if (source.reviewDeleteAllPagesAlertDeleteButtonTitle !== undefined) {
      this.reviewDeleteAllPagesAlertDeleteButtonTitle =
        source.reviewDeleteAllPagesAlertDeleteButtonTitle;
    }
    if (source.reviewDeletePageAlertDeleteRetakeButtonTitle !== undefined) {
      this.reviewDeletePageAlertDeleteRetakeButtonTitle =
        source.reviewDeletePageAlertDeleteRetakeButtonTitle;
    }
    if (source.reviewDeletePageAlertCancelButtonTitle !== undefined) {
      this.reviewDeletePageAlertCancelButtonTitle = source.reviewDeletePageAlertCancelButtonTitle;
    }
    if (source.reviewDeleteAllPagesAlertCancelButtonTitle !== undefined) {
      this.reviewDeleteAllPagesAlertCancelButtonTitle =
        source.reviewDeleteAllPagesAlertCancelButtonTitle;
    }
    if (source.acknowledgementTitle !== undefined) {
      this.acknowledgementTitle = source.acknowledgementTitle;
    }
    if (source.reorderPageTitle !== undefined) {
      this.reorderPageTitle = source.reorderPageTitle;
    }
    if (source.reorderPageGuidanceTitle !== undefined) {
      this.reorderPageGuidanceTitle = source.reorderPageGuidanceTitle;
    }
    if (source.reorderPageText !== undefined) {
      this.reorderPageText = source.reorderPageText;
    }
    if (source.reorderTopBarConfirmButtonTitle !== undefined) {
      this.reorderTopBarConfirmButtonTitle = source.reorderTopBarConfirmButtonTitle;
    }
    if (source.reorderTopBarCancelButtonTitle !== undefined) {
      this.reorderTopBarCancelButtonTitle = source.reorderTopBarCancelButtonTitle;
    }
    if (source.croppingTopBarConfirmButtonTitle !== undefined) {
      this.croppingTopBarConfirmButtonTitle = source.croppingTopBarConfirmButtonTitle;
    }
    if (source.croppingTopBarCancelButtonTitle !== undefined) {
      this.croppingTopBarCancelButtonTitle = source.croppingTopBarCancelButtonTitle;
    }
    if (source.croppingDetectButtonTitle !== undefined) {
      this.croppingDetectButtonTitle = source.croppingDetectButtonTitle;
    }
    if (source.croppingRotateButtonTitle !== undefined) {
      this.croppingRotateButtonTitle = source.croppingRotateButtonTitle;
    }
    if (source.croppingResetButtonTitle !== undefined) {
      this.croppingResetButtonTitle = source.croppingResetButtonTitle;
    }
    if (source.croppingScreenTitle !== undefined) {
      this.croppingScreenTitle = source.croppingScreenTitle;
    }
    if (source.accessibilityDescriptionCameraTopBarIntroButton !== undefined) {
      this.accessibilityDescriptionCameraTopBarIntroButton =
        source.accessibilityDescriptionCameraTopBarIntroButton;
    }
    if (source.accessibilityDescriptionCameraTopBarCancelButton !== undefined) {
      this.accessibilityDescriptionCameraTopBarCancelButton =
        source.accessibilityDescriptionCameraTopBarCancelButton;
    }
    if (source.accessibilityDescriptionCameraImportButton !== undefined) {
      this.accessibilityDescriptionCameraImportButton =
        source.accessibilityDescriptionCameraImportButton;
    }
    if (source.accessibilityDescriptionCameraTorchOnButton !== undefined) {
      this.accessibilityDescriptionCameraTorchOnButton =
        source.accessibilityDescriptionCameraTorchOnButton;
    }
    if (source.accessibilityDescriptionCameraTorchOffButton !== undefined) {
      this.accessibilityDescriptionCameraTorchOffButton =
        source.accessibilityDescriptionCameraTorchOffButton;
    }
    if (source.accessibilityDescriptionCameraShutterButton !== undefined) {
      this.accessibilityDescriptionCameraShutterButton =
        source.accessibilityDescriptionCameraShutterButton;
    }
    if (source.accessibilityDescriptionCameraAutoSnapButton !== undefined) {
      this.accessibilityDescriptionCameraAutoSnapButton =
        source.accessibilityDescriptionCameraAutoSnapButton;
    }
    if (source.accessibilityDescriptionCameraManualSnapButton !== undefined) {
      this.accessibilityDescriptionCameraManualSnapButton =
        source.accessibilityDescriptionCameraManualSnapButton;
    }
    if (source.accessibilityDescriptionCameraPreviewButton !== undefined) {
      this.accessibilityDescriptionCameraPreviewButton =
        source.accessibilityDescriptionCameraPreviewButton;
    }
    if (source.accessibilityDescriptionCameraIntroDoneButton !== undefined) {
      this.accessibilityDescriptionCameraIntroDoneButton =
        source.accessibilityDescriptionCameraIntroDoneButton;
    }
    if (source.accessibilityDescriptionAcknowledgementRetakeButton !== undefined) {
      this.accessibilityDescriptionAcknowledgementRetakeButton =
        source.accessibilityDescriptionAcknowledgementRetakeButton;
    }
    if (source.accessibilityDescriptionAcknowledgementAcceptButton !== undefined) {
      this.accessibilityDescriptionAcknowledgementAcceptButton =
        source.accessibilityDescriptionAcknowledgementAcceptButton;
    }
    if (source.accessibilityDescriptionCroppingTopBarConfirmButton !== undefined) {
      this.accessibilityDescriptionCroppingTopBarConfirmButton =
        source.accessibilityDescriptionCroppingTopBarConfirmButton;
    }
    if (source.accessibilityDescriptionCroppingTopBarCancelButton !== undefined) {
      this.accessibilityDescriptionCroppingTopBarCancelButton =
        source.accessibilityDescriptionCroppingTopBarCancelButton;
    }
    if (source.accessibilityDescriptionCroppingDetectButton !== undefined) {
      this.accessibilityDescriptionCroppingDetectButton =
        source.accessibilityDescriptionCroppingDetectButton;
    }
    if (source.accessibilityDescriptionCroppingRotateButton !== undefined) {
      this.accessibilityDescriptionCroppingRotateButton =
        source.accessibilityDescriptionCroppingRotateButton;
    }
    if (source.accessibilityDescriptionCroppingResetButton !== undefined) {
      this.accessibilityDescriptionCroppingResetButton =
        source.accessibilityDescriptionCroppingResetButton;
    }
    if (source.accessibilityDescriptionReorderTopBarConfirmButton !== undefined) {
      this.accessibilityDescriptionReorderTopBarConfirmButton =
        source.accessibilityDescriptionReorderTopBarConfirmButton;
    }
    if (source.accessibilityDescriptionReorderTopBarCancelButton !== undefined) {
      this.accessibilityDescriptionReorderTopBarCancelButton =
        source.accessibilityDescriptionReorderTopBarCancelButton;
    }
    if (source.accessibilityDescriptionReviewNextPageButton !== undefined) {
      this.accessibilityDescriptionReviewNextPageButton =
        source.accessibilityDescriptionReviewNextPageButton;
    }
    if (source.accessibilityDescriptionReviewPreviousPageButton !== undefined) {
      this.accessibilityDescriptionReviewPreviousPageButton =
        source.accessibilityDescriptionReviewPreviousPageButton;
    }
    if (source.accessibilityDescriptionReviewAddButton !== undefined) {
      this.accessibilityDescriptionReviewAddButton = source.accessibilityDescriptionReviewAddButton;
    }
    if (source.accessibilityDescriptionReviewRetakeButton !== undefined) {
      this.accessibilityDescriptionReviewRetakeButton =
        source.accessibilityDescriptionReviewRetakeButton;
    }
    if (source.accessibilityDescriptionReviewCropButton !== undefined) {
      this.accessibilityDescriptionReviewCropButton =
        source.accessibilityDescriptionReviewCropButton;
    }
    if (source.accessibilityDescriptionReviewRotateButton !== undefined) {
      this.accessibilityDescriptionReviewRotateButton =
        source.accessibilityDescriptionReviewRotateButton;
    }
    if (source.accessibilityDescriptionReviewDeleteButton !== undefined) {
      this.accessibilityDescriptionReviewDeleteButton =
        source.accessibilityDescriptionReviewDeleteButton;
    }
    if (source.accessibilityDescriptionReviewSubmitButton !== undefined) {
      this.accessibilityDescriptionReviewSubmitButton =
        source.accessibilityDescriptionReviewSubmitButton;
    }
    if (source.accessibilityDescriptionReviewMoreButton !== undefined) {
      this.accessibilityDescriptionReviewMoreButton =
        source.accessibilityDescriptionReviewMoreButton;
    }
    if (source.accessibilityDescriptionReviewDeleteAllButton !== undefined) {
      this.accessibilityDescriptionReviewDeleteAllButton =
        source.accessibilityDescriptionReviewDeleteAllButton;
    }
    if (source.accessibilityDescriptionReviewReorderPagesButton !== undefined) {
      this.accessibilityDescriptionReviewReorderPagesButton =
        source.accessibilityDescriptionReviewReorderPagesButton;
    }
    if (source.accessibilityDescriptionZoomOverlayCancelButton !== undefined) {
      this.accessibilityDescriptionZoomOverlayCancelButton =
        source.accessibilityDescriptionZoomOverlayCancelButton;
    }
    if (source.accessibilityDescriptionReviewZoomButton !== undefined) {
      this.accessibilityDescriptionReviewZoomButton =
        source.accessibilityDescriptionReviewZoomButton;
    }
    if (source.accessibilityDescriptionReviewTopBarBackButton !== undefined) {
      this.accessibilityDescriptionReviewTopBarBackButton =
        source.accessibilityDescriptionReviewTopBarBackButton;
    }
    if (source.accessibilityDescriptionReviewDeletePageAlertConfirmButton !== undefined) {
      this.accessibilityDescriptionReviewDeletePageAlertConfirmButton =
        source.accessibilityDescriptionReviewDeletePageAlertConfirmButton;
    }
    if (source.accessibilityDescriptionReviewDeleteAllPagesAlertDeleteButton !== undefined) {
      this.accessibilityDescriptionReviewDeleteAllPagesAlertDeleteButton =
        source.accessibilityDescriptionReviewDeleteAllPagesAlertDeleteButton;
    }
    if (source.accessibilityDescriptionReviewDeletePageAlertDeleteRetakeButton !== undefined) {
      this.accessibilityDescriptionReviewDeletePageAlertDeleteRetakeButton =
        source.accessibilityDescriptionReviewDeletePageAlertDeleteRetakeButton;
    }
    if (source.accessibilityDescriptionReviewDeletePageAlertCancelButton !== undefined) {
      this.accessibilityDescriptionReviewDeletePageAlertCancelButton =
        source.accessibilityDescriptionReviewDeletePageAlertCancelButton;
    }
    if (source.accessibilityDescriptionReviewDeleteAllPagesAlertCancelButton !== undefined) {
      this.accessibilityDescriptionReviewDeleteAllPagesAlertCancelButton =
        source.accessibilityDescriptionReviewDeleteAllPagesAlertCancelButton;
    }
    if (source.accessibilityDescriptionCameraLimitReachedOkButton !== undefined) {
      this.accessibilityDescriptionCameraLimitReachedOkButton =
        source.accessibilityDescriptionCameraLimitReachedOkButton;
    }
    if (source.accessibilityDescriptionCameraCancelYesButton !== undefined) {
      this.accessibilityDescriptionCameraCancelYesButton =
        source.accessibilityDescriptionCameraCancelYesButton;
    }
    if (source.accessibilityDescriptionCameraCancelNoButton !== undefined) {
      this.accessibilityDescriptionCameraCancelNoButton =
        source.accessibilityDescriptionCameraCancelNoButton;
    }
    if (source.cameraPermissionEnableCameraTitle !== undefined) {
      this.cameraPermissionEnableCameraTitle = source.cameraPermissionEnableCameraTitle;
    }
    if (source.cameraPermissionEnableCameraExplanation !== undefined) {
      this.cameraPermissionEnableCameraExplanation = source.cameraPermissionEnableCameraExplanation;
    }
    if (source.cameraPermissionEnableCameraButton !== undefined) {
      this.cameraPermissionEnableCameraButton = source.cameraPermissionEnableCameraButton;
    }
    if (source.cameraPermissionCloseButton !== undefined) {
      this.cameraPermissionCloseButton = source.cameraPermissionCloseButton;
    }
    if (source.accessibilityDescriptionCameraPermissionEnableCameraButton !== undefined) {
      this.accessibilityDescriptionCameraPermissionEnableCameraButton =
        source.accessibilityDescriptionCameraPermissionEnableCameraButton;
    }
    if (source.accessibilityDescriptionCameraPermissionCloseButton !== undefined) {
      this.accessibilityDescriptionCameraPermissionCloseButton =
        source.accessibilityDescriptionCameraPermissionCloseButton;
    }
  }
}
