// react-native-scanbot-sdk/src/ui_v2/document/ReorderPagesScreenConfiguration.ts
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/document/ReorderPagesScreenConfiguration.yaml

import {
  BackgroundStyle,
  ButtonConfiguration,
  ForegroundStyle,
  StyledText,
} from '../../ui_v2/common/Common';
import { UserGuidanceConfiguration } from '../../ui_v2/common/UserGuidanceConfiguration';
import { DeepPartial, PartiallyConstructible } from '../../utils/utils';

/**
Configuration of the screen for reordering the scanned pages.
*/
export class ReorderPagesScreenConfiguration extends PartiallyConstructible {
  /**
    Configuration of the 'cancel' button, located in the top bar.
    */
  public topBarBackButton: ButtonConfiguration = new ButtonConfiguration({
    visible: true,
    text: '?reorderTopBarCancelButtonTitle',
    accessibilityDescription: '?accessibilityDescriptionReorderTopBarCancelButton',
    background: new BackgroundStyle({
      strokeColor: '#00000000',
      fillColor: '#00000000',
      strokeWidth: 0.0,
    }),
    foreground: new ForegroundStyle({
      iconVisible: true,
      color: '?sbColorOnPrimary',
      useShadow: false,
    }),
  });
  /**
    Configuration of the title, located in the top bar.
    */
  public topBarTitle: StyledText = new StyledText({
    text: '?reorderPageTitle',
    color: '?sbColorOnPrimary',
  });
  /**
    Configuration of the 'confirm' button, located in the top bar.
    */
  public topBarConfirmButton: ButtonConfiguration = new ButtonConfiguration({
    visible: true,
    text: '?reorderTopBarConfirmButtonTitle',
    accessibilityDescription: '?accessibilityDescriptionReorderTopBarConfirmButton',
    background: new BackgroundStyle({
      strokeColor: '#00000000',
      fillColor: '#00000000',
      strokeWidth: 0.0,
    }),
    foreground: new ForegroundStyle({
      iconVisible: true,
      color: '?sbColorOnPrimary',
      useShadow: false,
    }),
  });
  /**
    The background color of the reorder pages screen.
    
    Default is "?sbColorOutline"
    */
  public backgroundColor: string = '?sbColorOutline';
  /**
    Configuration of the static user guidance, located just below the top bar.
    */
  public guidance: UserGuidanceConfiguration = new UserGuidanceConfiguration({
    title: new StyledText({ text: '?reorderPageGuidanceTitle', color: '?sbColorOnPrimary' }),
    background: new BackgroundStyle({
      strokeColor: '#00000000',
      fillColor: '?sbColorSurfaceLow',
      strokeWidth: 0.0,
    }),
  });
  /**
    Configuration of the title below each page object.
    */
  public pageTextStyle: StyledText = new StyledText({
    text: '?reorderPageText',
    color: '?sbColorOnSurface',
  });

  /** @param source {@displayType `DeepPartial<ReorderPagesScreenConfiguration>`} */
  public constructor(source: DeepPartial<ReorderPagesScreenConfiguration> = {}) {
    super();
    if (source.topBarBackButton !== undefined) {
      this.topBarBackButton = new ButtonConfiguration(source.topBarBackButton);
    }
    if (source.topBarTitle !== undefined) {
      this.topBarTitle = new StyledText(source.topBarTitle);
    }
    if (source.topBarConfirmButton !== undefined) {
      this.topBarConfirmButton = new ButtonConfiguration(source.topBarConfirmButton);
    }
    if (source.backgroundColor !== undefined) {
      this.backgroundColor = source.backgroundColor;
    }
    if (source.guidance !== undefined) {
      this.guidance = new UserGuidanceConfiguration(source.guidance);
    }
    if (source.pageTextStyle !== undefined) {
      this.pageTextStyle = new StyledText(source.pageTextStyle);
    }
  }
}
