// react-native-scanbot-sdk/lib/module/ui_v2/dde/DocumentDataExtractorIntroScreenConfiguration.js
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/documentdataextractor/DocumentDataExtractorIntroScreenConfiguration.yaml

import { BackgroundStyle, ButtonConfiguration, ForegroundStyle, StyledText } from '../../ui_v2/common/Common';
import { PartiallyConstructible } from '../../utils/utils';

/**
The image for the introduction screen of a document data extractor screen.
*/
/** @internal */
export let DocumentDataExtractorIntroImage;
(function (_DocumentDataExtractorIntroImage) {
  function From(source) {
    const _type = source._type;
    switch (_type) {
      case 'DocumentDataIntroDefaultImage':
        return new DocumentDataIntroDefaultImage(source);
      case 'DocumentDataIntroNoImage':
        return new DocumentDataIntroNoImage(source);
      case 'DocumentDataIntroCustomImage':
        return new DocumentDataIntroCustomImage(source);
      default:
        throw `Unknown child class name: ${_type}`;
    }
  }
  _DocumentDataExtractorIntroImage.From = From;
})(DocumentDataExtractorIntroImage || (DocumentDataExtractorIntroImage = {}));
/**
The image for the introduction screen featuring a document.
*/
export class DocumentDataIntroDefaultImage extends PartiallyConstructible {
  _type = 'DocumentDataIntroDefaultImage';

  /** @param source {@displayType `DeepPartial<DocumentDataIntroDefaultImage>`} */
  constructor(source = {}) {
    super();
  }
}

/**
No image for the introduction screen.
*/
export class DocumentDataIntroNoImage extends PartiallyConstructible {
  _type = 'DocumentDataIntroNoImage';

  /** @param source {@displayType `DeepPartial<DocumentDataIntroNoImage>`} */
  constructor(source = {}) {
    super();
  }
}

/**
A custom image for the introduction screen.
*/
export class DocumentDataIntroCustomImage extends PartiallyConstructible {
  _type = 'DocumentDataIntroCustomImage';
  /**
    The web or file URI to the image.
    */

  /** @param source {@displayType `DeepPartial<DocumentDataIntroCustomImage>`} */
  constructor(source = {}) {
    super();
    if (source.uri !== undefined) {
      this.uri = source.uri;
    } else {
      throw new Error('uri must be present in constructor argument');
    }
  }
}

/**
Configuration of the introduction screen for the document data extractor.
*/
export class DocumentDataExtractorIntroScreenConfiguration extends PartiallyConstructible {
  /**
    The image for the introduction screen.
    */
  image = new DocumentDataIntroDefaultImage({});
  /**
    The background color of the introduction screen.
    
    Default is "?sbColorSurface"
    */
  backgroundColor = '?sbColorSurface';
  /**
    The divider color of the introduction screen.
    
    Default is "?sbColorOutline"
    */
  dividerColor = '?sbColorOutline';
  /**
    The handle color of the introduction screen.
    
    Default is "?sbColorOutline"
    */
  handlerColor = '?sbColorOutline';
  /**
    Determines whether the introduction screen should automatically be shown or not when the scanning session starts.
    
    Default is false
    */
  showAutomatically = false;
  /**
    The title of the introduction screen, located in the top bar.
    */
  title = new StyledText({
    text: '?introScreenTitle',
    color: '?sbColorOnSurface'
  });
  /**
    The text explanation of the introduction screen.
    */
  explanation = new StyledText({
    text: '?introScreenText',
    color: '?sbColorOnSurface'
  });
  /**
    Configuration of the 'Done' / 'Start scanning' button.
    */
  doneButton = new ButtonConfiguration({
    text: '?introScreenDoneButton',
    accessibilityDescription: '?accessibilityDescriptionIntroScreenDoneButton',
    background: new BackgroundStyle({
      strokeColor: '#00000000',
      fillColor: '?sbColorPrimary',
      strokeWidth: 0.0
    }),
    foreground: new ForegroundStyle({
      color: '?sbColorOnPrimary'
    })
  });

  /** @param source {@displayType `DeepPartial<DocumentDataExtractorIntroScreenConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.image !== undefined) {
      this.image = DocumentDataExtractorIntroImage.From(source.image);
    }
    if (source.backgroundColor !== undefined) {
      this.backgroundColor = source.backgroundColor;
    }
    if (source.dividerColor !== undefined) {
      this.dividerColor = source.dividerColor;
    }
    if (source.handlerColor !== undefined) {
      this.handlerColor = source.handlerColor;
    }
    if (source.showAutomatically !== undefined) {
      this.showAutomatically = source.showAutomatically;
    }
    if (source.title !== undefined) {
      this.title = new StyledText(source.title);
    }
    if (source.explanation !== undefined) {
      this.explanation = new StyledText(source.explanation);
    }
    if (source.doneButton !== undefined) {
      this.doneButton = new ButtonConfiguration(source.doneButton);
    }
  }
}
