// react-native-scanbot-sdk/lib/commonjs/analytics/Analytics.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AnalyticsFeature = exports.AnalyticsEventName = exports.AnalyticsEventFactory = exports.AnalyticsEvent = exports.AnalyticsCategory = void 0;
var _JsonSerializationTypes = require("../utils/json/JsonSerializationTypes");
var _utils = require("../utils/utils");
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from native/schemas/analytics/Analytics.yaml

/**
The feature that is being tracked.
*/
class AnalyticsFeature extends _utils.PartiallyConstructible {
  /** @param source {@displayType `DeepPartial<AnalyticsFeature>`} */
  constructor(source = {}) {
    super();
  }
  serialize(config = new _JsonSerializationTypes.ToJsonConfiguration()) {
    return {};
  }
}
exports.AnalyticsFeature = AnalyticsFeature;
(function (_AnalyticsFeature) {
  const documentScanner = _AnalyticsFeature.documentScanner = 'rtuui_document_scanner';
})(AnalyticsFeature || (exports.AnalyticsFeature = AnalyticsFeature = {}));
/**
The category that is being tracked.
*/
class AnalyticsCategory extends _utils.PartiallyConstructible {
  /** @param source {@displayType `DeepPartial<AnalyticsCategory>`} */
  constructor(source = {}) {
    super();
  }
  serialize(config = new _JsonSerializationTypes.ToJsonConfiguration()) {
    return {};
  }
}
exports.AnalyticsCategory = AnalyticsCategory;
(function (_AnalyticsCategory) {
  const acknowledge = _AnalyticsCategory.acknowledge = 'acknowledge';
})(AnalyticsCategory || (exports.AnalyticsCategory = AnalyticsCategory = {}));
/**
The event that is being tracked.
*/
class AnalyticsEvent extends _utils.PartiallyConstructible {
  /**
    The feature associated with the event.
    */

  /**
    The category of the event.
    */

  /**
    The name of the event.
    */

  /**
    The value associated with the event.
    */
  values = [];

  /** @param source {@displayType `DeepPartial<AnalyticsEvent>`} */
  constructor(source = {}) {
    super();
    if (source.feature !== undefined) {
      this.feature = source.feature;
    } else {
      throw new Error('feature must be present in constructor argument');
    }
    if (source.category !== undefined) {
      this.category = source.category;
    } else {
      throw new Error('category must be present in constructor argument');
    }
    if (source.name !== undefined) {
      this.name = source.name;
    } else {
      throw new Error('name must be present in constructor argument');
    }
    if (source.values !== undefined) {
      this.values = source.values.map(it => {
        return it;
      });
    }
  }
  serialize(config = new _JsonSerializationTypes.ToJsonConfiguration()) {
    return {
      feature: this.feature,
      category: this.category,
      name: this.name,
      values: this.values.map(it => {
        return it;
      })
    };
  }
}

/**
The name of the event.
*/
exports.AnalyticsEvent = AnalyticsEvent;
class AnalyticsEventName extends _utils.PartiallyConstructible {
  /** @param source {@displayType `DeepPartial<AnalyticsEventName>`} */
  constructor(source = {}) {
    super();
  }
  serialize(config = new _JsonSerializationTypes.ToJsonConfiguration()) {
    return {};
  }
}
exports.AnalyticsEventName = AnalyticsEventName;
(function (_AnalyticsEventName) {
  const importedImageQualityResult = _AnalyticsEventName.importedImageQualityResult = 'imported_image_quality_result';
  const scannedImageQualityResult = _AnalyticsEventName.scannedImageQualityResult = 'scanned_image_quality_result';
  const proceedImage = _AnalyticsEventName.proceedImage = 'proceed_with_image';
  const retakeImage = _AnalyticsEventName.retakeImage = 'retake_image';
})(AnalyticsEventName || (exports.AnalyticsEventName = AnalyticsEventName = {}));
/**
A predefined analytics event.
*/
class AnalyticsEventFactory extends _utils.PartiallyConstructible {
  /** @param source {@displayType `DeepPartial<AnalyticsEventFactory>`} */
  constructor(source = {}) {
    super();
  }
  serialize(config = new _JsonSerializationTypes.ToJsonConfiguration()) {
    return {};
  }
}
exports.AnalyticsEventFactory = AnalyticsEventFactory;
(function (_AnalyticsEventFactory) {
  const scannedImageQualityResult = _AnalyticsEventFactory.scannedImageQualityResult = new AnalyticsEvent({
    feature: 'rtuui_document_scanner',
    category: 'acknowledge',
    name: 'scanned_image_quality_result'
  });
  const importedImageQualityResult = _AnalyticsEventFactory.importedImageQualityResult = new AnalyticsEvent({
    feature: 'rtuui_document_scanner',
    category: 'acknowledge',
    name: 'imported_image_quality_result'
  });
  const proceedWithImage = _AnalyticsEventFactory.proceedWithImage = new AnalyticsEvent({
    feature: 'rtuui_document_scanner',
    category: 'acknowledge',
    name: 'proceed_with_image'
  });
  const retakeImage = _AnalyticsEventFactory.retakeImage = new AnalyticsEvent({
    feature: 'rtuui_document_scanner',
    category: 'acknowledge',
    name: 'retake_image'
  });
})(AnalyticsEventFactory || (exports.AnalyticsEventFactory = AnalyticsEventFactory = {}));
