// react-native-scanbot-sdk/lib/commonjs/ui_v2/dde/DocumentDataExtractorIntroScreenConfiguration.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DocumentDataIntroNoImage = exports.DocumentDataIntroDefaultImage = exports.DocumentDataIntroCustomImage = exports.DocumentDataExtractorIntroScreenConfiguration = exports.DocumentDataExtractorIntroImage = void 0;
var _Common = require("../../ui_v2/common/Common");
var _utils = require("../../utils/utils");
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/documentdataextractor/DocumentDataExtractorIntroScreenConfiguration.yaml
/**
The image for the introduction screen of a document data extractor screen.
*/
/** @internal */
let DocumentDataExtractorIntroImage = exports.DocumentDataExtractorIntroImage = void 0;
(function (_DocumentDataExtractorIntroImage) {
  function From(source) {
    const _type = source._type;
    switch (_type) {
      case 'DocumentDataIntroDefaultImage':
        return new DocumentDataIntroDefaultImage(source);
      case 'DocumentDataIntroNoImage':
        return new DocumentDataIntroNoImage(source);
      case 'DocumentDataIntroCustomImage':
        return new DocumentDataIntroCustomImage(source);
      default:
        throw `Unknown child class name: ${_type}`;
    }
  }
  _DocumentDataExtractorIntroImage.From = From;
})(DocumentDataExtractorIntroImage || (exports.DocumentDataExtractorIntroImage = DocumentDataExtractorIntroImage = {}));
/**
The image for the introduction screen featuring a document.
*/
class DocumentDataIntroDefaultImage extends _utils.PartiallyConstructible {
  _type = 'DocumentDataIntroDefaultImage';

  /** @param source {@displayType `DeepPartial<DocumentDataIntroDefaultImage>`} */
  constructor(source = {}) {
    super();
  }
}

/**
No image for the introduction screen.
*/
exports.DocumentDataIntroDefaultImage = DocumentDataIntroDefaultImage;
class DocumentDataIntroNoImage extends _utils.PartiallyConstructible {
  _type = 'DocumentDataIntroNoImage';

  /** @param source {@displayType `DeepPartial<DocumentDataIntroNoImage>`} */
  constructor(source = {}) {
    super();
  }
}

/**
A custom image for the introduction screen.
*/
exports.DocumentDataIntroNoImage = DocumentDataIntroNoImage;
class DocumentDataIntroCustomImage extends _utils.PartiallyConstructible {
  _type = 'DocumentDataIntroCustomImage';
  /**
    The web or file URI to the image.
    */

  /** @param source {@displayType `DeepPartial<DocumentDataIntroCustomImage>`} */
  constructor(source = {}) {
    super();
    if (source.uri !== undefined) {
      this.uri = source.uri;
    } else {
      throw new Error('uri must be present in constructor argument');
    }
  }
}

/**
Configuration of the introduction screen for the document data extractor.
*/
exports.DocumentDataIntroCustomImage = DocumentDataIntroCustomImage;
class DocumentDataExtractorIntroScreenConfiguration extends _utils.PartiallyConstructible {
  /**
    The image for the introduction screen.
    */
  image = new DocumentDataIntroDefaultImage({});
  /**
    The background color of the introduction screen.
    
    Default is "?sbColorSurface"
    */
  backgroundColor = '?sbColorSurface';
  /**
    The divider color of the introduction screen.
    
    Default is "?sbColorOutline"
    */
  dividerColor = '?sbColorOutline';
  /**
    The handle color of the introduction screen.
    
    Default is "?sbColorOutline"
    */
  handlerColor = '?sbColorOutline';
  /**
    Determines whether the introduction screen should automatically be shown or not when the scanning session starts.
    
    Default is false
    */
  showAutomatically = false;
  /**
    The title of the introduction screen, located in the top bar.
    */
  title = new _Common.StyledText({
    text: '?introScreenTitle',
    color: '?sbColorOnSurface'
  });
  /**
    The text explanation of the introduction screen.
    */
  explanation = new _Common.StyledText({
    text: '?introScreenText',
    color: '?sbColorOnSurface'
  });
  /**
    Configuration of the 'Done' / 'Start scanning' button.
    */
  doneButton = new _Common.ButtonConfiguration({
    text: '?introScreenDoneButton',
    accessibilityDescription: '?accessibilityDescriptionIntroScreenDoneButton',
    background: new _Common.BackgroundStyle({
      strokeColor: '#00000000',
      fillColor: '?sbColorPrimary',
      strokeWidth: 0.0
    }),
    foreground: new _Common.ForegroundStyle({
      color: '?sbColorOnPrimary'
    })
  });

  /** @param source {@displayType `DeepPartial<DocumentDataExtractorIntroScreenConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.image !== undefined) {
      this.image = DocumentDataExtractorIntroImage.From(source.image);
    }
    if (source.backgroundColor !== undefined) {
      this.backgroundColor = source.backgroundColor;
    }
    if (source.dividerColor !== undefined) {
      this.dividerColor = source.dividerColor;
    }
    if (source.handlerColor !== undefined) {
      this.handlerColor = source.handlerColor;
    }
    if (source.showAutomatically !== undefined) {
      this.showAutomatically = source.showAutomatically;
    }
    if (source.title !== undefined) {
      this.title = new _Common.StyledText(source.title);
    }
    if (source.explanation !== undefined) {
      this.explanation = new _Common.StyledText(source.explanation);
    }
    if (source.doneButton !== undefined) {
      this.doneButton = new _Common.ButtonConfiguration(source.doneButton);
    }
  }
}
exports.DocumentDataExtractorIntroScreenConfiguration = DocumentDataExtractorIntroScreenConfiguration;
