// react-native-scanbot-barcode-scanner-sdk/lib/module/types/core/licensing/LicensingTypes.js
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from core/schemas/LicensingTypes.yaml

import { ToJsonConfiguration } from '../common/JsonSerializationTypes';

/**
Status of the license plate scanner.

- `OKAY`:
   License is valid and accepted.
- `TRIAL`:
   No license set yet. The SDK is in trial mode.
- `OKAY_EXPIRING_SOON`:
   License is valid but will expire soon. Make sure to update your license before it expires to ensure uninterrupted function.
- `FAILURE_NOT_SET`:
   No license set. On platforms that support trial mode, it means that the SDK's trial mode is over.
- `FAILURE_CORRUPTED`:
   No license active. The set license was unreadable or has an invalid format.
- `FAILURE_WRONG_OS`:
   No license active. The set license does not cover the current operating system.
- `FAILURE_APP_ID_MISMATCH`:
   No license active. The set license does not cover the current app's bundle identifier.
- `FAILURE_EXPIRED`:
   No license active. The set license is valid but it has expired.
- `FAILURE_SERVER`:
   No license active. The online license check failed.
- `FAILURE_VERSION`:
   No license active. The set license does not cover the current SDK version.
- `FAILURE_INACTIVE`:
   No license active. The set license is not yet active.
*/

/** @hidden */
export const LicenseStatusValues = ['OKAY', 'TRIAL', 'OKAY_EXPIRING_SOON', 'FAILURE_NOT_SET', 'FAILURE_CORRUPTED', 'FAILURE_WRONG_OS', 'FAILURE_APP_ID_MISMATCH', 'FAILURE_EXPIRED', 'FAILURE_SERVER', 'FAILURE_VERSION', 'FAILURE_INACTIVE'];

/**
Description of error that occurred during device deregistration.
*/
export class DeviceDeregistrationError {
  /**
    Sequential number of the device registration. Starts with 0. Each time a call to initializeLicense with a new license is made, the device ID.
    */

  /**
    Error message describing the error.
    */

  /** @param source {@displayType `DeepPartial<DeviceDeregistrationError>`} */
  constructor(source = {}) {
    if (source.deviceId !== undefined) {
      this.deviceId = source.deviceId;
    } else {
      throw new Error('deviceId must be present in constructor argument');
    }
    if (source.errorMessage !== undefined) {
      this.errorMessage = source.errorMessage;
    } else {
      throw new Error('errorMessage must be present in constructor argument');
    }
  }
  serialize(config = new ToJsonConfiguration()) {
    return {
      deviceId: this.deviceId,
      errorMessage: this.errorMessage
    };
  }
}

/**
Information about the SDK license.
*/
export class LicenseInfo {
  /**
    License status.
    */

  /**
    True if the license status corresponds to a valid license.
    */

  /**
    Message describing the license status.
    */

  /**
    Expiration timestamp of the license or trial period in seconds since the UNIX epoch.
    */

  /**
    Expiration date of the license or trial period in ISO 8601 date-time format.
    */

  /**
    Additional error message in case of a license error.
    */

  /**
    True if an online license check is in progress.
    */

  /**
    The currently active license. May be different from the license used to initialize
    the SDK if the online license check has completed and the server has sent
    an updated license.
    */

  /**
    Number of devices the license is valid for, if the license is floating.
    */
  devices = null;
  /**
    Time in milliseconds until the next online license check will be performed. -1 if no online license check is scheduled.
    */

  /**
    List of errors that occurred during device deregistration. If for some device deregistration completed successfully, it will not be listed here. For each device at most one entry will be present.
    */

  /** @param source {@displayType `DeepPartial<LicenseInfo>`} */
  constructor(source = {}) {
    if (source.status !== undefined) {
      this.status = source.status;
    } else {
      throw new Error('status must be present in constructor argument');
    }
    if (source.isValid !== undefined) {
      this.isValid = source.isValid;
    } else {
      throw new Error('isValid must be present in constructor argument');
    }
    if (source.licenseStatusMessage !== undefined) {
      this.licenseStatusMessage = source.licenseStatusMessage;
    } else {
      throw new Error('licenseStatusMessage must be present in constructor argument');
    }
    if (source.expirationTimestamp !== undefined) {
      this.expirationTimestamp = source.expirationTimestamp;
    } else {
      throw new Error('expirationTimestamp must be present in constructor argument');
    }
    if (source.expirationDateString !== undefined) {
      this.expirationDateString = source.expirationDateString;
    } else {
      throw new Error('expirationDateString must be present in constructor argument');
    }
    if (source.errorMessage !== undefined) {
      this.errorMessage = source.errorMessage;
    } else {
      throw new Error('errorMessage must be present in constructor argument');
    }
    if (source.onlineLicenseCheckInProgress !== undefined) {
      this.onlineLicenseCheckInProgress = source.onlineLicenseCheckInProgress;
    } else {
      throw new Error('onlineLicenseCheckInProgress must be present in constructor argument');
    }
    if (source.license !== undefined) {
      this.license = source.license;
    } else {
      throw new Error('license must be present in constructor argument');
    }
    if (source.devices !== undefined) {
      this.devices = source.devices != null ? source.devices : null;
    }
    if (source.nextOnlineLicenseCheckInMs !== undefined) {
      this.nextOnlineLicenseCheckInMs = source.nextOnlineLicenseCheckInMs;
    } else {
      throw new Error('nextOnlineLicenseCheckInMs must be present in constructor argument');
    }
    if (source.deviceDeregistationErrors !== undefined) {
      this.deviceDeregistationErrors = source.deviceDeregistationErrors.map(it => {
        return new DeviceDeregistrationError(it);
      });
    } else {
      throw new Error('deviceDeregistationErrors must be present in constructor argument');
    }
  }
  serialize(config = new ToJsonConfiguration()) {
    return {
      status: this.status,
      isValid: this.isValid,
      licenseStatusMessage: this.licenseStatusMessage,
      expirationTimestamp: this.expirationTimestamp,
      expirationDateString: this.expirationDateString,
      errorMessage: this.errorMessage,
      onlineLicenseCheckInProgress: this.onlineLicenseCheckInProgress,
      license: this.license,
      devices: this.devices != null ? this.devices : null,
      nextOnlineLicenseCheckInMs: this.nextOnlineLicenseCheckInMs,
      deviceDeregistationErrors: this.deviceDeregistationErrors.map(it => {
        return it.serialize(config);
      })
    };
  }
}

/**
License features.

- `BARCODE_SCANNER`:
   Barcode scanner.
- `CHECK_SCANNER`:
   Check scanner.
- `CREDIT_CARD_SCANNER`:
   Credit card scanner.
- `IMAGE_PROCESSING`:
   Image processing.
- `DOCUMENT_SCANNER`:
   Document scanner.
- `LICENSE_PLATE_SCANNER`:
   License plate scanner.
- `MEDICAL_CERTIFICATE_SCANNER`:
   Medical certificate scanner.
- `MRZ_SCANNER`:
   MRZ scanner.
- `OCR`:
   OCR.
- `TEXT_PATTERN_SCANNER`:
   Text pattern scanner.
- `PDF_CREATION`:
   PDF creation.
- `ID_CARD_SCANNER`:
   ID card scanner.
- `DRIVER_LICENSE_SCANNER`:
   Driver license scanner.
- `EHIC_SCANNER`:
   EHIC scanner.
- `DOCUMENT_CLASSIFICATION`:
   Document classification.
*/

/** @hidden */
export const FeatureValues = ['BARCODE_SCANNER', 'CHECK_SCANNER', 'CREDIT_CARD_SCANNER', 'IMAGE_PROCESSING', 'DOCUMENT_SCANNER', 'LICENSE_PLATE_SCANNER', 'MEDICAL_CERTIFICATE_SCANNER', 'MRZ_SCANNER', 'OCR', 'TEXT_PATTERN_SCANNER', 'PDF_CREATION', 'ID_CARD_SCANNER', 'DRIVER_LICENSE_SCANNER', 'EHIC_SCANNER', 'DOCUMENT_CLASSIFICATION'];
