// react-native-scanbot-sdk/lib/commonjs/ui_v2/document/CroppingTextLocalization.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CroppingTextLocalization = void 0;
var _utils = require("../../utils/utils");
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/document/CroppingTextLocalization.yaml

/**
Configuration of all the strings for the standalone cropping screen.
*/
class CroppingTextLocalization extends _utils.PartiallyConstructible {
  /**
    The cropping progress overlay title.
    
    Default is "Please wait..."
    */
  cameraProgressOverlayTitle = 'Please wait...';
  /**
    The 'cancel' button title, located in the top bar of the cropping screen.
    
    Default is "Cancel"
    */
  croppingTopBarCancelButtonTitle = 'Cancel';
  /**
    The 'confirm' button title, located in the top bar of the cropping screen.
    
    Default is "Done"
    */
  croppingTopBarConfirmButtonTitle = 'Done';
  /**
    The 'detect document' button title, located in the bottom bar of the cropping screen.
    
    Default is "Detect"
    */
  croppingDetectButtonTitle = 'Detect';
  /**
    The 'rotate page' button title, located in the bottom bar of the cropping screen.
    
    Default is "Rotate"
    */
  croppingRotateButtonTitle = 'Rotate';
  /**
    The 'reset detection' button title, located in the bottom bar of the cropping screen.
    
    Default is "Reset"
    */
  croppingResetButtonTitle = 'Reset';
  /**
    The cropping screen title, located in the top bar.
    
    Default is "Crop"
    */
  croppingScreenTitle = 'Crop';
  /**
    The accessibility hint for the 'confirm' button, located in the top bar of the cropping screen.
    
    Default is "Tap to apply changes"
    */
  accessibilityDescriptionCroppingTopBarConfirmButton = 'Tap to apply changes';
  /**
    The accessibility hint for the 'cancel' button, located in the top bar of the cropping screen.
    
    Default is "Tap to close screen without applying changes"
    */
  accessibilityDescriptionCroppingTopBarCancelButton = 'Tap to close screen without applying changes';
  /**
    The accessibility hint for the 'detect document' button, located in the bottom bar of the cropping screen.
    
    Default is "Tap to detect document on the image"
    */
  accessibilityDescriptionCroppingDetectButton = 'Tap to detect document on the image';
  /**
    The accessibility hint for the 'rotate page' button, located in the bottom bar of the cropping screen.
    
    Default is "Tap to rotate document"
    */
  accessibilityDescriptionCroppingRotateButton = 'Tap to rotate document';
  /**
    The accessibility hint for the 'reset detection' button, located in the bottom bar of the cropping screen.
    
    Default is "Tap to reset document contour to whole image"
    */
  accessibilityDescriptionCroppingResetButton = 'Tap to reset document contour to whole image';

  /** @param source {@displayType `DeepPartial<CroppingTextLocalization>`} */
  constructor(source = {}) {
    super();
    if (source.cameraProgressOverlayTitle !== undefined) {
      this.cameraProgressOverlayTitle = source.cameraProgressOverlayTitle;
    }
    if (source.croppingTopBarCancelButtonTitle !== undefined) {
      this.croppingTopBarCancelButtonTitle = source.croppingTopBarCancelButtonTitle;
    }
    if (source.croppingTopBarConfirmButtonTitle !== undefined) {
      this.croppingTopBarConfirmButtonTitle = source.croppingTopBarConfirmButtonTitle;
    }
    if (source.croppingDetectButtonTitle !== undefined) {
      this.croppingDetectButtonTitle = source.croppingDetectButtonTitle;
    }
    if (source.croppingRotateButtonTitle !== undefined) {
      this.croppingRotateButtonTitle = source.croppingRotateButtonTitle;
    }
    if (source.croppingResetButtonTitle !== undefined) {
      this.croppingResetButtonTitle = source.croppingResetButtonTitle;
    }
    if (source.croppingScreenTitle !== undefined) {
      this.croppingScreenTitle = source.croppingScreenTitle;
    }
    if (source.accessibilityDescriptionCroppingTopBarConfirmButton !== undefined) {
      this.accessibilityDescriptionCroppingTopBarConfirmButton = source.accessibilityDescriptionCroppingTopBarConfirmButton;
    }
    if (source.accessibilityDescriptionCroppingTopBarCancelButton !== undefined) {
      this.accessibilityDescriptionCroppingTopBarCancelButton = source.accessibilityDescriptionCroppingTopBarCancelButton;
    }
    if (source.accessibilityDescriptionCroppingDetectButton !== undefined) {
      this.accessibilityDescriptionCroppingDetectButton = source.accessibilityDescriptionCroppingDetectButton;
    }
    if (source.accessibilityDescriptionCroppingRotateButton !== undefined) {
      this.accessibilityDescriptionCroppingRotateButton = source.accessibilityDescriptionCroppingRotateButton;
    }
    if (source.accessibilityDescriptionCroppingResetButton !== undefined) {
      this.accessibilityDescriptionCroppingResetButton = source.accessibilityDescriptionCroppingResetButton;
    }
  }
}
exports.CroppingTextLocalization = CroppingTextLocalization;
