// react-native-scanbot-sdk/lib/commonjs/ui_v2/document/ReorderPagesScreenConfiguration.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ReorderPagesScreenConfiguration = void 0;
var _Common = require("../../ui_v2/common/Common");
var _UserGuidanceConfiguration = require("../../ui_v2/common/UserGuidanceConfiguration");
var _utils = require("../../utils/utils");
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/document/ReorderPagesScreenConfiguration.yaml

/**
Configuration of the screen for reordering the scanned pages.
*/
class ReorderPagesScreenConfiguration extends _utils.PartiallyConstructible {
  /**
    Configuration of the 'cancel' button, located in the top bar.
    */
  topBarBackButton = new _Common.ButtonConfiguration({
    visible: true,
    text: '?reorderTopBarCancelButtonTitle',
    accessibilityDescription: '?accessibilityDescriptionReorderTopBarCancelButton',
    background: new _Common.BackgroundStyle({
      strokeColor: '#00000000',
      fillColor: '#00000000',
      strokeWidth: 0.0
    }),
    foreground: new _Common.ForegroundStyle({
      iconVisible: true,
      color: '?sbColorOnPrimary',
      useShadow: false
    })
  });
  /**
    Configuration of the title, located in the top bar.
    */
  topBarTitle = new _Common.StyledText({
    text: '?reorderPageTitle',
    color: '?sbColorOnPrimary'
  });
  /**
    Configuration of the 'confirm' button, located in the top bar.
    */
  topBarConfirmButton = new _Common.ButtonConfiguration({
    visible: true,
    text: '?reorderTopBarConfirmButtonTitle',
    accessibilityDescription: '?accessibilityDescriptionReorderTopBarConfirmButton',
    background: new _Common.BackgroundStyle({
      strokeColor: '#00000000',
      fillColor: '#00000000',
      strokeWidth: 0.0
    }),
    foreground: new _Common.ForegroundStyle({
      iconVisible: true,
      color: '?sbColorOnPrimary',
      useShadow: false
    })
  });
  /**
    The background color of the reorder pages screen.
    
    Default is "?sbColorOutline"
    */
  backgroundColor = '?sbColorOutline';
  /**
    Configuration of the static user guidance, located just below the top bar.
    */
  guidance = new _UserGuidanceConfiguration.UserGuidanceConfiguration({
    title: new _Common.StyledText({
      text: '?reorderPageGuidanceTitle',
      color: '?sbColorOnPrimary'
    }),
    background: new _Common.BackgroundStyle({
      strokeColor: '#00000000',
      fillColor: '?sbColorSurfaceLow',
      strokeWidth: 0.0
    })
  });
  /**
    Configuration of the title below each page object.
    */
  pageTextStyle = new _Common.StyledText({
    text: '?reorderPageText',
    color: '?sbColorOnSurface'
  });

  /** @param source {@displayType `DeepPartial<ReorderPagesScreenConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.topBarBackButton !== undefined) {
      this.topBarBackButton = new _Common.ButtonConfiguration(source.topBarBackButton);
    }
    if (source.topBarTitle !== undefined) {
      this.topBarTitle = new _Common.StyledText(source.topBarTitle);
    }
    if (source.topBarConfirmButton !== undefined) {
      this.topBarConfirmButton = new _Common.ButtonConfiguration(source.topBarConfirmButton);
    }
    if (source.backgroundColor !== undefined) {
      this.backgroundColor = source.backgroundColor;
    }
    if (source.guidance !== undefined) {
      this.guidance = new _UserGuidanceConfiguration.UserGuidanceConfiguration(source.guidance);
    }
    if (source.pageTextStyle !== undefined) {
      this.pageTextStyle = new _Common.StyledText(source.pageTextStyle);
    }
  }
}
exports.ReorderPagesScreenConfiguration = ReorderPagesScreenConfiguration;
