// react-native-scanbot-sdk/lib/commonjs/ui_v2/check/CheckScannerScreenConfiguration.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CheckScannerScreenConfiguration = void 0;
var _CheckScannerTypes = require("../../check/CheckScannerTypes");
var _CheckScannerIntroScreenConfiguration = require("../../ui_v2/check/CheckScannerIntroScreenConfiguration");
var _CheckScannerScreenTextLocalization = require("../../ui_v2/check/CheckScannerScreenTextLocalization");
var _CheckScannerUserGuidance = require("../../ui_v2/check/CheckScannerUserGuidance");
var _CheckScanningProgressConfiguration = require("../../ui_v2/check/CheckScanningProgressConfiguration");
var _ActionBarConfiguration = require("../../ui_v2/common/ActionBarConfiguration");
var _CameraConfiguration = require("../../ui_v2/common/CameraConfiguration");
var _CameraPermission = require("../../ui_v2/common/CameraPermission");
var _Common = require("../../ui_v2/common/Common");
var _ScanCompletionOverlay = require("../../ui_v2/common/ScanCompletionOverlay");
var _ScanbotAlertDialog = require("../../ui_v2/common/ScanbotAlertDialog");
var _TopBarConfiguration = require("../../ui_v2/common/TopBarConfiguration");
var _UserGuidanceConfiguration = require("../../ui_v2/common/UserGuidanceConfiguration");
var _ViewFinderConfiguration = require("../../ui_v2/common/ViewFinderConfiguration");
var _Geometry = require("../../utils/geometry/Geometry");
var _utils = require("../../utils/utils");
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/check/CheckScannerScreenConfiguration.yaml

/**
Configuration of the screen for detecting check data.
*/
class CheckScannerScreenConfiguration extends _utils.PartiallyConstructible {
  /**
    Version number of the configuration object.
    
    Default is "1.0"
    */
  version = '1.0';
  /**
    The configuration object should be applied for this screen.
    
    Default is "CheckScanner"
    */
  screen = 'CheckScanner';
  /**
    Configuration of the all strings for check scanner screen.
    */
  localization = new _CheckScannerScreenTextLocalization.CheckScannerScreenTextLocalization({});
  /**
    Define the screen's base color values from which other colors are derived.
    */
  palette = new _Common.Palette({});
  /**
    The background color of the check scanner screen.
    
    Default is "?sbColorSurfaceLow"
    */
  backgroundColor = '?sbColorSurfaceLow';
  /**
    Configuration of the camera behavior.
    */
  cameraConfiguration = new _CameraConfiguration.CameraConfiguration({
    orientationLockMode: 'PORTRAIT'
  });
  /**
    Configuration of the dialog for requesting camera permissions.
    */
  cameraPermission = new _CameraPermission.CameraPermissionScreen({
    statusBarMode: 'DARK',
    background: '?sbColorSurface',
    iconBackground: '?sbColorOutline',
    icon: new _Common.IconStyle({
      visible: true,
      color: '?sbColorOnSurface'
    }),
    enableCameraButton: new _Common.ButtonConfiguration({
      visible: true,
      text: '?cameraPermissionEnableCameraButton',
      accessibilityDescription: '?accessibilityDescriptionCameraPermissionEnableCameraButton',
      background: new _Common.BackgroundStyle({
        strokeColor: '?sbColorPrimary',
        fillColor: '?sbColorPrimary',
        strokeWidth: 0.0
      }),
      foreground: new _Common.ForegroundStyle({
        iconVisible: false,
        color: '?sbColorOnPrimary',
        useShadow: false
      })
    }),
    closeButton: new _Common.ButtonConfiguration({
      visible: true,
      text: '?cameraPermissionCloseButton',
      accessibilityDescription: '?accessibilityDescriptionCameraPermissionCloseButton',
      background: new _Common.BackgroundStyle({
        strokeColor: '#00000000',
        fillColor: '#00000000',
        strokeWidth: 0.0
      }),
      foreground: new _Common.ForegroundStyle({
        iconVisible: false,
        color: '?sbColorPrimary',
        useShadow: false
      })
    }),
    enableCameraTitle: new _Common.StyledText({
      text: '?cameraPermissionEnableCameraTitle',
      color: '?sbColorOnSurface'
    }),
    enableCameraExplanation: new _Common.StyledText({
      text: '?cameraPermissionEnableCameraExplanation',
      color: '?sbColorOnSurfaceVariant'
    })
  });
  /**
    Configuration of the top user guidance.
    */
  topUserGuidance = new _UserGuidanceConfiguration.UserGuidanceConfiguration({
    title: new _Common.StyledText({
      text: '?topUserGuidance'
    })
  });
  /**
    Configuration of the user guidance below finder view.
    */
  scanStatusUserGuidance = new _CheckScannerUserGuidance.CheckScannerGuidanceConfiguration({});
  /**
    Configuration of the top bar on the check scanner screen.
    */
  topBar = new _TopBarConfiguration.TopBarConfiguration({
    title: new _Common.StyledText({
      visible: true,
      text: '?topBarTitle'
    }),
    cancelButton: new _Common.ButtonConfiguration({
      text: '?topBarCancelButton',
      accessibilityDescription: '?accessibilityDescriptionCancelButton',
      background: new _Common.BackgroundStyle({
        strokeColor: '#00000000',
        fillColor: '#00000000',
        strokeWidth: 0.0
      }),
      foreground: new _Common.ForegroundStyle({
        color: '?sbColorOnPrimary'
      })
    })
  });
  /**
    Configuration of the button in the top bar that opens the introduction screen.
    */
  topBarOpenIntroScreenButton = new _Common.IconButton({
    color: '?sbColorOnPrimary',
    accessibilityDescription: '?accessibilityDescriptionOpenIntroScreenButton'
  });
  /**
    Configuration of the introduction screen for the check scanner.
    */
  introScreen = new _CheckScannerIntroScreenConfiguration.CheckScannerIntroScreenConfiguration({});
  /**
    Configuration of the action bar.
    */
  actionBar = new _ActionBarConfiguration.ActionBarConfiguration({
    flipCameraButton: new _Common.RoundButton({
      visible: false,
      accessibilityDescription: '?accessibilityDescriptionFlipCameraButton',
      backgroundColor: '?sbColorSurfaceHigh',
      foregroundColor: '?sbColorOnPrimary',
      activeBackgroundColor: '?sbColorWarning',
      activeForegroundColor: '#1C1B1F'
    })
  });
  /**
    Configuration of the view finder.
    */
  viewFinder = new _ViewFinderConfiguration.PermanentViewFinderConfiguration({
    style: new _ViewFinderConfiguration.FinderCorneredStyle({
      strokeColor: '?sbColorSurface',
      strokeWidth: 2.0
    }),
    aspectRatio: new _Geometry.AspectRatio({
      width: 2.14,
      height: 1.0
    }),
    minimumInsets: new _Common.EdgeInsets({
      top: 24.0,
      left: 24.0,
      bottom: 24.0,
      right: 24.0
    })
  });
  /**
    Configuration for the check recognizer.
    */
  scannerConfiguration = new _CheckScannerTypes.CheckScannerConfiguration({});
  /**
    Flag to capture high resolution image.
    
    Default is false
    */
  captureHighResolutionImage = false;
  /**
    Flag to show or hide the preset button.
    
    Default is true
    */
  exampleOverlayVisible = true;
  /**
    Configuration of visualization of the scanning progress.
    */
  scanningProgress = new _CheckScanningProgressConfiguration.CheckScanningProgressConfiguration({});
  /**
    Timeout for the no Check found state. If the no Check found state is active longer than this value, the alert will be presented.
    
    Default is 5000
    */
  noCheckFoundTimeout = 5000;
  /**
    Timeout for the accumulation state. If the accumulation state is active longer than this value, the alert will be presented.
    
    Default is 3000
    */
  accumulationTimeout = 3000;
  /**
    Configuration of the success overlay.
    */
  successOverlay = new _ScanCompletionOverlay.ScanCompletionOverlay({});
  /**
    Configuration of the alert dialog that will be presented when no Check is detected.
    */
  noCheckDetectedAlertDialog = new _ScanbotAlertDialog.ScanbotAlertDialog({
    title: new _Common.StyledText({
      text: '?checkNoCheckDetectedAlertTitle',
      color: '?sbColorOnSurface'
    }),
    subtitle: new _Common.StyledText({
      text: '?checkNoCheckDetectedAlertSubtitle',
      color: '?sbColorOnSurfaceVariant'
    }),
    sheetColor: '?sbColorSurface',
    modalOverlayColor: '?sbColorModalOverlay',
    dividerColor: '?sbColorOutline',
    okButton: new _Common.ButtonConfiguration({
      visible: true,
      text: '?checkAlertRetryButton',
      accessibilityDescription: '?accessibilityDescriptionRetryButton',
      background: new _Common.BackgroundStyle({
        strokeColor: '?sbColorPrimary',
        fillColor: '?sbColorPrimary',
        strokeWidth: 1.0
      }),
      foreground: new _Common.ForegroundStyle({
        iconVisible: true,
        color: '?sbColorOnPrimary',
        useShadow: false
      })
    }),
    cancelButton: new _Common.ButtonConfiguration({
      visible: true,
      text: '?checkAlertCloseScannerButton',
      accessibilityDescription: '?accessibilityDescriptionCloseScannerButton',
      background: new _Common.BackgroundStyle({
        strokeColor: '#00000000',
        fillColor: '#00000000',
        strokeWidth: 1.0
      }),
      foreground: new _Common.ForegroundStyle({
        iconVisible: false,
        color: '?sbColorPrimary',
        useShadow: false
      })
    })
  });
  /**
    Configuration of the alert dialog that will be presented when scanned image is not good.
    */
  imageNotGoodAlertDialog = new _ScanbotAlertDialog.ScanbotAlertDialog({
    title: new _Common.StyledText({
      text: '?checkImageNotGoodAlertTitle',
      color: '?sbColorOnSurface'
    }),
    subtitle: new _Common.StyledText({
      text: '?checkImageNotGoodAlertSubtitle',
      color: '?sbColorOnSurfaceVariant'
    }),
    sheetColor: '?sbColorSurface',
    modalOverlayColor: '?sbColorModalOverlay',
    dividerColor: '?sbColorOutline',
    okButton: new _Common.ButtonConfiguration({
      visible: true,
      text: '?checkAlertRetryButton',
      accessibilityDescription: '?accessibilityDescriptionRetryButton',
      background: new _Common.BackgroundStyle({
        strokeColor: '?sbColorPrimary',
        fillColor: '?sbColorPrimary',
        strokeWidth: 1.0
      }),
      foreground: new _Common.ForegroundStyle({
        iconVisible: true,
        color: '?sbColorOnPrimary',
        useShadow: false
      })
    }),
    cancelButton: new _Common.ButtonConfiguration({
      visible: true,
      text: '?checkAlertCloseScannerButton',
      accessibilityDescription: '?accessibilityDescriptionCloseScannerButton',
      background: new _Common.BackgroundStyle({
        strokeColor: '#00000000',
        fillColor: '#00000000',
        strokeWidth: 1.0
      }),
      foreground: new _Common.ForegroundStyle({
        iconVisible: false,
        color: '?sbColorPrimary',
        useShadow: false
      })
    })
  });
  /**
    Configuration of the scan confirmation sound.
    */
  sound = new _Common.Sound({});
  /**
    Configuration of the vibration feedback.
    */
  vibration = new _Common.Vibration({});
  /**
    If enabled, the screen will not turn off while the check scanner is active. (Android only).
    
    Default is true
    */
  keepScreenOn = true;

  /** @param source {@displayType `DeepPartial<CheckScannerScreenConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.version !== undefined) {
      this.version = source.version;
    }
    if (source.screen !== undefined) {
      this.screen = source.screen;
    }
    if (source.localization !== undefined) {
      this.localization = new _CheckScannerScreenTextLocalization.CheckScannerScreenTextLocalization(source.localization);
    }
    if (source.palette !== undefined) {
      this.palette = new _Common.Palette(source.palette);
    }
    if (source.backgroundColor !== undefined) {
      this.backgroundColor = source.backgroundColor;
    }
    if (source.cameraConfiguration !== undefined) {
      this.cameraConfiguration = new _CameraConfiguration.CameraConfiguration(source.cameraConfiguration);
    }
    if (source.cameraPermission !== undefined) {
      this.cameraPermission = new _CameraPermission.CameraPermissionScreen(source.cameraPermission);
    }
    if (source.topUserGuidance !== undefined) {
      this.topUserGuidance = new _UserGuidanceConfiguration.UserGuidanceConfiguration(source.topUserGuidance);
    }
    if (source.scanStatusUserGuidance !== undefined) {
      this.scanStatusUserGuidance = new _CheckScannerUserGuidance.CheckScannerGuidanceConfiguration(source.scanStatusUserGuidance);
    }
    if (source.topBar !== undefined) {
      this.topBar = new _TopBarConfiguration.TopBarConfiguration(source.topBar);
    }
    if (source.topBarOpenIntroScreenButton !== undefined) {
      this.topBarOpenIntroScreenButton = new _Common.IconButton(source.topBarOpenIntroScreenButton);
    }
    if (source.introScreen !== undefined) {
      this.introScreen = new _CheckScannerIntroScreenConfiguration.CheckScannerIntroScreenConfiguration(source.introScreen);
    }
    if (source.actionBar !== undefined) {
      this.actionBar = new _ActionBarConfiguration.ActionBarConfiguration(source.actionBar);
    }
    if (source.viewFinder !== undefined) {
      this.viewFinder = new _ViewFinderConfiguration.PermanentViewFinderConfiguration(source.viewFinder);
    }
    if (source.scannerConfiguration !== undefined) {
      this.scannerConfiguration = new _CheckScannerTypes.CheckScannerConfiguration(source.scannerConfiguration);
    }
    if (source.captureHighResolutionImage !== undefined) {
      this.captureHighResolutionImage = source.captureHighResolutionImage;
    }
    if (source.exampleOverlayVisible !== undefined) {
      this.exampleOverlayVisible = source.exampleOverlayVisible;
    }
    if (source.scanningProgress !== undefined) {
      this.scanningProgress = new _CheckScanningProgressConfiguration.CheckScanningProgressConfiguration(source.scanningProgress);
    }
    if (source.noCheckFoundTimeout !== undefined) {
      this.noCheckFoundTimeout = source.noCheckFoundTimeout;
    }
    if (source.accumulationTimeout !== undefined) {
      this.accumulationTimeout = source.accumulationTimeout;
    }
    if (source.successOverlay !== undefined) {
      this.successOverlay = new _ScanCompletionOverlay.ScanCompletionOverlay(source.successOverlay);
    }
    if (source.noCheckDetectedAlertDialog !== undefined) {
      this.noCheckDetectedAlertDialog = new _ScanbotAlertDialog.ScanbotAlertDialog(source.noCheckDetectedAlertDialog);
    }
    if (source.imageNotGoodAlertDialog !== undefined) {
      this.imageNotGoodAlertDialog = new _ScanbotAlertDialog.ScanbotAlertDialog(source.imageNotGoodAlertDialog);
    }
    if (source.sound !== undefined) {
      this.sound = new _Common.Sound(source.sound);
    }
    if (source.vibration !== undefined) {
      this.vibration = new _Common.Vibration(source.vibration);
    }
    if (source.keepScreenOn !== undefined) {
      this.keepScreenOn = source.keepScreenOn;
    }
  }
}
exports.CheckScannerScreenConfiguration = CheckScannerScreenConfiguration;
