// react-native-scanbot-sdk/lib/commonjs/ui_v2/check/CheckScannerUserGuidance.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CheckScannerGuidanceStates = exports.CheckScannerGuidanceConfiguration = void 0;
var _Common = require("../../ui_v2/common/Common");
var _utils = require("../../utils/utils");
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/check/CheckScannerUserGuidance.yaml

/**
Configure the captions of the user guidance hints for different scanning states.
*/
class CheckScannerGuidanceStates extends _utils.PartiallyConstructible {
  /**
    The user guidance text displayed when no check is found. Initial state.
    
    Default is "?checkUserGuidanceNoCheckFound"
    */
  noCheckFound = '?checkUserGuidanceNoCheckFound';
  /**
    The user guidance text displayed when a check presence was detected and scanning is in progress.
    
    Default is "?checkUserGuidanceScanningProgress"
    */
  scanningProgress = '?checkUserGuidanceScanningProgress';
  /**
    The user guidance text displayed when it is too dark to capture an adequate image.
    
    Default is "?checkUserGuidanceTooDark"
    */
  tooDark = '?checkUserGuidanceTooDark';
  /**
    The user guidance text displayed when the detected check is too far away and appears too small.
    
    Default is "?checkUserGuidanceTooSmall"
    */
  tooSmall = '?checkUserGuidanceTooSmall';
  /**
    The user guidance text displayed when the detected check is not in a good perspective (device tilted).
    
    Default is "?checkUserGuidanceBadPerspective"
    */
  badPerspective = '?checkUserGuidanceBadPerspective';

  /** @param source {@displayType `DeepPartial<CheckScannerGuidanceStates>`} */
  constructor(source = {}) {
    super();
    if (source.noCheckFound !== undefined) {
      this.noCheckFound = source.noCheckFound;
    }
    if (source.scanningProgress !== undefined) {
      this.scanningProgress = source.scanningProgress;
    }
    if (source.tooDark !== undefined) {
      this.tooDark = source.tooDark;
    }
    if (source.tooSmall !== undefined) {
      this.tooSmall = source.tooSmall;
    }
    if (source.badPerspective !== undefined) {
      this.badPerspective = source.badPerspective;
    }
  }
}

/**
Configuration of the hints guiding users through the check scanning process.
*/
exports.CheckScannerGuidanceStates = CheckScannerGuidanceStates;
class CheckScannerGuidanceConfiguration extends _utils.PartiallyConstructible {
  /**
    Determines whether the user guidance hints should be visible (enabled).
    
    Default is true
    */
  visibility = true;
  /**
    Configure the text style for the user guidance hints.
    */
  title = new _Common.StyledText({
    text: '?checkUserGuidanceNoCheckFound',
    color: '?sbColorOnPrimary'
  });
  /**
    Configure the background style for the user guidance hints.
    */
  background = new _Common.BackgroundStyle({
    strokeColor: '#00000000',
    fillColor: '?sbColorSurfaceLow'
  });
  /**
    Configure the captions of the user guidance hints for different states.
    */
  statesTitles = new CheckScannerGuidanceStates({});

  /** @param source {@displayType `DeepPartial<CheckScannerGuidanceConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.visibility !== undefined) {
      this.visibility = source.visibility;
    }
    if (source.title !== undefined) {
      this.title = new _Common.StyledText(source.title);
    }
    if (source.background !== undefined) {
      this.background = new _Common.BackgroundStyle(source.background);
    }
    if (source.statesTitles !== undefined) {
      this.statesTitles = new CheckScannerGuidanceStates(source.statesTitles);
    }
  }
}
exports.CheckScannerGuidanceConfiguration = CheckScannerGuidanceConfiguration;
