// react-native-scanbot-sdk/lib/commonjs/ui_v2/check/CheckScannerUIResult.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CheckScannerUiResult = void 0;
var _GenericDocument = require("../../documents/GenericDocument");
var _image = require("../../imageRef/image");
var _JsonSerializationTypes = require("../../utils/json/JsonSerializationTypes");
var _utils = require("../../utils/utils");
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/check/CheckScannerUIResult.yaml

/**
Configuration of the check scanner RTU UI result.
*/
class CheckScannerUiResult extends _utils.PartiallyConstructible {
  /**
    The status of the check recognition step.
    */

  /**
    Crop of the check if documentDetectionMode is set to DETECT_AND_CROP_DOCUMENT. Will be non-empty, only if check recognition succeeded.
    */

  /**
    Generic document containing check data. Not present, if status is FAIL.
    */

  /** @param source {@displayType `DeepPartial<CheckScannerUiResult>`} */
  constructor(source = {}) {
    super();
    if (source.recognitionStatus !== undefined) {
      this.recognitionStatus = source.recognitionStatus;
    } else {
      throw new Error('recognitionStatus must be present in constructor argument');
    }
    if (source.croppedImage !== undefined) {
      this.croppedImage = source.croppedImage != null ? _image.ImageRef.From(source.croppedImage) : null;
    } else {
      throw new Error('croppedImage must be present in constructor argument');
    }
    if (source.check !== undefined) {
      this.check = source.check != null ? new _GenericDocument.GenericDocument(source.check) : null;
    } else {
      throw new Error('check must be present in constructor argument');
    }
  }
  async serialize(config = new _JsonSerializationTypes.ToJsonConfiguration()) {
    return {
      recognitionStatus: this.recognitionStatus,
      croppedImage: config.serializeImages ? this.croppedImage != null ? await this.croppedImage.serialize(config.imageSerializationMode) : null : undefined,
      check: this.check != null ? await this.check.serialize(config) : null
    };
  }
  _released = false;
  release() {
    if (this._released) {
      return;
    }
    {
      if (this.croppedImage != null) {
        this.croppedImage.release();
      }
    }
    {
      if (this.check != null) {
        this.check.release();
      }
    }
    this._released = true;
  }
  async encodeImages() {
    if (this.croppedImage != null) {
      await this.croppedImage.encodeInPlace();
    }
    if (this.check != null) {
      await this.check.encodeImages();
    }
  }
}
exports.CheckScannerUiResult = CheckScannerUiResult;
