// react-native-scanbot-sdk/src/barcode/BarcodeDocumentTypes.ts
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from core/schemas/BarcodeDocumentTypes.yaml

import { BarcodeFormat } from '../barcode/BarcodeTypes';
import { ToJsonConfiguration } from '../utils/json/JsonSerializationTypes';
import { DeepPartial, PartiallyConstructible } from '../utils/utils';

/**
Type of barcode document format used.

- `AAMVA`:
   American Association of Motor Vehicle Administrators barcode document.
- `BOARDING_PASS`:
   Boarding pass barcode document.
- `DE_MEDICAL_PLAN`:
   German medication plan barcode document.
- `MEDICAL_CERTIFICATE`:
   German medical certificate barcode document.
- `ID_CARD_PDF_417`:
   ID card barcode document.
- `SEPA`:
   SEPA barcode (aka GiroCode) document.
- `SWISS_QR`:
   Swiss QR barcode document.
- `VCARD`:
   VCard barcode document.
- `GS1`:
   GS1 barcode document.
- `HIBC`:
   Health industry barcode document.
*/
export type BarcodeDocumentFormat =
  | 'AAMVA'
  | 'BOARDING_PASS'
  | 'DE_MEDICAL_PLAN'
  | 'MEDICAL_CERTIFICATE'
  | 'ID_CARD_PDF_417'
  | 'SEPA'
  | 'SWISS_QR'
  | 'VCARD'
  | 'GS1'
  | 'HIBC';

export const BarcodeDocumentFormatValues: BarcodeDocumentFormat[] = [
  'AAMVA',
  'BOARDING_PASS',
  'DE_MEDICAL_PLAN',
  'MEDICAL_CERTIFICATE',
  'ID_CARD_PDF_417',
  'SEPA',
  'SWISS_QR',
  'VCARD',
  'GS1',
  'HIBC',
];

/**
Barcode document formats.
*/
export class BarcodeDocumentFormats extends PartiallyConstructible {
  /** @param source {@displayType `DeepPartial<BarcodeDocumentFormats>`} */
  public constructor(source: DeepPartial<BarcodeDocumentFormats> = {}) {
    super();
  }

  public serialize(
    config: ToJsonConfiguration = new ToJsonConfiguration()
  ): DeepPartial<BarcodeDocumentFormats> {
    return {};
  }
}

export namespace BarcodeDocumentFormats {
  /**
    All barcode document formats.
    */
  export const all: BarcodeDocumentFormat[] = [
    'AAMVA',
    'BOARDING_PASS',
    'DE_MEDICAL_PLAN',
    'MEDICAL_CERTIFICATE',
    'ID_CARD_PDF_417',
    'SEPA',
    'SWISS_QR',
    'VCARD',
    'GS1',
    'HIBC',
  ];
  /**
    Barcode types that are used to encode documents.
    */
  export const documentBarcodeFormats: { [key in BarcodeDocumentFormat]: BarcodeFormat[] } = {
    AAMVA: ['PDF_417'],
    ID_CARD_PDF_417: ['PDF_417'],
    SEPA: ['QR_CODE'],
    MEDICAL_CERTIFICATE: ['PDF_417'],
    DE_MEDICAL_PLAN: ['DATA_MATRIX'],
    BOARDING_PASS: ['PDF_417', 'AZTEC', 'QR_CODE'],
    VCARD: ['QR_CODE', 'DATA_MATRIX'],
    SWISS_QR: ['QR_CODE'],
    GS1: [
      'CODE_128',
      'AZTEC',
      'DATA_MATRIX',
      'DATABAR',
      'DATABAR_EXPANDED',
      'DATABAR_LIMITED',
      'EAN_13',
      'EAN_8',
      'GS1_COMPOSITE',
      'ITF',
      'MICRO_PDF_417',
      'PDF_417',
      'QR_CODE',
      'UPC_A',
      'UPC_E',
    ],
    HIBC: ['CODE_39', 'CODE_128', 'AZTEC', 'DATA_MATRIX', 'MICRO_PDF_417', 'PDF_417', 'QR_CODE'],
  };
}
