// react-native-scanbot-barcode-scanner-sdk/src/types/core/barcode/BarcodeDocumentTypes.ts
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from core/schemas/BarcodeDocumentTypes.yaml

import { ToJsonConfiguration } from '../common/JsonSerializationTypes';
import { GenericDocument } from '../generic_document/GenericDocument';
import { DeepPartial } from '../utils/utils';
import { BarcodeFormat } from './BarcodeTypes';

/**
Type of barcode document format used.

- `AAMVA`:
   American Association of Motor Vehicle Administrators barcode document.
- `BOARDING_PASS`:
   Boarding pass barcode document.
- `DE_MEDICAL_PLAN`:
   German medication plan barcode document.
- `MEDICAL_CERTIFICATE`:
   German medical certificate barcode document.
- `ID_CARD_PDF_417`:
   ID card barcode document.
- `SEPA`:
   SEPA barcode (aka GiroCode) document.
- `SWISS_QR`:
   Swiss QR barcode document.
- `VCARD`:
   VCard barcode document.
- `GS1`:
   GS1 barcode document.
- `HIBC`:
   Health industry barcode document.
- `BRITISH_COLUMBIA_DRIVER_LICENSE`:
   British Columbia driver license barcode document.
*/
export type BarcodeDocumentFormat =
  /**
    American Association of Motor Vehicle Administrators barcode document.
    */
  | 'AAMVA'
  /**
    Boarding pass barcode document.
    */
  | 'BOARDING_PASS'
  /**
    German medication plan barcode document.
    */
  | 'DE_MEDICAL_PLAN'
  /**
    German medical certificate barcode document.
    */
  | 'MEDICAL_CERTIFICATE'
  /**
    ID card barcode document.
    */
  | 'ID_CARD_PDF_417'
  /**
    SEPA barcode (aka GiroCode) document.
    */
  | 'SEPA'
  /**
    Swiss QR barcode document.
    */
  | 'SWISS_QR'
  /**
    VCard barcode document.
    */
  | 'VCARD'
  /**
    GS1 barcode document.
    */
  | 'GS1'
  /**
    Health industry barcode document.
    */
  | 'HIBC'
  /**
    British Columbia driver license barcode document.
    */
  | 'BRITISH_COLUMBIA_DRIVER_LICENSE';

/** @hidden */
export const BarcodeDocumentFormatValues: ReadonlyArray<BarcodeDocumentFormat> = [
  'AAMVA',
  'BOARDING_PASS',
  'DE_MEDICAL_PLAN',
  'MEDICAL_CERTIFICATE',
  'ID_CARD_PDF_417',
  'SEPA',
  'SWISS_QR',
  'VCARD',
  'GS1',
  'HIBC',
  'BRITISH_COLUMBIA_DRIVER_LICENSE',
] as const;

/**
Barcode document formats.
*/
export class BarcodeDocumentFormats {
  /** @param source {@displayType `DeepPartial<BarcodeDocumentFormats>`} */
  public constructor(source: DeepPartial<BarcodeDocumentFormats> = {}) {}

  public serialize(
    config: ToJsonConfiguration = new ToJsonConfiguration()
  ): DeepPartial<BarcodeDocumentFormats> {
    return {};
  }
}

export namespace BarcodeDocumentFormats {
  /**
    All barcode document formats.
    */
  export const all: BarcodeDocumentFormat[] = [
    'AAMVA',
    'BOARDING_PASS',
    'DE_MEDICAL_PLAN',
    'MEDICAL_CERTIFICATE',
    'ID_CARD_PDF_417',
    'SEPA',
    'SWISS_QR',
    'VCARD',
    'GS1',
    'HIBC',
    'BRITISH_COLUMBIA_DRIVER_LICENSE',
  ];
  /**
    Barcode types that are used to encode documents.
    */
  export const documentBarcodeFormats: { [key in BarcodeDocumentFormat]: BarcodeFormat[] } = {
    AAMVA: ['PDF_417'],
    ID_CARD_PDF_417: ['PDF_417'],
    SEPA: ['QR_CODE'],
    MEDICAL_CERTIFICATE: ['PDF_417'],
    DE_MEDICAL_PLAN: ['DATA_MATRIX'],
    BOARDING_PASS: ['PDF_417', 'AZTEC', 'QR_CODE'],
    VCARD: ['QR_CODE', 'DATA_MATRIX'],
    SWISS_QR: ['QR_CODE'],
    GS1: [
      'CODE_128',
      'AZTEC',
      'DATA_MATRIX',
      'DATABAR',
      'DATABAR_EXPANDED',
      'DATABAR_LIMITED',
      'EAN_13',
      'EAN_8',
      'GS1_COMPOSITE',
      'ITF',
      'MICRO_PDF_417',
      'PDF_417',
      'QR_CODE',
      'UPC_A',
      'UPC_E',
    ],
    HIBC: ['CODE_39', 'CODE_128', 'AZTEC', 'DATA_MATRIX', 'MICRO_PDF_417', 'PDF_417', 'QR_CODE'],
    BRITISH_COLUMBIA_DRIVER_LICENSE: ['PDF_417'],
  };
}

/**
The result of barcode document parsing.
*/
export class BarcodeDocumentParserResult {
  /**
    True if any document was parsed.
    */
  public readonly success: boolean;
  /**
    In case of success, the parsed document.
    */
  public readonly parsedDocument: GenericDocument | null;

  /** @param source {@displayType `DeepPartial<BarcodeDocumentParserResult>`} */
  public constructor(source: DeepPartial<BarcodeDocumentParserResult> = {}) {
    if (source.success !== undefined) {
      this.success = source.success;
    } else {
      throw new Error('success must be present in constructor argument');
    }
    if (source.parsedDocument !== undefined) {
      this.parsedDocument =
        source.parsedDocument != null ? new GenericDocument(source.parsedDocument) : null;
    } else {
      throw new Error('parsedDocument must be present in constructor argument');
    }
  }

  public async serialize(
    config: ToJsonConfiguration = new ToJsonConfiguration()
  ): Promise<DeepPartial<BarcodeDocumentParserResult>> {
    return {
      success: this.success,
      parsedDocument:
        this.parsedDocument != null ? await this.parsedDocument.serialize(config) : null,
    };
  }

  private _released: Boolean = false;

  public release() {
    if (this._released) {
      return;
    }
    {
      if (this.parsedDocument != null) {
        this.parsedDocument.release();
      }
    }
    this._released = true;
  }

  public async encodeImages(): Promise<void> {
    if (this.parsedDocument != null) {
      await this.parsedDocument.encodeImages();
    }
  }
}
