// react-native-scanbot-barcode-scanner-sdk/lib/commonjs/types/core/ui_v2/common/CameraPermission.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CameraPermissionScreen = void 0;
var _Common = require("./Common");
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/common/CameraPermission.yaml

/**
Configuration of the camera permission request view.
*/
class CameraPermissionScreen {
  /**
    Determines the visual mode for displaying the contents of the status bar.
    
    Default is DARK
    */
  statusBarMode = 'DARK';
  /**
    The background color of the camera permission request.
    
    Default is "?sbColorSurface"
    */
  background = '?sbColorSurface';
  /**
    The background color of the icon used in the camera permission request.
    
    Default is "?sbColorOutline"
    */
  iconBackground = '?sbColorOutline';
  /**
    Configuration of the icon used in the camera permission request.
    */
  icon = new _Common.IconStyle({
    color: '?sbColorOnSurface'
  });
  /**
    Configuration of the camera permission request's confirmation button.
    */
  enableCameraButton = new _Common.ButtonConfiguration({
    text: '?cameraPermissionEnableCameraButton',
    background: new _Common.BackgroundStyle({
      strokeColor: '?sbColorPrimary',
      fillColor: '?sbColorPrimary',
      strokeWidth: 0.0
    }),
    foreground: new _Common.ForegroundStyle({
      iconVisible: false,
      color: '?sbColorOnPrimary'
    })
  });
  /**
    Configuration of the camera permission request's close button.
    */
  closeButton = new _Common.ButtonConfiguration({
    text: '?cameraPermissionCloseButton',
    background: new _Common.BackgroundStyle({
      strokeColor: '#00000000',
      fillColor: '#00000000',
      strokeWidth: 0.0
    }),
    foreground: new _Common.ForegroundStyle({
      iconVisible: false,
      color: '?sbColorPrimary'
    })
  });
  /**
    Configuration of the camera permission request's title.
    */
  enableCameraTitle = new _Common.StyledText({
    text: '?cameraPermissionEnableCameraTitle',
    color: '?sbColorOnSurface'
  });
  /**
    Configuration of the camera permission request's explanatory text.
    */
  enableCameraExplanation = new _Common.StyledText({
    text: '?cameraPermissionEnableCameraExplanation',
    color: '?sbColorOnSurfaceVariant'
  });

  /** @param source {@displayType `DeepPartial<CameraPermissionScreen>`} */
  constructor(source = {}) {
    if (source.statusBarMode !== undefined) {
      this.statusBarMode = source.statusBarMode;
    }
    if (source.background !== undefined) {
      this.background = source.background;
    }
    if (source.iconBackground !== undefined) {
      this.iconBackground = source.iconBackground;
    }
    if (source.icon !== undefined) {
      this.icon = new _Common.IconStyle(source.icon);
    }
    if (source.enableCameraButton !== undefined) {
      this.enableCameraButton = new _Common.ButtonConfiguration(source.enableCameraButton);
    }
    if (source.closeButton !== undefined) {
      this.closeButton = new _Common.ButtonConfiguration(source.closeButton);
    }
    if (source.enableCameraTitle !== undefined) {
      this.enableCameraTitle = new _Common.StyledText(source.enableCameraTitle);
    }
    if (source.enableCameraExplanation !== undefined) {
      this.enableCameraExplanation = new _Common.StyledText(source.enableCameraExplanation);
    }
  }
}
exports.CameraPermissionScreen = CameraPermissionScreen;
