// react-native-scanbot-barcode-scanner-sdk/lib/commonjs/types/core/ui_v2/common/ScanCompletionOverlay.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ScanCompletionOverlay = void 0;
var _Common = require("./Common");
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/common/ScanCompletionOverlay.yaml

/**
Configuration of the overlay to be shown after the successful scan.
*/
class ScanCompletionOverlay {
  /**
    A caption below the icon.
    */
  message = new _Common.StyledText({
    text: '?completionOverlaySuccessMessage',
    color: '?sbColorOnPrimary'
  });
  /**
    Color of the icon.
    
    Default is "?sbColorOnPrimary"
    */
  iconColor = '?sbColorOnPrimary';
  /**
    Background color of the overlay.
    
    Default is "?sbColorSurfaceHigh"
    */
  overlayBackgroundColor = '?sbColorSurfaceHigh';
  /**
    Timeout in milliseconds after which the overlay is automatically dismissed.
    
    Default is 1000
    */
  timeout = 1000;

  /** @param source {@displayType `DeepPartial<ScanCompletionOverlay>`} */
  constructor(source = {}) {
    if (source.message !== undefined) {
      this.message = new _Common.StyledText(source.message);
    }
    if (source.iconColor !== undefined) {
      this.iconColor = source.iconColor;
    }
    if (source.overlayBackgroundColor !== undefined) {
      this.overlayBackgroundColor = source.overlayBackgroundColor;
    }
    if (source.timeout !== undefined) {
      this.timeout = source.timeout;
    }
  }
}
exports.ScanCompletionOverlay = ScanCompletionOverlay;
