// react-native-scanbot-barcode-scanner-sdk/lib/commonjs/types/core/ui_v2/common/ViewFinderConfiguration.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ViewFinderConfiguration = exports.PermanentViewFinderConfiguration = exports.FinderStyle = exports.FinderStrokedStyle = exports.FinderCorneredStyle = exports.BaseViewFinderConfiguration = void 0;
var _Geometry = require("../../geometry/Geometry");
var _Common = require("./Common");
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/common/ViewFinderConfiguration.yaml
/**
Base configuration of the scanning interface’s viewfinder, serving as guidance to the user.
*/
/** @internal */
let BaseViewFinderConfiguration = exports.BaseViewFinderConfiguration = void 0;
(function (_BaseViewFinderConfiguration) {
  function from(source) {
    const _type = source._type;
    switch (_type) {
      case 'ViewFinderConfiguration':
        return new ViewFinderConfiguration(source);
      case 'PermanentViewFinderConfiguration':
        return new PermanentViewFinderConfiguration(source);
      default:
        throw new Error(`Unknown child class name: ${_type}`);
    }
  }
  _BaseViewFinderConfiguration.from = from;
})(BaseViewFinderConfiguration || (exports.BaseViewFinderConfiguration = BaseViewFinderConfiguration = {}));
/**
Configuration of the scanning interface’s viewfinder, allowing it to be shown or hidden as guidance to the user.
*/
class ViewFinderConfiguration {
  _type = 'ViewFinderConfiguration';
  /**
    The visual appearance of the viewfinder.
    */
  style = new FinderCorneredStyle({
    strokeColor: '?sbColorSurface',
    strokeWidth: 2.0,
    cornerRadius: 10.0
  });
  /**
    The color of the viewfinder overlay.
    
    Default is "?sbColorSurfaceLow"
    */
  overlayColor = '?sbColorSurfaceLow';
  /**
    The viewfinder's aspect ratio.
    */
  aspectRatio = new _Geometry.AspectRatio({
    width: 1.0,
    height: 1.0
  });
  /**
    The minimum insets of the viewfinder.
    */
  minimumInsets = new _Common.EdgeInsets({
    top: 16.0,
    left: 16.0,
    bottom: 16.0,
    right: 16.0
  });
  /**
    The preferred height of the viewfinder.
    
    Default is -1.0
    */
  preferredHeight = -1.0;
  /**
    Whether the viewfinder is visible.
    
    Default is true
    */
  visible = true;

  /** @param source {@displayType `DeepPartial<ViewFinderConfiguration>`} */
  constructor(source = {}) {
    if (source.style !== undefined) {
      this.style = FinderStyle.from(source.style);
    }
    if (source.overlayColor !== undefined) {
      this.overlayColor = source.overlayColor;
    }
    if (source.aspectRatio !== undefined) {
      this.aspectRatio = new _Geometry.AspectRatio(source.aspectRatio);
    }
    if (source.minimumInsets !== undefined) {
      this.minimumInsets = new _Common.EdgeInsets(source.minimumInsets);
    }
    if (source.preferredHeight !== undefined) {
      this.preferredHeight = source.preferredHeight;
    }
    if (source.visible !== undefined) {
      this.visible = source.visible;
    }
  }
}

/**
Configuration of the scanning interface’s viewfinder, which is always visible to guide the user.
*/
exports.ViewFinderConfiguration = ViewFinderConfiguration;
class PermanentViewFinderConfiguration {
  _type = 'PermanentViewFinderConfiguration';
  /**
    The visual appearance of the viewfinder.
    */
  style = new FinderCorneredStyle({
    strokeColor: '?sbColorSurface',
    strokeWidth: 2.0,
    cornerRadius: 10.0
  });
  /**
    The color of the viewfinder overlay.
    
    Default is "?sbColorSurfaceLow"
    */
  overlayColor = '?sbColorSurfaceLow';
  /**
    The viewfinder's aspect ratio.
    */
  aspectRatio = new _Geometry.AspectRatio({
    width: 1.0,
    height: 1.0
  });
  /**
    The minimum insets of the viewfinder.
    */
  minimumInsets = new _Common.EdgeInsets({
    top: 16.0,
    left: 16.0,
    bottom: 16.0,
    right: 16.0
  });
  /**
    The preferred height of the viewfinder.
    
    Default is -1.0
    */
  preferredHeight = -1.0;

  /** @param source {@displayType `DeepPartial<PermanentViewFinderConfiguration>`} */
  constructor(source = {}) {
    if (source.style !== undefined) {
      this.style = FinderStyle.from(source.style);
    }
    if (source.overlayColor !== undefined) {
      this.overlayColor = source.overlayColor;
    }
    if (source.aspectRatio !== undefined) {
      this.aspectRatio = new _Geometry.AspectRatio(source.aspectRatio);
    }
    if (source.minimumInsets !== undefined) {
      this.minimumInsets = new _Common.EdgeInsets(source.minimumInsets);
    }
    if (source.preferredHeight !== undefined) {
      this.preferredHeight = source.preferredHeight;
    }
  }
}

/**
The visual appearance of the viewfinder.
*/
exports.PermanentViewFinderConfiguration = PermanentViewFinderConfiguration;
/** @internal */
let FinderStyle = exports.FinderStyle = void 0;
(function (_FinderStyle) {
  function from(source) {
    const _type = source._type;
    switch (_type) {
      case 'FinderCorneredStyle':
        return new FinderCorneredStyle(source);
      case 'FinderStrokedStyle':
        return new FinderStrokedStyle(source);
      default:
        throw new Error(`Unknown child class name: ${_type}`);
    }
  }
  _FinderStyle.from = from;
})(FinderStyle || (exports.FinderStyle = FinderStyle = {}));
/**
A variant of the viewfinder displaying only the four corners of the scanning area.
*/
class FinderCorneredStyle {
  _type = 'FinderCorneredStyle';
  /**
    The color of the viewfinder corner's outlines.
    
    Default is "#FFFFFFFF"
    */
  strokeColor = '#FFFFFFFF';
  /**
    The width of the viewfinder corner's outlines.
    
    Default is 3.0
    */
  strokeWidth = 3.0;
  /**
    The radius of the viewfinder's corners.
    
    Default is 10.0
    */
  cornerRadius = 10.0;

  /** @param source {@displayType `DeepPartial<FinderCorneredStyle>`} */
  constructor(source = {}) {
    if (source.strokeColor !== undefined) {
      this.strokeColor = source.strokeColor;
    }
    if (source.strokeWidth !== undefined) {
      this.strokeWidth = source.strokeWidth;
    }
    if (source.cornerRadius !== undefined) {
      this.cornerRadius = source.cornerRadius;
    }
  }
}

/**
A variant of the viewfinder displaying a full outline of the scanning area.
*/
exports.FinderCorneredStyle = FinderCorneredStyle;
class FinderStrokedStyle {
  _type = 'FinderStrokedStyle';
  /**
    The color of the viewfinder corner's outlines.
    
    Default is "#FFFFFFFF"
    */
  strokeColor = '#FFFFFFFF';
  /**
    The width of the viewfinder corner's outlines.
    
    Default is 3.0
    */
  strokeWidth = 3.0;
  /**
    The radius of the viewfinder's corners.
    
    Default is 10.0
    */
  cornerRadius = 10.0;

  /** @param source {@displayType `DeepPartial<FinderStrokedStyle>`} */
  constructor(source = {}) {
    if (source.strokeColor !== undefined) {
      this.strokeColor = source.strokeColor;
    }
    if (source.strokeWidth !== undefined) {
      this.strokeWidth = source.strokeWidth;
    }
    if (source.cornerRadius !== undefined) {
      this.cornerRadius = source.cornerRadius;
    }
  }
}
exports.FinderStrokedStyle = FinderStrokedStyle;
