// react-native-scanbot-sdk/lib/module/analytics/Analytics.js
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from native/schemas/analytics/Analytics.yaml

import { ToJsonConfiguration } from '../utils/json/JsonSerializationTypes';
import { PartiallyConstructible } from '../utils/utils';

/**
The feature that is being tracked.
*/
export class AnalyticsFeature extends PartiallyConstructible {
  /** @param source {@displayType `DeepPartial<AnalyticsFeature>`} */
  constructor(source = {}) {
    super();
  }
  serialize(config = new ToJsonConfiguration()) {
    return {};
  }
}
(function (_AnalyticsFeature) {
  const documentScanner = _AnalyticsFeature.documentScanner = 'rtuui_document_scanner';
})(AnalyticsFeature || (AnalyticsFeature = {}));
/**
The category that is being tracked.
*/
export class AnalyticsCategory extends PartiallyConstructible {
  /** @param source {@displayType `DeepPartial<AnalyticsCategory>`} */
  constructor(source = {}) {
    super();
  }
  serialize(config = new ToJsonConfiguration()) {
    return {};
  }
}
(function (_AnalyticsCategory) {
  const acknowledge = _AnalyticsCategory.acknowledge = 'acknowledge';
})(AnalyticsCategory || (AnalyticsCategory = {}));
/**
The event that is being tracked.
*/
export class AnalyticsEvent extends PartiallyConstructible {
  /**
    The feature associated with the event.
    */

  /**
    The category of the event.
    */

  /**
    The name of the event.
    */

  /**
    The value associated with the event.
    */
  values = [];

  /** @param source {@displayType `DeepPartial<AnalyticsEvent>`} */
  constructor(source = {}) {
    super();
    if (source.feature !== undefined) {
      this.feature = source.feature;
    } else {
      throw new Error('feature must be present in constructor argument');
    }
    if (source.category !== undefined) {
      this.category = source.category;
    } else {
      throw new Error('category must be present in constructor argument');
    }
    if (source.name !== undefined) {
      this.name = source.name;
    } else {
      throw new Error('name must be present in constructor argument');
    }
    if (source.values !== undefined) {
      this.values = source.values.map(it => {
        return it;
      });
    }
  }
  serialize(config = new ToJsonConfiguration()) {
    return {
      feature: this.feature,
      category: this.category,
      name: this.name,
      values: this.values.map(it => {
        return it;
      })
    };
  }
}

/**
The name of the event.
*/
export class AnalyticsEventName extends PartiallyConstructible {
  /** @param source {@displayType `DeepPartial<AnalyticsEventName>`} */
  constructor(source = {}) {
    super();
  }
  serialize(config = new ToJsonConfiguration()) {
    return {};
  }
}
(function (_AnalyticsEventName) {
  const importedImageQualityResult = _AnalyticsEventName.importedImageQualityResult = 'imported_image_quality_result';
  const scannedImageQualityResult = _AnalyticsEventName.scannedImageQualityResult = 'scanned_image_quality_result';
  const proceedImage = _AnalyticsEventName.proceedImage = 'proceed_with_image';
  const retakeImage = _AnalyticsEventName.retakeImage = 'retake_image';
})(AnalyticsEventName || (AnalyticsEventName = {}));
/**
A predefined analytics event.
*/
export class AnalyticsEventFactory extends PartiallyConstructible {
  /** @param source {@displayType `DeepPartial<AnalyticsEventFactory>`} */
  constructor(source = {}) {
    super();
  }
  serialize(config = new ToJsonConfiguration()) {
    return {};
  }
}
(function (_AnalyticsEventFactory) {
  const scannedImageQualityResult = _AnalyticsEventFactory.scannedImageQualityResult = new AnalyticsEvent({
    feature: 'rtuui_document_scanner',
    category: 'acknowledge',
    name: 'scanned_image_quality_result'
  });
  const importedImageQualityResult = _AnalyticsEventFactory.importedImageQualityResult = new AnalyticsEvent({
    feature: 'rtuui_document_scanner',
    category: 'acknowledge',
    name: 'imported_image_quality_result'
  });
  const proceedWithImage = _AnalyticsEventFactory.proceedWithImage = new AnalyticsEvent({
    feature: 'rtuui_document_scanner',
    category: 'acknowledge',
    name: 'proceed_with_image'
  });
  const retakeImage = _AnalyticsEventFactory.retakeImage = new AnalyticsEvent({
    feature: 'rtuui_document_scanner',
    category: 'acknowledge',
    name: 'retake_image'
  });
})(AnalyticsEventFactory || (AnalyticsEventFactory = {}));
