// react-native-scanbot-sdk/lib/commonjs/ui_v2/mrz/MRZScannerScreenTextLocalization.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MrzScannerScreenTextLocalization = void 0;
var _utils = require("../../utils/utils");
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/mrz/MRZScannerScreenTextLocalization.yaml

/**
Configuration of the all strings for MRZ scanner screen.
*/
class MrzScannerScreenTextLocalization extends _utils.PartiallyConstructible {
  /**
    Title for the top bar.
    
    Default is "MRZ Scanner"
    */
  topBarTitle = 'MRZ Scanner';
  /**
    Cancel button text for the top bar.
    
    Default is "Cancel"
    */
  topBarCancelButton = 'Cancel';
  /**
    Text for the top user guidance caption.
    
    Default is "Scan your Identity Document"
    */
  topUserGuidance = 'Scan your Identity Document';
  /**
    Text for the user guidance caption below the finder view.
    
    Default is "Scan the MRZ"
    */
  finderViewUserGuidance = 'Scan the MRZ';
  /**
    Title for the introduction screen.
    
    Default is "How to scan an MRZ"
    */
  introScreenTitle = 'How to scan an MRZ';
  /**
    Start scanning button text for the introduction screen.
    
    Default is "Start Scanning"
    */
  introScreenDoneButton = 'Start Scanning';
  /**
    The text explanation for the introduction screen.
    
    Default is "The Machine Readable Zone (MRZ) is a special code on your ID document (such as a passport or ID card) that contains your personal information in a machine-readable format.\n\nTo scan it, simply hold your camera over the document, so that it aligns with the MRZ section. Once scanned, the data will be automatically processed, and you will be directed to the results screen.\n\nPress 'Start Scanning' to begin."
    */
  introScreenText = "The Machine Readable Zone (MRZ) is a special code on your ID document (such as a passport or ID card) that contains your personal information in a machine-readable format.\n\nTo scan it, simply hold your camera over the document, so that it aligns with the MRZ section. Once scanned, the data will be automatically processed, and you will be directed to the results screen.\n\nPress 'Start Scanning' to begin.";
  /**
    Caption for the success overlay.
    
    Default is "Scanned successfully"
    */
  completionOverlaySuccessMessage = 'Scanned successfully';
  /**
    Accessibility description for the 'open introduction screen' button in top bar.
    
    Default is "Open introduction screen."
    */
  accessibilityDescriptionOpenIntroScreenButton = 'Open introduction screen.';
  /**
    Accessibility description for the 'done'/'start scanning' button in the introduction screen.
    
    Default is "Start scanning"
    */
  accessibilityDescriptionIntroScreenDoneButton = 'Start scanning';
  /**
    Accessibility description for the 'cancel' button in the top bar.
    
    Default is "Cancel"
    */
  accessibilityDescriptionCancelButton = 'Cancel';
  /**
    Accessibility description for the flash button.
    
    Default is "Toggle flash"
    */
  accessibilityDescriptionFlashButton = 'Toggle flash';
  /**
    Accessibility description for the zoom button.
    
    Default is "Toggle camera zoom"
    */
  accessibilityDescriptionZoomButton = 'Toggle camera zoom';
  /**
    Accessibility description for the flip camera button.
    
    Default is "Flip camera"
    */
  accessibilityDescriptionFlipCameraButton = 'Flip camera';
  /**
    The title of the camera permission dialog.
    
    Default is "Camera permission denied!"
    */
  cameraPermissionEnableCameraTitle = 'Camera permission denied!';
  /**
    The explanation text of the camera permission dialog.
    
    Default is "Please allow the usage of the camera to start the scanning process."
    */
  cameraPermissionEnableCameraExplanation = 'Please allow the usage of the camera to start the scanning process.';
  /**
    The 'enable' button title of the camera permission dialog.
    
    Default is "Grant permission"
    */
  cameraPermissionEnableCameraButton = 'Grant permission';
  /**
    The 'close' button title of the camera permission dialog.
    
    Default is "Close"
    */
  cameraPermissionCloseButton = 'Close';
  /**
    The accessibility hint for the 'enable' button of the camera permission dialog.
    
    Default is "Tap to grant camera permission"
    */
  accessibilityDescriptionCameraPermissionEnableCameraButton = 'Tap to grant camera permission';
  /**
    The accessibility hint for the 'close' button of the camera permission dialog.
    
    Default is "Close screen without granting permission"
    */
  accessibilityDescriptionCameraPermissionCloseButton = 'Close screen without granting permission';

  /** @param source {@displayType `DeepPartial<MrzScannerScreenTextLocalization>`} */
  constructor(source = {}) {
    super();
    if (source.topBarTitle !== undefined) {
      this.topBarTitle = source.topBarTitle;
    }
    if (source.topBarCancelButton !== undefined) {
      this.topBarCancelButton = source.topBarCancelButton;
    }
    if (source.topUserGuidance !== undefined) {
      this.topUserGuidance = source.topUserGuidance;
    }
    if (source.finderViewUserGuidance !== undefined) {
      this.finderViewUserGuidance = source.finderViewUserGuidance;
    }
    if (source.introScreenTitle !== undefined) {
      this.introScreenTitle = source.introScreenTitle;
    }
    if (source.introScreenDoneButton !== undefined) {
      this.introScreenDoneButton = source.introScreenDoneButton;
    }
    if (source.introScreenText !== undefined) {
      this.introScreenText = source.introScreenText;
    }
    if (source.completionOverlaySuccessMessage !== undefined) {
      this.completionOverlaySuccessMessage = source.completionOverlaySuccessMessage;
    }
    if (source.accessibilityDescriptionOpenIntroScreenButton !== undefined) {
      this.accessibilityDescriptionOpenIntroScreenButton = source.accessibilityDescriptionOpenIntroScreenButton;
    }
    if (source.accessibilityDescriptionIntroScreenDoneButton !== undefined) {
      this.accessibilityDescriptionIntroScreenDoneButton = source.accessibilityDescriptionIntroScreenDoneButton;
    }
    if (source.accessibilityDescriptionCancelButton !== undefined) {
      this.accessibilityDescriptionCancelButton = source.accessibilityDescriptionCancelButton;
    }
    if (source.accessibilityDescriptionFlashButton !== undefined) {
      this.accessibilityDescriptionFlashButton = source.accessibilityDescriptionFlashButton;
    }
    if (source.accessibilityDescriptionZoomButton !== undefined) {
      this.accessibilityDescriptionZoomButton = source.accessibilityDescriptionZoomButton;
    }
    if (source.accessibilityDescriptionFlipCameraButton !== undefined) {
      this.accessibilityDescriptionFlipCameraButton = source.accessibilityDescriptionFlipCameraButton;
    }
    if (source.cameraPermissionEnableCameraTitle !== undefined) {
      this.cameraPermissionEnableCameraTitle = source.cameraPermissionEnableCameraTitle;
    }
    if (source.cameraPermissionEnableCameraExplanation !== undefined) {
      this.cameraPermissionEnableCameraExplanation = source.cameraPermissionEnableCameraExplanation;
    }
    if (source.cameraPermissionEnableCameraButton !== undefined) {
      this.cameraPermissionEnableCameraButton = source.cameraPermissionEnableCameraButton;
    }
    if (source.cameraPermissionCloseButton !== undefined) {
      this.cameraPermissionCloseButton = source.cameraPermissionCloseButton;
    }
    if (source.accessibilityDescriptionCameraPermissionEnableCameraButton !== undefined) {
      this.accessibilityDescriptionCameraPermissionEnableCameraButton = source.accessibilityDescriptionCameraPermissionEnableCameraButton;
    }
    if (source.accessibilityDescriptionCameraPermissionCloseButton !== undefined) {
      this.accessibilityDescriptionCameraPermissionCloseButton = source.accessibilityDescriptionCameraPermissionCloseButton;
    }
  }
}
exports.MrzScannerScreenTextLocalization = MrzScannerScreenTextLocalization;
