// react-native-scanbot-sdk/lib/commonjs/ui_v2/mrz/MRZScannerIntroScreenConfiguration.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MrzScannerIntroScreenConfiguration = exports.MrzScannerIntroImage = exports.MrzIntroNoImage = exports.MrzIntroDefaultImage = exports.MrzIntroCustomImage = void 0;
var _Common = require("../../ui_v2/common/Common");
var _utils = require("../../utils/utils");
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/mrz/MRZScannerIntroScreenConfiguration.yaml
/**
Determines the image for the introduction screen.
*/
/** @internal */
let MrzScannerIntroImage = exports.MrzScannerIntroImage = void 0;
(function (_MrzScannerIntroImage) {
  function From(source) {
    const _type = source._type;
    switch (_type) {
      case 'MrzIntroDefaultImage':
        return new MrzIntroDefaultImage(source);
      case 'MrzIntroNoImage':
        return new MrzIntroNoImage(source);
      case 'MrzIntroCustomImage':
        return new MrzIntroCustomImage(source);
      default:
        throw `Unknown child class name: ${_type}`;
    }
  }
  _MrzScannerIntroImage.From = From;
})(MrzScannerIntroImage || (exports.MrzScannerIntroImage = MrzScannerIntroImage = {}));
/**
No image for the introduction screen.
*/
class MrzIntroDefaultImage extends _utils.PartiallyConstructible {
  _type = 'MrzIntroDefaultImage';

  /** @param source {@displayType `DeepPartial<MrzIntroDefaultImage>`} */
  constructor(source = {}) {
    super();
  }
}

/**
No image for the introduction screen.
*/
exports.MrzIntroDefaultImage = MrzIntroDefaultImage;
class MrzIntroNoImage extends _utils.PartiallyConstructible {
  _type = 'MrzIntroNoImage';

  /** @param source {@displayType `DeepPartial<MrzIntroNoImage>`} */
  constructor(source = {}) {
    super();
  }
}

/**
A custom image for the introduction screen.
*/
exports.MrzIntroNoImage = MrzIntroNoImage;
class MrzIntroCustomImage extends _utils.PartiallyConstructible {
  _type = 'MrzIntroCustomImage';
  /**
    The web or file URI to the image.
    */

  /** @param source {@displayType `DeepPartial<MrzIntroCustomImage>`} */
  constructor(source = {}) {
    super();
    if (source.uri !== undefined) {
      this.uri = source.uri;
    } else {
      throw new Error('uri must be present in constructor argument');
    }
  }
}

/**
Configuration of the introduction screen for the MRZ scanner.
*/
exports.MrzIntroCustomImage = MrzIntroCustomImage;
class MrzScannerIntroScreenConfiguration extends _utils.PartiallyConstructible {
  /**
    The image for the introduction screen.
    */
  image = new MrzIntroDefaultImage({});
  /**
    The background color of the introduction screen.
    
    Default is "?sbColorSurface"
    */
  backgroundColor = '?sbColorSurface';
  /**
    The divider color of the introduction screen.
    
    Default is "?sbColorOutline"
    */
  dividerColor = '?sbColorOutline';
  /**
    The handle color of the introduction screen.
    
    Default is "?sbColorOutline"
    */
  handlerColor = '?sbColorOutline';
  /**
    Determines whether the introduction screen should automatically be shown or not when the scanning session starts.
    
    Default is false
    */
  showAutomatically = false;
  /**
    The title of the introduction screen, located in the top bar.
    */
  title = new _Common.StyledText({
    text: '?introScreenTitle',
    color: '?sbColorOnSurface'
  });
  /**
    The text explanation of the introduction screen.
    */
  explanation = new _Common.StyledText({
    text: '?introScreenText',
    color: '?sbColorOnSurface'
  });
  /**
    Configuration of the 'Done' / 'Start scanning' button.
    */
  doneButton = new _Common.ButtonConfiguration({
    text: '?introScreenDoneButton',
    accessibilityDescription: '?accessibilityDescriptionIntroScreenDoneButton',
    background: new _Common.BackgroundStyle({
      strokeColor: '#00000000',
      fillColor: '?sbColorPrimary',
      strokeWidth: 0.0
    }),
    foreground: new _Common.ForegroundStyle({
      color: '?sbColorOnPrimary'
    })
  });

  /** @param source {@displayType `DeepPartial<MrzScannerIntroScreenConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.image !== undefined) {
      this.image = MrzScannerIntroImage.From(source.image);
    }
    if (source.backgroundColor !== undefined) {
      this.backgroundColor = source.backgroundColor;
    }
    if (source.dividerColor !== undefined) {
      this.dividerColor = source.dividerColor;
    }
    if (source.handlerColor !== undefined) {
      this.handlerColor = source.handlerColor;
    }
    if (source.showAutomatically !== undefined) {
      this.showAutomatically = source.showAutomatically;
    }
    if (source.title !== undefined) {
      this.title = new _Common.StyledText(source.title);
    }
    if (source.explanation !== undefined) {
      this.explanation = new _Common.StyledText(source.explanation);
    }
    if (source.doneButton !== undefined) {
      this.doneButton = new _Common.ButtonConfiguration(source.doneButton);
    }
  }
}
exports.MrzScannerIntroScreenConfiguration = MrzScannerIntroScreenConfiguration;
