// react-native-scanbot-barcode-scanner-sdk/src/types/core/licensing/LicensingTypes.ts
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from core/schemas/LicensingTypes.yaml

import { ToJsonConfiguration } from '../common/JsonSerializationTypes';
import { DeepPartial } from '../utils/utils';

/**
Status of the license plate scanner.

- `OKAY`:
   License is valid and accepted.
- `TRIAL`:
   No license set yet. The SDK is in trial mode.
- `OKAY_EXPIRING_SOON`:
   License is valid but will expire soon. Make sure to update your license before it expires to ensure uninterrupted function.
- `FAILURE_NOT_SET`:
   No license set. On platforms that support trial mode, it means that the SDK's trial mode is over.
- `FAILURE_CORRUPTED`:
   No license active. The set license was unreadable or has an invalid format.
- `FAILURE_WRONG_OS`:
   No license active. The set license does not cover the current operating system.
- `FAILURE_APP_ID_MISMATCH`:
   No license active. The set license does not cover the current app's bundle identifier.
- `FAILURE_EXPIRED`:
   No license active. The set license is valid but it has expired.
- `FAILURE_SERVER`:
   No license active. The online license check failed.
- `FAILURE_VERSION`:
   No license active. The set license does not cover the current SDK version.
- `FAILURE_INACTIVE`:
   No license active. The set license is not yet active.
*/
export type LicenseStatus =
  /**
    License is valid and accepted.
    */
  | 'OKAY'
  /**
    No license set yet. The SDK is in trial mode.
    */
  | 'TRIAL'
  /**
    License is valid but will expire soon. Make sure to update your license before it expires to ensure uninterrupted function.
    */
  | 'OKAY_EXPIRING_SOON'
  /**
    No license set. On platforms that support trial mode, it means that the SDK's trial mode is over.
    */
  | 'FAILURE_NOT_SET'
  /**
    No license active. The set license was unreadable or has an invalid format.
    */
  | 'FAILURE_CORRUPTED'
  /**
    No license active. The set license does not cover the current operating system.
    */
  | 'FAILURE_WRONG_OS'
  /**
    No license active. The set license does not cover the current app's bundle identifier.
    */
  | 'FAILURE_APP_ID_MISMATCH'
  /**
    No license active. The set license is valid but it has expired.
    */
  | 'FAILURE_EXPIRED'
  /**
    No license active. The online license check failed.
    */
  | 'FAILURE_SERVER'
  /**
    No license active. The set license does not cover the current SDK version.
    */
  | 'FAILURE_VERSION'
  /**
    No license active. The set license is not yet active.
    */
  | 'FAILURE_INACTIVE';

/** @hidden */
export const LicenseStatusValues: ReadonlyArray<LicenseStatus> = [
  'OKAY',
  'TRIAL',
  'OKAY_EXPIRING_SOON',
  'FAILURE_NOT_SET',
  'FAILURE_CORRUPTED',
  'FAILURE_WRONG_OS',
  'FAILURE_APP_ID_MISMATCH',
  'FAILURE_EXPIRED',
  'FAILURE_SERVER',
  'FAILURE_VERSION',
  'FAILURE_INACTIVE',
] as const;

/**
Description of error that occurred during device deregistration.
*/
export class DeviceDeregistrationError {
  /**
    Sequential number of the device registration. Starts with 0. Each time a call to initializeLicense with a new license is made, the device ID.
    */
  public readonly deviceId: number;
  /**
    Error message describing the error.
    */
  public readonly errorMessage: string;

  /** @param source {@displayType `DeepPartial<DeviceDeregistrationError>`} */
  public constructor(source: DeepPartial<DeviceDeregistrationError> = {}) {
    if (source.deviceId !== undefined) {
      this.deviceId = source.deviceId;
    } else {
      throw new Error('deviceId must be present in constructor argument');
    }
    if (source.errorMessage !== undefined) {
      this.errorMessage = source.errorMessage;
    } else {
      throw new Error('errorMessage must be present in constructor argument');
    }
  }

  public serialize(
    config: ToJsonConfiguration = new ToJsonConfiguration()
  ): DeepPartial<DeviceDeregistrationError> {
    return {
      deviceId: this.deviceId,
      errorMessage: this.errorMessage,
    };
  }
}

/**
Information about the SDK license.
*/
export class LicenseInfo {
  /**
    License status.
    */
  public readonly status: LicenseStatus;
  /**
    True if the license status corresponds to a valid license.
    */
  public readonly isValid: boolean;
  /**
    Message describing the license status.
    */
  public readonly licenseStatusMessage: string;
  /**
    Expiration timestamp of the license or trial period in seconds since the UNIX epoch.
    */
  public readonly expirationTimestamp: number;
  /**
    Expiration date of the license or trial period in ISO 8601 date-time format.
    */
  public readonly expirationDateString: string;
  /**
    Additional error message in case of a license error.
    */
  public readonly errorMessage: string;
  /**
    True if an online license check is in progress.
    */
  public readonly onlineLicenseCheckInProgress: boolean;
  /**
    The currently active license. May be different from the license used to initialize
    the SDK if the online license check has completed and the server has sent
    an updated license.
    */
  public readonly license: string;
  /**
    Number of devices the license is valid for, if the license is floating.
    */
  public readonly devices: number | null = null;
  /**
    Time in milliseconds until the next online license check will be performed. -1 if no online license check is scheduled.
    */
  public readonly nextOnlineLicenseCheckInMs: number;
  /**
    List of errors that occurred during device deregistration. If for some device deregistration completed successfully, it will not be listed here. For each device at most one entry will be present.
    */
  public readonly deviceDeregistationErrors: DeviceDeregistrationError[];

  /** @param source {@displayType `DeepPartial<LicenseInfo>`} */
  public constructor(source: DeepPartial<LicenseInfo> = {}) {
    if (source.status !== undefined) {
      this.status = source.status;
    } else {
      throw new Error('status must be present in constructor argument');
    }
    if (source.isValid !== undefined) {
      this.isValid = source.isValid;
    } else {
      throw new Error('isValid must be present in constructor argument');
    }
    if (source.licenseStatusMessage !== undefined) {
      this.licenseStatusMessage = source.licenseStatusMessage;
    } else {
      throw new Error('licenseStatusMessage must be present in constructor argument');
    }
    if (source.expirationTimestamp !== undefined) {
      this.expirationTimestamp = source.expirationTimestamp;
    } else {
      throw new Error('expirationTimestamp must be present in constructor argument');
    }
    if (source.expirationDateString !== undefined) {
      this.expirationDateString = source.expirationDateString;
    } else {
      throw new Error('expirationDateString must be present in constructor argument');
    }
    if (source.errorMessage !== undefined) {
      this.errorMessage = source.errorMessage;
    } else {
      throw new Error('errorMessage must be present in constructor argument');
    }
    if (source.onlineLicenseCheckInProgress !== undefined) {
      this.onlineLicenseCheckInProgress = source.onlineLicenseCheckInProgress;
    } else {
      throw new Error('onlineLicenseCheckInProgress must be present in constructor argument');
    }
    if (source.license !== undefined) {
      this.license = source.license;
    } else {
      throw new Error('license must be present in constructor argument');
    }
    if (source.devices !== undefined) {
      this.devices = source.devices != null ? source.devices : null;
    }
    if (source.nextOnlineLicenseCheckInMs !== undefined) {
      this.nextOnlineLicenseCheckInMs = source.nextOnlineLicenseCheckInMs;
    } else {
      throw new Error('nextOnlineLicenseCheckInMs must be present in constructor argument');
    }
    if (source.deviceDeregistationErrors !== undefined) {
      this.deviceDeregistationErrors = source.deviceDeregistationErrors.map(
        (it: DeepPartial<DeviceDeregistrationError>) => {
          return new DeviceDeregistrationError(it);
        }
      );
    } else {
      throw new Error('deviceDeregistationErrors must be present in constructor argument');
    }
  }

  public serialize(
    config: ToJsonConfiguration = new ToJsonConfiguration()
  ): DeepPartial<LicenseInfo> {
    return {
      status: this.status,
      isValid: this.isValid,
      licenseStatusMessage: this.licenseStatusMessage,
      expirationTimestamp: this.expirationTimestamp,
      expirationDateString: this.expirationDateString,
      errorMessage: this.errorMessage,
      onlineLicenseCheckInProgress: this.onlineLicenseCheckInProgress,
      license: this.license,
      devices: this.devices != null ? this.devices : null,
      nextOnlineLicenseCheckInMs: this.nextOnlineLicenseCheckInMs,
      deviceDeregistationErrors: this.deviceDeregistationErrors.map((it: any) => {
        return it.serialize(config);
      }),
    };
  }
}

/**
License features.

- `BARCODE_SCANNER`:
   Barcode scanner.
- `CHECK_SCANNER`:
   Check scanner.
- `CREDIT_CARD_SCANNER`:
   Credit card scanner.
- `IMAGE_PROCESSING`:
   Image processing.
- `DOCUMENT_SCANNER`:
   Document scanner.
- `LICENSE_PLATE_SCANNER`:
   License plate scanner.
- `MEDICAL_CERTIFICATE_SCANNER`:
   Medical certificate scanner.
- `MRZ_SCANNER`:
   MRZ scanner.
- `OCR`:
   OCR.
- `TEXT_PATTERN_SCANNER`:
   Text pattern scanner.
- `PDF_CREATION`:
   PDF creation.
- `ID_CARD_SCANNER`:
   ID card scanner.
- `DRIVER_LICENSE_SCANNER`:
   Driver license scanner.
- `EHIC_SCANNER`:
   EHIC scanner.
- `DOCUMENT_CLASSIFICATION`:
   Document classification.
*/
export type Feature =
  /**
    Barcode scanner.
    */
  | 'BARCODE_SCANNER'
  /**
    Check scanner.
    */
  | 'CHECK_SCANNER'
  /**
    Credit card scanner.
    */
  | 'CREDIT_CARD_SCANNER'
  /**
    Image processing.
    */
  | 'IMAGE_PROCESSING'
  /**
    Document scanner.
    */
  | 'DOCUMENT_SCANNER'
  /**
    License plate scanner.
    */
  | 'LICENSE_PLATE_SCANNER'
  /**
    Medical certificate scanner.
    */
  | 'MEDICAL_CERTIFICATE_SCANNER'
  /**
    MRZ scanner.
    */
  | 'MRZ_SCANNER'
  /**
    OCR.
    */
  | 'OCR'
  /**
    Text pattern scanner.
    */
  | 'TEXT_PATTERN_SCANNER'
  /**
    PDF creation.
    */
  | 'PDF_CREATION'
  /**
    ID card scanner.
    */
  | 'ID_CARD_SCANNER'
  /**
    Driver license scanner.
    */
  | 'DRIVER_LICENSE_SCANNER'
  /**
    EHIC scanner.
    */
  | 'EHIC_SCANNER'
  /**
    Document classification.
    */
  | 'DOCUMENT_CLASSIFICATION';

/** @hidden */
export const FeatureValues: ReadonlyArray<Feature> = [
  'BARCODE_SCANNER',
  'CHECK_SCANNER',
  'CREDIT_CARD_SCANNER',
  'IMAGE_PROCESSING',
  'DOCUMENT_SCANNER',
  'LICENSE_PLATE_SCANNER',
  'MEDICAL_CERTIFICATE_SCANNER',
  'MRZ_SCANNER',
  'OCR',
  'TEXT_PATTERN_SCANNER',
  'PDF_CREATION',
  'ID_CARD_SCANNER',
  'DRIVER_LICENSE_SCANNER',
  'EHIC_SCANNER',
  'DOCUMENT_CLASSIFICATION',
] as const;
