// react-native-scanbot-sdk/src/document_data_extractor/DocumentDataExtractorTypes.ts
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from core/schemas/DocumentDataExtractorTypes.yaml

import {
  DocumentDataExtractorCommonConfiguration,
  DocumentDataExtractorConfigurationElement,
} from '../document_data_extractor/DocumentDataExtractorConfigurationTypes';
import { DocumentDetectionResult } from '../document_scanner/DocumentScannerTypes';
import { CommonFieldType } from '../documents/CommonFieldType';
import { GenericDocument } from '../documents/GenericDocument';
import { ResultAccumulationConfiguration } from '../frame_accumulation/FrameAccumulationTypes';
import { ImageRef } from '../imageRef/image';
import { ToJsonConfiguration } from '../utils/json/JsonSerializationTypes';
import { DeepPartial, PartiallyConstructible } from '../utils/utils';

/**
The status of the extraction process.

- `SUCCESS`:
   The document was extracted successfully.
- `ERROR_NOTHING_FOUND`:
   No document was detected.
- `ERROR_BAD_CROP`:
   A document was detected, but it was at an angle/distance that was too large.
- `ERROR_UNKNOWN_DOCUMENT`:
   A document was detected, but it was not extracted as a supported document.
- `ERROR_UNACCEPTABLE_DOCUMENT`:
   A document was detected as a supported document, but it was not part of the accepted documents.
- `ERROR_STILL_FOCUSSING`:
   The document is still being focused.
- `INCOMPLETE_VALIDATION`:
   All fields were extracted, but some of them failed validation.
*/
export type DocumentDataExtractionStatus =
  | 'SUCCESS'
  | 'ERROR_NOTHING_FOUND'
  | 'ERROR_BAD_CROP'
  | 'ERROR_UNKNOWN_DOCUMENT'
  | 'ERROR_UNACCEPTABLE_DOCUMENT'
  | 'ERROR_STILL_FOCUSSING'
  | 'INCOMPLETE_VALIDATION';

export const DocumentDataExtractionStatusValues: DocumentDataExtractionStatus[] = [
  'SUCCESS',
  'ERROR_NOTHING_FOUND',
  'ERROR_BAD_CROP',
  'ERROR_UNKNOWN_DOCUMENT',
  'ERROR_UNACCEPTABLE_DOCUMENT',
  'ERROR_STILL_FOCUSSING',
  'INCOMPLETE_VALIDATION',
];

/**
Contains the result of running the generic document extractor.
*/
export class DocumentDataExtractionResult extends PartiallyConstructible {
  /**
    The status of the extraction process.
    */
  public readonly status: DocumentDataExtractionStatus;
  /**
    The extracted document.
    */
  public readonly document: GenericDocument | null;
  /**
    Result of the document detection in the input image.
    */
  public readonly documentDetectionResult: DocumentDetectionResult;
  /**
    Crop of the document if it was detected.
    */
  public readonly croppedImage: ImageRef | null = null;

  /** @param source {@displayType `DeepPartial<DocumentDataExtractionResult>`} */
  public constructor(source: DeepPartial<DocumentDataExtractionResult> = {}) {
    super();
    if (source.status !== undefined) {
      this.status = source.status;
    } else {
      throw new Error('status must be present in constructor argument');
    }
    if (source.document !== undefined) {
      this.document = source.document != null ? new GenericDocument(source.document) : null;
    } else {
      throw new Error('document must be present in constructor argument');
    }
    if (source.documentDetectionResult !== undefined) {
      this.documentDetectionResult = new DocumentDetectionResult(source.documentDetectionResult);
    } else {
      throw new Error('documentDetectionResult must be present in constructor argument');
    }
    if (source.croppedImage !== undefined) {
      this.croppedImage = source.croppedImage != null ? ImageRef.From(source.croppedImage) : null;
    }
  }

  public async serialize(
    config: ToJsonConfiguration = new ToJsonConfiguration()
  ): Promise<DeepPartial<DocumentDataExtractionResult>> {
    return {
      status: this.status,
      document: this.document != null ? await this.document.serialize(config) : null,
      documentDetectionResult: this.documentDetectionResult.serialize(config),
      croppedImage: config.serializeImages
        ? this.croppedImage != null
          ? await this.croppedImage.serialize(config.imageSerializationMode)
          : null
        : undefined,
    };
  }

  private _released: Boolean = false;

  public release() {
    if (this._released) {
      return;
    }
    {
      if (this.document != null) {
        this.document.release();
      }
    }
    {
      if (this.croppedImage != null) {
        this.croppedImage.release();
      }
    }
    this._released = true;
  }

  public async encodeImages(): Promise<void> {
    if (this.document != null) {
      await this.document.encodeImages();
    }
    if (this.croppedImage != null) {
      await this.croppedImage.encodeInPlace();
    }
  }
}

/**
The extraction mode.

- `LIVE`:
   The extractor will attempt to accumulate results over multiple frames.
- `SINGLE_SHOT`:
   The extractor will only use the current frame and will spend additional time to ensure the best possible result.
*/
export type DocumentDataExtractionMode = 'LIVE' | 'SINGLE_SHOT';

export const DocumentDataExtractionModeValues: DocumentDataExtractionMode[] = [
  'LIVE',
  'SINGLE_SHOT',
];

/**
Parameters for the extraction process.
*/
export class DocumentDataFrameExtractionParameters extends PartiallyConstructible {
  /**
    The extraction mode.
    
    Default is LIVE
    */
  public mode: DocumentDataExtractionMode = 'LIVE';

  /** @param source {@displayType `DeepPartial<DocumentDataFrameExtractionParameters>`} */
  public constructor(source: DeepPartial<DocumentDataFrameExtractionParameters> = {}) {
    super();
    if (source.mode !== undefined) {
      this.mode = source.mode;
    }
  }
}

/**
Level of trust in the authenticity of the extracted document.

- `TRUSTED`:
   The extracted document is trusted, i.e. the scanner may assume that it's authentic. This allows certain optimizations that make extraction faster.
- `UNTRUSTED`:
   The extracted document is not trusted, i.e. its authenticity cannot be assumed. All fields will be read out independently and can be further verified. Extracting untrusted documents takes slightly longer, as the extractor is not allowed to employ certain time-saving optimizations.
*/
export type DocumentTrustMode = 'TRUSTED' | 'UNTRUSTED';

export const DocumentTrustModeValues: DocumentTrustMode[] = ['TRUSTED', 'UNTRUSTED'];

/**
Configuration for the document data extractor.
*/
export class DocumentDataExtractorConfiguration extends PartiallyConstructible {
  /**
    Configuration for how to accumulate results.
    */
  public resultAccumulationConfig: ResultAccumulationConfiguration =
    new ResultAccumulationConfiguration({});
  /**
    Normalized names of the fields to exclude from the result.
    */
  public fieldExcludeList: string[] = [];
  /**
    List of configuration elements for the document data extractor.
    */
  public configurations: DocumentDataExtractorConfigurationElement[] = [
    new DocumentDataExtractorCommonConfiguration({}),
  ];
  /**
    If true, crops of the detected documents and fields will be returned in the result.
    
    Default is false
    */
  public returnCrops: boolean = false;
  /**
    Level of trust in the authenticity of the extracted document.
    
    Default is TRUSTED
    */
  public documentTrustMode: DocumentTrustMode = 'TRUSTED';

  /** @param source {@displayType `DeepPartial<DocumentDataExtractorConfiguration>`} */
  public constructor(source: DeepPartial<DocumentDataExtractorConfiguration> = {}) {
    super();
    if (source.resultAccumulationConfig !== undefined) {
      this.resultAccumulationConfig = new ResultAccumulationConfiguration(
        source.resultAccumulationConfig
      );
    }
    if (source.fieldExcludeList !== undefined) {
      this.fieldExcludeList = source.fieldExcludeList.map((it: any) => {
        return it;
      });
    }
    if (source.configurations !== undefined) {
      this.configurations = source.configurations.map((it: any) => {
        return DocumentDataExtractorConfigurationElement.From(it);
      });
    }
    if (source.returnCrops !== undefined) {
      this.returnCrops = source.returnCrops;
    }
    if (source.documentTrustMode !== undefined) {
      this.documentTrustMode = source.documentTrustMode;
    }
  }
}

/**
The status of the consistency verification.

- `OK`:
   All values that correspond to the same logical field are exactly equal.
- `OK_PUNCTUATION_UPPER_CASE_CLEANED`:
   All values that correspond to the same logical field are equal after punctuation is removed and all characters are converted to upper case.
- `OK_DEBURRED`:
   All values that correspond to the same logical field are equal after punctuation is removed, all characters are converted to upper case and diacritics are replaced with their base characters.
- `OK_DEBURRED_EXTENDED`:
   Same as OK_DEBURRED but allows replacing diacritics not only with their base characters but also with some more complex combinations. E.g., according to ICAO 9303 MRZ standard Ü can be replaced with U or UE or UXX.
- `OK_TRUNCATED`:
   In addition to OK_DEBURRED_EXTENDED, allows one string to be a prefix of another string. Correctly works with complex diacritics replacements. In particular, if one string is MUEL and another is MÜLLER, then they are considered equal up to truncation.
- `INCONSISTENT`:
   The values that correspond to the same logical field are inconsistent.
*/
export type ValueConsistency =
  | 'OK'
  | 'OK_PUNCTUATION_UPPER_CASE_CLEANED'
  | 'OK_DEBURRED'
  | 'OK_DEBURRED_EXTENDED'
  | 'OK_TRUNCATED'
  | 'INCONSISTENT';

export const ValueConsistencyValues: ValueConsistency[] = [
  'OK',
  'OK_PUNCTUATION_UPPER_CASE_CLEANED',
  'OK_DEBURRED',
  'OK_DEBURRED_EXTENDED',
  'OK_TRUNCATED',
  'INCONSISTENT',
];

/**
Report of the field consistency verification.
*/
export class FieldConsistencyVerificationReport extends PartiallyConstructible {
  /**
    Logical field that is being verified.
    */
  public readonly fieldType: CommonFieldType;
  /**
    The status of the consistency verification.
    */
  public readonly consistency: ValueConsistency;
  /**
    Full names of the fields that were considered during the verification.
    */
  public readonly fieldFullNames: string[];

  /** @param source {@displayType `DeepPartial<FieldConsistencyVerificationReport>`} */
  public constructor(source: DeepPartial<FieldConsistencyVerificationReport> = {}) {
    super();
    if (source.fieldType !== undefined) {
      this.fieldType = source.fieldType;
    } else {
      throw new Error('fieldType must be present in constructor argument');
    }
    if (source.consistency !== undefined) {
      this.consistency = source.consistency;
    } else {
      throw new Error('consistency must be present in constructor argument');
    }
    if (source.fieldFullNames !== undefined) {
      this.fieldFullNames = source.fieldFullNames.map((it: any) => {
        return it;
      });
    } else {
      throw new Error('fieldFullNames must be present in constructor argument');
    }
  }

  public serialize(
    config: ToJsonConfiguration = new ToJsonConfiguration()
  ): DeepPartial<FieldConsistencyVerificationReport> {
    return {
      fieldType: this.fieldType,
      consistency: this.consistency,
      fieldFullNames: this.fieldFullNames.map((it: any) => {
        return it;
      }),
    };
  }
}

/**
Report of the document consistency verification. Shows whether information in the fields that correspond to the same logical field is consistent.
*/
export class ConsistencyVerificationReport extends PartiallyConstructible {
  /**
    Overall status of the consistency verification. Corresponds to the most severe status of all fields.
    */
  public readonly overallConsistency: ValueConsistency;
  /**
    List of reports for each field that was verified.
    */
  public readonly fieldDetails: FieldConsistencyVerificationReport[];

  /** @param source {@displayType `DeepPartial<ConsistencyVerificationReport>`} */
  public constructor(source: DeepPartial<ConsistencyVerificationReport> = {}) {
    super();
    if (source.overallConsistency !== undefined) {
      this.overallConsistency = source.overallConsistency;
    } else {
      throw new Error('overallConsistency must be present in constructor argument');
    }
    if (source.fieldDetails !== undefined) {
      this.fieldDetails = source.fieldDetails.map((it: any) => {
        return new FieldConsistencyVerificationReport(it);
      });
    } else {
      throw new Error('fieldDetails must be present in constructor argument');
    }
  }

  public serialize(
    config: ToJsonConfiguration = new ToJsonConfiguration()
  ): DeepPartial<ConsistencyVerificationReport> {
    return {
      overallConsistency: this.overallConsistency,
      fieldDetails: this.fieldDetails.map((it: any) => {
        return it.serialize(config);
      }),
    };
  }
}

/**
The result of the value semantics verification.

- `SKIPPED`:
   Verification of the value semantics was skipped. For example, because the field is empty.
- `VALID`:
   No errors were detected during semantics verification. For example, the date of birth is in the past, the date of expiry is in the future.
- `INVALID`:
   Semantically invalid value. For example, the date of birth is in the future, the date of expiry is in the past.
*/
export type ValueSemantics = 'SKIPPED' | 'VALID' | 'INVALID';

export const ValueSemanticsValues: ValueSemantics[] = ['SKIPPED', 'VALID', 'INVALID'];

/**
Report of the field value semantics verification.
*/
export class FieldValueSemantics extends PartiallyConstructible {
  /**
    Full name of the field that was considered during the verification.
    */
  public readonly fieldFullName: string;
  /**
    Shows whether the data in the field is semantically invalid, e.g. the date of birth is in the future.
    */
  public readonly invalidSemantics: boolean;

  /** @param source {@displayType `DeepPartial<FieldValueSemantics>`} */
  public constructor(source: DeepPartial<FieldValueSemantics> = {}) {
    super();
    if (source.fieldFullName !== undefined) {
      this.fieldFullName = source.fieldFullName;
    } else {
      throw new Error('fieldFullName must be present in constructor argument');
    }
    if (source.invalidSemantics !== undefined) {
      this.invalidSemantics = source.invalidSemantics;
    } else {
      throw new Error('invalidSemantics must be present in constructor argument');
    }
  }

  public serialize(
    config: ToJsonConfiguration = new ToJsonConfiguration()
  ): DeepPartial<FieldValueSemantics> {
    return {
      fieldFullName: this.fieldFullName,
      invalidSemantics: this.invalidSemantics,
    };
  }
}

/**
Report of a logical field value semantics verification. If some logical field, e.g. expiry date is present in multiple places in the document, it can occur that some entries are semantically correct and some are semantically incorrect. This report shows the overall status of the logical field value semantics verification and details for each field that was verified.
*/
export class CompositeFieldValueSemantics extends PartiallyConstructible {
  /**
    Overall status of the logical value semantics verification. Corresponds to the most severe status of all fields, i.e. if at least one field is semantically incorrect, the overall status is semantically incorrect. If no fields are present, the overall status is SKIPPED.
    */
  public readonly overallFieldValueSemantics: ValueSemantics;
  /**
    List of value semantics for each field corresponding to the same logical field.
    */
  public readonly fieldDetails: FieldValueSemantics[];

  /** @param source {@displayType `DeepPartial<CompositeFieldValueSemantics>`} */
  public constructor(source: DeepPartial<CompositeFieldValueSemantics> = {}) {
    super();
    if (source.overallFieldValueSemantics !== undefined) {
      this.overallFieldValueSemantics = source.overallFieldValueSemantics;
    } else {
      throw new Error('overallFieldValueSemantics must be present in constructor argument');
    }
    if (source.fieldDetails !== undefined) {
      this.fieldDetails = source.fieldDetails.map((it: any) => {
        return new FieldValueSemantics(it);
      });
    } else {
      throw new Error('fieldDetails must be present in constructor argument');
    }
  }

  public serialize(
    config: ToJsonConfiguration = new ToJsonConfiguration()
  ): DeepPartial<CompositeFieldValueSemantics> {
    return {
      overallFieldValueSemantics: this.overallFieldValueSemantics,
      fieldDetails: this.fieldDetails.map((it: any) => {
        return it.serialize(config);
      }),
    };
  }
}

/**
Report of the document value semantics verification.
*/
export class SemanticsVerificationReport extends PartiallyConstructible {
  /**
    Overall status of the value semantics verification. Corresponds to the most severe status of all fields, i.e. if at least one field is semantically incorrect, the overall status is semantically incorrect. If some verifications are skipped, they do not impact the overall status.
    */
  public readonly overallInvalidSemantics: boolean;
  /**
    Status of the expiry date semantics verification. Shows whether the expiry date in the document is in the future.
    */
  public readonly expiryDateSemantics: CompositeFieldValueSemantics;
  /**
    Status of the issue date semantics verification. Shows whether the issue date in the document is in the past.
    */
  public readonly issueDateSemantics: CompositeFieldValueSemantics;
  /**
    Status of the birth date semantics verification. Shows whether the birth date in the document is in the past.
    */
  public readonly birthDateSemantics: CompositeFieldValueSemantics;
  /**
    Result of verification whether the issue date is before the expiry date.
    */
  public readonly issueDateBeforeExpiryDate: ValueSemantics;
  /**
    Result of verification whether the birth date is before the issue date.
    */
  public readonly birthDateBeforeIssueDate: ValueSemantics;
  /**
    Result of verification whether the birth date is before the expiry date.
    */
  public readonly birthDateBeforeExpiryDate: ValueSemantics;

  /** @param source {@displayType `DeepPartial<SemanticsVerificationReport>`} */
  public constructor(source: DeepPartial<SemanticsVerificationReport> = {}) {
    super();
    if (source.overallInvalidSemantics !== undefined) {
      this.overallInvalidSemantics = source.overallInvalidSemantics;
    } else {
      throw new Error('overallInvalidSemantics must be present in constructor argument');
    }
    if (source.expiryDateSemantics !== undefined) {
      this.expiryDateSemantics = new CompositeFieldValueSemantics(source.expiryDateSemantics);
    } else {
      throw new Error('expiryDateSemantics must be present in constructor argument');
    }
    if (source.issueDateSemantics !== undefined) {
      this.issueDateSemantics = new CompositeFieldValueSemantics(source.issueDateSemantics);
    } else {
      throw new Error('issueDateSemantics must be present in constructor argument');
    }
    if (source.birthDateSemantics !== undefined) {
      this.birthDateSemantics = new CompositeFieldValueSemantics(source.birthDateSemantics);
    } else {
      throw new Error('birthDateSemantics must be present in constructor argument');
    }
    if (source.issueDateBeforeExpiryDate !== undefined) {
      this.issueDateBeforeExpiryDate = source.issueDateBeforeExpiryDate;
    } else {
      throw new Error('issueDateBeforeExpiryDate must be present in constructor argument');
    }
    if (source.birthDateBeforeIssueDate !== undefined) {
      this.birthDateBeforeIssueDate = source.birthDateBeforeIssueDate;
    } else {
      throw new Error('birthDateBeforeIssueDate must be present in constructor argument');
    }
    if (source.birthDateBeforeExpiryDate !== undefined) {
      this.birthDateBeforeExpiryDate = source.birthDateBeforeExpiryDate;
    } else {
      throw new Error('birthDateBeforeExpiryDate must be present in constructor argument');
    }
  }

  public serialize(
    config: ToJsonConfiguration = new ToJsonConfiguration()
  ): DeepPartial<SemanticsVerificationReport> {
    return {
      overallInvalidSemantics: this.overallInvalidSemantics,
      expiryDateSemantics: this.expiryDateSemantics.serialize(config),
      issueDateSemantics: this.issueDateSemantics.serialize(config),
      birthDateSemantics: this.birthDateSemantics.serialize(config),
      issueDateBeforeExpiryDate: this.issueDateBeforeExpiryDate,
      birthDateBeforeIssueDate: this.birthDateBeforeIssueDate,
      birthDateBeforeExpiryDate: this.birthDateBeforeExpiryDate,
    };
  }
}

/**
Report of the suspicious data field verification.
*/
export class FieldSuspiciousDataVerificationReport extends PartiallyConstructible {
  /**
    Full name of the field that was considered during the verification.
    */
  public readonly fieldFullName: string;
  /**
    Shows whether the data in the field is suspicious, e.g. contains word "Sample" in the name.
    */
  public readonly isSuspicious: boolean;

  /** @param source {@displayType `DeepPartial<FieldSuspiciousDataVerificationReport>`} */
  public constructor(source: DeepPartial<FieldSuspiciousDataVerificationReport> = {}) {
    super();
    if (source.fieldFullName !== undefined) {
      this.fieldFullName = source.fieldFullName;
    } else {
      throw new Error('fieldFullName must be present in constructor argument');
    }
    if (source.isSuspicious !== undefined) {
      this.isSuspicious = source.isSuspicious;
    } else {
      throw new Error('isSuspicious must be present in constructor argument');
    }
  }

  public serialize(
    config: ToJsonConfiguration = new ToJsonConfiguration()
  ): DeepPartial<FieldSuspiciousDataVerificationReport> {
    return {
      fieldFullName: this.fieldFullName,
      isSuspicious: this.isSuspicious,
    };
  }
}

/**
Report of the suspicious data document verification.
*/
export class SuspiciousDataVerificationReport extends PartiallyConstructible {
  /**
    Overall status of the suspicious data verification. Corresponds to the most severe status of all fields, i.e. if at least one field is suspicious, the overall status is suspicious.
    */
  public readonly overallSuspiciousData: boolean;
  /**
    List of reports for each field that was verified.
    */
  public readonly fieldDetails: FieldSuspiciousDataVerificationReport[];

  /** @param source {@displayType `DeepPartial<SuspiciousDataVerificationReport>`} */
  public constructor(source: DeepPartial<SuspiciousDataVerificationReport> = {}) {
    super();
    if (source.overallSuspiciousData !== undefined) {
      this.overallSuspiciousData = source.overallSuspiciousData;
    } else {
      throw new Error('overallSuspiciousData must be present in constructor argument');
    }
    if (source.fieldDetails !== undefined) {
      this.fieldDetails = source.fieldDetails.map((it: any) => {
        return new FieldSuspiciousDataVerificationReport(it);
      });
    } else {
      throw new Error('fieldDetails must be present in constructor argument');
    }
  }

  public serialize(
    config: ToJsonConfiguration = new ToJsonConfiguration()
  ): DeepPartial<SuspiciousDataVerificationReport> {
    return {
      overallSuspiciousData: this.overallSuspiciousData,
      fieldDetails: this.fieldDetails.map((it: any) => {
        return it.serialize(config);
      }),
    };
  }
}

/**
Report which shows whether the data in the field that can be validated is valid.
*/
export class FieldInvalidDataVerificationReport extends PartiallyConstructible {
  /**
    Full name of the field that was considered during the verification.
    */
  public readonly fieldFullName: string;
  /**
    Shows whether the data in the field is valid, e.g. if MRZ check digits are correct.
    */
  public readonly isInvalid: boolean;

  /** @param source {@displayType `DeepPartial<FieldInvalidDataVerificationReport>`} */
  public constructor(source: DeepPartial<FieldInvalidDataVerificationReport> = {}) {
    super();
    if (source.fieldFullName !== undefined) {
      this.fieldFullName = source.fieldFullName;
    } else {
      throw new Error('fieldFullName must be present in constructor argument');
    }
    if (source.isInvalid !== undefined) {
      this.isInvalid = source.isInvalid;
    } else {
      throw new Error('isInvalid must be present in constructor argument');
    }
  }

  public serialize(
    config: ToJsonConfiguration = new ToJsonConfiguration()
  ): DeepPartial<FieldInvalidDataVerificationReport> {
    return {
      fieldFullName: this.fieldFullName,
      isInvalid: this.isInvalid,
    };
  }
}

/**
Report of the field validation verification.
*/
export class InvalidDataVerificationReport extends PartiallyConstructible {
  /**
    Overall status of the document verification. Corresponds to the most severe status of all fields, i.e. if at least one field is invalid, the overall status is invalid.
    */
  public readonly overallIsInvalid: boolean;
  /**
    List of reports for each field that was verified.
    */
  public readonly fieldDetails: FieldInvalidDataVerificationReport[];

  /** @param source {@displayType `DeepPartial<InvalidDataVerificationReport>`} */
  public constructor(source: DeepPartial<InvalidDataVerificationReport> = {}) {
    super();
    if (source.overallIsInvalid !== undefined) {
      this.overallIsInvalid = source.overallIsInvalid;
    } else {
      throw new Error('overallIsInvalid must be present in constructor argument');
    }
    if (source.fieldDetails !== undefined) {
      this.fieldDetails = source.fieldDetails.map((it: any) => {
        return new FieldInvalidDataVerificationReport(it);
      });
    } else {
      throw new Error('fieldDetails must be present in constructor argument');
    }
  }

  public serialize(
    config: ToJsonConfiguration = new ToJsonConfiguration()
  ): DeepPartial<InvalidDataVerificationReport> {
    return {
      overallIsInvalid: this.overallIsInvalid,
      fieldDetails: this.fieldDetails.map((it: any) => {
        return it.serialize(config);
      }),
    };
  }
}

/**
Overall status of the document verification.

- `OK`:
   All the verifications passed successfully.
- `OK_WITH_WARNINGS`:
   No verifications failed, but some of them returned warnings.
- `INVALID`:
   At least one verification failed.
*/
export type DocumentVerificationStatus = 'OK' | 'OK_WITH_WARNINGS' | 'INVALID';

export const DocumentVerificationStatusValues: DocumentVerificationStatus[] = [
  'OK',
  'OK_WITH_WARNINGS',
  'INVALID',
];

/**
Report of the document verification.
*/
export class DocumentVerificationReport extends PartiallyConstructible {
  /**
    Overall status of the document verification.
    */
  public readonly overallVerificationStatus: DocumentVerificationStatus;
  /**
    Report of the document consistency verification. Shows whether information in the fields that correspond to the same logical field and is present in different parts of the document is consistent.
    */
  public readonly consistencyVerificationReport: ConsistencyVerificationReport;
  /**
    Report of the semantics verification. Shows whether the values in the fields are semantically correct.
    */
  public readonly semanticsVerificationReport: SemanticsVerificationReport;
  /**
    Shows whether the data in the document is suspicious, e.g. contains word "Sample" in the name.
    */
  public readonly suspiciousDataVerificationReport: SuspiciousDataVerificationReport;
  /**
    Some fields in the documents allow validation, for example, MRZ contains check digits to verify the data. This report shows whether there are any invalid fields in the document.
    */
  public readonly invalidDataVerificationReport: InvalidDataVerificationReport;

  /** @param source {@displayType `DeepPartial<DocumentVerificationReport>`} */
  public constructor(source: DeepPartial<DocumentVerificationReport> = {}) {
    super();
    if (source.overallVerificationStatus !== undefined) {
      this.overallVerificationStatus = source.overallVerificationStatus;
    } else {
      throw new Error('overallVerificationStatus must be present in constructor argument');
    }
    if (source.consistencyVerificationReport !== undefined) {
      this.consistencyVerificationReport = new ConsistencyVerificationReport(
        source.consistencyVerificationReport
      );
    } else {
      throw new Error('consistencyVerificationReport must be present in constructor argument');
    }
    if (source.semanticsVerificationReport !== undefined) {
      this.semanticsVerificationReport = new SemanticsVerificationReport(
        source.semanticsVerificationReport
      );
    } else {
      throw new Error('semanticsVerificationReport must be present in constructor argument');
    }
    if (source.suspiciousDataVerificationReport !== undefined) {
      this.suspiciousDataVerificationReport = new SuspiciousDataVerificationReport(
        source.suspiciousDataVerificationReport
      );
    } else {
      throw new Error('suspiciousDataVerificationReport must be present in constructor argument');
    }
    if (source.invalidDataVerificationReport !== undefined) {
      this.invalidDataVerificationReport = new InvalidDataVerificationReport(
        source.invalidDataVerificationReport
      );
    } else {
      throw new Error('invalidDataVerificationReport must be present in constructor argument');
    }
  }

  public serialize(
    config: ToJsonConfiguration = new ToJsonConfiguration()
  ): DeepPartial<DocumentVerificationReport> {
    return {
      overallVerificationStatus: this.overallVerificationStatus,
      consistencyVerificationReport: this.consistencyVerificationReport.serialize(config),
      semanticsVerificationReport: this.semanticsVerificationReport.serialize(config),
      suspiciousDataVerificationReport: this.suspiciousDataVerificationReport.serialize(config),
      invalidDataVerificationReport: this.invalidDataVerificationReport.serialize(config),
    };
  }
}
