// react-native-scanbot-sdk/lib/commonjs/ui_v2/common/ActionBarConfiguration.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ActionBarConfiguration = void 0;
var _Common = require("../../ui_v2/common/Common");
var _utils = require("../../utils/utils");
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/common/ActionBarConfiguration.yaml

/**
Configuration of the buttons available in the action bar.
*/
class ActionBarConfiguration extends _utils.PartiallyConstructible {
  /**
    Configuration of the 'flash' button.
    */
  flashButton = new _Common.RoundButton({
    visible: true,
    accessibilityDescription: '?accessibilityDescriptionFlashButton',
    backgroundColor: '?sbColorSurfaceHigh',
    foregroundColor: '?sbColorOnPrimary',
    activeBackgroundColor: '?sbColorWarning',
    activeForegroundColor: '#1C1B1F'
  });
  /**
    Configuration of the 'zoom' button.
    */
  zoomButton = new _Common.RoundButton({
    visible: true,
    accessibilityDescription: '?accessibilityDescriptionZoomButton',
    backgroundColor: '?sbColorSurfaceHigh',
    foregroundColor: '?sbColorOnPrimary',
    activeBackgroundColor: '?sbColorSurfaceHigh',
    activeForegroundColor: '?sbColorOnPrimary'
  });
  /**
    Configuration of the 'flip camera' button.
    */
  flipCameraButton = new _Common.RoundButton({
    visible: true,
    accessibilityDescription: '?accessibilityDescriptionFlipCameraButton',
    backgroundColor: '?sbColorSurfaceHigh',
    foregroundColor: '?sbColorOnPrimary',
    activeBackgroundColor: '?sbColorWarning',
    activeForegroundColor: '#1C1B1F'
  });

  /** @param source {@displayType `DeepPartial<ActionBarConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.flashButton !== undefined) {
      this.flashButton = new _Common.RoundButton(source.flashButton);
    }
    if (source.zoomButton !== undefined) {
      this.zoomButton = new _Common.RoundButton(source.zoomButton);
    }
    if (source.flipCameraButton !== undefined) {
      this.flipCameraButton = new _Common.RoundButton(source.flipCameraButton);
    }
  }
}
exports.ActionBarConfiguration = ActionBarConfiguration;
