// react-native-scanbot-sdk/lib/module/ui_v2/creditcard/CreditCardScannerIntroScreenConfiguration.js
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from rtu-ui-v2/schemas/creditcard/CreditCardScannerIntroScreenConfiguration.yaml

import { BackgroundStyle, ButtonConfiguration, ForegroundStyle, StyledText } from '../../ui_v2/common/Common';
import { PartiallyConstructible } from '../../utils/utils';

/**
The image for the introduction screen of a credit card scanner screen.
*/
/** @internal */
export let CreditCardScannerIntroImage;
(function (_CreditCardScannerIntroImage) {
  function From(source) {
    const _type = source._type;
    switch (_type) {
      case 'CreditCardIntroOneSideImage':
        return new CreditCardIntroOneSideImage(source);
      case 'CreditCardIntroTwoSidesImage':
        return new CreditCardIntroTwoSidesImage(source);
      case 'CreditCardNoImage':
        return new CreditCardNoImage(source);
      case 'CreditCardIntroCustomImage':
        return new CreditCardIntroCustomImage(source);
      default:
        throw `Unknown child class name: ${_type}`;
    }
  }
  _CreditCardScannerIntroImage.From = From;
})(CreditCardScannerIntroImage || (CreditCardScannerIntroImage = {}));
/**
The image for the introduction screen with one side of a credit card.
*/
export class CreditCardIntroOneSideImage extends PartiallyConstructible {
  _type = 'CreditCardIntroOneSideImage';

  /** @param source {@displayType `DeepPartial<CreditCardIntroOneSideImage>`} */
  constructor(source = {}) {
    super();
  }
}

/**
The image for the introduction screen with two sides of a credit card.
*/
export class CreditCardIntroTwoSidesImage extends PartiallyConstructible {
  _type = 'CreditCardIntroTwoSidesImage';

  /** @param source {@displayType `DeepPartial<CreditCardIntroTwoSidesImage>`} */
  constructor(source = {}) {
    super();
  }
}

/**
No image for the introduction screen.
*/
export class CreditCardNoImage extends PartiallyConstructible {
  _type = 'CreditCardNoImage';

  /** @param source {@displayType `DeepPartial<CreditCardNoImage>`} */
  constructor(source = {}) {
    super();
  }
}

/**
A custom image for the introduction screen.
*/
export class CreditCardIntroCustomImage extends PartiallyConstructible {
  _type = 'CreditCardIntroCustomImage';
  /**
    The web or file URI to the image.
    */

  /** @param source {@displayType `DeepPartial<CreditCardIntroCustomImage>`} */
  constructor(source = {}) {
    super();
    if (source.uri !== undefined) {
      this.uri = source.uri;
    } else {
      throw new Error('uri must be present in constructor argument');
    }
  }
}

/**
Configuration of the introduction screen for the credit card scanner.
*/
export class CreditCardScannerIntroScreenConfiguration extends PartiallyConstructible {
  /**
    The image for the introduction screen.
    */
  image = new CreditCardIntroOneSideImage({});
  /**
    The background color of the introduction screen.
    
    Default is "?sbColorSurface"
    */
  backgroundColor = '?sbColorSurface';
  /**
    The divider color of the introduction screen.
    
    Default is "?sbColorOutline"
    */
  dividerColor = '?sbColorOutline';
  /**
    The handle color of the introduction screen.
    
    Default is "?sbColorOutline"
    */
  handlerColor = '?sbColorOutline';
  /**
    Determines whether the introduction screen should automatically be shown or not when the scanning session starts.
    
    Default is false
    */
  showAutomatically = false;
  /**
    The title of the introduction screen, located in the top bar.
    */
  title = new StyledText({
    text: '?introScreenTitle',
    color: '?sbColorOnSurface'
  });
  /**
    The text explanation of the introduction screen.
    */
  explanation = new StyledText({
    text: '?introScreenText',
    color: '?sbColorOnSurface'
  });
  /**
    Configuration of the 'Done' / 'Start scanning' button.
    */
  doneButton = new ButtonConfiguration({
    text: '?introScreenDoneButton',
    accessibilityDescription: '?accessibilityDescriptionIntroScreenDoneButton',
    background: new BackgroundStyle({
      strokeColor: '#00000000',
      fillColor: '?sbColorPrimary',
      strokeWidth: 0.0
    }),
    foreground: new ForegroundStyle({
      color: '?sbColorOnPrimary'
    })
  });

  /** @param source {@displayType `DeepPartial<CreditCardScannerIntroScreenConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.image !== undefined) {
      this.image = CreditCardScannerIntroImage.From(source.image);
    }
    if (source.backgroundColor !== undefined) {
      this.backgroundColor = source.backgroundColor;
    }
    if (source.dividerColor !== undefined) {
      this.dividerColor = source.dividerColor;
    }
    if (source.handlerColor !== undefined) {
      this.handlerColor = source.handlerColor;
    }
    if (source.showAutomatically !== undefined) {
      this.showAutomatically = source.showAutomatically;
    }
    if (source.title !== undefined) {
      this.title = new StyledText(source.title);
    }
    if (source.explanation !== undefined) {
      this.explanation = new StyledText(source.explanation);
    }
    if (source.doneButton !== undefined) {
      this.doneButton = new ButtonConfiguration(source.doneButton);
    }
  }
}
