// react-native-scanbot-sdk/lib/commonjs/document_data_extractor/DocumentDataExtractorConfigurationTypes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MrzFallbackConfiguration = exports.GlareDetectionConfiguration = exports.EuropeanHealthInsuranceCardIssuingCountryValues = exports.EuropeanHealthInsuranceCardConfiguration = exports.DocumentDataExtractorConfigurationElement = exports.DocumentDataExtractorCommonConfiguration = exports.DeAddressParserConfiguration = exports.DateValidationConfiguration = exports.AutoFocusConfiguration = void 0;
var _utils = require("../utils/utils");
/// Auto-generated with ScanbotSDKCodegenV3. Modifications will be overwritten.
/// Generated from core/schemas/DocumentDataExtractorConfigurationTypes.yaml
/**
Base class for all generic document configuration elements.
*/
/** @internal */
let DocumentDataExtractorConfigurationElement = exports.DocumentDataExtractorConfigurationElement = void 0;
(function (_DocumentDataExtractorConfigurationElement) {
  function From(source) {
    const _type = source._type;
    switch (_type) {
      case 'DateValidationConfiguration':
        return new DateValidationConfiguration(source);
      case 'DeAddressParserConfiguration':
        return new DeAddressParserConfiguration(source);
      case 'EuropeanHealthInsuranceCardConfiguration':
        return new EuropeanHealthInsuranceCardConfiguration(source);
      case 'MRZFallbackConfiguration':
        return new MrzFallbackConfiguration(source);
      case 'AutoFocusConfiguration':
        return new AutoFocusConfiguration(source);
      case 'GlareDetectionConfiguration':
        return new GlareDetectionConfiguration(source);
      case 'DocumentDataExtractorCommonConfiguration':
        return new DocumentDataExtractorCommonConfiguration(source);
      default:
        throw `Unknown child class name: ${_type}`;
    }
  }
  _DocumentDataExtractorConfigurationElement.From = From;
})(DocumentDataExtractorConfigurationElement || (exports.DocumentDataExtractorConfigurationElement = DocumentDataExtractorConfigurationElement = {}));
/**
Date validation configuration.
This configuration element does not enable the scanning of any document types by itself.
Add to recognizer configuration to require date validation checks for specific document types.
*/
class DateValidationConfiguration extends _utils.PartiallyConstructible {
  _type = 'DateValidationConfiguration';
  /**
    Minimum year that is considered valid.
    
    Default is 0
    */
  minYear = 0;
  /**
    Maximum year that is considered valid.
    
    Default is 2999
    */
  maxYear = 2999;
  /**
    Name, FullName, or NormalizedName of field type for which the year bounds should be applied.
    */

  /** @param source {@displayType `DeepPartial<DateValidationConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.minYear !== undefined) {
      this.minYear = source.minYear;
    }
    if (source.maxYear !== undefined) {
      this.maxYear = source.maxYear;
    }
    if (source.fieldTypeName !== undefined) {
      this.fieldTypeName = source.fieldTypeName;
    } else {
      throw new Error('fieldTypeName must be present in constructor argument');
    }
  }
}

/**
German Address Field validation configuration.
This configuration does not enable the scanning of any document types by itself.
Add to recognizer configuration to require German Address validation checks for specific document types.
*/
exports.DateValidationConfiguration = DateValidationConfiguration;
class DeAddressParserConfiguration extends _utils.PartiallyConstructible {
  _type = 'DeAddressParserConfiguration';
  /**
    List of generic document formats for which German Address validation has to be applied.
    */

  /**
    binary file with address database.
    
    Default is ""
    */
  binaryFileAddressDatabase = '';

  /** @param source {@displayType `DeepPartial<DeAddressParserConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.documentTypes !== undefined) {
      this.documentTypes = source.documentTypes.map(it => {
        return it;
      });
    } else {
      throw new Error('documentTypes must be present in constructor argument');
    }
    if (source.binaryFileAddressDatabase !== undefined) {
      this.binaryFileAddressDatabase = source.binaryFileAddressDatabase;
    }
  }
}

/**
European Health Insurance Card (EHIC) issuing country.

- `AUSTRIA`:
   Austria (Validation on the personal identification number is performed as described in "site:www.sozialversicherung.at Was ist die Versicherungsnummer?").
- `BELGIUM`:
   Belgium.
- `BULGARIA`:
   Bulgaria.
- `CROATIA`:
   Croatia.
- `CYPRUS`:
   Cyprus.
- `CZECH_REPUBLIC`:
   Czech Republic.
- `DENMARK`:
   Denmark.
- `ESTONIA`:
   Estonia.
- `FINLAND`:
   Finland.
- `FRANCE`:
   France.
- `GERMANY`:
   Germany (Validation performed according to "Spezifikation für Musterkarten und Testkarten (eGK, HBA, SMC), Anhang A" and "GS1 Struktur der Kenn-Nummer (ICCSN) der elektronischen Gesundheitskarte").
- `GREECE`:
   Greece.
- `HUNGARY`:
   Hungary.
- `IRELAND`:
   Ireland.
- `ITALY`:
   Italy.
- `LATVIA`:
   Latvia.
- `LITHUANIA`:
   Lithuania.
- `LUXEMBOURG`:
   Luxembourg.
- `MALTA`:
   Malta.
- `NETHERLANDS`:
   Netherlands.
- `POLAND`:
   Poland.
- `PORTUGAL`:
   Portugal.
- `ROMANIA`:
   Romania.
- `SLOVAKIA`:
   Slovakia.
- `SLOVENIA`:
   Slovenia.
- `SPAIN`:
   Spain.
- `SWEDEN`:
   Sweden.
- `SWITZERLAND`:
   Switzerland.
*/
exports.DeAddressParserConfiguration = DeAddressParserConfiguration;
const EuropeanHealthInsuranceCardIssuingCountryValues = exports.EuropeanHealthInsuranceCardIssuingCountryValues = ['AUSTRIA', 'BELGIUM', 'BULGARIA', 'CROATIA', 'CYPRUS', 'CZECH_REPUBLIC', 'DENMARK', 'ESTONIA', 'FINLAND', 'FRANCE', 'GERMANY', 'GREECE', 'HUNGARY', 'IRELAND', 'ITALY', 'LATVIA', 'LITHUANIA', 'LUXEMBOURG', 'MALTA', 'NETHERLANDS', 'POLAND', 'PORTUGAL', 'ROMANIA', 'SLOVAKIA', 'SLOVENIA', 'SPAIN', 'SWEDEN', 'SWITZERLAND'];

/**
European Health Insurance Card (EHIC) configuration element. Add to extractor configuration to scan EHICs.
*/
class EuropeanHealthInsuranceCardConfiguration extends _utils.PartiallyConstructible {
  _type = 'EuropeanHealthInsuranceCardConfiguration';
  /**
    If an expected country is selected, validation rules for the given country are used,
    and if the expected country cannot be inferred or the inferred country doesn't  match the given country,
    the result of extract() will be IncompleteValidation.
    */
  expectedCountry = null;

  /** @param source {@displayType `DeepPartial<EuropeanHealthInsuranceCardConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.expectedCountry !== undefined) {
      this.expectedCountry = source.expectedCountry != null ? EuropeanHealthInsuranceCardIssuingCountryValues.includes(source.expectedCountry) ? source.expectedCountry : null : null;
    }
  }
}

/**
MRZ fallback configuration element.
This configuration element does not enable the scanning of any document types by itself.
Add to extractor configuration to enable recognizing the MRZ only, specifically for documents that are otherwise not supported.
Note that this may enable scanning of documents that were not enabled through the accepted document types.
*/
exports.EuropeanHealthInsuranceCardConfiguration = EuropeanHealthInsuranceCardConfiguration;
class MrzFallbackConfiguration extends _utils.PartiallyConstructible {
  _type = 'MRZFallbackConfiguration';
  /**
    List of ISO 3166-1 alpha-3 country codes for which the MRZ fallback is enabled. (e.g. "DEU" for Germany).
    */
  acceptedCountries = [];
  /**
    List of MRZ document types for which MRZ fallback is enabled.
    */
  acceptedMRZTypes = [];

  /** @param source {@displayType `DeepPartial<MrzFallbackConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.acceptedCountries !== undefined) {
      this.acceptedCountries = source.acceptedCountries.map(it => {
        return it;
      });
    }
    if (source.acceptedMRZTypes !== undefined) {
      this.acceptedMRZTypes = source.acceptedMRZTypes.map(it => {
        return it;
      });
    }
  }
}

/**
Auto-focus configuration.
The recognizer will ignore frames that are not focused. This option is enabled by default.
This configuration has no effect in SINGLE_SHOT mode.
*/
exports.MrzFallbackConfiguration = MrzFallbackConfiguration;
class AutoFocusConfiguration extends _utils.PartiallyConstructible {
  _type = 'AutoFocusConfiguration';
  /**
    Sensitivity threshold for detecting that autofocus is in progress.
    The value should be in the range [0.0, 1.0].
    The lower the value, the more frames might be discarded as not focused. 
    The higher the value, the earlier the focus might be considered as focused.
    If set to 0.0, the feature is disabled.
    
    Default is 0.02
    */
  sensitivityThreshold = 0.02;

  /** @param source {@displayType `DeepPartial<AutoFocusConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.sensitivityThreshold !== undefined) {
      this.sensitivityThreshold = source.sensitivityThreshold;
    }
  }
}

/**
Glare detection configuration.
The recognizer will detect whether fields are obstructed by glare. This feature is enabled by default.
This configuration has no effect in SINGLE_SHOT mode.
*/
exports.AutoFocusConfiguration = AutoFocusConfiguration;
class GlareDetectionConfiguration extends _utils.PartiallyConstructible {
  _type = 'GlareDetectionConfiguration';
  /**
    Glare detection threshold.
    The value should be in the range [0.0, 1.0].
    Determines how much glare is acceptable relative to the text size. So 0.5 means that the glare can be up to 50% of the text size.
    If set to 0.0, the feature is disabled.
    
    Default is 0.5
    */
  glareDetectionThreshold = 0.5;

  /** @param source {@displayType `DeepPartial<GlareDetectionConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.glareDetectionThreshold !== undefined) {
      this.glareDetectionThreshold = source.glareDetectionThreshold;
    }
  }
}

/**
Convenience configuration element for enabling the scanning of multiple document types with a common configuration.
Add to extractor configuration to enable the scanning of multiple document types.
*/
exports.GlareDetectionConfiguration = GlareDetectionConfiguration;
class DocumentDataExtractorCommonConfiguration extends _utils.PartiallyConstructible {
  _type = 'DocumentDataExtractorCommonConfiguration';
  /**
    List of document types to scan. By default, the list is empty.
    */
  acceptedDocumentTypes = ['DeIdCardFront', 'DeIdCardBack', 'DeHealthInsuranceCardFront', 'DePassport', 'DeResidencePermitFront', 'DeResidencePermitBack', 'EuropeanHealthInsuranceCard', 'EuropeanDriverLicenseFront', 'EuropeanDriverLicenseBack'];

  /** @param source {@displayType `DeepPartial<DocumentDataExtractorCommonConfiguration>`} */
  constructor(source = {}) {
    super();
    if (source.acceptedDocumentTypes !== undefined) {
      this.acceptedDocumentTypes = source.acceptedDocumentTypes.map(it => {
        return it;
      });
    }
  }
}
exports.DocumentDataExtractorCommonConfiguration = DocumentDataExtractorCommonConfiguration;
