// react-native-scanbot-barcode-scanner-sdk/lib/commonjs/types/base/errors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UnknownError = exports.TimeoutError = exports.SBError = exports.ProcessError = exports.OutOfMemoryError = exports.NullPointerError = exports.InvalidLicenseError = exports.InvalidImageRefError = exports.InvalidDataError = exports.InvalidArgumentError = exports.IllegalStateError = exports.IOError = exports.ComponentUnavailableError = void 0;
exports.createSBError = createSBError;
class SBError extends Error {
  constructor(message) {
    super(message);
  }
}
exports.SBError = SBError;
class InvalidLicenseError extends SBError {
  type = 'InvalidLicense';
  constructor(message) {
    super(message);
  }
}
exports.InvalidLicenseError = InvalidLicenseError;
class NullPointerError extends SBError {
  type = 'NullPointer';
  constructor(message) {
    super(message);
  }
}
exports.NullPointerError = NullPointerError;
class InvalidArgumentError extends SBError {
  type = 'InvalidArgument';
  constructor(message) {
    super(message);
  }
}
exports.InvalidArgumentError = InvalidArgumentError;
class InvalidImageRefError extends SBError {
  type = 'InvalidImageRef';
  constructor(message) {
    super(message);
  }
}
exports.InvalidImageRefError = InvalidImageRefError;
class ComponentUnavailableError extends SBError {
  type = 'ComponentUnavailable';
  constructor(message) {
    super(message);
  }
}
exports.ComponentUnavailableError = ComponentUnavailableError;
class IllegalStateError extends SBError {
  type = 'IllegalState';
  constructor(message) {
    super(message);
  }
}
exports.IllegalStateError = IllegalStateError;
class IOError extends SBError {
  type = 'IOError';
  constructor(message) {
    super(message);
  }
}
exports.IOError = IOError;
class InvalidDataError extends SBError {
  type = 'InvalidData';
  constructor(message) {
    super(message);
  }
}
exports.InvalidDataError = InvalidDataError;
class OutOfMemoryError extends SBError {
  type = 'OutOfMemory';
  constructor(message) {
    super(message);
  }
}
exports.OutOfMemoryError = OutOfMemoryError;
class TimeoutError extends SBError {
  type = 'Timeout';
  constructor(message) {
    super(message);
  }
}
exports.TimeoutError = TimeoutError;
class UnknownError extends SBError {
  type = 'Unknown';
  constructor(message) {
    super(message);
  }
}
exports.UnknownError = UnknownError;
class ProcessError extends SBError {
  type = 'ProcessError';
  constructor(message, code) {
    super(message);
    this.code = code;
  }
}

/**
 * @internal
 * @hidden
 */
exports.ProcessError = ProcessError;
function createSBError(err) {
  const code = Number(err.code);
  const message = err.message || err.errorMessage || 'An unknown error occurred';
  switch (code) {
    case 1:
      return new UnknownError(message);
    case 2:
      return new InvalidLicenseError(message);
    case 3:
      return new NullPointerError(message);
    case 4:
      return new InvalidArgumentError(message);
    case 5:
      return new InvalidImageRefError(message);
    case 6:
      return new ComponentUnavailableError(message);
    case 7:
      return new IllegalStateError(message);
    case 8:
      return new IOError(message);
    case 9:
      return new InvalidDataError(message);
    case 11:
      return new OutOfMemoryError(message);
    case 12:
      return new TimeoutError(message);
    default:
      {
        if (code >= 100) {
          return new ProcessError(message, code);
        }
        return new UnknownError(message);
      }
  }
}
