// react-native-scanbot-barcode-scanner-sdk/lib/commonjs/types/base/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.handleImageInput = handleImageInput;
exports.mapRTUUIResult = mapRTUUIResult;
exports.withSBErrorHandling = withSBErrorHandling;
var _core = require("../core");
var _errors = require("./errors");
/**
 * @internal
 * @hidden
 */
function mapRTUUIResult(result, LClass) {
  if (result.status === 'OK') {
    return {
      ...result,
      data: new LClass(result.data)
    };
  } else {
    return result;
  }
}

/**
 * @internal
 * @hidden
 */

function handleImageInput(image) {
  const handleInput = input => {
    if (input instanceof _core.ImageRef) {
      return input.uniqueId ? {
        uniqueId: input.uniqueId
      } : {
        buffer: input.buffer
      };
    } else {
      return {
        imageFileUri: input
      };
    }
  };
  return Array.isArray(image) ? image.map(handleInput) : handleInput(image);
}

/**
 * @internal
 * @hidden
 */
async function withSBErrorHandling(fn) {
  try {
    return await fn();
  } catch (error) {
    throw (0, _errors.createSBError)(error);
  }
}
